{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "from itertools import groupby\n",
    "from operator import itemgetter\n",
    "import subprocess\n",
    "import sys\n",
    "import os\n",
    "import csv\n",
    "import matplotlib.pyplot as plt\n",
    "from matplotlib import animation\n",
    "import seaborn as sns\n",
    "plt.rcParams['figure.figsize'] = (20.0, 10.0)\n",
    "plt.rcParams['font.family'] = \"serif\"\n",
    "%matplotlib inline\n",
    "\n",
    "gff3Cols=[\"seqid\",\"source\",\"type\",\"start\",\"end\",\"score\",\"strand\",\"phase\",\"attributes\"]\n",
    "novelSources=['CHESS','StringTie','FANTOM']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_known_only=pd.read_csv(\"./chess2.02.gff\",sep=\"\\t\",names=gff3Cols)\n",
    "df_known_only.dropna(inplace=True,axis=0)\n",
    "df_known_only.reset_index(inplace=True,drop=True)\n",
    "df_known_only[\"start\"]=df_known_only[\"start\"].astype(int)\n",
    "df_known_only[\"end\"]=df_known_only[\"end\"].astype(int)\n",
    "df_known_only=df_known_only[~(df_known_only['source'].isin(novelSources))].reset_index(drop=True)\n",
    "df_known_only[\"id\"]=df_known_only.attributes.str.split(\"ID=\",expand=True)[1].str.split(\";\",expand=True)[0]\n",
    "df_known_only[\"parent\"]=df_known_only.attributes.str.split(\"Parent=\",expand=True)[1].str.split(\";\",expand=True)[0]\n",
    "df_known_only[\"geneID\"]=np.where(df_known_only[\"type\"].isin(['transcript','exon','CDS']),df_known_only.parent.str.extract('(CHS.(\\d)*)',expand=True)[0],df_known_only['id'])\n",
    "tmp=df_known_only[df_known_only[\"type\"]==\"gene\"].reset_index(drop=True)\n",
    "tmp[\"gene_type\"]=tmp.attributes.str.split(\"GENE_TYPE=\",expand=True)[1].str.split(\";\",expand=True)[0].str.strip(\"\\n\")\n",
    "tmp=tmp[tmp[\"gene_type\"]==\"protein_coding\"].reset_index(drop=True)\n",
    "setProtGenes_known_only=set(tmp[\"geneID\"])\n",
    "del tmp\n",
    "df_known_only=df_known_only[df_known_only[\"geneID\"].isin(setProtGenes_known_only)].reset_index(drop=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqid</th>\n",
       "      <th>source</th>\n",
       "      <th>type</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>phase</th>\n",
       "      <th>attributes</th>\n",
       "      <th>id</th>\n",
       "      <th>parent</th>\n",
       "      <th>geneID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>gene</td>\n",
       "      <td>69091</td>\n",
       "      <td>70008</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=CHS.6;GENE_TYPE=protein_coding;STATUS=known...</td>\n",
       "      <td>CHS.6</td>\n",
       "      <td>None</td>\n",
       "      <td>CHS.6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>transcript</td>\n",
       "      <td>69091</td>\n",
       "      <td>70008</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=CHS.6.1;Parent=CHS.6;gene_name=OR4F5;Dbxref...</td>\n",
       "      <td>CHS.6.1</td>\n",
       "      <td>CHS.6</td>\n",
       "      <td>CHS.6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>CDS</td>\n",
       "      <td>69091</td>\n",
       "      <td>70008</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>Parent=CHS.6.1</td>\n",
       "      <td>None</td>\n",
       "      <td>CHS.6.1</td>\n",
       "      <td>CHS.6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>exon</td>\n",
       "      <td>69091</td>\n",
       "      <td>70008</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>Parent=CHS.6.1</td>\n",
       "      <td>None</td>\n",
       "      <td>CHS.6.1</td>\n",
       "      <td>CHS.6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>gene</td>\n",
       "      <td>181049</td>\n",
       "      <td>184258</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=CHS.9;GENE_TYPE=protein_coding;STATUS=known...</td>\n",
       "      <td>CHS.9</td>\n",
       "      <td>None</td>\n",
       "      <td>CHS.9</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  seqid      source        type   start     end score strand phase  \\\n",
       "0  chr1      RefSeq        gene   69091   70008     .      +     .   \n",
       "1  chr1  BestRefSeq  transcript   69091   70008     .      +     .   \n",
       "2  chr1  BestRefSeq         CDS   69091   70008     .      +     0   \n",
       "3  chr1  BestRefSeq        exon   69091   70008     .      +     .   \n",
       "4  chr1      RefSeq        gene  181049  184258     .      +     .   \n",
       "\n",
       "                                          attributes       id   parent geneID  \n",
       "0  ID=CHS.6;GENE_TYPE=protein_coding;STATUS=known...    CHS.6     None  CHS.6  \n",
       "1  ID=CHS.6.1;Parent=CHS.6;gene_name=OR4F5;Dbxref...  CHS.6.1    CHS.6  CHS.6  \n",
       "2                                     Parent=CHS.6.1     None  CHS.6.1  CHS.6  \n",
       "3                                     Parent=CHS.6.1     None  CHS.6.1  CHS.6  \n",
       "4  ID=CHS.9;GENE_TYPE=protein_coding;STATUS=known...    CHS.9     None  CHS.9  "
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_known_only.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "# now we need to create a subset of exons from known transcripts\n",
    "cdsParents_known_only=set(df_known_only[df_known_only[\"type\"]==\"CDS\"][\"parent\"])\n",
    "transIDs_known_only=set(df_known_only[df_known_only[\"type\"]==\"transcript\"][\"id\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>geneID</th>\n",
       "      <th>smin</th>\n",
       "      <th>smax</th>\n",
       "      <th>emin</th>\n",
       "      <th>emax</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CHS.10000</td>\n",
       "      <td>112961247</td>\n",
       "      <td>113275267</td>\n",
       "      <td>112961664</td>\n",
       "      <td>113278436</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CHS.10003</td>\n",
       "      <td>113314529</td>\n",
       "      <td>113385926</td>\n",
       "      <td>113314618</td>\n",
       "      <td>113385952</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CHS.10004</td>\n",
       "      <td>113386014</td>\n",
       "      <td>113398964</td>\n",
       "      <td>113388069</td>\n",
       "      <td>113400418</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CHS.10005</td>\n",
       "      <td>113409595</td>\n",
       "      <td>113475076</td>\n",
       "      <td>113410920</td>\n",
       "      <td>113475279</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CHS.10008</td>\n",
       "      <td>113687546</td>\n",
       "      <td>113706222</td>\n",
       "      <td>113688274</td>\n",
       "      <td>113709811</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      geneID       smin       smax       emin       emax\n",
       "0  CHS.10000  112961247  113275267  112961664  113278436\n",
       "1  CHS.10003  113314529  113385926  113314618  113385952\n",
       "2  CHS.10004  113386014  113398964  113388069  113400418\n",
       "3  CHS.10005  113409595  113475076  113410920  113475279\n",
       "4  CHS.10008  113687546  113706222  113688274  113709811"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "knownDF=df_known_only[(df_known_only['type']=='exon')&(df_known_only[\"parent\"].isin(cdsParents_known_only))].reset_index(drop=True)\n",
    "# technically all we need are the mins and maxs of known transcripts\n",
    "# and the corresponding mins and maxs of the novel isoforms\n",
    "knownDF=knownDF.groupby(\"geneID\").agg({'start':['min','max'],'end':['min','max']}).reset_index()\n",
    "knownDF.columns=['geneID','smin','smax','emin','emax']\n",
    "knownDF.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqid</th>\n",
       "      <th>source</th>\n",
       "      <th>type</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>phase</th>\n",
       "      <th>attributes</th>\n",
       "      <th>parent</th>\n",
       "      <th>geneID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>StringTie</td>\n",
       "      <td>exon</td>\n",
       "      <td>732160</td>\n",
       "      <td>732211</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>Parent=CHS.23.9</td>\n",
       "      <td>CHS.23.9</td>\n",
       "      <td>CHS.23</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>StringTie</td>\n",
       "      <td>exon</td>\n",
       "      <td>732981</td>\n",
       "      <td>733213</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>Parent=CHS.23.9</td>\n",
       "      <td>CHS.23.9</td>\n",
       "      <td>CHS.23</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>StringTie</td>\n",
       "      <td>exon</td>\n",
       "      <td>736713</td>\n",
       "      <td>736770</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>Parent=CHS.23.9</td>\n",
       "      <td>CHS.23.9</td>\n",
       "      <td>CHS.23</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>StringTie</td>\n",
       "      <td>exon</td>\n",
       "      <td>740187</td>\n",
       "      <td>740293</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>Parent=CHS.23.9</td>\n",
       "      <td>CHS.23.9</td>\n",
       "      <td>CHS.23</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>StringTie</td>\n",
       "      <td>exon</td>\n",
       "      <td>733082</td>\n",
       "      <td>733213</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>Parent=CHS.23.10</td>\n",
       "      <td>CHS.23.10</td>\n",
       "      <td>CHS.23</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  seqid     source  type   start     end score strand phase        attributes  \\\n",
       "0  chr1  StringTie  exon  732160  732211     .      -     .   Parent=CHS.23.9   \n",
       "1  chr1  StringTie  exon  732981  733213     .      -     .   Parent=CHS.23.9   \n",
       "2  chr1  StringTie  exon  736713  736770     .      -     .   Parent=CHS.23.9   \n",
       "3  chr1  StringTie  exon  740187  740293     .      -     .   Parent=CHS.23.9   \n",
       "4  chr1  StringTie  exon  733082  733213     .      -     .  Parent=CHS.23.10   \n",
       "\n",
       "      parent  geneID  \n",
       "0   CHS.23.9  CHS.23  \n",
       "1   CHS.23.9  CHS.23  \n",
       "2   CHS.23.9  CHS.23  \n",
       "3   CHS.23.9  CHS.23  \n",
       "4  CHS.23.10  CHS.23  "
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# now we need to get the same dataframe but for novel isoforms only\n",
    "# computing same mins and maxs\n",
    "# then we need to merge the two dataframes together in such a way, that each novel transcript is represented\n",
    "# which implies that certain known isoforms will be repeated if multiple novel are annotated for the same gene\n",
    "df_novel=pd.read_csv(\"./chess2.02.gff\",sep=\"\\t\",names=gff3Cols)\n",
    "df_novel.dropna(inplace=True,axis=0)\n",
    "df_novel.reset_index(inplace=True,drop=True)\n",
    "df_novel=df_novel[df_novel[\"type\"]==\"exon\"].reset_index(drop=True)\n",
    "df_novel[\"start\"]=df_novel[\"start\"].astype(int)\n",
    "df_novel[\"end\"]=df_novel[\"end\"].astype(int)\n",
    "df_novel=df_novel[(df_novel['source'].isin(novelSources))].reset_index(drop=True)\n",
    "df_novel[\"parent\"]=df_novel.attributes.str.split(\"Parent=\",expand=True)[1].str.split(\";\",expand=True)[0]\n",
    "df_novel[\"geneID\"]=df_novel.parent.str.extract('(CHS.(\\d)*)',expand=True)[0]\n",
    "df_novel=df_novel[df_novel[\"geneID\"].isin(set(knownDF[\"geneID\"]))].reset_index(drop=True)\n",
    "df_novel.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>parent</th>\n",
       "      <th>smin</th>\n",
       "      <th>smax</th>\n",
       "      <th>emin</th>\n",
       "      <th>emax</th>\n",
       "      <th>geneID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CHS.10000.18</td>\n",
       "      <td>113204284</td>\n",
       "      <td>113262846</td>\n",
       "      <td>113204504</td>\n",
       "      <td>113265237</td>\n",
       "      <td>CHS.10000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CHS.10000.26</td>\n",
       "      <td>113246078</td>\n",
       "      <td>113275267</td>\n",
       "      <td>113246370</td>\n",
       "      <td>113280040</td>\n",
       "      <td>CHS.10000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CHS.10000.27</td>\n",
       "      <td>113246078</td>\n",
       "      <td>113275267</td>\n",
       "      <td>113246370</td>\n",
       "      <td>113280283</td>\n",
       "      <td>CHS.10000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CHS.10000.29</td>\n",
       "      <td>113246154</td>\n",
       "      <td>113262846</td>\n",
       "      <td>113246370</td>\n",
       "      <td>113265354</td>\n",
       "      <td>CHS.10000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CHS.10000.30</td>\n",
       "      <td>113246154</td>\n",
       "      <td>113262846</td>\n",
       "      <td>113246370</td>\n",
       "      <td>113265354</td>\n",
       "      <td>CHS.10000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         parent       smin       smax       emin       emax     geneID\n",
       "0  CHS.10000.18  113204284  113262846  113204504  113265237  CHS.10000\n",
       "1  CHS.10000.26  113246078  113275267  113246370  113280040  CHS.10000\n",
       "2  CHS.10000.27  113246078  113275267  113246370  113280283  CHS.10000\n",
       "3  CHS.10000.29  113246154  113262846  113246370  113265354  CHS.10000\n",
       "4  CHS.10000.30  113246154  113262846  113246370  113265354  CHS.10000"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "novelDF=df_novel.groupby(\"parent\").agg({'start':['min','max'],'end':['min','max'],'geneID':'min'}).reset_index()\n",
    "novelDF.columns=['parent','smin','smax','emin','emax','geneID']\n",
    "novelDF.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>geneID</th>\n",
       "      <th>smin_x</th>\n",
       "      <th>smax_x</th>\n",
       "      <th>emin_x</th>\n",
       "      <th>emax_x</th>\n",
       "      <th>parent</th>\n",
       "      <th>smin_y</th>\n",
       "      <th>smax_y</th>\n",
       "      <th>emin_y</th>\n",
       "      <th>emax_y</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CHS.10000</td>\n",
       "      <td>112961247</td>\n",
       "      <td>113275267</td>\n",
       "      <td>112961664</td>\n",
       "      <td>113278436</td>\n",
       "      <td>CHS.10000.18</td>\n",
       "      <td>113204284.0</td>\n",
       "      <td>113262846.0</td>\n",
       "      <td>113204504.0</td>\n",
       "      <td>113265237.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CHS.10000</td>\n",
       "      <td>112961247</td>\n",
       "      <td>113275267</td>\n",
       "      <td>112961664</td>\n",
       "      <td>113278436</td>\n",
       "      <td>CHS.10000.26</td>\n",
       "      <td>113246078.0</td>\n",
       "      <td>113275267.0</td>\n",
       "      <td>113246370.0</td>\n",
       "      <td>113280040.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CHS.10000</td>\n",
       "      <td>112961247</td>\n",
       "      <td>113275267</td>\n",
       "      <td>112961664</td>\n",
       "      <td>113278436</td>\n",
       "      <td>CHS.10000.27</td>\n",
       "      <td>113246078.0</td>\n",
       "      <td>113275267.0</td>\n",
       "      <td>113246370.0</td>\n",
       "      <td>113280283.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CHS.10000</td>\n",
       "      <td>112961247</td>\n",
       "      <td>113275267</td>\n",
       "      <td>112961664</td>\n",
       "      <td>113278436</td>\n",
       "      <td>CHS.10000.29</td>\n",
       "      <td>113246154.0</td>\n",
       "      <td>113262846.0</td>\n",
       "      <td>113246370.0</td>\n",
       "      <td>113265354.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CHS.10000</td>\n",
       "      <td>112961247</td>\n",
       "      <td>113275267</td>\n",
       "      <td>112961664</td>\n",
       "      <td>113278436</td>\n",
       "      <td>CHS.10000.30</td>\n",
       "      <td>113246154.0</td>\n",
       "      <td>113262846.0</td>\n",
       "      <td>113246370.0</td>\n",
       "      <td>113265354.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      geneID     smin_x     smax_x     emin_x     emax_x        parent  \\\n",
       "0  CHS.10000  112961247  113275267  112961664  113278436  CHS.10000.18   \n",
       "1  CHS.10000  112961247  113275267  112961664  113278436  CHS.10000.26   \n",
       "2  CHS.10000  112961247  113275267  112961664  113278436  CHS.10000.27   \n",
       "3  CHS.10000  112961247  113275267  112961664  113278436  CHS.10000.29   \n",
       "4  CHS.10000  112961247  113275267  112961664  113278436  CHS.10000.30   \n",
       "\n",
       "        smin_y       smax_y       emin_y       emax_y  \n",
       "0  113204284.0  113262846.0  113204504.0  113265237.0  \n",
       "1  113246078.0  113275267.0  113246370.0  113280040.0  \n",
       "2  113246078.0  113275267.0  113246370.0  113280283.0  \n",
       "3  113246154.0  113262846.0  113246370.0  113265354.0  \n",
       "4  113246154.0  113262846.0  113246370.0  113265354.0  "
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# now merge the two together (novel with the known transcripts)\n",
    "resDF=knownDF.merge(novelDF,on='geneID',how='outer',indicator=True)\n",
    "assert len(resDF[(resDF['_merge']=='right_only')])==0,\"something's wrong with the novel transcripts. Some don't match the known isoforms \"\n",
    "resDF=resDF[resDF['_merge']=='both'].reset_index(drop=True)\n",
    "resDF.drop('_merge',inplace=True,axis=1)\n",
    "resDF.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>geneID</th>\n",
       "      <th>smin_known</th>\n",
       "      <th>smax_known</th>\n",
       "      <th>emin_known</th>\n",
       "      <th>emax_known</th>\n",
       "      <th>parent</th>\n",
       "      <th>smin_novel</th>\n",
       "      <th>smax_novel</th>\n",
       "      <th>emin_novel</th>\n",
       "      <th>emax_novel</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CHS.10000</td>\n",
       "      <td>112961247</td>\n",
       "      <td>113275267</td>\n",
       "      <td>112961664</td>\n",
       "      <td>113278436</td>\n",
       "      <td>CHS.10000.18</td>\n",
       "      <td>113204284.0</td>\n",
       "      <td>113262846.0</td>\n",
       "      <td>113204504.0</td>\n",
       "      <td>113265237.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CHS.10000</td>\n",
       "      <td>112961247</td>\n",
       "      <td>113275267</td>\n",
       "      <td>112961664</td>\n",
       "      <td>113278436</td>\n",
       "      <td>CHS.10000.26</td>\n",
       "      <td>113246078.0</td>\n",
       "      <td>113275267.0</td>\n",
       "      <td>113246370.0</td>\n",
       "      <td>113280040.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CHS.10000</td>\n",
       "      <td>112961247</td>\n",
       "      <td>113275267</td>\n",
       "      <td>112961664</td>\n",
       "      <td>113278436</td>\n",
       "      <td>CHS.10000.27</td>\n",
       "      <td>113246078.0</td>\n",
       "      <td>113275267.0</td>\n",
       "      <td>113246370.0</td>\n",
       "      <td>113280283.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CHS.10000</td>\n",
       "      <td>112961247</td>\n",
       "      <td>113275267</td>\n",
       "      <td>112961664</td>\n",
       "      <td>113278436</td>\n",
       "      <td>CHS.10000.29</td>\n",
       "      <td>113246154.0</td>\n",
       "      <td>113262846.0</td>\n",
       "      <td>113246370.0</td>\n",
       "      <td>113265354.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CHS.10000</td>\n",
       "      <td>112961247</td>\n",
       "      <td>113275267</td>\n",
       "      <td>112961664</td>\n",
       "      <td>113278436</td>\n",
       "      <td>CHS.10000.30</td>\n",
       "      <td>113246154.0</td>\n",
       "      <td>113262846.0</td>\n",
       "      <td>113246370.0</td>\n",
       "      <td>113265354.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      geneID  smin_known  smax_known  emin_known  emax_known        parent  \\\n",
       "0  CHS.10000   112961247   113275267   112961664   113278436  CHS.10000.18   \n",
       "1  CHS.10000   112961247   113275267   112961664   113278436  CHS.10000.26   \n",
       "2  CHS.10000   112961247   113275267   112961664   113278436  CHS.10000.27   \n",
       "3  CHS.10000   112961247   113275267   112961664   113278436  CHS.10000.29   \n",
       "4  CHS.10000   112961247   113275267   112961664   113278436  CHS.10000.30   \n",
       "\n",
       "    smin_novel   smax_novel   emin_novel   emax_novel  \n",
       "0  113204284.0  113262846.0  113204504.0  113265237.0  \n",
       "1  113246078.0  113275267.0  113246370.0  113280040.0  \n",
       "2  113246078.0  113275267.0  113246370.0  113280283.0  \n",
       "3  113246154.0  113262846.0  113246370.0  113265354.0  \n",
       "4  113246154.0  113262846.0  113246370.0  113265354.0  "
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "resDF.columns=['geneID','smin_known','smax_known','emin_known','emax_known',\n",
    "               'parent','smin_novel','smax_novel','emin_novel','emax_novel']\n",
    "resDF.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>geneID</th>\n",
       "      <th>smin_known</th>\n",
       "      <th>smax_known</th>\n",
       "      <th>emin_known</th>\n",
       "      <th>emax_known</th>\n",
       "      <th>parent</th>\n",
       "      <th>smin_novel</th>\n",
       "      <th>smax_novel</th>\n",
       "      <th>emin_novel</th>\n",
       "      <th>emax_novel</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CHS.10904</td>\n",
       "      <td>8950044</td>\n",
       "      <td>9215276</td>\n",
       "      <td>8950236</td>\n",
       "      <td>9215657</td>\n",
       "      <td>CHS.10904.15</td>\n",
       "      <td>9244635.0</td>\n",
       "      <td>9247581.0</td>\n",
       "      <td>9246548.0</td>\n",
       "      <td>9247608.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CHS.10904</td>\n",
       "      <td>8950044</td>\n",
       "      <td>9215276</td>\n",
       "      <td>8950236</td>\n",
       "      <td>9215657</td>\n",
       "      <td>CHS.10904.17</td>\n",
       "      <td>9244635.0</td>\n",
       "      <td>9260614.0</td>\n",
       "      <td>9246548.0</td>\n",
       "      <td>9263160.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CHS.11417</td>\n",
       "      <td>42326126</td>\n",
       "      <td>42446592</td>\n",
       "      <td>42326229</td>\n",
       "      <td>42448621</td>\n",
       "      <td>CHS.11417.44</td>\n",
       "      <td>42238469.0</td>\n",
       "      <td>42239137.0</td>\n",
       "      <td>42238727.0</td>\n",
       "      <td>42241129.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CHS.11615</td>\n",
       "      <td>50979783</td>\n",
       "      <td>51028511</td>\n",
       "      <td>50981803</td>\n",
       "      <td>51028566</td>\n",
       "      <td>CHS.11615.3</td>\n",
       "      <td>50952264.0</td>\n",
       "      <td>50969390.0</td>\n",
       "      <td>50953808.0</td>\n",
       "      <td>50970219.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CHS.11626</td>\n",
       "      <td>51238724</td>\n",
       "      <td>51246061</td>\n",
       "      <td>51238920</td>\n",
       "      <td>51246717</td>\n",
       "      <td>CHS.11626.25</td>\n",
       "      <td>51247766.0</td>\n",
       "      <td>51248479.0</td>\n",
       "      <td>51247978.0</td>\n",
       "      <td>51249650.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>CHS.11936</td>\n",
       "      <td>57434685</td>\n",
       "      <td>57449277</td>\n",
       "      <td>57435199</td>\n",
       "      <td>57452062</td>\n",
       "      <td>CHS.11936.5</td>\n",
       "      <td>57431191.0</td>\n",
       "      <td>57433703.0</td>\n",
       "      <td>57431967.0</td>\n",
       "      <td>57434408.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>CHS.12135</td>\n",
       "      <td>69359710</td>\n",
       "      <td>69390147</td>\n",
       "      <td>69360023</td>\n",
       "      <td>69390870</td>\n",
       "      <td>CHS.12135.8</td>\n",
       "      <td>69416433.0</td>\n",
       "      <td>69424001.0</td>\n",
       "      <td>69416627.0</td>\n",
       "      <td>69424837.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>CHS.12461</td>\n",
       "      <td>95217746</td>\n",
       "      <td>95300165</td>\n",
       "      <td>95217886</td>\n",
       "      <td>95302790</td>\n",
       "      <td>CHS.12461.67</td>\n",
       "      <td>95305942.0</td>\n",
       "      <td>95306937.0</td>\n",
       "      <td>95306028.0</td>\n",
       "      <td>95310639.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>CHS.13013</td>\n",
       "      <td>121400083</td>\n",
       "      <td>121423386</td>\n",
       "      <td>121400218</td>\n",
       "      <td>121424352</td>\n",
       "      <td>CHS.13013.16</td>\n",
       "      <td>121375425.0</td>\n",
       "      <td>121391970.0</td>\n",
       "      <td>121375816.0</td>\n",
       "      <td>121392070.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>CHS.14025</td>\n",
       "      <td>48075728</td>\n",
       "      <td>48095060</td>\n",
       "      <td>48077311</td>\n",
       "      <td>48095141</td>\n",
       "      <td>CHS.14025.8</td>\n",
       "      <td>48053331.0</td>\n",
       "      <td>48058783.0</td>\n",
       "      <td>48053491.0</td>\n",
       "      <td>48059622.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>CHS.14448</td>\n",
       "      <td>87671328</td>\n",
       "      <td>87675381</td>\n",
       "      <td>87672209</td>\n",
       "      <td>87679619</td>\n",
       "      <td>CHS.14448.6</td>\n",
       "      <td>87679888.0</td>\n",
       "      <td>87695419.0</td>\n",
       "      <td>87691235.0</td>\n",
       "      <td>87696416.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>CHS.14806</td>\n",
       "      <td>113177539</td>\n",
       "      <td>113208599</td>\n",
       "      <td>113177976</td>\n",
       "      <td>113209466</td>\n",
       "      <td>CHS.14806.2</td>\n",
       "      <td>113165003.0</td>\n",
       "      <td>113175378.0</td>\n",
       "      <td>113165183.0</td>\n",
       "      <td>113176331.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>CHS.15061</td>\n",
       "      <td>23555982</td>\n",
       "      <td>23558695</td>\n",
       "      <td>23556343</td>\n",
       "      <td>23559581</td>\n",
       "      <td>CHS.15061.8</td>\n",
       "      <td>23513207.0</td>\n",
       "      <td>23518865.0</td>\n",
       "      <td>23518396.0</td>\n",
       "      <td>23520831.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>CHS.16509</td>\n",
       "      <td>104085658</td>\n",
       "      <td>104112524</td>\n",
       "      <td>104085852</td>\n",
       "      <td>104112749</td>\n",
       "      <td>CHS.16509.35</td>\n",
       "      <td>104117141.0</td>\n",
       "      <td>104117482.0</td>\n",
       "      <td>104117252.0</td>\n",
       "      <td>104128281.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>CHS.16509</td>\n",
       "      <td>104085658</td>\n",
       "      <td>104112524</td>\n",
       "      <td>104085852</td>\n",
       "      <td>104112749</td>\n",
       "      <td>CHS.16509.36</td>\n",
       "      <td>104117179.0</td>\n",
       "      <td>104117390.0</td>\n",
       "      <td>104117252.0</td>\n",
       "      <td>104126141.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>CHS.16509</td>\n",
       "      <td>104085658</td>\n",
       "      <td>104112524</td>\n",
       "      <td>104085852</td>\n",
       "      <td>104112749</td>\n",
       "      <td>CHS.16509.37</td>\n",
       "      <td>104117501.0</td>\n",
       "      <td>104119835.0</td>\n",
       "      <td>104119798.0</td>\n",
       "      <td>104125736.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>CHS.16971</td>\n",
       "      <td>34379068</td>\n",
       "      <td>34407524</td>\n",
       "      <td>34381613</td>\n",
       "      <td>34407674</td>\n",
       "      <td>CHS.16971.19</td>\n",
       "      <td>34415382.0</td>\n",
       "      <td>34437398.0</td>\n",
       "      <td>34419343.0</td>\n",
       "      <td>34437465.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>CHS.17277</td>\n",
       "      <td>45631148</td>\n",
       "      <td>45690973</td>\n",
       "      <td>45631230</td>\n",
       "      <td>45691294</td>\n",
       "      <td>CHS.17277.19</td>\n",
       "      <td>45697041.0</td>\n",
       "      <td>45701436.0</td>\n",
       "      <td>45699635.0</td>\n",
       "      <td>45703171.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>CHS.17900</td>\n",
       "      <td>75598083</td>\n",
       "      <td>75626239</td>\n",
       "      <td>75598681</td>\n",
       "      <td>75626378</td>\n",
       "      <td>CHS.17900.25</td>\n",
       "      <td>75588978.0</td>\n",
       "      <td>75593717.0</td>\n",
       "      <td>75592470.0</td>\n",
       "      <td>75596585.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>CHS.17960</td>\n",
       "      <td>78167468</td>\n",
       "      <td>78234371</td>\n",
       "      <td>78168485</td>\n",
       "      <td>78234707</td>\n",
       "      <td>CHS.17960.16</td>\n",
       "      <td>78244203.0</td>\n",
       "      <td>78245269.0</td>\n",
       "      <td>78244853.0</td>\n",
       "      <td>78245686.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>CHS.18044</td>\n",
       "      <td>80975744</td>\n",
       "      <td>80989579</td>\n",
       "      <td>80979477</td>\n",
       "      <td>80989878</td>\n",
       "      <td>CHS.18044.3</td>\n",
       "      <td>80947327.0</td>\n",
       "      <td>80951120.0</td>\n",
       "      <td>80948898.0</td>\n",
       "      <td>80951765.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>CHS.18044</td>\n",
       "      <td>80975744</td>\n",
       "      <td>80989579</td>\n",
       "      <td>80979477</td>\n",
       "      <td>80989878</td>\n",
       "      <td>CHS.18044.4</td>\n",
       "      <td>80947327.0</td>\n",
       "      <td>80957473.0</td>\n",
       "      <td>80948898.0</td>\n",
       "      <td>80957870.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>CHS.18044</td>\n",
       "      <td>80975744</td>\n",
       "      <td>80989579</td>\n",
       "      <td>80979477</td>\n",
       "      <td>80989878</td>\n",
       "      <td>CHS.18044.5</td>\n",
       "      <td>80947327.0</td>\n",
       "      <td>80959152.0</td>\n",
       "      <td>80948898.0</td>\n",
       "      <td>80959508.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>CHS.18289</td>\n",
       "      <td>92617448</td>\n",
       "      <td>92667898</td>\n",
       "      <td>92619459</td>\n",
       "      <td>92668043</td>\n",
       "      <td>CHS.18289.19</td>\n",
       "      <td>92714674.0</td>\n",
       "      <td>92733985.0</td>\n",
       "      <td>92714822.0</td>\n",
       "      <td>92734218.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>CHS.18289</td>\n",
       "      <td>92617448</td>\n",
       "      <td>92667898</td>\n",
       "      <td>92619459</td>\n",
       "      <td>92668043</td>\n",
       "      <td>CHS.18289.21</td>\n",
       "      <td>92728899.0</td>\n",
       "      <td>92733985.0</td>\n",
       "      <td>92728905.0</td>\n",
       "      <td>92734213.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>CHS.18502</td>\n",
       "      <td>5553</td>\n",
       "      <td>357560</td>\n",
       "      <td>5791</td>\n",
       "      <td>357692</td>\n",
       "      <td>CHS.18502.6</td>\n",
       "      <td>850.0</td>\n",
       "      <td>4255.0</td>\n",
       "      <td>953.0</td>\n",
       "      <td>5052.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>CHS.19065</td>\n",
       "      <td>10528422</td>\n",
       "      <td>10580549</td>\n",
       "      <td>10533092</td>\n",
       "      <td>10580716</td>\n",
       "      <td>CHS.19065.14</td>\n",
       "      <td>10522391.0</td>\n",
       "      <td>10528092.0</td>\n",
       "      <td>10522438.0</td>\n",
       "      <td>10528421.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>CHS.20010</td>\n",
       "      <td>58392394</td>\n",
       "      <td>58404499</td>\n",
       "      <td>58392787</td>\n",
       "      <td>58406144</td>\n",
       "      <td>CHS.20010.12</td>\n",
       "      <td>58295080.0</td>\n",
       "      <td>58295391.0</td>\n",
       "      <td>58295309.0</td>\n",
       "      <td>58295817.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>CHS.20010</td>\n",
       "      <td>58392394</td>\n",
       "      <td>58404499</td>\n",
       "      <td>58392787</td>\n",
       "      <td>58406144</td>\n",
       "      <td>CHS.20010.14</td>\n",
       "      <td>58295327.0</td>\n",
       "      <td>58329863.0</td>\n",
       "      <td>58295534.0</td>\n",
       "      <td>58329891.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>CHS.20405</td>\n",
       "      <td>79568021</td>\n",
       "      <td>79598785</td>\n",
       "      <td>79571350</td>\n",
       "      <td>79600725</td>\n",
       "      <td>CHS.20405.2</td>\n",
       "      <td>79202625.0</td>\n",
       "      <td>79229310.0</td>\n",
       "      <td>79209607.0</td>\n",
       "      <td>79229589.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>206</th>\n",
       "      <td>CHS.55637</td>\n",
       "      <td>65144195</td>\n",
       "      <td>65194706</td>\n",
       "      <td>65144438</td>\n",
       "      <td>65195538</td>\n",
       "      <td>CHS.55637.14</td>\n",
       "      <td>65196261.0</td>\n",
       "      <td>65271228.0</td>\n",
       "      <td>65196570.0</td>\n",
       "      <td>65275018.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>207</th>\n",
       "      <td>CHS.56081</td>\n",
       "      <td>92844883</td>\n",
       "      <td>92877890</td>\n",
       "      <td>92848551</td>\n",
       "      <td>92878060</td>\n",
       "      <td>CHS.56081.18</td>\n",
       "      <td>92829164.0</td>\n",
       "      <td>92831379.0</td>\n",
       "      <td>92831112.0</td>\n",
       "      <td>92834261.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>208</th>\n",
       "      <td>CHS.56429</td>\n",
       "      <td>109136188</td>\n",
       "      <td>109166901</td>\n",
       "      <td>109137627</td>\n",
       "      <td>109167291</td>\n",
       "      <td>CHS.56429.13</td>\n",
       "      <td>109129684.0</td>\n",
       "      <td>109130141.0</td>\n",
       "      <td>109129723.0</td>\n",
       "      <td>109135968.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>209</th>\n",
       "      <td>CHS.5648</td>\n",
       "      <td>18545207</td>\n",
       "      <td>18659232</td>\n",
       "      <td>18546145</td>\n",
       "      <td>18659264</td>\n",
       "      <td>CHS.5648.2</td>\n",
       "      <td>18512583.0</td>\n",
       "      <td>18535513.0</td>\n",
       "      <td>18513660.0</td>\n",
       "      <td>18536097.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>210</th>\n",
       "      <td>CHS.56649</td>\n",
       "      <td>121178137</td>\n",
       "      <td>121201639</td>\n",
       "      <td>121181573</td>\n",
       "      <td>121202087</td>\n",
       "      <td>CHS.56649.11</td>\n",
       "      <td>121211074.0</td>\n",
       "      <td>121222910.0</td>\n",
       "      <td>121214330.0</td>\n",
       "      <td>121223019.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>211</th>\n",
       "      <td>CHS.58072</td>\n",
       "      <td>53176273</td>\n",
       "      <td>53224740</td>\n",
       "      <td>53176622</td>\n",
       "      <td>53225422</td>\n",
       "      <td>CHS.58072.1</td>\n",
       "      <td>53150008.0</td>\n",
       "      <td>53163263.0</td>\n",
       "      <td>53159806.0</td>\n",
       "      <td>53163496.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>212</th>\n",
       "      <td>CHS.58072</td>\n",
       "      <td>53176273</td>\n",
       "      <td>53224740</td>\n",
       "      <td>53176622</td>\n",
       "      <td>53225422</td>\n",
       "      <td>CHS.58072.2</td>\n",
       "      <td>53152587.0</td>\n",
       "      <td>53163263.0</td>\n",
       "      <td>53153353.0</td>\n",
       "      <td>53172721.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>213</th>\n",
       "      <td>CHS.58432</td>\n",
       "      <td>101488044</td>\n",
       "      <td>101488444</td>\n",
       "      <td>101488099</td>\n",
       "      <td>101495811</td>\n",
       "      <td>CHS.58432.1</td>\n",
       "      <td>101417918.0</td>\n",
       "      <td>101447494.0</td>\n",
       "      <td>101418141.0</td>\n",
       "      <td>101448891.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>214</th>\n",
       "      <td>CHS.58432</td>\n",
       "      <td>101488044</td>\n",
       "      <td>101488444</td>\n",
       "      <td>101488099</td>\n",
       "      <td>101495811</td>\n",
       "      <td>CHS.58432.10</td>\n",
       "      <td>101418280.0</td>\n",
       "      <td>101447494.0</td>\n",
       "      <td>101418370.0</td>\n",
       "      <td>101456917.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>215</th>\n",
       "      <td>CHS.58432</td>\n",
       "      <td>101488044</td>\n",
       "      <td>101488444</td>\n",
       "      <td>101488099</td>\n",
       "      <td>101495811</td>\n",
       "      <td>CHS.58432.3</td>\n",
       "      <td>101418264.0</td>\n",
       "      <td>101447083.0</td>\n",
       "      <td>101418370.0</td>\n",
       "      <td>101457856.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>216</th>\n",
       "      <td>CHS.5862</td>\n",
       "      <td>30844634</td>\n",
       "      <td>31032171</td>\n",
       "      <td>30845573</td>\n",
       "      <td>31032406</td>\n",
       "      <td>CHS.5862.62</td>\n",
       "      <td>30820096.0</td>\n",
       "      <td>30821250.0</td>\n",
       "      <td>30820444.0</td>\n",
       "      <td>30821313.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>217</th>\n",
       "      <td>CHS.58841</td>\n",
       "      <td>135285787</td>\n",
       "      <td>135343261</td>\n",
       "      <td>135287846</td>\n",
       "      <td>135344109</td>\n",
       "      <td>CHS.58841.2</td>\n",
       "      <td>135248590.0</td>\n",
       "      <td>135255612.0</td>\n",
       "      <td>135249301.0</td>\n",
       "      <td>135255671.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>218</th>\n",
       "      <td>CHS.58841</td>\n",
       "      <td>135285787</td>\n",
       "      <td>135343261</td>\n",
       "      <td>135287846</td>\n",
       "      <td>135344109</td>\n",
       "      <td>CHS.58841.4</td>\n",
       "      <td>135248941.0</td>\n",
       "      <td>135252299.0</td>\n",
       "      <td>135249301.0</td>\n",
       "      <td>135252530.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>219</th>\n",
       "      <td>CHS.58841</td>\n",
       "      <td>135285787</td>\n",
       "      <td>135343261</td>\n",
       "      <td>135287846</td>\n",
       "      <td>135344109</td>\n",
       "      <td>CHS.58841.5</td>\n",
       "      <td>135248942.0</td>\n",
       "      <td>135276075.0</td>\n",
       "      <td>135249301.0</td>\n",
       "      <td>135276095.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>220</th>\n",
       "      <td>CHS.58841</td>\n",
       "      <td>135285787</td>\n",
       "      <td>135343261</td>\n",
       "      <td>135287846</td>\n",
       "      <td>135344109</td>\n",
       "      <td>CHS.58841.6</td>\n",
       "      <td>135248945.0</td>\n",
       "      <td>135255268.0</td>\n",
       "      <td>135249301.0</td>\n",
       "      <td>135255442.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>221</th>\n",
       "      <td>CHS.6543</td>\n",
       "      <td>75397830</td>\n",
       "      <td>75401105</td>\n",
       "      <td>75400374</td>\n",
       "      <td>75401894</td>\n",
       "      <td>CHS.6543.2</td>\n",
       "      <td>75279528.0</td>\n",
       "      <td>75284247.0</td>\n",
       "      <td>75279911.0</td>\n",
       "      <td>75284331.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>222</th>\n",
       "      <td>CHS.6867</td>\n",
       "      <td>94545767</td>\n",
       "      <td>94601528</td>\n",
       "      <td>94545907</td>\n",
       "      <td>94602099</td>\n",
       "      <td>CHS.6867.19</td>\n",
       "      <td>94604801.0</td>\n",
       "      <td>94609853.0</td>\n",
       "      <td>94608026.0</td>\n",
       "      <td>94613935.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>223</th>\n",
       "      <td>CHS.7412</td>\n",
       "      <td>122980040</td>\n",
       "      <td>122990174</td>\n",
       "      <td>122980695</td>\n",
       "      <td>122990407</td>\n",
       "      <td>CHS.7412.14</td>\n",
       "      <td>122954458.0</td>\n",
       "      <td>122964406.0</td>\n",
       "      <td>122954568.0</td>\n",
       "      <td>122978608.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>224</th>\n",
       "      <td>CHS.7618</td>\n",
       "      <td>133278630</td>\n",
       "      <td>133308823</td>\n",
       "      <td>133279901</td>\n",
       "      <td>133309156</td>\n",
       "      <td>CHS.7618.20</td>\n",
       "      <td>133309910.0</td>\n",
       "      <td>133311620.0</td>\n",
       "      <td>133309960.0</td>\n",
       "      <td>133312336.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>225</th>\n",
       "      <td>CHS.7618</td>\n",
       "      <td>133278630</td>\n",
       "      <td>133308823</td>\n",
       "      <td>133279901</td>\n",
       "      <td>133309156</td>\n",
       "      <td>CHS.7618.21</td>\n",
       "      <td>133311842.0</td>\n",
       "      <td>133312194.0</td>\n",
       "      <td>133312166.0</td>\n",
       "      <td>133312336.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>226</th>\n",
       "      <td>CHS.7925</td>\n",
       "      <td>5645442</td>\n",
       "      <td>5685016</td>\n",
       "      <td>5645863</td>\n",
       "      <td>5685109</td>\n",
       "      <td>CHS.7925.1</td>\n",
       "      <td>5586300.0</td>\n",
       "      <td>5624826.0</td>\n",
       "      <td>5591023.0</td>\n",
       "      <td>5624853.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>227</th>\n",
       "      <td>CHS.8020</td>\n",
       "      <td>8104931</td>\n",
       "      <td>8168866</td>\n",
       "      <td>8110407</td>\n",
       "      <td>8169088</td>\n",
       "      <td>CHS.8020.1</td>\n",
       "      <td>8092965.0</td>\n",
       "      <td>8094769.0</td>\n",
       "      <td>8094159.0</td>\n",
       "      <td>8097209.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>228</th>\n",
       "      <td>CHS.8020</td>\n",
       "      <td>8104931</td>\n",
       "      <td>8168866</td>\n",
       "      <td>8110407</td>\n",
       "      <td>8169088</td>\n",
       "      <td>CHS.8020.14</td>\n",
       "      <td>8092967.0</td>\n",
       "      <td>8094738.0</td>\n",
       "      <td>8094159.0</td>\n",
       "      <td>8097217.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>229</th>\n",
       "      <td>CHS.8020</td>\n",
       "      <td>8104931</td>\n",
       "      <td>8168866</td>\n",
       "      <td>8110407</td>\n",
       "      <td>8169088</td>\n",
       "      <td>CHS.8020.6</td>\n",
       "      <td>8092966.0</td>\n",
       "      <td>8098460.0</td>\n",
       "      <td>8094159.0</td>\n",
       "      <td>8098589.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>230</th>\n",
       "      <td>CHS.8889</td>\n",
       "      <td>61430042</td>\n",
       "      <td>61445941</td>\n",
       "      <td>61430182</td>\n",
       "      <td>61446767</td>\n",
       "      <td>CHS.8889.22</td>\n",
       "      <td>61454522.0</td>\n",
       "      <td>61472002.0</td>\n",
       "      <td>61455205.0</td>\n",
       "      <td>61474335.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>231</th>\n",
       "      <td>CHS.8961</td>\n",
       "      <td>62665252</td>\n",
       "      <td>62666396</td>\n",
       "      <td>62665332</td>\n",
       "      <td>62667451</td>\n",
       "      <td>CHS.8961.7</td>\n",
       "      <td>62670042.0</td>\n",
       "      <td>62671284.0</td>\n",
       "      <td>62670353.0</td>\n",
       "      <td>62671563.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>232</th>\n",
       "      <td>CHS.9218</td>\n",
       "      <td>66435075</td>\n",
       "      <td>66438710</td>\n",
       "      <td>66436112</td>\n",
       "      <td>66438860</td>\n",
       "      <td>CHS.9218.12</td>\n",
       "      <td>66461269.0</td>\n",
       "      <td>66465913.0</td>\n",
       "      <td>66465234.0</td>\n",
       "      <td>66466712.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>233</th>\n",
       "      <td>CHS.9672</td>\n",
       "      <td>86791059</td>\n",
       "      <td>86807884</td>\n",
       "      <td>86791195</td>\n",
       "      <td>86811233</td>\n",
       "      <td>CHS.9672.12</td>\n",
       "      <td>86823421.0</td>\n",
       "      <td>86951216.0</td>\n",
       "      <td>86823593.0</td>\n",
       "      <td>86953207.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>234</th>\n",
       "      <td>CHS.9672</td>\n",
       "      <td>86791059</td>\n",
       "      <td>86807884</td>\n",
       "      <td>86791195</td>\n",
       "      <td>86811233</td>\n",
       "      <td>CHS.9672.13</td>\n",
       "      <td>86920393.0</td>\n",
       "      <td>86951216.0</td>\n",
       "      <td>86920465.0</td>\n",
       "      <td>86952909.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>235</th>\n",
       "      <td>CHS.9672</td>\n",
       "      <td>86791059</td>\n",
       "      <td>86807884</td>\n",
       "      <td>86791195</td>\n",
       "      <td>86811233</td>\n",
       "      <td>CHS.9672.16</td>\n",
       "      <td>86940785.0</td>\n",
       "      <td>86951216.0</td>\n",
       "      <td>86941307.0</td>\n",
       "      <td>86951765.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>236 rows × 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        geneID  smin_known  smax_known  emin_known  emax_known        parent  \\\n",
       "0    CHS.10904     8950044     9215276     8950236     9215657  CHS.10904.15   \n",
       "1    CHS.10904     8950044     9215276     8950236     9215657  CHS.10904.17   \n",
       "2    CHS.11417    42326126    42446592    42326229    42448621  CHS.11417.44   \n",
       "3    CHS.11615    50979783    51028511    50981803    51028566   CHS.11615.3   \n",
       "4    CHS.11626    51238724    51246061    51238920    51246717  CHS.11626.25   \n",
       "5    CHS.11936    57434685    57449277    57435199    57452062   CHS.11936.5   \n",
       "6    CHS.12135    69359710    69390147    69360023    69390870   CHS.12135.8   \n",
       "7    CHS.12461    95217746    95300165    95217886    95302790  CHS.12461.67   \n",
       "8    CHS.13013   121400083   121423386   121400218   121424352  CHS.13013.16   \n",
       "9    CHS.14025    48075728    48095060    48077311    48095141   CHS.14025.8   \n",
       "10   CHS.14448    87671328    87675381    87672209    87679619   CHS.14448.6   \n",
       "11   CHS.14806   113177539   113208599   113177976   113209466   CHS.14806.2   \n",
       "12   CHS.15061    23555982    23558695    23556343    23559581   CHS.15061.8   \n",
       "13   CHS.16509   104085658   104112524   104085852   104112749  CHS.16509.35   \n",
       "14   CHS.16509   104085658   104112524   104085852   104112749  CHS.16509.36   \n",
       "15   CHS.16509   104085658   104112524   104085852   104112749  CHS.16509.37   \n",
       "16   CHS.16971    34379068    34407524    34381613    34407674  CHS.16971.19   \n",
       "17   CHS.17277    45631148    45690973    45631230    45691294  CHS.17277.19   \n",
       "18   CHS.17900    75598083    75626239    75598681    75626378  CHS.17900.25   \n",
       "19   CHS.17960    78167468    78234371    78168485    78234707  CHS.17960.16   \n",
       "20   CHS.18044    80975744    80989579    80979477    80989878   CHS.18044.3   \n",
       "21   CHS.18044    80975744    80989579    80979477    80989878   CHS.18044.4   \n",
       "22   CHS.18044    80975744    80989579    80979477    80989878   CHS.18044.5   \n",
       "23   CHS.18289    92617448    92667898    92619459    92668043  CHS.18289.19   \n",
       "24   CHS.18289    92617448    92667898    92619459    92668043  CHS.18289.21   \n",
       "25   CHS.18502        5553      357560        5791      357692   CHS.18502.6   \n",
       "26   CHS.19065    10528422    10580549    10533092    10580716  CHS.19065.14   \n",
       "27   CHS.20010    58392394    58404499    58392787    58406144  CHS.20010.12   \n",
       "28   CHS.20010    58392394    58404499    58392787    58406144  CHS.20010.14   \n",
       "29   CHS.20405    79568021    79598785    79571350    79600725   CHS.20405.2   \n",
       "..         ...         ...         ...         ...         ...           ...   \n",
       "206  CHS.55637    65144195    65194706    65144438    65195538  CHS.55637.14   \n",
       "207  CHS.56081    92844883    92877890    92848551    92878060  CHS.56081.18   \n",
       "208  CHS.56429   109136188   109166901   109137627   109167291  CHS.56429.13   \n",
       "209   CHS.5648    18545207    18659232    18546145    18659264    CHS.5648.2   \n",
       "210  CHS.56649   121178137   121201639   121181573   121202087  CHS.56649.11   \n",
       "211  CHS.58072    53176273    53224740    53176622    53225422   CHS.58072.1   \n",
       "212  CHS.58072    53176273    53224740    53176622    53225422   CHS.58072.2   \n",
       "213  CHS.58432   101488044   101488444   101488099   101495811   CHS.58432.1   \n",
       "214  CHS.58432   101488044   101488444   101488099   101495811  CHS.58432.10   \n",
       "215  CHS.58432   101488044   101488444   101488099   101495811   CHS.58432.3   \n",
       "216   CHS.5862    30844634    31032171    30845573    31032406   CHS.5862.62   \n",
       "217  CHS.58841   135285787   135343261   135287846   135344109   CHS.58841.2   \n",
       "218  CHS.58841   135285787   135343261   135287846   135344109   CHS.58841.4   \n",
       "219  CHS.58841   135285787   135343261   135287846   135344109   CHS.58841.5   \n",
       "220  CHS.58841   135285787   135343261   135287846   135344109   CHS.58841.6   \n",
       "221   CHS.6543    75397830    75401105    75400374    75401894    CHS.6543.2   \n",
       "222   CHS.6867    94545767    94601528    94545907    94602099   CHS.6867.19   \n",
       "223   CHS.7412   122980040   122990174   122980695   122990407   CHS.7412.14   \n",
       "224   CHS.7618   133278630   133308823   133279901   133309156   CHS.7618.20   \n",
       "225   CHS.7618   133278630   133308823   133279901   133309156   CHS.7618.21   \n",
       "226   CHS.7925     5645442     5685016     5645863     5685109    CHS.7925.1   \n",
       "227   CHS.8020     8104931     8168866     8110407     8169088    CHS.8020.1   \n",
       "228   CHS.8020     8104931     8168866     8110407     8169088   CHS.8020.14   \n",
       "229   CHS.8020     8104931     8168866     8110407     8169088    CHS.8020.6   \n",
       "230   CHS.8889    61430042    61445941    61430182    61446767   CHS.8889.22   \n",
       "231   CHS.8961    62665252    62666396    62665332    62667451    CHS.8961.7   \n",
       "232   CHS.9218    66435075    66438710    66436112    66438860   CHS.9218.12   \n",
       "233   CHS.9672    86791059    86807884    86791195    86811233   CHS.9672.12   \n",
       "234   CHS.9672    86791059    86807884    86791195    86811233   CHS.9672.13   \n",
       "235   CHS.9672    86791059    86807884    86791195    86811233   CHS.9672.16   \n",
       "\n",
       "      smin_novel   smax_novel   emin_novel   emax_novel  \n",
       "0      9244635.0    9247581.0    9246548.0    9247608.0  \n",
       "1      9244635.0    9260614.0    9246548.0    9263160.0  \n",
       "2     42238469.0   42239137.0   42238727.0   42241129.0  \n",
       "3     50952264.0   50969390.0   50953808.0   50970219.0  \n",
       "4     51247766.0   51248479.0   51247978.0   51249650.0  \n",
       "5     57431191.0   57433703.0   57431967.0   57434408.0  \n",
       "6     69416433.0   69424001.0   69416627.0   69424837.0  \n",
       "7     95305942.0   95306937.0   95306028.0   95310639.0  \n",
       "8    121375425.0  121391970.0  121375816.0  121392070.0  \n",
       "9     48053331.0   48058783.0   48053491.0   48059622.0  \n",
       "10    87679888.0   87695419.0   87691235.0   87696416.0  \n",
       "11   113165003.0  113175378.0  113165183.0  113176331.0  \n",
       "12    23513207.0   23518865.0   23518396.0   23520831.0  \n",
       "13   104117141.0  104117482.0  104117252.0  104128281.0  \n",
       "14   104117179.0  104117390.0  104117252.0  104126141.0  \n",
       "15   104117501.0  104119835.0  104119798.0  104125736.0  \n",
       "16    34415382.0   34437398.0   34419343.0   34437465.0  \n",
       "17    45697041.0   45701436.0   45699635.0   45703171.0  \n",
       "18    75588978.0   75593717.0   75592470.0   75596585.0  \n",
       "19    78244203.0   78245269.0   78244853.0   78245686.0  \n",
       "20    80947327.0   80951120.0   80948898.0   80951765.0  \n",
       "21    80947327.0   80957473.0   80948898.0   80957870.0  \n",
       "22    80947327.0   80959152.0   80948898.0   80959508.0  \n",
       "23    92714674.0   92733985.0   92714822.0   92734218.0  \n",
       "24    92728899.0   92733985.0   92728905.0   92734213.0  \n",
       "25         850.0       4255.0        953.0       5052.0  \n",
       "26    10522391.0   10528092.0   10522438.0   10528421.0  \n",
       "27    58295080.0   58295391.0   58295309.0   58295817.0  \n",
       "28    58295327.0   58329863.0   58295534.0   58329891.0  \n",
       "29    79202625.0   79229310.0   79209607.0   79229589.0  \n",
       "..           ...          ...          ...          ...  \n",
       "206   65196261.0   65271228.0   65196570.0   65275018.0  \n",
       "207   92829164.0   92831379.0   92831112.0   92834261.0  \n",
       "208  109129684.0  109130141.0  109129723.0  109135968.0  \n",
       "209   18512583.0   18535513.0   18513660.0   18536097.0  \n",
       "210  121211074.0  121222910.0  121214330.0  121223019.0  \n",
       "211   53150008.0   53163263.0   53159806.0   53163496.0  \n",
       "212   53152587.0   53163263.0   53153353.0   53172721.0  \n",
       "213  101417918.0  101447494.0  101418141.0  101448891.0  \n",
       "214  101418280.0  101447494.0  101418370.0  101456917.0  \n",
       "215  101418264.0  101447083.0  101418370.0  101457856.0  \n",
       "216   30820096.0   30821250.0   30820444.0   30821313.0  \n",
       "217  135248590.0  135255612.0  135249301.0  135255671.0  \n",
       "218  135248941.0  135252299.0  135249301.0  135252530.0  \n",
       "219  135248942.0  135276075.0  135249301.0  135276095.0  \n",
       "220  135248945.0  135255268.0  135249301.0  135255442.0  \n",
       "221   75279528.0   75284247.0   75279911.0   75284331.0  \n",
       "222   94604801.0   94609853.0   94608026.0   94613935.0  \n",
       "223  122954458.0  122964406.0  122954568.0  122978608.0  \n",
       "224  133309910.0  133311620.0  133309960.0  133312336.0  \n",
       "225  133311842.0  133312194.0  133312166.0  133312336.0  \n",
       "226    5586300.0    5624826.0    5591023.0    5624853.0  \n",
       "227    8092965.0    8094769.0    8094159.0    8097209.0  \n",
       "228    8092967.0    8094738.0    8094159.0    8097217.0  \n",
       "229    8092966.0    8098460.0    8094159.0    8098589.0  \n",
       "230   61454522.0   61472002.0   61455205.0   61474335.0  \n",
       "231   62670042.0   62671284.0   62670353.0   62671563.0  \n",
       "232   66461269.0   66465913.0   66465234.0   66466712.0  \n",
       "233   86823421.0   86951216.0   86823593.0   86953207.0  \n",
       "234   86920393.0   86951216.0   86920465.0   86952909.0  \n",
       "235   86940785.0   86951216.0   86941307.0   86951765.0  \n",
       "\n",
       "[236 rows x 10 columns]"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# and finally we shall use this information to write a query that will answer our question\n",
    "# which novel isoforms are incompatible with the known exon coordinates\n",
    "# basically, which ones are outside any known annotation\n",
    "finalRes=resDF[(resDF['emax_novel']<resDF['smin_known'])|(resDF['smin_novel']>resDF['emax_known'])].reset_index(drop=True)\n",
    "finalRes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "236\n",
      "165\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>geneID</th>\n",
       "      <th>smin_known</th>\n",
       "      <th>smax_known</th>\n",
       "      <th>emin_known</th>\n",
       "      <th>emax_known</th>\n",
       "      <th>parent</th>\n",
       "      <th>smin_novel</th>\n",
       "      <th>smax_novel</th>\n",
       "      <th>emin_novel</th>\n",
       "      <th>emax_novel</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>115</th>\n",
       "      <td>CHS.36728</td>\n",
       "      <td>17962572</td>\n",
       "      <td>18267018</td>\n",
       "      <td>17962694</td>\n",
       "      <td>18268922</td>\n",
       "      <td>CHS.36728.3</td>\n",
       "      <td>18330788.0</td>\n",
       "      <td>18342247.0</td>\n",
       "      <td>18330962.0</td>\n",
       "      <td>18344196.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        geneID  smin_known  smax_known  emin_known  emax_known       parent  \\\n",
       "115  CHS.36728    17962572    18267018    17962694    18268922  CHS.36728.3   \n",
       "\n",
       "     smin_novel  smax_novel  emin_novel  emax_novel  \n",
       "115  18330788.0  18342247.0  18330962.0  18344196.0  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>geneID</th>\n",
       "      <th>smin_known</th>\n",
       "      <th>smax_known</th>\n",
       "      <th>emin_known</th>\n",
       "      <th>emax_known</th>\n",
       "      <th>parent</th>\n",
       "      <th>smin_novel</th>\n",
       "      <th>smax_novel</th>\n",
       "      <th>emin_novel</th>\n",
       "      <th>emax_novel</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>177</th>\n",
       "      <td>CHS.51125</td>\n",
       "      <td>88759368</td>\n",
       "      <td>89333363</td>\n",
       "      <td>88760084</td>\n",
       "      <td>89337057</td>\n",
       "      <td>CHS.51125.3</td>\n",
       "      <td>88758022.0</td>\n",
       "      <td>88758970.0</td>\n",
       "      <td>88758521.0</td>\n",
       "      <td>88759367.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        geneID  smin_known  smax_known  emin_known  emax_known       parent  \\\n",
       "177  CHS.51125    88759368    89333363    88760084    89337057  CHS.51125.3   \n",
       "\n",
       "     smin_novel  smax_novel  emin_novel  emax_novel  \n",
       "177  88758022.0  88758970.0  88758521.0  88759367.0  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "print(len(set(finalRes[\"parent\"])))\n",
    "print(len(set(finalRes[\"geneID\"])))\n",
    "display(finalRes[finalRes[\"parent\"]==\"CHS.36728.3\"])\n",
    "display(finalRes[finalRes['parent']=='CHS.51125.3'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>geneID</th>\n",
       "      <th>smin_known</th>\n",
       "      <th>smax_known</th>\n",
       "      <th>emin_known</th>\n",
       "      <th>emax_known</th>\n",
       "      <th>parent</th>\n",
       "      <th>smin_novel</th>\n",
       "      <th>smax_novel</th>\n",
       "      <th>emin_novel</th>\n",
       "      <th>emax_novel</th>\n",
       "      <th>dist</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CHS.10904</td>\n",
       "      <td>8950044</td>\n",
       "      <td>9215276</td>\n",
       "      <td>8950236</td>\n",
       "      <td>9215657</td>\n",
       "      <td>CHS.10904.15</td>\n",
       "      <td>9244635.0</td>\n",
       "      <td>9247581.0</td>\n",
       "      <td>9246548.0</td>\n",
       "      <td>9247608.0</td>\n",
       "      <td>28978.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CHS.10904</td>\n",
       "      <td>8950044</td>\n",
       "      <td>9215276</td>\n",
       "      <td>8950236</td>\n",
       "      <td>9215657</td>\n",
       "      <td>CHS.10904.17</td>\n",
       "      <td>9244635.0</td>\n",
       "      <td>9260614.0</td>\n",
       "      <td>9246548.0</td>\n",
       "      <td>9263160.0</td>\n",
       "      <td>28978.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CHS.11417</td>\n",
       "      <td>42326126</td>\n",
       "      <td>42446592</td>\n",
       "      <td>42326229</td>\n",
       "      <td>42448621</td>\n",
       "      <td>CHS.11417.44</td>\n",
       "      <td>42238469.0</td>\n",
       "      <td>42239137.0</td>\n",
       "      <td>42238727.0</td>\n",
       "      <td>42241129.0</td>\n",
       "      <td>84997.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CHS.11615</td>\n",
       "      <td>50979783</td>\n",
       "      <td>51028511</td>\n",
       "      <td>50981803</td>\n",
       "      <td>51028566</td>\n",
       "      <td>CHS.11615.3</td>\n",
       "      <td>50952264.0</td>\n",
       "      <td>50969390.0</td>\n",
       "      <td>50953808.0</td>\n",
       "      <td>50970219.0</td>\n",
       "      <td>9564.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CHS.11626</td>\n",
       "      <td>51238724</td>\n",
       "      <td>51246061</td>\n",
       "      <td>51238920</td>\n",
       "      <td>51246717</td>\n",
       "      <td>CHS.11626.25</td>\n",
       "      <td>51247766.0</td>\n",
       "      <td>51248479.0</td>\n",
       "      <td>51247978.0</td>\n",
       "      <td>51249650.0</td>\n",
       "      <td>1049.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      geneID  smin_known  smax_known  emin_known  emax_known        parent  \\\n",
       "0  CHS.10904     8950044     9215276     8950236     9215657  CHS.10904.15   \n",
       "1  CHS.10904     8950044     9215276     8950236     9215657  CHS.10904.17   \n",
       "2  CHS.11417    42326126    42446592    42326229    42448621  CHS.11417.44   \n",
       "3  CHS.11615    50979783    51028511    50981803    51028566   CHS.11615.3   \n",
       "4  CHS.11626    51238724    51246061    51238920    51246717  CHS.11626.25   \n",
       "\n",
       "   smin_novel  smax_novel  emin_novel  emax_novel     dist  \n",
       "0   9244635.0   9247581.0   9246548.0   9247608.0  28978.0  \n",
       "1   9244635.0   9260614.0   9246548.0   9263160.0  28978.0  \n",
       "2  42238469.0  42239137.0  42238727.0  42241129.0  84997.0  \n",
       "3  50952264.0  50969390.0  50953808.0  50970219.0   9564.0  \n",
       "4  51247766.0  51248479.0  51247978.0  51249650.0   1049.0  "
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# lastly, what is the maximum distance from the novel to the known\n",
    "finalRes[\"dist\"]=np.where(finalRes['emax_novel']<finalRes['smin_known'],finalRes['smin_known']-finalRes['emax_novel'],finalRes['smin_novel']-finalRes['emax_known'])\n",
    "finalRes.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>geneID</th>\n",
       "      <th>smin_known</th>\n",
       "      <th>smax_known</th>\n",
       "      <th>emin_known</th>\n",
       "      <th>emax_known</th>\n",
       "      <th>parent</th>\n",
       "      <th>smin_novel</th>\n",
       "      <th>smax_novel</th>\n",
       "      <th>emin_novel</th>\n",
       "      <th>emax_novel</th>\n",
       "      <th>dist</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CHS.20405</td>\n",
       "      <td>79568021</td>\n",
       "      <td>79598785</td>\n",
       "      <td>79571350</td>\n",
       "      <td>79600725</td>\n",
       "      <td>CHS.20405.5</td>\n",
       "      <td>79212668.0</td>\n",
       "      <td>79218845.0</td>\n",
       "      <td>79212901.0</td>\n",
       "      <td>79219521.0</td>\n",
       "      <td>348500.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CHS.20405</td>\n",
       "      <td>79568021</td>\n",
       "      <td>79598785</td>\n",
       "      <td>79571350</td>\n",
       "      <td>79600725</td>\n",
       "      <td>CHS.20405.2</td>\n",
       "      <td>79202625.0</td>\n",
       "      <td>79229310.0</td>\n",
       "      <td>79209607.0</td>\n",
       "      <td>79229589.0</td>\n",
       "      <td>338432.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CHS.20405</td>\n",
       "      <td>79568021</td>\n",
       "      <td>79598785</td>\n",
       "      <td>79571350</td>\n",
       "      <td>79600725</td>\n",
       "      <td>CHS.20405.3</td>\n",
       "      <td>79202625.0</td>\n",
       "      <td>79229310.0</td>\n",
       "      <td>79209607.0</td>\n",
       "      <td>79229608.0</td>\n",
       "      <td>338413.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CHS.36195</td>\n",
       "      <td>49614495</td>\n",
       "      <td>49656624</td>\n",
       "      <td>49615490</td>\n",
       "      <td>49657542</td>\n",
       "      <td>CHS.36195.5</td>\n",
       "      <td>49435804.0</td>\n",
       "      <td>49450194.0</td>\n",
       "      <td>49441212.0</td>\n",
       "      <td>49450273.0</td>\n",
       "      <td>164222.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CHS.9672</td>\n",
       "      <td>86791059</td>\n",
       "      <td>86807884</td>\n",
       "      <td>86791195</td>\n",
       "      <td>86811233</td>\n",
       "      <td>CHS.9672.16</td>\n",
       "      <td>86940785.0</td>\n",
       "      <td>86951216.0</td>\n",
       "      <td>86941307.0</td>\n",
       "      <td>86951765.0</td>\n",
       "      <td>129552.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      geneID  smin_known  smax_known  emin_known  emax_known       parent  \\\n",
       "0  CHS.20405    79568021    79598785    79571350    79600725  CHS.20405.5   \n",
       "1  CHS.20405    79568021    79598785    79571350    79600725  CHS.20405.2   \n",
       "2  CHS.20405    79568021    79598785    79571350    79600725  CHS.20405.3   \n",
       "3  CHS.36195    49614495    49656624    49615490    49657542  CHS.36195.5   \n",
       "4   CHS.9672    86791059    86807884    86791195    86811233  CHS.9672.16   \n",
       "\n",
       "   smin_novel  smax_novel  emin_novel  emax_novel      dist  \n",
       "0  79212668.0  79218845.0  79212901.0  79219521.0  348500.0  \n",
       "1  79202625.0  79229310.0  79209607.0  79229589.0  338432.0  \n",
       "2  79202625.0  79229310.0  79209607.0  79229608.0  338413.0  \n",
       "3  49435804.0  49450194.0  49441212.0  49450273.0  164222.0  \n",
       "4  86940785.0  86951216.0  86941307.0  86951765.0  129552.0  "
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "finalRes.sort_values(by='dist',ascending=False,inplace=True)\n",
    "finalRes.reset_index(drop=True,inplace=True)\n",
    "finalRes.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<matplotlib.axes._subplots.AxesSubplot at 0x7f30c9c35748>"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAXwAAAD7CAYAAABpJS8eAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADl0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uIDIuMi4yLCBodHRwOi8vbWF0cGxvdGxpYi5vcmcvhp/UCwAAFXhJREFUeJzt3X+U3XV95/HnW0GJDEGy4MCGllRA1x9ZYGfYniwgd5buUk05W+1uWU9glyN1drsulSIK0shSN9VQE380W6up9g8lu7MtnJ64ppZaZaLLoUrSVbP+oqCx6tEMmhIYjBTwvX/cbw6XcYbc+X7n3jt3P8/HOTnc72e+P17feyevfPl+v/feyEwkSf//e9agA0iS+sPCl6RCWPiSVAgLX5IKYeFLUiEsfEkqhIUvSYWw8CWpEBa+JBXimEEHOOLkk0/ONWvW1Fr20Ucf5fjjj1/aQD0yTFnBvL00TFnBvL1UN+vevXt/kJmndL1AZi6LP2NjY1nXXXfdVXvZfhumrJnm7aVhyppp3l6qmxXYk4voWU/pSFIhLHxJKoSFL0mFsPAlqRAWviQVwsKXpEJY+JJUCAtfkgph4UtSIZbNRys0se+7h7jqxl21lt2/ef0Sp5Gk5ckjfEkqhIUvSYWw8CWpEF2dw4+IU4FNwDmZeX41FsA11SxrgOdn5uuqn70ZWAmcBPxFZn5siXNLkhap24u2FwI7gXM7xq4AHsrMjwBExD+u/vvzwERmvioijgW+EhGfycyHljC3JGmRujqlk5m3A4/MGd4ArIqI34iIdwCz1fgvAfdUyz0OfBV4xdLElSTVFe3P0O9ixogWsCUzx6vprwL/IzPfHhEvAj4OvAR4P/DVzHxvNd9twHRmfmiedU4CkwCjo6NjU1NTtXZi5uAhDhyutShrV59Yb8GaZmdnGRkZ6es2mzBv7wxTVjBvL9XNOjExsfdIJ3ejyX34DwOfA8jM+yJiJfAzwAxwQsd8K6uxn5KZ24HtAOPj49lqtWoF2bZjJ1v31duV/RvqbbOu6elp6u7nIJi3d4YpK5i3l/qVtcldOp8CXghQlf2zge/TPtJfV40fA7wU+EyzmJKkproq/Ii4GLgSOC0iNkbECuBW4NyIuAl4D/DvM/PHmfk54K7qvP424Dov2ErS4HV1HiQzdwO75wwfBv7DAvO/q2EuSdIS841XklQIC1+SCmHhS1IhLHxJKoSFL0mFsPAlqRAWviQVwsKXpEJY+JJUCAtfkgph4UtSISx8SSqEhS9JhbDwJakQFr4kFcLCl6RCWPiSVIijFn5EnBoRH4qIe+f52YaIyIgY6Rj7hYh4f0TcEhH/ZakDS5Lq6eYrDi8EdgLndg5GxEtof0F559jzgA8AL8vMxyLijoi4JDM/tVSBJUn1HPUIPzNvBx7pHKuK/S3Ab8+ZfR3wrcx8rJq+G1i/BDklSQ1FZh59pogWsCUzx6vp9wDbMvMbEZHACZk5GxGvBS7PzF+u5vs1oJWZVyyw3klgEmB0dHRsamqq1k7MHDzEgcO1FmXt6hPrLVjT7OwsIyMjR59xmTBv7wxTVjBvL9XNOjExsfdIL3ejm1M6TxMRPwOcBPxqRBwZvi4i/gyYAU7omH1lNTavzNwObAcYHx/PVqu12DgAbNuxk637Fr0rAOzfUG+bdU1PT1N3PwfBvL0zTFnBvL3Ur6yLbsnM/DZw1ZHpiHgn8O7qCP95wBkR8dzqtM4FwPuXKqwkqb5u7tK5GLgSOC0iNkbEimr8lIjYWM32lohYnZk/An4d+L2I2AR8yQu2krQ8HPUIPzN3A7vnGX8Q2FT96Rz/JPDJpQooSVoavvFKkgph4UtSISx8SSqEhS9JhbDwJakQFr4kFcLCl6RCWPiSVAgLX5IKYeFLUiEsfEkqhIUvSYWw8CWpEBa+JBXCwpekQlj4klQIC1+SCtHVd9pGxKm0v9nqnMw8vxq7ATgV+D4wBtycmV+rfnYFcB7wJPBAZn6wB9klSYvQ7ZeYXwjsBM7tGBsBrsvMjIjLgXcBl0XE6cD1wHnVz+6NiE9n5t8saXJJ0qJ0dUonM28HHpkz9rbMzI71zFaPLwX2dvzsHuCVS5BVktRAPNXLR5kxogVsyczxOePPAT4GvCEzH4iItwKjmXlt9fNNAJm5cZ51TgKTAKOjo2NTU1O1dmLm4CEOHK61KGtXn1hvwZpmZ2cZGRnp6zabMG/vDFNWMG8v1c06MTGxd24nP5NuT+nMqyr7PwB+KzMfqIZngLM6ZlsJ3D/f8pm5HdgOMD4+nq1Wq1aObTt2snVfvV3Zv6HeNuuanp6m7n4Ognl7Z5iygnl7qV9Za9+lExErgA8C787MvRHxK9WP7gTGIiKq6XXAJ5rFlCQ11e1dOhcDVwKnRcRGYCuwA3g58HNVtx8P3JGZ34mILcB7IuJJ4ENesJWkweuq8DNzN7B7zvBrnmH+24DbGuSSJC0x33glSYWw8CWpEBa+JBXCwpekQlj4klQIC1+SCmHhS1IhLHxJKoSFL0mFsPAlqRAWviQVwsKXpEJY+JJUCAtfkgph4UtSISx8SSqEhS9Jhej2Kw5PBTYB52Tm+dXYccAW4LvA2cDmzLyv+tkVwHnAk8ADmfnBHmSXJC1CV4UPXAjsBM7tGLsW+NvM/N2IWAt8GLgoIk4HrgfOy8yMiHsj4tN+r60kDVZXp3Qy83bgkTnD64F7qp/vA86JiJXApcDezMxqvnuAVy5NXElSXfFULx9lxogWsCUzx6vprwOXZ+YXqunvAC3g3wCjmXltNb4JIDM3zrPOSWASYHR0dGxqaqrWTswcPMSBw7UWZe3qE+stWNPs7CwjIyN93WYT5u2dYcoK5u2lulknJib2HunkbnR7Smc+M8AJHdMrq7EZ4Kw54/fPt4LM3A5sBxgfH89Wq1UryLYdO9m6r96u7N9Qb5t1TU9PU3c/B8G8vTNMWcG8vdSvrE3u0tkFrAOozuF/MTMfBu4ExiIiqvnWAZ9olFKS1FhXhR8RFwNXAqdFxMaIWAG8DzgjIjYCbwKuBsjM79C+e+c9EbEV+JAXbCVp8Lo6D5KZu4Hd8/zoDQvMfxtwW4NckqQl5huvJKkQFr4kFcLCl6RCWPiSVAgLX5IKYeFLUiEsfEkqhIUvSYWw8CWpEBa+JBXCwpekQlj4klQIC1+SCmHhS1IhLHxJKoSFL0mFsPAlqRBNvsSciHgzsAb4AXA27a85XAFsBr5Rjd2UmQeaxZQkNVW78CPiVOCtwMmZ+ZOI2Am8BrgI+MvM/OOIuIz299teuSRpJUm1NTml8yPg74GV1fQI8GVgPXBPNXZ3NS1JGrDIzPoLR1wJXAF8DwjaX2r+Q2A0Mx+KiGOAx4FjM/OJeZafBCYBRkdHx6ampmrlmDl4iAOH6+3D2tUn1luwptnZWUZGRvq6zSbM2zvDlBXM20t1s05MTOzNzPFu569d+BFxLvAR4J9k5hMRsRV4Engt8M8y89sRsQq4PzNXHW194+PjuWfPnlpZtu3YydZ99c5O7d/c3/8BmZ6eptVq9XWbTZi3d4YpK5i3l+pmjYhFFX6TUzqrgYMdR+7fA44DdgHrqrELqmlJ0oA1uUvnz4FXVUf2DwEvB64FHgNujYgXAWcC1zdOKUlqrHbhZ+aTtM/Zz+f1ddcrSeoN33glSYWw8CWpEBa+JBXCwpekQlj4klQIC1+SCmHhS1IhLHxJKoSFL0mFsPAlqRAWviQVwsKXpEJY+JJUCAtfkgph4UtSISx8SSqEhS9JhWjyFYcARMSLaX9x+WHgYuAWYAZ4G3A/sAZ4U2bONt2WJKm+RoUfEc8G3g1clpk/iYiPAE8AHwVuzszPR8Q1wA20/wGQJA1I01M65wMBXBMRbwUuo/2F5hPAvdU8dwPrG25HktRQZGb9hSMuBz4ArMnMQxFxG/Bl4IbMfH41z1nAdGaePs/yk8AkwOjo6NjU1FStHDMHD3HgcL19WLv6xHoL1jQ7O8vIyEhft9mEeXtnmLKCeXupbtaJiYm9mTne7fxNz+E/DHwtMw9V0/8buAhYERGR7X9NVtI+p/9TMnM7sB1gfHw8W61WrRDbduxk6756u7J/Q71t1jU9PU3d/RwE8/bOMGUF8/ZSv7I2PaXzOeAfVOfyAc6gfYR/F+3TPQAXALsabkeS1FCjI/zMPBgRNwDvjYgHgVOAtwP/Hbg5Iv4l8LPAdY2TSpIaaXxbZmb+KfCnc4b3A69rum5J0tLxjVeSVAgLX5IKYeFLUiEsfEkqhIUvSYWw8CWpEBa+JBXCwpekQlj4klQIC1+SCmHhS1IhLHxJKoSFL0mFsPAlqRAWviQVwsKXpEJY+JJUiMaFHxErIuJLEbGlmj4uIv5bRLw1Iv4oIl7UPKYkqanGX3EIbAL+T8f0tcDfZubvRsRa4MPARUuwHUlSA42O8CPiSuBu4Jsdw+uBewAycx9wTkSsbLIdSVJzkZn1Fox4KXBFZt4UEbcAI5l5fUR8Hbg8M79QzfcdoJWZ98+zjklgEmB0dHRsamqqVpaZg4c4cLjWoqxdfWK9BWuanZ1lZGSkr9tswry9M0xZwby9VDfrxMTE3swc73b+Jqd0Xg38OCJuBC4EnhMR1wIzwAkd862sxn5KZm4HtgOMj49nq9WqFWTbjp1s3VdvV/ZvqLfNuqanp6m7n4Ng3t4Zpqxg3l7qV9bahZ+Zv3PkcUQcR/sI/73V43XAZ6tz+F/MzIebR5UkNdH4om1E/ArwCtpH+K8F3gdsiYiNwFnA1U23IUlqrnHhZ+YdwB1zht/QdL2SpKXlG68kqRAWviQVYineeDXU1ty4q/ay+zevX8IkktRbHuFLUiEsfEkqhIUvSYWw8CWpEBa+JBXCwpekQlj4klQIC1+SCmHhS1IhLHxJKoSFL0mFsPAlqRAWviQVwsKXpEI0+njkiDgT2AT8NXA68MPMfHtErAI2A98AzgZuyswDTcNKkupr+nn4q4CpzNwJEBFfiYhdwOuBv8zMP46Iy4AtwJUNtyVJaqDRKZ3MvPdI2Xes71FgPXBPNXZ3NS1JGqDIzKVZUcSrgVZmvjEiHgNGM/OhiDgGeBw4NjOfmLPMJDAJMDo6OjY1NVVr2zMHD3HgcLP8daxdfeKil5mdnWVkZKQHaXrDvL0zTFnBvL1UN+vExMTezBzvdv4l+YrDiJgAJoBrq6EZ4ATgIWAl8Hdzyx4gM7cD2wHGx8ez1WrV2v62HTvZuq//39a4f0Nr0ctMT09Tdz8Hwby9M0xZwby91K+sjVsyItYDFwFvBE6LiDOAXcA64NvABdW0JGmAmt6lMwb8T2APcBdwPPD7wE3ArRHxIuBM4PqGOSVJDTUq/MzcCyx04un1TdYtSVpavvFKkgph4UtSISx8SSqEhS9Jhej/zesCYM2Nze5U3b/ZNy9LWhwLv4E6pf2mtU9wVcOyl6Q6PKUjSYWw8CWpEBa+JBXCwpekQlj4klQIC1+SCmHhS1IhLHxJKoSFL0mFsPAlqRB+tIIkNdTks7H6+blYPSv8iPgF4DW0v9A8M/O3e7WtEjX98LVuLeVn//iBb9Jg9aTwI+J5wAeAl2XmYxFxR0Rckpmf6sX2JElH16sj/HXAtzLzsWr6bmA9YOEXbJAfCT0s/8st9VJk5tKvNOK1wOWZ+cvV9K8Brcy8Ys58k8BkNfli4Os1N3ky8IOay/bbMGUF8/bSMGUF8/ZS3axnZOYp3c7cqyP8GeCEjumV1djTZOZ2YHvTjUXEnswcb7qefhimrGDeXhqmrGDeXupX1l7dlnkPcEZEPLeavgDwWz8kaYB6coSfmT+KiF8Hfi8iHgS+5AVbSRqsnt2WmZmfBD7Zq/XP0fi0UB8NU1Ywby8NU1Ywby/1JWtPLtpKkpYfP1pBkgox1B+tMMh380bEXwE/riafzMxLImIVsBn4BnA2cFNmHqjmfzPtu5VOAv4iMz9WjZ8LvAH4JvAC4PrMfCIijgO2AN+t1rU5M+9bRL5TgU3AOZl5fjW24Doj4grgPOBJ4IHM/GA1vgZ4G3A/sAZ4U2bORsSzgHcAs8AZwIcz86+qZRb9uiyQ9yrgP/LU8/zhzPzooPNGxJlV1r8GTgd+mJlv79frv9C+18h7C9DqmPV3qlOxA8tbvU7/C/gc8BzgTOB1wIpl+twulPeG5fbcApCZQ/kHeB7tv9TPrabvAC7p4/ZvmWfsA8CvVo8vAz5aPf554M+qx8cCfwM8Hwjg/wKnVj/bClxdPb4ReEv1eC3w2UXm+9dVhj0dY/Ouk3YJfIGnTvHdC5xdPf5z4J9Wj68B/mv1+N8C768erwLuA55d93VZIO9VwJp55h1oXuB84F91TH8FGOvH6/9M+14j70/9Dg86L+2zDhs7pncCG5bxc7tQ3mX33GbmUJ/SWejdvP2yNiJuiIhbIuLIdtfTviV1bp5fOjKemY8DXwVeAbwQWJGZ359nmfUdy+wDzomIld2Gy8zbgUfmDC+0zkuBvVn91lTzvDIijgUmaP8iPVO+g7SPwl9GzddlgbwA/zkiro+Im6sjaAadNzPvzcydHUPPAh6lP6//vPteMy8R8VvV83tD9ZEoA82bmT/JzE1VtmNol9rXWb7P7UJ5l91zC8N9SucFPL0gHq7G+uXWzPx8RDwb+ExEPDIn08PASdUvwQtov7Bzsz7Iwvuw0P493CDzQutcaPxk4HDHL1Q3+U5ZYLyO3cCuzHwwIl4F/AlwyXLKGxGvBu7MzK9FRD9e/0a/93Py/gmwPzMfjYj/BGwDrl4OeSPiUuA3gY9n5p7l/tzOk/cwy/C5HeYj/K7ezdsrmfn56r9PAp+lfWTZmWkl8HeZ+cQzZH2mfejF/i02xw+AFRERi8i3ZLkz85uZ+WA1+Wng4uof2GWRNyImaL/uvznPNnr1+i9Z3sz8cmY+Wv3408A/b7DtJc2bmXdm5i8CP1cV5rJ+bufmXa7P7TAX/sDezRsR/ygiru4YOpv2eeBdtE8RzM3z8SPj1VHJS4HP0L4Adbi6YDl3mV0dy6wFvpiZTY7un2mddwJjHUW5DvhE9b+cd9E+B/xM+VYBxwFfZglfl4h4Z/V8Qfs5/mb1D+zA81an8S4F3gicGhHr6M/rP+++18kbEe/qmOXI7/BA80bESztOkUL7AuYLWabP7UJ5l+NzC0N+H35E/AvaF/seBB7PPt2lExH/EPh92nc9rKR98eU62hdfbgW+Rftq/Y359DsJTqr+fCKffmX+mmqZVTx1ZX4F7Svz3wPOAt6Ri7tL52Lg3wG/CPwB7YtALLTO6mr/OO2r/ffl0+96uZn2L+TPAtflU3e9vBP4UTX+h/nUXS+Lfl0WyDsJvJz2X6K1wPs6tjGwvBExRvt0055q6Hjavw8fow+v/0L7XiPvi2lftJ6pnt+bO7YxkLzRvqPoXbT/bh0LvAT4DeDvl+lzu1DeNy635xaGvPAlSd0b5lM6kqRFsPAlqRAWviQVwsKXpEJY+JJUCAtfkgph4UtSISx8SSrE/wOj8ejjZelEQwAAAABJRU5ErkJggg==\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "finalRes['dist'].hist(bins=20)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# interestingly there are other known annotated transcripts for these genes\n",
    "# which may in fact overlap novel isoforms\n",
    "# however, those transcripts do not have a corresponding CDS\n",
    "# so perhaps it would be a better idea to separate them out into distinct genes????"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>cds</th>\n",
       "      <th>geneID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CHS.37542.44</td>\n",
       "      <td>1</td>\n",
       "      <td>CHS.37542</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CHS.1638.6</td>\n",
       "      <td>1</td>\n",
       "      <td>CHS.1638</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CHS.40717.13</td>\n",
       "      <td>1</td>\n",
       "      <td>CHS.40717</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CHS.34617.5</td>\n",
       "      <td>1</td>\n",
       "      <td>CHS.34617</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CHS.47197.10</td>\n",
       "      <td>1</td>\n",
       "      <td>CHS.47197</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             id  cds     geneID\n",
       "0  CHS.37542.44    1  CHS.37542\n",
       "1    CHS.1638.6    1   CHS.1638\n",
       "2  CHS.40717.13    1  CHS.40717\n",
       "3   CHS.34617.5    1  CHS.34617\n",
       "4  CHS.47197.10    1  CHS.47197"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# lastly, let's investigate cases where all known transcripts have a CDS - this way if something is outside - very susplicious\n",
    "\n",
    "# what we should do is for each CDS group (cds that belong to the same gene)\n",
    "# we need to intersect it with all known transcritpts for that gene\n",
    "# if the set difference is 0 - that means all known transcripts in a gene have a CDS annotated\n",
    "transIDs=pd.DataFrame([transIDs_known_only]).T\n",
    "transIDs.columns=['id']\n",
    "transIDs['cds']=transIDs['id'].isin(cdsParents_known_only)\n",
    "transIDs['cds']=transIDs['cds'].astype(int)\n",
    "transIDs['geneID']=transIDs.id.str.extract('(CHS.(\\d)*)',expand=True)[0]\n",
    "transIDs.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "9870\n"
     ]
    }
   ],
   "source": [
    "grouped=transIDs[['geneID','cds']].groupby(by='geneID').mean()\n",
    "grouped.reset_index(inplace=True)\n",
    "cdsOnlyGenes=set(grouped[grouped['cds']==1][\"geneID\"])\n",
    "print(len(cdsOnlyGenes))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "45\n",
      "36\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>geneID</th>\n",
       "      <th>smin_known</th>\n",
       "      <th>smax_known</th>\n",
       "      <th>emin_known</th>\n",
       "      <th>emax_known</th>\n",
       "      <th>parent</th>\n",
       "      <th>smin_novel</th>\n",
       "      <th>smax_novel</th>\n",
       "      <th>emin_novel</th>\n",
       "      <th>emax_novel</th>\n",
       "      <th>dist</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CHS.36728</td>\n",
       "      <td>17962572</td>\n",
       "      <td>18267018</td>\n",
       "      <td>17962694</td>\n",
       "      <td>18268922</td>\n",
       "      <td>CHS.36728.3</td>\n",
       "      <td>18330788.0</td>\n",
       "      <td>18342247.0</td>\n",
       "      <td>18330962.0</td>\n",
       "      <td>18344196.0</td>\n",
       "      <td>61866.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      geneID  smin_known  smax_known  emin_known  emax_known       parent  \\\n",
       "2  CHS.36728    17962572    18267018    17962694    18268922  CHS.36728.3   \n",
       "\n",
       "   smin_novel  smax_novel  emin_novel  emax_novel     dist  \n",
       "2  18330788.0  18342247.0  18330962.0  18344196.0  61866.0  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>geneID</th>\n",
       "      <th>smin_known</th>\n",
       "      <th>smax_known</th>\n",
       "      <th>emin_known</th>\n",
       "      <th>emax_known</th>\n",
       "      <th>parent</th>\n",
       "      <th>smin_novel</th>\n",
       "      <th>smax_novel</th>\n",
       "      <th>emin_novel</th>\n",
       "      <th>emax_novel</th>\n",
       "      <th>dist</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>CHS.51125</td>\n",
       "      <td>88759368</td>\n",
       "      <td>89333363</td>\n",
       "      <td>88760084</td>\n",
       "      <td>89337057</td>\n",
       "      <td>CHS.51125.3</td>\n",
       "      <td>88758022.0</td>\n",
       "      <td>88758970.0</td>\n",
       "      <td>88758521.0</td>\n",
       "      <td>88759367.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       geneID  smin_known  smax_known  emin_known  emax_known       parent  \\\n",
       "25  CHS.51125    88759368    89333363    88760084    89337057  CHS.51125.3   \n",
       "\n",
       "    smin_novel  smax_novel  emin_novel  emax_novel  dist  \n",
       "25  88758022.0  88758970.0  88758521.0  88759367.0   1.0  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>geneID</th>\n",
       "      <th>smin_known</th>\n",
       "      <th>smax_known</th>\n",
       "      <th>emin_known</th>\n",
       "      <th>emax_known</th>\n",
       "      <th>parent</th>\n",
       "      <th>smin_novel</th>\n",
       "      <th>smax_novel</th>\n",
       "      <th>emin_novel</th>\n",
       "      <th>emax_novel</th>\n",
       "      <th>dist</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CHS.51962</td>\n",
       "      <td>138076432</td>\n",
       "      <td>138116620</td>\n",
       "      <td>138076611</td>\n",
       "      <td>138118304</td>\n",
       "      <td>CHS.51962.13</td>\n",
       "      <td>138001784.0</td>\n",
       "      <td>138004556.0</td>\n",
       "      <td>138002390.0</td>\n",
       "      <td>138007888.0</td>\n",
       "      <td>68544.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CHS.51962</td>\n",
       "      <td>138076432</td>\n",
       "      <td>138116620</td>\n",
       "      <td>138076611</td>\n",
       "      <td>138118304</td>\n",
       "      <td>CHS.51962.15</td>\n",
       "      <td>138002270.0</td>\n",
       "      <td>138008838.0</td>\n",
       "      <td>138002390.0</td>\n",
       "      <td>138011346.0</td>\n",
       "      <td>65086.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CHS.36728</td>\n",
       "      <td>17962572</td>\n",
       "      <td>18267018</td>\n",
       "      <td>17962694</td>\n",
       "      <td>18268922</td>\n",
       "      <td>CHS.36728.3</td>\n",
       "      <td>18330788.0</td>\n",
       "      <td>18342247.0</td>\n",
       "      <td>18330962.0</td>\n",
       "      <td>18344196.0</td>\n",
       "      <td>61866.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CHS.51962</td>\n",
       "      <td>138076432</td>\n",
       "      <td>138116620</td>\n",
       "      <td>138076611</td>\n",
       "      <td>138118304</td>\n",
       "      <td>CHS.51962.14</td>\n",
       "      <td>138002214.0</td>\n",
       "      <td>138021187.0</td>\n",
       "      <td>138002390.0</td>\n",
       "      <td>138022179.0</td>\n",
       "      <td>54253.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CHS.40180</td>\n",
       "      <td>37453454</td>\n",
       "      <td>37590088</td>\n",
       "      <td>37453830</td>\n",
       "      <td>37596146</td>\n",
       "      <td>CHS.40180.9</td>\n",
       "      <td>37615856.0</td>\n",
       "      <td>37622829.0</td>\n",
       "      <td>37619710.0</td>\n",
       "      <td>37624376.0</td>\n",
       "      <td>19710.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      geneID  smin_known  smax_known  emin_known  emax_known        parent  \\\n",
       "0  CHS.51962   138076432   138116620   138076611   138118304  CHS.51962.13   \n",
       "1  CHS.51962   138076432   138116620   138076611   138118304  CHS.51962.15   \n",
       "2  CHS.36728    17962572    18267018    17962694    18268922   CHS.36728.3   \n",
       "3  CHS.51962   138076432   138116620   138076611   138118304  CHS.51962.14   \n",
       "4  CHS.40180    37453454    37590088    37453830    37596146   CHS.40180.9   \n",
       "\n",
       "    smin_novel   smax_novel   emin_novel   emax_novel     dist  \n",
       "0  138001784.0  138004556.0  138002390.0  138007888.0  68544.0  \n",
       "1  138002270.0  138008838.0  138002390.0  138011346.0  65086.0  \n",
       "2   18330788.0   18342247.0   18330962.0   18344196.0  61866.0  \n",
       "3  138002214.0  138021187.0  138002390.0  138022179.0  54253.0  \n",
       "4   37615856.0   37622829.0   37619710.0   37624376.0  19710.0  "
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# now need to select those from the finalRes dataframe\n",
    "cdsOnlyOutOfBounds=finalRes[finalRes['geneID'].isin(cdsOnlyGenes)].reset_index(drop=True)\n",
    "print(len(set(cdsOnlyOutOfBounds[\"parent\"])))\n",
    "print(len(set(cdsOnlyOutOfBounds[\"geneID\"])))\n",
    "display(cdsOnlyOutOfBounds[cdsOnlyOutOfBounds[\"parent\"]==\"CHS.36728.3\"])\n",
    "display(cdsOnlyOutOfBounds[cdsOnlyOutOfBounds['parent']=='CHS.51125.3'])\n",
    "cdsOnlyOutOfBounds.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqid</th>\n",
       "      <th>source</th>\n",
       "      <th>type</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>phase</th>\n",
       "      <th>attributes</th>\n",
       "      <th>parent</th>\n",
       "      <th>geneID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>exon</td>\n",
       "      <td>69091</td>\n",
       "      <td>70008</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>Parent=CHS.6.1</td>\n",
       "      <td>CHS.6.1</td>\n",
       "      <td>CHS.6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>Gnomon</td>\n",
       "      <td>exon</td>\n",
       "      <td>181049</td>\n",
       "      <td>181110</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>Parent=CHS.9.1</td>\n",
       "      <td>CHS.9.1</td>\n",
       "      <td>CHS.9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>Gnomon</td>\n",
       "      <td>exon</td>\n",
       "      <td>182529</td>\n",
       "      <td>182746</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>Parent=CHS.9.1</td>\n",
       "      <td>CHS.9.1</td>\n",
       "      <td>CHS.9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>Gnomon</td>\n",
       "      <td>exon</td>\n",
       "      <td>183114</td>\n",
       "      <td>183240</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>Parent=CHS.9.1</td>\n",
       "      <td>CHS.9.1</td>\n",
       "      <td>CHS.9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>Gnomon</td>\n",
       "      <td>exon</td>\n",
       "      <td>183922</td>\n",
       "      <td>184258</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>Parent=CHS.9.1</td>\n",
       "      <td>CHS.9.1</td>\n",
       "      <td>CHS.9</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  seqid      source  type   start     end score strand phase      attributes  \\\n",
       "0  chr1  BestRefSeq  exon   69091   70008     .      +     .  Parent=CHS.6.1   \n",
       "1  chr1      Gnomon  exon  181049  181110     .      +     .  Parent=CHS.9.1   \n",
       "2  chr1      Gnomon  exon  182529  182746     .      +     .  Parent=CHS.9.1   \n",
       "3  chr1      Gnomon  exon  183114  183240     .      +     .  Parent=CHS.9.1   \n",
       "4  chr1      Gnomon  exon  183922  184258     .      +     .  Parent=CHS.9.1   \n",
       "\n",
       "    parent geneID  \n",
       "0  CHS.6.1  CHS.6  \n",
       "1  CHS.9.1  CHS.9  \n",
       "2  CHS.9.1  CHS.9  \n",
       "3  CHS.9.1  CHS.9  \n",
       "4  CHS.9.1  CHS.9  "
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# now we need to compute the same statistics for the known isoforms of known genes\n",
    "# to see how frequently there are exons that do not overlap\n",
    "\n",
    "df_known_exons=pd.read_csv(\"./chess2.02.gff\",sep=\"\\t\",names=gff3Cols)\n",
    "df_known_exons.dropna(1inplace=True,axis=0)\n",
    "df_known_exons.reset_index(inplace=True,drop=True)\n",
    "df_known_exons=df_known_exons[df_known_exons[\"type\"]==\"exon\"].reset_index(drop=True)\n",
    "df_known_exons[\"start\"]=df_known_exons[\"start\"].astype(int)\n",
    "df_known_exons[\"end\"]=df_known_exons[\"end\"].astype(int)\n",
    "df_known_exons=df_known_exons[~(df_known_exons['source'].isin(novelSources))].reset_index(drop=True)\n",
    "df_known_exons[\"parent\"]=df_known_exons.attributes.str.split(\"Parent=\",expand=True)[1].str.split(\";\",expand=True)[0]\n",
    "df_known_exons[\"geneID\"]=df_known_exons.parent.str.extract('(CHS.(\\d)*)',expand=True)[0]\n",
    "df_known_exons=df_known_exons[df_known_exons[\"geneID\"].isin(set(knownDF[\"geneID\"]))].reset_index(drop=True)\n",
    "df_known_exons.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>parent</th>\n",
       "      <th>smin</th>\n",
       "      <th>smax</th>\n",
       "      <th>emin</th>\n",
       "      <th>emax</th>\n",
       "      <th>geneID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CHS.10000.1</td>\n",
       "      <td>112961247</td>\n",
       "      <td>113262846</td>\n",
       "      <td>112961664</td>\n",
       "      <td>113265197</td>\n",
       "      <td>CHS.10000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CHS.10000.10</td>\n",
       "      <td>112961368</td>\n",
       "      <td>113262846</td>\n",
       "      <td>112961664</td>\n",
       "      <td>113263012</td>\n",
       "      <td>CHS.10000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CHS.10000.12</td>\n",
       "      <td>112961425</td>\n",
       "      <td>113040926</td>\n",
       "      <td>112961664</td>\n",
       "      <td>113041216</td>\n",
       "      <td>CHS.10000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CHS.10000.13</td>\n",
       "      <td>112961439</td>\n",
       "      <td>113133449</td>\n",
       "      <td>112961664</td>\n",
       "      <td>113133951</td>\n",
       "      <td>CHS.10000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CHS.10000.2</td>\n",
       "      <td>112961247</td>\n",
       "      <td>113262846</td>\n",
       "      <td>112961664</td>\n",
       "      <td>113265197</td>\n",
       "      <td>CHS.10000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         parent       smin       smax       emin       emax     geneID\n",
       "0   CHS.10000.1  112961247  113262846  112961664  113265197  CHS.10000\n",
       "1  CHS.10000.10  112961368  113262846  112961664  113263012  CHS.10000\n",
       "2  CHS.10000.12  112961425  113040926  112961664  113041216  CHS.10000\n",
       "3  CHS.10000.13  112961439  113133449  112961664  113133951  CHS.10000\n",
       "4   CHS.10000.2  112961247  113262846  112961664  113265197  CHS.10000"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "knownExonDF=df_known_exons.groupby(\"parent\").agg({'start':['min','max'],'end':['min','max'],'geneID':'min'}).reset_index()\n",
    "knownExonDF.columns=['parent','smin','smax','emin','emax','geneID']\n",
    "knownExonDF.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>geneID</th>\n",
       "      <th>smin_known</th>\n",
       "      <th>smax_known</th>\n",
       "      <th>emin_known</th>\n",
       "      <th>emax_known</th>\n",
       "      <th>parent</th>\n",
       "      <th>smin_novel</th>\n",
       "      <th>smax_novel</th>\n",
       "      <th>emin_novel</th>\n",
       "      <th>emax_novel</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CHS.10363</td>\n",
       "      <td>126123806</td>\n",
       "      <td>126140119</td>\n",
       "      <td>126124484</td>\n",
       "      <td>126140638</td>\n",
       "      <td>CHS.10363.7</td>\n",
       "      <td>126100505</td>\n",
       "      <td>126100505</td>\n",
       "      <td>126102413</td>\n",
       "      <td>126102413</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CHS.10621</td>\n",
       "      <td>280057</td>\n",
       "      <td>388927</td>\n",
       "      <td>285662</td>\n",
       "      <td>389455</td>\n",
       "      <td>CHS.10621.17</td>\n",
       "      <td>273954</td>\n",
       "      <td>273954</td>\n",
       "      <td>277123</td>\n",
       "      <td>277123</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CHS.10680</td>\n",
       "      <td>2812621</td>\n",
       "      <td>2824090</td>\n",
       "      <td>2812856</td>\n",
       "      <td>2826197</td>\n",
       "      <td>CHS.10680.11</td>\n",
       "      <td>2830425</td>\n",
       "      <td>2836745</td>\n",
       "      <td>2830885</td>\n",
       "      <td>2836992</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CHS.10698</td>\n",
       "      <td>3613366</td>\n",
       "      <td>3753015</td>\n",
       "      <td>3615439</td>\n",
       "      <td>3753216</td>\n",
       "      <td>CHS.10698.15</td>\n",
       "      <td>3610307</td>\n",
       "      <td>3610307</td>\n",
       "      <td>3612003</td>\n",
       "      <td>3612003</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CHS.10725</td>\n",
       "      <td>4590072</td>\n",
       "      <td>4613515</td>\n",
       "      <td>4590105</td>\n",
       "      <td>4613888</td>\n",
       "      <td>CHS.10725.31</td>\n",
       "      <td>4567955</td>\n",
       "      <td>4572846</td>\n",
       "      <td>4568048</td>\n",
       "      <td>4573198</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      geneID  smin_known  smax_known  emin_known  emax_known        parent  \\\n",
       "0  CHS.10363   126123806   126140119   126124484   126140638   CHS.10363.7   \n",
       "1  CHS.10621      280057      388927      285662      389455  CHS.10621.17   \n",
       "2  CHS.10680     2812621     2824090     2812856     2826197  CHS.10680.11   \n",
       "3  CHS.10698     3613366     3753015     3615439     3753216  CHS.10698.15   \n",
       "4  CHS.10725     4590072     4613515     4590105     4613888  CHS.10725.31   \n",
       "\n",
       "   smin_novel  smax_novel  emin_novel  emax_novel  \n",
       "0   126100505   126100505   126102413   126102413  \n",
       "1      273954      273954      277123      277123  \n",
       "2     2830425     2836745     2830885     2836992  \n",
       "3     3610307     3610307     3612003     3612003  \n",
       "4     4567955     4572846     4568048     4573198  "
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# now merge the two together (novel with the known transcripts)\n",
    "knownResDF=knownDF.merge(knownExonDF,on='geneID',how='outer',indicator=True)\n",
    "assert len(knownResDF[(knownResDF['_merge']=='right_only')])==0,\"something's wrong with the novel transcripts. Some don't match the known isoforms \"\n",
    "knownResDF=knownResDF[knownResDF['_merge']=='both'].reset_index(drop=True)\n",
    "knownResDF.drop('_merge',inplace=True,axis=1)\n",
    "knownResDF.columns=['geneID','smin_known','smax_known','emin_known','emax_known',\n",
    "               'parent','smin_novel','smax_novel','emin_novel','emax_novel']\n",
    "finalKnownResDF=knownResDF[(knownResDF['emax_novel']<knownResDF['smin_known'])|(knownResDF['smin_novel']>knownResDF['emax_known'])].reset_index(drop=True)\n",
    "finalKnownResDF.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "288\n",
      "221\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>geneID</th>\n",
       "      <th>smin_known</th>\n",
       "      <th>smax_known</th>\n",
       "      <th>emin_known</th>\n",
       "      <th>emax_known</th>\n",
       "      <th>parent</th>\n",
       "      <th>smin_novel</th>\n",
       "      <th>smax_novel</th>\n",
       "      <th>emin_novel</th>\n",
       "      <th>emax_novel</th>\n",
       "      <th>dist</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CHS.10363</td>\n",
       "      <td>126123806</td>\n",
       "      <td>126140119</td>\n",
       "      <td>126124484</td>\n",
       "      <td>126140638</td>\n",
       "      <td>CHS.10363.7</td>\n",
       "      <td>126100505</td>\n",
       "      <td>126100505</td>\n",
       "      <td>126102413</td>\n",
       "      <td>126102413</td>\n",
       "      <td>21393</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CHS.10621</td>\n",
       "      <td>280057</td>\n",
       "      <td>388927</td>\n",
       "      <td>285662</td>\n",
       "      <td>389455</td>\n",
       "      <td>CHS.10621.17</td>\n",
       "      <td>273954</td>\n",
       "      <td>273954</td>\n",
       "      <td>277123</td>\n",
       "      <td>277123</td>\n",
       "      <td>2934</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CHS.10680</td>\n",
       "      <td>2812621</td>\n",
       "      <td>2824090</td>\n",
       "      <td>2812856</td>\n",
       "      <td>2826197</td>\n",
       "      <td>CHS.10680.11</td>\n",
       "      <td>2830425</td>\n",
       "      <td>2836745</td>\n",
       "      <td>2830885</td>\n",
       "      <td>2836992</td>\n",
       "      <td>4228</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CHS.10698</td>\n",
       "      <td>3613366</td>\n",
       "      <td>3753015</td>\n",
       "      <td>3615439</td>\n",
       "      <td>3753216</td>\n",
       "      <td>CHS.10698.15</td>\n",
       "      <td>3610307</td>\n",
       "      <td>3610307</td>\n",
       "      <td>3612003</td>\n",
       "      <td>3612003</td>\n",
       "      <td>1363</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CHS.10725</td>\n",
       "      <td>4590072</td>\n",
       "      <td>4613515</td>\n",
       "      <td>4590105</td>\n",
       "      <td>4613888</td>\n",
       "      <td>CHS.10725.31</td>\n",
       "      <td>4567955</td>\n",
       "      <td>4572846</td>\n",
       "      <td>4568048</td>\n",
       "      <td>4573198</td>\n",
       "      <td>16874</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      geneID  smin_known  smax_known  emin_known  emax_known        parent  \\\n",
       "0  CHS.10363   126123806   126140119   126124484   126140638   CHS.10363.7   \n",
       "1  CHS.10621      280057      388927      285662      389455  CHS.10621.17   \n",
       "2  CHS.10680     2812621     2824090     2812856     2826197  CHS.10680.11   \n",
       "3  CHS.10698     3613366     3753015     3615439     3753216  CHS.10698.15   \n",
       "4  CHS.10725     4590072     4613515     4590105     4613888  CHS.10725.31   \n",
       "\n",
       "   smin_novel  smax_novel  emin_novel  emax_novel   dist  \n",
       "0   126100505   126100505   126102413   126102413  21393  \n",
       "1      273954      273954      277123      277123   2934  \n",
       "2     2830425     2836745     2830885     2836992   4228  \n",
       "3     3610307     3610307     3612003     3612003   1363  \n",
       "4     4567955     4572846     4568048     4573198  16874  "
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print(len(set(finalKnownResDF[\"parent\"])))\n",
    "print(len(set(finalKnownResDF[\"geneID\"])))\n",
    "# make sure all is correct by seeing if any transcripts overlap novel\n",
    "# if difference is the same as the number of known transcripts - all good\n",
    "assert len(set(finalKnownResDF[\"parent\"]).difference(set(finalRes['parent'])))==len(set(finalKnownResDF[\"parent\"])),\"novel transcrits present\"\n",
    "finalKnownResDF[\"dist\"]=np.where(finalKnownResDF['emax_novel']<finalKnownResDF['smin_known'],finalKnownResDF['smin_known']-finalKnownResDF['emax_novel'],finalKnownResDF['smin_novel']-finalKnownResDF['emax_known'])\n",
    "finalKnownResDF.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>geneID</th>\n",
       "      <th>smin_known</th>\n",
       "      <th>smax_known</th>\n",
       "      <th>emin_known</th>\n",
       "      <th>emax_known</th>\n",
       "      <th>parent</th>\n",
       "      <th>smin_novel</th>\n",
       "      <th>smax_novel</th>\n",
       "      <th>emin_novel</th>\n",
       "      <th>emax_novel</th>\n",
       "      <th>dist</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CHS.20405</td>\n",
       "      <td>79568021</td>\n",
       "      <td>79598785</td>\n",
       "      <td>79571350</td>\n",
       "      <td>79600725</td>\n",
       "      <td>CHS.20405.6</td>\n",
       "      <td>79212711</td>\n",
       "      <td>79229310</td>\n",
       "      <td>79212901</td>\n",
       "      <td>79229453</td>\n",
       "      <td>338568</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CHS.12017</td>\n",
       "      <td>63142759</td>\n",
       "      <td>63149867</td>\n",
       "      <td>63145386</td>\n",
       "      <td>63152810</td>\n",
       "      <td>CHS.12017.8</td>\n",
       "      <td>63292625</td>\n",
       "      <td>63359876</td>\n",
       "      <td>63292936</td>\n",
       "      <td>63360037</td>\n",
       "      <td>139815</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CHS.9672</td>\n",
       "      <td>86791059</td>\n",
       "      <td>86807884</td>\n",
       "      <td>86791195</td>\n",
       "      <td>86811233</td>\n",
       "      <td>CHS.9672.17</td>\n",
       "      <td>86940786</td>\n",
       "      <td>86951216</td>\n",
       "      <td>86940913</td>\n",
       "      <td>86951407</td>\n",
       "      <td>129553</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CHS.9672</td>\n",
       "      <td>86791059</td>\n",
       "      <td>86807884</td>\n",
       "      <td>86791195</td>\n",
       "      <td>86811233</td>\n",
       "      <td>CHS.9672.18</td>\n",
       "      <td>86940786</td>\n",
       "      <td>86951216</td>\n",
       "      <td>86941027</td>\n",
       "      <td>86951736</td>\n",
       "      <td>129553</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CHS.46819</td>\n",
       "      <td>80106610</td>\n",
       "      <td>80345975</td>\n",
       "      <td>80106683</td>\n",
       "      <td>80346270</td>\n",
       "      <td>CHS.46819.26</td>\n",
       "      <td>80466958</td>\n",
       "      <td>80466958</td>\n",
       "      <td>80469080</td>\n",
       "      <td>80469080</td>\n",
       "      <td>120688</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      geneID  smin_known  smax_known  emin_known  emax_known        parent  \\\n",
       "0  CHS.20405    79568021    79598785    79571350    79600725   CHS.20405.6   \n",
       "1  CHS.12017    63142759    63149867    63145386    63152810   CHS.12017.8   \n",
       "2   CHS.9672    86791059    86807884    86791195    86811233   CHS.9672.17   \n",
       "3   CHS.9672    86791059    86807884    86791195    86811233   CHS.9672.18   \n",
       "4  CHS.46819    80106610    80345975    80106683    80346270  CHS.46819.26   \n",
       "\n",
       "   smin_novel  smax_novel  emin_novel  emax_novel    dist  \n",
       "0    79212711    79229310    79212901    79229453  338568  \n",
       "1    63292625    63359876    63292936    63360037  139815  \n",
       "2    86940786    86951216    86940913    86951407  129553  \n",
       "3    86940786    86951216    86941027    86951736  129553  \n",
       "4    80466958    80466958    80469080    80469080  120688  "
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "finalKnownResDF.sort_values(by='dist',ascending=False,inplace=True)\n",
    "finalKnownResDF.reset_index(drop=True,inplace=True)\n",
    "finalKnownResDF.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAYUAAAD7CAYAAACFfIhNAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADl0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uIDIuMi4yLCBodHRwOi8vbWF0cGxvdGxpYi5vcmcvhp/UCwAAEWdJREFUeJzt3HuQnXV9x/H3NyTVSAglo2wULGiMWuwOMrvoZLSyW9qJEmxLp612gA5T6o71UhCwIKJDNWIoyVhMrZrKTGeE6Xr7I5QUEUuWOE4UEkebChrRxFZas2gEXIwUwrd/nCe/nKx7ssu57Tn6fs1kcp7feS6f5+zJ+eS5nI3MRJIkgAXzHUCS1DssBUlSYSlIkgpLQZJUWAqSpMJSkCQVC2ebISJWAGuBrwEnAz/OzPdFxLXASN2sH8jMO6tl3gksBU4AvpCZt7Y5tySpA2YtBWAZMJ6ZmwEi4r6I2AKQmSPTZ46IVwKjmXlORCwC7ouIbZn5cBtzS5I6YNZSyMx7pw0tAB4DiIh3A48DxwAbM/NnwLnA9mrZJyLifuA1gEcLktTj5nKkUETEecAdmfmtiPgMsDczH4uItwAbgYuBE4H76xZ7tBqbvq4xYAxg8eLFQ89//vOb3AV46qmnWLCgvy6PmLk7+i1zv+UFM3fLTJl37979o8x8Tls3lJlz+gOMAh8GFszw3EuBPdXj9wPvqXvuVuD3j7buoaGhbMXWrVtbWn4+mLk7+i1zv+XNNHO3zJQZ2JFz/Ayf6585VWVErAFWA5cAyyNiVUTcUDfLSuCB6vFtwKpquYXAacC2FnpLktQlc7n7aAj4FLAD2AocC3wEeDIibgQmgUHgrQCZ+dWI2BoR11G7++iy9CKzJPWFuVxo3gkseTorzcwbZp9LktRr+utKiySpoywFSVJhKUiSCktBklRYCpKkwlKQJBVP69dc9KpdDz7CRVdtaWrZvevWtDmNJPUvjxQkSYWlIEkqLAVJUmEpSJIKS0GSVFgKkqTCUpAkFZaCJKmwFCRJhaUgSSosBUlSYSlIkgpLQZJUWAqSpMJSkCQVloIkqbAUJEmFpSBJKiwFSVJhKUiSCktBklRYCpKkwlKQJBWWgiSpsBQkSYWlIEkqLAVJUmEpSJIKS0GSVCycbYaIWAGsBb4GnAz8ODPfFxHLgHXA94CVwNWZua9a5p3AUuAE4AuZeWuH8kuS2mjWUgCWAeOZuRkgIu6LiC3Am4AvZuanI+L1wHrgwoh4JTCamedExCLgvojYlpkPd2onJEntMevpo8y891Ah1C3zGLAG2F6NfbmaBjj30HhmPgHcD7ymXYElSZ0TmTn3mSPOA0Yy85KIeBwYyMyHI2Ih8ASwCPgIcH9m/n21zM3ARGZ+Ytq6xoAxgIGBgaHx8fGmd2Jy/yPsO9DcsoMnHd/0dlsxNTXFkiVL5mXbzTJz5/VbXjBzt8yUeXR0dGdmDrdzO3M5fQRARIwCo8Cl1dAkcBzwMLXrBz/JzCcj4tD4IUureY+QmZuATQDDw8M5MjLSTH4ANt6ymQ275rwrR9h7fvPbbcXExASt7PN8MHPn9VteMHO3dCvznO4+iog1wGrgEmB5RKwCtgCrqlleVU0D3HZovDqCOA3Y1sbMkqQOmcvdR0PAp4AdwFbgWGqniK4Gro+IFwMrgCsAMvOrEbE1Iq6jdvfRZV5klqT+MGspZOZOoNHJtzc1WOaGVkJJkuaHX16TJBWWgiSpsBQkSYWlIEkqLAVJUmEpSJIKS0GSVFgKkqTCUpAkFZaCJKmwFCRJhaUgSSosBUlSYSlIkgpLQZJUWAqSpMJSkCQVloIkqbAUJEmFpSBJKiwFSVJhKUiSCktBklRYCpKkwlKQJBWWgiSpsBQkSYWlIEkqLAVJUmEpSJIKS0GSVFgKkqTCUpAkFZaCJKmwFCRJxcLZZoiI5cBa4PTMPLMauwh4M/DzarabMvOT1XMXAGcAB4HvZubHO5BbktQBs5YC8GpgM/DyaeNvzMy99QMRcTJwBXBGZmZE3BsRd2Xmd9qSVpLUUbOWQmZ+NiJGZnjqbRHxQ+BZwD9k5n5gNbAzM7OaZzvwOsBSkKQ+EIc/v48yU60U1mfmcDX9AmAqMx+KiHOAyzPz7Ih4FzCQmZdW860FyMxrZljnGDAGMDAwMDQ+Pt70Tkzuf4R9B5pbdvCk45vebiumpqZYsmTJvGy7WWbuvH7LC2bulpkyj46O7jz0udwuczl99Asyc0/d5F3ArRFxDDAJvKjuuaXAAw3WsQnYBDA8PJwjIyPNRAFg4y2b2bCrqV1h7/nNb7cVExMTtLLP88HMnddvecHM3dKtzE3dfRQRH4yIQ5/CK4E9mXkQuAMYioionlsF3N56TElSN8zl7qOzgAuB50bENcAG4IfARyNiDzBYPU9m/iAi1gMfioiDwCe8yCxJ/WMuF5rvBu6eNnzjUea/Gbi5xVySpHngl9ckSYWlIEkqLAVJUmEpSJIKS0GSVFgKkqTCUpAkFZaCJKmwFCRJhaUgSSosBUlSYSlIkgpLQZJUWAqSpMJSkCQVloIkqbAUJEmFpSBJKiwFSVJhKUiSCktBklRYCpKkwlKQJBWWgiSpsBQkSYWlIEkqLAVJUmEpSJIKS0GSVFgKkqTCUpAkFZaCJKmwFCRJhaUgSSosBUlSYSlIkoqFs80QEcuBtcDpmXlmNfZMYD3wILASWJeZu6vnLgDOAA4C383Mj3couySpzWYtBeDVwGbg5XVjlwL/lZl/FxGDwE3Ab0fEycAVwBmZmRFxb0TclZnfaXtySVLbzXr6KDM/C/x02vAaYHv1/C7g9IhYCqwGdmZmVvNtB17XvriSpE6Kw5/fR5kpYgRYn5nD1fS3gTdk5ter6R8AI8CfAAOZeWk1vhYgM6+ZYZ1jwBjAwMDA0Pj4eNM7Mbn/EfYdaG7ZwZOOb3q7rZiammLJkiXzsu1mmbnz+i0vmLlbZso8Ojq689DncrvM5fTRTCaB4+qml1Zjk8CLpo0/MNMKMnMTsAlgeHg4R0ZGmowCG2/ZzIZdze3K3vOb324rJiYmaGWf54OZO6/f8oKZu6VbmZu9+2gLsAqguqbwjcx8FLgDGIqIqOZbBdzeckpJUlfMWgoRcRZwIfDciLgmIhYDNwKnRMQ1wOXAxQCZ+QNqdyV9KCI2AJ/wIrMk9Y9Zz7lk5t3A3TM89dYG898M3NxiLknSPPDLa5KkwlKQJBWWgiSpsBQkSYWlIEkqLAVJUmEpSJIKS0GSVFgKkqTCUpAkFZaCJKmwFCRJhaUgSSosBUlSYSlIkgpLQZJUWAqSpMJSkCQVloIkqbAUJEmFpSBJKiwFSVJhKUiSCktBklRYCpKkwlKQJBWWgiSpsBQkSYWlIEkqLAVJUmEpSJIKS0GSVFgKkqTCUpAkFZaCJKlY2OoKIuIrwM+ryYOZeXZELAPWAd8DVgJXZ+a+VrclSeqslksB+HxmXjtt7Drgi5n56Yh4PbAeuLAN25IkdVA7Th8NRsSVEXFtRKypxtYA26vHX66mJUk9LjKztRVEvCIz74mIY4BtwLuAO4GBzHw4IhYCTwCLMvPJuuXGgDGAgYGBofHx8aYzTO5/hH0Hmlt28KTjm95uK6ampliyZMm8bLtZZu68fssLZu6WmTKPjo7uzMzhdm6n5dNHmXlP9ffBiPgSMApMAscBDwNLgZ/UF0I1/yZgE8Dw8HCOjIw0nWHjLZvZsKu5Xdl7fvPbbcXExASt7PN8MHPn9VteMHO3dCtzS6ePIuKlEXFx3dBK4AFgC7CqGntVNS1J6nGtHik8CpwbEc+jdkTw38C/ALcD10fEi4EVwBUtbkeS1AUtlUJm/g9w3gxP7Qfe1Mq6JUnd55fXJEmFpSBJKiwFSVJhKUiSCktBklRYCpKkwlKQJBWWgiSpsBQkSYWlIEkqLAVJUmEpSJIKS0GSVFgKkqTCUpAkFZaCJKmwFCRJhaUgSSosBUlSYSlIkgpLQZJULJzvAPPt1Ku2NL3s3nVr2phEkuafRwqSpMJSkCQVloIkqbAUJEmFpSBJKiwFSVJhKUiSCktBklRYCpKkwlKQJBW/8r/mohWt/IqMf37tsW1MIknt4ZGCJKnwSGGe7HrwES6ap1/G1+wRzuWDTzLS9FYl9QNLoU+1cupKkhrpWClExO8CfwRMApmZf9upbak/+GvKpd7XkVKIiGcBHwNelpmPR8TnIuLszPz3TmxP3eMRivTLrVNHCquA72fm49X0l4E1gKWgprRyHeSiq7Z4pKGO+mU6Co7MbP9KI/4MeENm/mE1/ZfASGZeUDfPGDBWTb4E+HYLm3w28KMWlp8PZu6Ofsvcb3nBzN0yU+ZTMvM57dxIp44UJoHj6qaXVmNFZm4CNrVjYxGxIzOH27GubjFzd/Rb5n7LC2bulm5l7tT3FLYDp0TEM6rpVwGejJakHteRI4XM/FlE/BXw4Yh4CPgPLzJLUu/r2C2pmXkncGen1j9NW05DdZmZu6PfMvdbXjBzt3Qlc0cuNEuS+pO/+0iSVPT1r7mY729NR8RXgJ9Xkwcz8+yIWAasA74HrASuzsx91fzvpHYn1gnAFzLz1mr85cBbgT3AicAVmflkRDwTWA88WK1rXWbubiLncmAtcHpmnlmNNVx3RFwAnAEcBL6bmR+vxk8F3gM8AJwKXJ6ZUxGxALgOmAJOAW7KzK9UyzT1M2qQ+SLgzRx+zW/KzE/2QuaIWFHl/RpwMvDjzHxft94Pjfa/yczXwhG/5uoD1engXsi8APhX4KvArwErgL8AFvfw69wo85U9+TpnZl/+AZ5F7R/6M6rpzwFndznDtTOMfQz40+rx64FPVo9fCfxb9XgR8B3g14EA/hNYXj23Abi4enwV8DfV40HgS03m/OMqy466sRnXTe3D4escPrV4L7Cyevx54BXV47cD768evxH4x+rxMmA3cEwrP6MGmS8CTp1h3nnPDJwJ/EHd9H3AUDfeD0fb/yYz/8L7uocyLwCuqZveDJzf469zo8w9+Tr38+mjRt+a7qbBiLgyIq6NiEPbXkPtltzpmc49NJ6ZTwD3A68BXggszswfzrDMmrpldgGnR8TSpxsyMz8L/HTacKN1rwZ2ZvUOquZ5XUQsAkapvamOlnM/tf/Jv4wWfkYNMgO8LSKuiIj3Vv8LpxcyZ+a9mbm5bmgB8BjdeT/MuP8tZCYi3l29zldWv7amVzI/lZlrq4wLqX3ofZvefp0bZe7J17mfTx+dyJEfGo9WY910fWbeExHHANsi4qfTcj0KnFC9EU6k9sOl7rkTgYdovB+N9vHRNmRvtO5G488GDtS9ueaS8zkNxpt1N7AlMx+KiHOAzwBn91rmiDgPuCMzvxUR3Xg/tPxvYVrmzwB7M/OxiHgLsBG4uJcyR8Rq4B3AbZm5ox9e5xkyH6AHX+d+PlKY9VvTnZaZ91R/HwS+RO1/pfW5lgI/ycwnaZz3aPvRyX18unl+BCyOiHgaOduaPzP3ZOZD1eRdwFlVIfdM5ogYpfY+eMcM2+nU+6GtmTPzm5n5WPX0XcDvtLD9jmTOzDsy87XAC6oP1J5/nadn7tXXuZ9LYV6/NR0RL42Ii+uGVlI7F72F2imI6ZluOzRe/Q/mNGAbtQtjB6oLq9OX2VK3zCDwjcxsx1HC0dZ9BzBU90G6Cri9OozdSu089NFyLgOeCXyTNv+MIuKD1WsHtdd7T1XIPZG5OoW4GrgEWB4Rq+jO+2HG/W82c0TcUDfLofd1T2SOiNPqTtVC7YLrC+nh17lR5l59nfv6ewoR8XvULkg+BDyRXbz7KCKeB3yE2p0bS6ldELqM2gWh64HvU7vL4Ko88i6IE6o/t+eRdxS8vVpmGYfvKFhM7Y6C/wVeBFyXzd19dBbw58BrgY9Su0BFo3VXdysMU7tbYXceeSfPe6m9OX8DuCwP38nzQeBn1fg/5eE7eZr6GTXIPAb8FrV/VIPAjXXbmdfMETFE7fTWjmroWGrvj1vpwvuh0f43mfkl1C64T1av83vrtjPfmVcAN1D7d7cI+E3gr4H/6+HXuVHmS3rxde7rUpAktVc/nz6SJLWZpSBJKiwFSVJhKUiSCktBklRYCpKkwlKQJBWWgiSp+H/7qDkHvmG45gAAAABJRU5ErkJggg==\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "<matplotlib.axes._subplots.AxesSubplot at 0x7f30be99fa20>"
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAXwAAAD7CAYAAABpJS8eAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADl0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uIDIuMi4yLCBodHRwOi8vbWF0cGxvdGxpYi5vcmcvhp/UCwAAFXhJREFUeJzt3X+U3XV95/HnW0GJDEGy4MCGllRA1x9ZYGfYniwgd5buUk05W+1uWU9glyN1drsulSIK0shSN9VQE380W6up9g8lu7MtnJ64ppZaZaLLoUrSVbP+oqCx6tEMmhIYjBTwvX/cbw6XcYbc+X7n3jt3P8/HOTnc72e+P17feyevfPl+v/feyEwkSf//e9agA0iS+sPCl6RCWPiSVAgLX5IKYeFLUiEsfEkqhIUvSYWw8CWpEBa+JBXimEEHOOLkk0/ONWvW1Fr20Ucf5fjjj1/aQD0yTFnBvL00TFnBvL1UN+vevXt/kJmndL1AZi6LP2NjY1nXXXfdVXvZfhumrJnm7aVhyppp3l6qmxXYk4voWU/pSFIhLHxJKoSFL0mFsPAlqRAWviQVwsKXpEJY+JJUCAtfkgph4UtSIZbNRys0se+7h7jqxl21lt2/ef0Sp5Gk5ckjfEkqhIUvSYWw8CWpEF2dw4+IU4FNwDmZeX41FsA11SxrgOdn5uuqn70ZWAmcBPxFZn5siXNLkhap24u2FwI7gXM7xq4AHsrMjwBExD+u/vvzwERmvioijgW+EhGfycyHljC3JGmRujqlk5m3A4/MGd4ArIqI34iIdwCz1fgvAfdUyz0OfBV4xdLElSTVFe3P0O9ixogWsCUzx6vprwL/IzPfHhEvAj4OvAR4P/DVzHxvNd9twHRmfmiedU4CkwCjo6NjU1NTtXZi5uAhDhyutShrV59Yb8GaZmdnGRkZ6es2mzBv7wxTVjBvL9XNOjExsfdIJ3ejyX34DwOfA8jM+yJiJfAzwAxwQsd8K6uxn5KZ24HtAOPj49lqtWoF2bZjJ1v31duV/RvqbbOu6elp6u7nIJi3d4YpK5i3l/qVtcldOp8CXghQlf2zge/TPtJfV40fA7wU+EyzmJKkproq/Ii4GLgSOC0iNkbECuBW4NyIuAl4D/DvM/PHmfk54K7qvP424Dov2ErS4HV1HiQzdwO75wwfBv7DAvO/q2EuSdIS841XklQIC1+SCmHhS1IhLHxJKoSFL0mFsPAlqRAWviQVwsKXpEJY+JJUCAtfkgph4UtSISx8SSqEhS9JhbDwJakQFr4kFcLCl6RCWPiSVIijFn5EnBoRH4qIe+f52YaIyIgY6Rj7hYh4f0TcEhH/ZakDS5Lq6eYrDi8EdgLndg5GxEtof0F559jzgA8AL8vMxyLijoi4JDM/tVSBJUn1HPUIPzNvBx7pHKuK/S3Ab8+ZfR3wrcx8rJq+G1i/BDklSQ1FZh59pogWsCUzx6vp9wDbMvMbEZHACZk5GxGvBS7PzF+u5vs1oJWZVyyw3klgEmB0dHRsamqq1k7MHDzEgcO1FmXt6hPrLVjT7OwsIyMjR59xmTBv7wxTVjBvL9XNOjExsfdIL3ejm1M6TxMRPwOcBPxqRBwZvi4i/gyYAU7omH1lNTavzNwObAcYHx/PVqu12DgAbNuxk637Fr0rAOzfUG+bdU1PT1N3PwfBvL0zTFnBvL3Ur6yLbsnM/DZw1ZHpiHgn8O7qCP95wBkR8dzqtM4FwPuXKqwkqb5u7tK5GLgSOC0iNkbEimr8lIjYWM32lohYnZk/An4d+L2I2AR8yQu2krQ8HPUIPzN3A7vnGX8Q2FT96Rz/JPDJpQooSVoavvFKkgph4UtSISx8SSqEhS9JhbDwJakQFr4kFcLCl6RCWPiSVAgLX5IKYeFLUiEsfEkqhIUvSYWw8CWpEBa+JBXCwpekQlj4klQIC1+SCtHVd9pGxKm0v9nqnMw8vxq7ATgV+D4wBtycmV+rfnYFcB7wJPBAZn6wB9klSYvQ7ZeYXwjsBM7tGBsBrsvMjIjLgXcBl0XE6cD1wHnVz+6NiE9n5t8saXJJ0qJ0dUonM28HHpkz9rbMzI71zFaPLwX2dvzsHuCVS5BVktRAPNXLR5kxogVsyczxOePPAT4GvCEzH4iItwKjmXlt9fNNAJm5cZ51TgKTAKOjo2NTU1O1dmLm4CEOHK61KGtXn1hvwZpmZ2cZGRnp6zabMG/vDFNWMG8v1c06MTGxd24nP5NuT+nMqyr7PwB+KzMfqIZngLM6ZlsJ3D/f8pm5HdgOMD4+nq1Wq1aObTt2snVfvV3Zv6HeNuuanp6m7n4Ognl7Z5iygnl7qV9Za9+lExErgA8C787MvRHxK9WP7gTGIiKq6XXAJ5rFlCQ11e1dOhcDVwKnRcRGYCuwA3g58HNVtx8P3JGZ34mILcB7IuJJ4ENesJWkweuq8DNzN7B7zvBrnmH+24DbGuSSJC0x33glSYWw8CWpEBa+JBXCwpekQlj4klQIC1+SCmHhS1IhLHxJKoSFL0mFsPAlqRAWviQVwsKXpEJY+JJUCAtfkgph4UtSISx8SSqEhS9Jhej2Kw5PBTYB52Tm+dXYccAW4LvA2cDmzLyv+tkVwHnAk8ADmfnBHmSXJC1CV4UPXAjsBM7tGLsW+NvM/N2IWAt8GLgoIk4HrgfOy8yMiHsj4tN+r60kDVZXp3Qy83bgkTnD64F7qp/vA86JiJXApcDezMxqvnuAVy5NXElSXfFULx9lxogWsCUzx6vprwOXZ+YXqunvAC3g3wCjmXltNb4JIDM3zrPOSWASYHR0dGxqaqrWTswcPMSBw7UWZe3qE+stWNPs7CwjIyN93WYT5u2dYcoK5u2lulknJib2HunkbnR7Smc+M8AJHdMrq7EZ4Kw54/fPt4LM3A5sBxgfH89Wq1UryLYdO9m6r96u7N9Qb5t1TU9PU3c/B8G8vTNMWcG8vdSvrE3u0tkFrAOozuF/MTMfBu4ExiIiqvnWAZ9olFKS1FhXhR8RFwNXAqdFxMaIWAG8DzgjIjYCbwKuBsjM79C+e+c9EbEV+JAXbCVp8Lo6D5KZu4Hd8/zoDQvMfxtwW4NckqQl5huvJKkQFr4kFcLCl6RCWPiSVAgLX5IKYeFLUiEsfEkqhIUvSYWw8CWpEBa+JBXCwpekQlj4klQIC1+SCmHhS1IhLHxJKoSFL0mFsPAlqRBNvsSciHgzsAb4AXA27a85XAFsBr5Rjd2UmQeaxZQkNVW78CPiVOCtwMmZ+ZOI2Am8BrgI+MvM/OOIuIz299teuSRpJUm1NTml8yPg74GV1fQI8GVgPXBPNXZ3NS1JGrDIzPoLR1wJXAF8DwjaX2r+Q2A0Mx+KiGOAx4FjM/OJeZafBCYBRkdHx6ampmrlmDl4iAOH6+3D2tUn1luwptnZWUZGRvq6zSbM2zvDlBXM20t1s05MTOzNzPFu569d+BFxLvAR4J9k5hMRsRV4Engt8M8y89sRsQq4PzNXHW194+PjuWfPnlpZtu3YydZ99c5O7d/c3/8BmZ6eptVq9XWbTZi3d4YpK5i3l+pmjYhFFX6TUzqrgYMdR+7fA44DdgHrqrELqmlJ0oA1uUvnz4FXVUf2DwEvB64FHgNujYgXAWcC1zdOKUlqrHbhZ+aTtM/Zz+f1ddcrSeoN33glSYWw8CWpEBa+JBXCwpekQlj4klQIC1+SCmHhS1IhLHxJKoSFL0mFsPAlqRAWviQVwsKXpEJY+JJUCAtfkgph4UtSISx8SSqEhS9JhWjyFYcARMSLaX9x+WHgYuAWYAZ4G3A/sAZ4U2bONt2WJKm+RoUfEc8G3g1clpk/iYiPAE8AHwVuzszPR8Q1wA20/wGQJA1I01M65wMBXBMRbwUuo/2F5hPAvdU8dwPrG25HktRQZGb9hSMuBz4ArMnMQxFxG/Bl4IbMfH41z1nAdGaePs/yk8AkwOjo6NjU1FStHDMHD3HgcL19WLv6xHoL1jQ7O8vIyEhft9mEeXtnmLKCeXupbtaJiYm9mTne7fxNz+E/DHwtMw9V0/8buAhYERGR7X9NVtI+p/9TMnM7sB1gfHw8W61WrRDbduxk6756u7J/Q71t1jU9PU3d/RwE8/bOMGUF8/ZSv7I2PaXzOeAfVOfyAc6gfYR/F+3TPQAXALsabkeS1FCjI/zMPBgRNwDvjYgHgVOAtwP/Hbg5Iv4l8LPAdY2TSpIaaXxbZmb+KfCnc4b3A69rum5J0tLxjVeSVAgLX5IKYeFLUiEsfEkqhIUvSYWw8CWpEBa+JBXCwpekQlj4klQIC1+SCmHhS1IhLHxJKoSFL0mFsPAlqRAWviQVwsKXpEJY+JJUiMaFHxErIuJLEbGlmj4uIv5bRLw1Iv4oIl7UPKYkqanGX3EIbAL+T8f0tcDfZubvRsRa4MPARUuwHUlSA42O8CPiSuBu4Jsdw+uBewAycx9wTkSsbLIdSVJzkZn1Fox4KXBFZt4UEbcAI5l5fUR8Hbg8M79QzfcdoJWZ98+zjklgEmB0dHRsamqqVpaZg4c4cLjWoqxdfWK9BWuanZ1lZGSkr9tswry9M0xZwby9VDfrxMTE3swc73b+Jqd0Xg38OCJuBC4EnhMR1wIzwAkd862sxn5KZm4HtgOMj49nq9WqFWTbjp1s3VdvV/ZvqLfNuqanp6m7n4Ng3t4Zpqxg3l7qV9bahZ+Zv3PkcUQcR/sI/73V43XAZ6tz+F/MzIebR5UkNdH4om1E/ArwCtpH+K8F3gdsiYiNwFnA1U23IUlqrnHhZ+YdwB1zht/QdL2SpKXlG68kqRAWviQVYineeDXU1ty4q/ay+zevX8IkktRbHuFLUiEsfEkqhIUvSYWw8CWpEBa+JBXCwpekQlj4klQIC1+SCmHhS1IhLHxJKoSFL0mFsPAlqRAWviQVwsKXpEI0+njkiDgT2AT8NXA68MPMfHtErAI2A98AzgZuyswDTcNKkupr+nn4q4CpzNwJEBFfiYhdwOuBv8zMP46Iy4AtwJUNtyVJaqDRKZ3MvPdI2Xes71FgPXBPNXZ3NS1JGqDIzKVZUcSrgVZmvjEiHgNGM/OhiDgGeBw4NjOfmLPMJDAJMDo6OjY1NVVr2zMHD3HgcLP8daxdfeKil5mdnWVkZKQHaXrDvL0zTFnBvL1UN+vExMTezBzvdv4l+YrDiJgAJoBrq6EZ4ATgIWAl8Hdzyx4gM7cD2wHGx8ez1WrV2v62HTvZuq//39a4f0Nr0ctMT09Tdz8Hwby9M0xZwby91K+sjVsyItYDFwFvBE6LiDOAXcA64NvABdW0JGmAmt6lMwb8T2APcBdwPPD7wE3ArRHxIuBM4PqGOSVJDTUq/MzcCyx04un1TdYtSVpavvFKkgph4UtSISx8SSqEhS9Jhej/zesCYM2Nze5U3b/ZNy9LWhwLv4E6pf2mtU9wVcOyl6Q6PKUjSYWw8CWpEBa+JBXCwpekQlj4klQIC1+SCmHhS1IhLHxJKoSFL0mFsPAlqRB+tIIkNdTks7H6+blYPSv8iPgF4DW0v9A8M/O3e7WtEjX98LVuLeVn//iBb9Jg9aTwI+J5wAeAl2XmYxFxR0Rckpmf6sX2JElH16sj/HXAtzLzsWr6bmA9YOEXbJAfCT0s/8st9VJk5tKvNOK1wOWZ+cvV9K8Brcy8Ys58k8BkNfli4Os1N3ky8IOay/bbMGUF8/bSMGUF8/ZS3axnZOYp3c7cqyP8GeCEjumV1djTZOZ2YHvTjUXEnswcb7qefhimrGDeXhqmrGDeXupX1l7dlnkPcEZEPLeavgDwWz8kaYB6coSfmT+KiF8Hfi8iHgS+5AVbSRqsnt2WmZmfBD7Zq/XP0fi0UB8NU1Ywby8NU1Ywby/1JWtPLtpKkpYfP1pBkgox1B+tMMh380bEXwE/riafzMxLImIVsBn4BnA2cFNmHqjmfzPtu5VOAv4iMz9WjZ8LvAH4JvAC4PrMfCIijgO2AN+t1rU5M+9bRL5TgU3AOZl5fjW24Doj4grgPOBJ4IHM/GA1vgZ4G3A/sAZ4U2bORsSzgHcAs8AZwIcz86+qZRb9uiyQ9yrgP/LU8/zhzPzooPNGxJlV1r8GTgd+mJlv79frv9C+18h7C9DqmPV3qlOxA8tbvU7/C/gc8BzgTOB1wIpl+twulPeG5fbcApCZQ/kHeB7tv9TPrabvAC7p4/ZvmWfsA8CvVo8vAz5aPf554M+qx8cCfwM8Hwjg/wKnVj/bClxdPb4ReEv1eC3w2UXm+9dVhj0dY/Ouk3YJfIGnTvHdC5xdPf5z4J9Wj68B/mv1+N8C768erwLuA55d93VZIO9VwJp55h1oXuB84F91TH8FGOvH6/9M+14j70/9Dg86L+2zDhs7pncCG5bxc7tQ3mX33GbmUJ/SWejdvP2yNiJuiIhbIuLIdtfTviV1bp5fOjKemY8DXwVeAbwQWJGZ359nmfUdy+wDzomIld2Gy8zbgUfmDC+0zkuBvVn91lTzvDIijgUmaP8iPVO+g7SPwl9GzddlgbwA/zkiro+Im6sjaAadNzPvzcydHUPPAh6lP6//vPteMy8R8VvV83tD9ZEoA82bmT/JzE1VtmNol9rXWb7P7UJ5l91zC8N9SucFPL0gHq7G+uXWzPx8RDwb+ExEPDIn08PASdUvwQtov7Bzsz7Iwvuw0P493CDzQutcaPxk4HDHL1Q3+U5ZYLyO3cCuzHwwIl4F/AlwyXLKGxGvBu7MzK9FRD9e/0a/93Py/gmwPzMfjYj/BGwDrl4OeSPiUuA3gY9n5p7l/tzOk/cwy/C5HeYj/K7ezdsrmfn56r9PAp+lfWTZmWkl8HeZ+cQzZH2mfejF/i02xw+AFRERi8i3ZLkz85uZ+WA1+Wng4uof2GWRNyImaL/uvznPNnr1+i9Z3sz8cmY+Wv3408A/b7DtJc2bmXdm5i8CP1cV5rJ+bufmXa7P7TAX/sDezRsR/ygiru4YOpv2eeBdtE8RzM3z8SPj1VHJS4HP0L4Adbi6YDl3mV0dy6wFvpiZTY7un2mddwJjHUW5DvhE9b+cd9E+B/xM+VYBxwFfZglfl4h4Z/V8Qfs5/mb1D+zA81an8S4F3gicGhHr6M/rP+++18kbEe/qmOXI7/BA80bESztOkUL7AuYLWabP7UJ5l+NzC0N+H35E/AvaF/seBB7PPt2lExH/EPh92nc9rKR98eU62hdfbgW+Rftq/Y359DsJTqr+fCKffmX+mmqZVTx1ZX4F7Svz3wPOAt6Ri7tL52Lg3wG/CPwB7YtALLTO6mr/OO2r/ffl0+96uZn2L+TPAtflU3e9vBP4UTX+h/nUXS+Lfl0WyDsJvJz2X6K1wPs6tjGwvBExRvt0055q6Hjavw8fow+v/0L7XiPvi2lftJ6pnt+bO7YxkLzRvqPoXbT/bh0LvAT4DeDvl+lzu1DeNy635xaGvPAlSd0b5lM6kqRFsPAlqRAWviQVwsKXpEJY+JJUCAtfkgph4UtSISx8SSrE/wOj8ejjZelEQwAAAABJRU5ErkJggg==\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "finalKnownResDF['dist'].hist(bins=20)\n",
    "plt.show()\n",
    "finalRes['dist'].hist(bins=20)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "9870\n",
      "0\n",
      "0\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>geneID</th>\n",
       "      <th>smin_known</th>\n",
       "      <th>smax_known</th>\n",
       "      <th>emin_known</th>\n",
       "      <th>emax_known</th>\n",
       "      <th>parent</th>\n",
       "      <th>smin_novel</th>\n",
       "      <th>smax_novel</th>\n",
       "      <th>emin_novel</th>\n",
       "      <th>emax_novel</th>\n",
       "      <th>dist</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [geneID, smin_known, smax_known, emin_known, emax_known, parent, smin_novel, smax_novel, emin_novel, emax_novel, dist]\n",
       "Index: []"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>geneID</th>\n",
       "      <th>smin_known</th>\n",
       "      <th>smax_known</th>\n",
       "      <th>emin_known</th>\n",
       "      <th>emax_known</th>\n",
       "      <th>parent</th>\n",
       "      <th>smin_novel</th>\n",
       "      <th>smax_novel</th>\n",
       "      <th>emin_novel</th>\n",
       "      <th>emax_novel</th>\n",
       "      <th>dist</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [geneID, smin_known, smax_known, emin_known, emax_known, parent, smin_novel, smax_novel, emin_novel, emax_novel, dist]\n",
       "Index: []"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>geneID</th>\n",
       "      <th>smin_known</th>\n",
       "      <th>smax_known</th>\n",
       "      <th>emin_known</th>\n",
       "      <th>emax_known</th>\n",
       "      <th>parent</th>\n",
       "      <th>smin_novel</th>\n",
       "      <th>smax_novel</th>\n",
       "      <th>emin_novel</th>\n",
       "      <th>emax_novel</th>\n",
       "      <th>dist</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [geneID, smin_known, smax_known, emin_known, emax_known, parent, smin_novel, smax_novel, emin_novel, emax_novel, dist]\n",
       "Index: []"
      ]
     },
     "execution_count": 39,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "transIDs=pd.DataFrame([transIDs_known_only]).T\n",
    "transIDs.columns=['id']\n",
    "transIDs['cds']=transIDs['id'].isin(cdsParents_known_only)\n",
    "transIDs['cds']=transIDs['cds'].astype(int)\n",
    "transIDs['geneID']=transIDs.id.str.extract('(CHS.(\\d)*)',expand=True)[0]\n",
    "grouped=transIDs[['geneID','cds']].groupby(by='geneID').mean()\n",
    "grouped.reset_index(inplace=True)\n",
    "cdsOnlyGenes=set(grouped[grouped['cds']==1][\"geneID\"])\n",
    "print(len(cdsOnlyGenes))\n",
    "# now need to select those from the finalRes dataframe\n",
    "cdsOnlyOutOfBounds=finalKnownResDF[finalKnownResDF['geneID'].isin(cdsOnlyGenes)].reset_index(drop=True)\n",
    "print(len(set(cdsOnlyOutOfBounds[\"parent\"])))\n",
    "print(len(set(cdsOnlyOutOfBounds[\"geneID\"])))\n",
    "display(cdsOnlyOutOfBounds[cdsOnlyOutOfBounds[\"parent\"]==\"CHS.36728.3\"])\n",
    "display(cdsOnlyOutOfBounds[cdsOnlyOutOfBounds['parent']=='CHS.51125.3'])\n",
    "cdsOnlyOutOfBounds.head()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
