{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import sys\n",
    "import os\n",
    "\n",
    "annFP=\"./chess2.1.gff\"\n",
    "genesFP=\"./chess2.1.genes\"\n",
    "transcriptsFP=\"./chess2.1.transcripts\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "total number of genes: 42611\n",
      "total number of genes: 46421\n",
      "total number of transcripts: 323266\n",
      "number of protein_coding genes: 20352\n",
      "number of protein_coding transcripts: 266339\n",
      "number of lncRNA genes: 18887\n",
      "number of lncRNA transcripts: 49892\n",
      "number of antisense_RNA genes: 2144\n",
      "number of antisense_RNA transcripts: 2688\n",
      "number of misc_RNA genes: 1228\n",
      "number of misc_RNA transcripts: 4347\n",
      "total number of novel genes: 4442\n",
      "number of novel protein_coding genes: 224\n",
      "number of novel protein_coding transcripts: 317\n",
      "number of novel lncRNA genes: 2671\n",
      "number of novel lncRNA transcripts: 3333\n",
      "number of novel antisense genes: 1546\n",
      "number of novel antisense transcripts: 1917\n",
      "number of novel misc genes: 1\n",
      "number of novel misc transcripts: 1\n"
     ]
    }
   ],
   "source": [
    "listGenes=[]\n",
    "def addGene(row):\n",
    "    global listGenes\n",
    "    ids=row[\"GFF_ID\"].split(\",\")\n",
    "    listGenes=listGenes+ids\n",
    "\n",
    "gffCols=[\"seqid\",\"source\",\"type\",\"start\",\"end\",\"score\",\"strand\",\"phase\",\"attributes\"]\n",
    "df1=pd.read_csv(annFP,sep=\"\\t\",comment=\"#\",names=gffCols)\n",
    "df1t=df1[df1[\"type\"]==\"transcript\"].reset_index(drop=True)\n",
    "df1t[\"tID\"]=df1t[\"attributes\"].str.split(\"ID=\",expand=True)[1].str.split(\";\",expand=True)[0]\n",
    "df1t[\"i2\"]=df1t[\"tID\"].str.extract('(CHS.(\\d)*.)',expand=True)[0]\n",
    "df1t[\"gID\"]=df1t[\"i2\"].str.extract('(CHS.(\\d)*)',expand=True)[0]\n",
    "df1t.drop([\"i2\"],axis=1,inplace=True)\n",
    "\n",
    "df1_genes=pd.read_csv(genesFP,sep=\"\\t\")\n",
    "print(\"total number of genes: \"+str(len(df1_genes)))\n",
    "listGenes=[]\n",
    "df1_genes.apply(lambda row: addGene(row),axis=1)\n",
    "setTotalGenes=set(listGenes)\n",
    "print(\"total number of genes: \"+str(len(setTotalGenes)))\n",
    "print(\"total number of transcripts: \"+str(len(set(df1t[\"tID\"]))))\n",
    "\n",
    "print(\"number of protein_coding genes: \"+str(len(df1_genes[df1_genes[\"Gene_Type\"]==\"protein_coding\"])))\n",
    "listGenes=[]\n",
    "df1_genes[df1_genes[\"Gene_Type\"]==\"protein_coding\"].apply(lambda row: addGene(row),axis=1)\n",
    "setProteinCodingGenes=set(listGenes)\n",
    "df1t_prot=df1t[df1t[\"gID\"].isin(setProteinCodingGenes)].reset_index(drop=True)\n",
    "print(\"number of protein_coding transcripts: \"+str(len(df1t_prot)))\n",
    "\n",
    "print(\"number of lncRNA genes: \"+str(len(df1_genes[df1_genes[\"Gene_Type\"]==\"lncRNA\"])))\n",
    "listGenes=[]\n",
    "df1_genes[df1_genes[\"Gene_Type\"]==\"lncRNA\"].apply(lambda row: addGene(row),axis=1)\n",
    "setlncRNAGenes=set(listGenes)\n",
    "df1t_lncRNA=df1t[df1t[\"gID\"].isin(setlncRNAGenes)].reset_index(drop=True)\n",
    "print(\"number of lncRNA transcripts: \"+str(len(df1t_lncRNA)))\n",
    "\n",
    "print(\"number of antisense_RNA genes: \"+str(len(df1_genes[df1_genes[\"Gene_Type\"]==\"antisense_RNA\"])))\n",
    "listGenes=[]\n",
    "df1_genes[df1_genes[\"Gene_Type\"]==\"antisense_RNA\"].apply(lambda row: addGene(row),axis=1)\n",
    "setantisenseRNAGenes=set(listGenes)\n",
    "df1t_antisenseRNA=df1t[df1t[\"gID\"].isin(setantisenseRNAGenes)].reset_index(drop=True)\n",
    "print(\"number of antisense_RNA transcripts: \"+str(len(df1t_antisenseRNA)))\n",
    "\n",
    "print(\"number of misc_RNA genes: \"+str(len(df1_genes[df1_genes[\"Gene_Type\"]==\"misc_RNA\"])))\n",
    "listGenes=[]\n",
    "df1_genes[df1_genes[\"Gene_Type\"]==\"misc_RNA\"].apply(lambda row: addGene(row),axis=1)\n",
    "setmiscRNAGenes=set(listGenes)\n",
    "df1t_miscRNA=df1t[df1t[\"gID\"].isin(setmiscRNAGenes)].reset_index(drop=True)\n",
    "print(\"number of misc_RNA transcripts: \"+str(len(df1t_miscRNA)))\n",
    "\n",
    "# print(\"number of other genes: \"+str(len(df1_genes)-(len(df1_genes[df1_genes[\"Gene_Type\"]==\"protein_coding\"])+len(df1_genes[df1_genes[\"Gene_Type\"]==\"lncRNA\"]))))\n",
    "# print(\"number of other transcripts: \"+str(len(set(df1t[\"tID\"]))-(len(df1t_prot)+len(df1t_lncRNA))))\n",
    "\n",
    "novel_genes=df1_genes[(df1_genes[\"Database\"]==\"CHESS\")|(df1_genes[\"Database\"]==\"FANTOM\")]\n",
    "print(\"total number of novel genes: \"+str(len(novel_genes)))\n",
    "print(\"number of novel protein_coding genes: \"+str(len(novel_genes[novel_genes[\"Gene_Type\"]==\"protein_coding\"])))\n",
    "listGenes=[]\n",
    "novel_genes[novel_genes[\"Gene_Type\"]==\"protein_coding\"].apply(lambda row: addGene(row),axis=1)\n",
    "setNovelprotGenes=set(listGenes)\n",
    "novel_transc_prot=df1t[df1t[\"gID\"].isin(setNovelprotGenes)].reset_index(drop=True)\n",
    "print(\"number of novel protein_coding transcripts: \"+str(len(novel_transc_prot)))\n",
    "\n",
    "print(\"number of novel lncRNA genes: \"+str(len(novel_genes[novel_genes[\"Gene_Type\"]==\"lncRNA\"])))\n",
    "listGenes=[]\n",
    "novel_genes[novel_genes[\"Gene_Type\"]==\"lncRNA\"].apply(lambda row: addGene(row),axis=1)\n",
    "setNovellncRNAGenes=set(listGenes)\n",
    "novel_transc_lncRNA=df1t[df1t[\"gID\"].isin(setNovellncRNAGenes)].reset_index(drop=True)\n",
    "print(\"number of novel lncRNA transcripts: \"+str(len(novel_transc_lncRNA)))\n",
    "\n",
    "print(\"number of novel antisense genes: \"+str(len(novel_genes[novel_genes[\"Gene_Type\"]==\"antisense_RNA\"])))\n",
    "listGenes=[]\n",
    "novel_genes[novel_genes[\"Gene_Type\"]==\"antisense_RNA\"].apply(lambda row: addGene(row),axis=1)\n",
    "setNovelantiRNAGenes=set(listGenes)\n",
    "novel_transc_antiRNA=df1t[df1t[\"gID\"].isin(setNovelantiRNAGenes)].reset_index(drop=True)\n",
    "print(\"number of novel antisense transcripts: \"+str(len(novel_transc_antiRNA)))\n",
    "\n",
    "print(\"number of novel misc genes: \"+str(len(novel_genes[novel_genes[\"Gene_Type\"]==\"misc_RNA\"])))\n",
    "listGenes=[]\n",
    "novel_genes[novel_genes[\"Gene_Type\"]==\"misc_RNA\"].apply(lambda row: addGene(row),axis=1)\n",
    "setNovelmiscRNAGenes=set(listGenes)\n",
    "novel_transc_miscRNA=df1t[df1t[\"gID\"].isin(setNovelmiscRNAGenes)].reset_index(drop=True)\n",
    "print(\"number of novel misc transcripts: \"+str(len(novel_transc_miscRNA)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "total number of genes: 42611\n",
      "total number of genes: 46421\n",
      "total number of transcripts: 323258\n",
      "total number of novel transcripts: 116156\n",
      "number of protein_coding genes: 20352\n",
      "number of protein_coding genes: 22880\n",
      "number of lncRNA genes: 18887\n",
      "number of lncRNA genes: 19907\n",
      "number of antisense_RNA genes: 2144\n",
      "number of antisense_RNA genes: 2145\n",
      "number of misc_RNA genes: 1228\n",
      "number of misc_RNA genes: 1489\n"
     ]
    }
   ],
   "source": [
    "# need to find polycistronic genes/transcripts\n",
    "# needs to be done based on refseq\n",
    "dontCount=[\"CHS.59388.1\",\n",
    "            \"CHS.51943.4\",\n",
    "            \"CHS.10180.8\",\n",
    "            \"CHS.26116.2\",\n",
    "            \"CHS.48289.2\",\n",
    "            \"CHS.48289.3\",\n",
    "            \"CHS.48749.2\",\n",
    "            \"CHS.48928.2\"]\n",
    "\n",
    "listGenes=[]\n",
    "def addGene(row):\n",
    "    global listGenes\n",
    "    ids=row[\"GFF_ID\"].split(\",\")\n",
    "    listGenes=listGenes+ids\n",
    "\n",
    "gffCols=[\"seqid\",\"source\",\"type\",\"start\",\"end\",\"score\",\"strand\",\"phase\",\"attributes\"]\n",
    "df1=pd.read_csv(annFP,sep=\"\\t\",comment=\"#\",names=gffCols)\n",
    "df1t=df1[df1[\"type\"]==\"transcript\"].reset_index(drop=True)\n",
    "df1t[\"tID\"]=df1t[\"attributes\"].str.split(\"ID=\",expand=True)[1].str.split(\";\",expand=True)[0]\n",
    "df1t[\"i2\"]=df1t[\"tID\"].str.extract('(CHS.(\\d)*.)',expand=True)[0]\n",
    "df1t[\"gID\"]=df1t[\"i2\"].str.extract('(CHS.(\\d)*)',expand=True)[0]\n",
    "df1t.drop([\"i2\"],axis=1,inplace=True)\n",
    "\n",
    "df1_genes=pd.read_csv(genesFP,sep=\"\\t\")\n",
    "print(\"total number of genes: \"+str(len(df1_genes)))\n",
    "listGenes=[]\n",
    "df1_genes.apply(lambda row: addGene(row),axis=1)\n",
    "setTotalGenes=set(listGenes)\n",
    "print(\"total number of genes: \"+str(len(setTotalGenes)))\n",
    "print(\"total number of transcripts: \"+str(len(set(df1t[~df1t[\"tID\"].isin(dontCount)][\"tID\"]))))\n",
    "print(\"total number of novel transcripts: \"+str(len(set(df1t[~(df1t[\"tID\"].isin(dontCount))&(df1t['source'].isin([\"FANTOM\",\"CHESS\",\"StringTie\"]))][\"tID\"]))))\n",
    "\n",
    "print(\"number of protein_coding genes: \"+str(len(df1_genes[df1_genes[\"Gene_Type\"]==\"protein_coding\"])))\n",
    "listGenes=[]\n",
    "df1_genes[df1_genes[\"Gene_Type\"]==\"protein_coding\"].apply(lambda row: addGene(row),axis=1)\n",
    "setProteinCodingGenes=set(listGenes)\n",
    "print(\"number of protein_coding genes: \"+str(len(setProteinCodingGenes)))\n",
    "# df1t_prot=df1t[(df1t[\"gID\"].isin(setProteinCodingGenes))&~(df1t[\"tID\"].isin(dontCount))].reset_index(drop=True)\n",
    "# print(\"number of protein_coding transcripts: \"+str(len(df1t_prot)))\n",
    "\n",
    "print(\"number of lncRNA genes: \"+str(len(df1_genes[df1_genes[\"Gene_Type\"]==\"lncRNA\"])))\n",
    "listGenes=[]\n",
    "df1_genes[df1_genes[\"Gene_Type\"]==\"lncRNA\"].apply(lambda row: addGene(row),axis=1)\n",
    "setlncRNAGenes=set(listGenes)\n",
    "print(\"number of lncRNA genes: \"+str(len(setlncRNAGenes)))\n",
    "\n",
    "print(\"number of antisense_RNA genes: \"+str(len(df1_genes[df1_genes[\"Gene_Type\"]==\"antisense_RNA\"])))\n",
    "listGenes=[]\n",
    "df1_genes[df1_genes[\"Gene_Type\"]==\"antisense_RNA\"].apply(lambda row: addGene(row),axis=1)\n",
    "setantisenseRNAGenes=set(listGenes)\n",
    "print(\"number of antisense_RNA genes: \"+str(len(setantisenseRNAGenes)))\n",
    "\n",
    "print(\"number of misc_RNA genes: \"+str(len(df1_genes[df1_genes[\"Gene_Type\"]==\"misc_RNA\"])))\n",
    "listGenes=[]\n",
    "df1_genes[df1_genes[\"Gene_Type\"]==\"misc_RNA\"].apply(lambda row: addGene(row),axis=1)\n",
    "setmiscRNAGenes=set(listGenes)\n",
    "print(\"number of misc_RNA genes: \"+str(len(setmiscRNAGenes)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'CHS.58457.7'"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(set(df1t[~(df1t[\"tID\"].isin(dontCount))&(df1t['source'].isin([\"FANTOM\",\"CHESS\",\"StringTie\"]))][\"tID\"]))\n",
    "novelIDs=set(df1t[~(df1t[\"tID\"].isin(dontCount))&(df1t['source'].isin([\"FANTOM\",\"CHESS\",\"StringTie\"]))][\"tID\"])\n",
    "list(novelIDs)[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "outFP=open(\"./chess2.1.novel.transcripts\",\"w+\")\n",
    "wrong=[]\n",
    "with open(\"./chess2.1.transcripts\",\"r\") as inFP:\n",
    "    for line in inFP.readlines():\n",
    "        lineCols=line.split(\"\\t\")\n",
    "        if \",\" in lineCols[0] or \";\" in lineCols[0]:\n",
    "            print(\"wrong\")\n",
    "            wrong.append(lineCols[0])\n",
    "        else:\n",
    "            if lineCols[0] in novelIDs:\n",
    "                outFP.write(line)\n",
    "outFP.close()\n",
    "inFP.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "outFP=open(\"./chess2.1.novel.transcripts_exons.gff\",\"w+\")\n",
    "wrong=[]\n",
    "with open(\"./chess2.1.gff\",\"r\") as inFP:\n",
    "    for line in inFP.readlines():\n",
    "        if line[0]==\"#\":\n",
    "            outFP.write(line)\n",
    "            continue\n",
    "        lineCols=line.split(\"\\t\")\n",
    "        if lineCols[2]==\"exon\":\n",
    "            tid=lineCols[8].split(\"Parent=\")[-1].split(\";\")[0].rstrip(\"\\n\")\n",
    "            if tid in novelIDs:\n",
    "                outFP.write(line)\n",
    "                continue\n",
    "        if lineCols[2]==\"transcript\":\n",
    "            tid=lineCols[8].split(\"ID=\")[-1].split(\";\")[0].rstrip(\"\\n\")\n",
    "            if tid in novelIDs:\n",
    "                outFP.write(line)\n",
    "                continue\n",
    "outFP.close()\n",
    "inFP.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "number of protein-coding transcripts: 266882\n",
      "number of protein-coding transcripts: 266331\n"
     ]
    }
   ],
   "source": [
    "dft=pd.read_csv(transcriptsFP,sep=\"\\t\")\n",
    "print(\"number of protein-coding transcripts: \"+str(len(dft[(dft[\"Gene_type\"]==\"protein_coding\")&~(dft[\"TranscriptID\"].isin(dontCount))])))\n",
    "dft_protein_coding=dft[(dft[\"CHESS_gene\"].isin(setProteinCodingGenes))&(dft[\"Gene_type\"]==\"protein_coding\")&~(dft[\"TranscriptID\"].isin(dontCount))]\n",
    "print(\"number of protein-coding transcripts: \"+str(len(dft_protein_coding)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "number of misc_RNA transcripts: 4347\n",
      "number of misc_RNA transcripts: 4347\n"
     ]
    }
   ],
   "source": [
    "print(\"number of misc_RNA transcripts: \"+str(len(dft[(dft[\"Gene_type\"]==\"misc_RNA\")&~(dft[\"TranscriptID\"].isin(dontCount))])))\n",
    "dft_misc_RNA=dft[(dft[\"CHESS_gene\"].isin(setmiscRNAGenes))&(dft[\"Gene_type\"]==\"misc_RNA\")&~(dft[\"TranscriptID\"].isin(dontCount))]\n",
    "print(\"number of misc_RNA transcripts: \"+str(len(dft_misc_RNA)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "number of antisense_RNA transcripts: 2694\n",
      "number of antisense_RNA transcripts: 2688\n"
     ]
    }
   ],
   "source": [
    "print(\"number of antisense_RNA transcripts: \"+str(len(dft[(dft[\"Gene_type\"]==\"antisense_RNA\")&~(dft[\"TranscriptID\"].isin(dontCount))])))\n",
    "dft_antisense_RNA=dft[(dft[\"CHESS_gene\"].isin(setantisenseRNAGenes))&(dft[\"Gene_type\"]==\"antisense_RNA\")&~(dft[\"TranscriptID\"].isin(dontCount))]\n",
    "print(\"number of antisense_RNA transcripts: \"+str(len(dft_antisense_RNA)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "number of lncRNA transcripts: 49886\n",
      "number of lncRNA transcripts: 49892\n"
     ]
    }
   ],
   "source": [
    "print(\"number of lncRNA transcripts: \"+str(len(dft[(dft[\"Gene_type\"]==\"lncRNA\")&~(dft[\"TranscriptID\"].isin(dontCount))])))\n",
    "dft_lncRNA=dft[(dft[\"CHESS_gene\"].isin(setlncRNAGenes))&~(dft[\"TranscriptID\"].isin(dontCount))]\n",
    "print(\"number of lncRNA transcripts: \"+str(len(set(dft_lncRNA[\"TranscriptID\"]))))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene_Type</th>\n",
       "      <th>Gene_Name</th>\n",
       "      <th>GFF_ID</th>\n",
       "      <th>Location</th>\n",
       "      <th>Database</th>\n",
       "      <th>RefSeq_GeneID</th>\n",
       "      <th>Description</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>protein_coding</td>\n",
       "      <td>LOC105371921</td>\n",
       "      <td>CHS.23496</td>\n",
       "      <td>chr17:80885776-80892131:-</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>105371921</td>\n",
       "      <td>uncharacterized LOC105371921</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>lncRNA</td>\n",
       "      <td>LOC107986127</td>\n",
       "      <td>CHS.38280</td>\n",
       "      <td>chr3:127744219-127751505:-</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>107986127</td>\n",
       "      <td>uncharacterized LOC107986127</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>protein_coding</td>\n",
       "      <td>RPS11</td>\n",
       "      <td>CHS.27254</td>\n",
       "      <td>chr19:49496365-49499712:+</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>6205</td>\n",
       "      <td>ribosomal protein S11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>protein_coding</td>\n",
       "      <td>CREB3L1</td>\n",
       "      <td>CHS.8579</td>\n",
       "      <td>chr11:46276634-46321430:+</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>90993</td>\n",
       "      <td>cAMP responsive element binding protein 3-like 1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>lncRNA</td>\n",
       "      <td>LOC101930053</td>\n",
       "      <td>CHS.54970</td>\n",
       "      <td>chr9:2385141-2494223:-</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>101930053</td>\n",
       "      <td>uncharacterized LOC101930053</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>protein_coding</td>\n",
       "      <td>PNMA1</td>\n",
       "      <td>CHS.15903</td>\n",
       "      <td>chr14:73711783-73714425:-</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>9240</td>\n",
       "      <td>paraneoplastic Ma antigen 1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>lncRNA</td>\n",
       "      <td>RP11-798K23.4</td>\n",
       "      <td>CHS.44935</td>\n",
       "      <td>chr5:179503322-179515579:-</td>\n",
       "      <td>GENCODE</td>\n",
       "      <td>-</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>protein_coding</td>\n",
       "      <td>TMEM216</td>\n",
       "      <td>CHS.8887</td>\n",
       "      <td>chr11:61391687-61402809:+</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>51259</td>\n",
       "      <td>transmembrane protein 216</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>lncRNA</td>\n",
       "      <td>TRAF3IP2-AS1</td>\n",
       "      <td>CHS.47202</td>\n",
       "      <td>chr6:111482713-111602930:+</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>643749</td>\n",
       "      <td>TRAF3IP2 antisense RNA 1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>protein_coding</td>\n",
       "      <td>C10orf90</td>\n",
       "      <td>CHS.7496</td>\n",
       "      <td>chr10:126421500-126671075:-</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>118611</td>\n",
       "      <td>chromosome 10 open reading frame 90</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>lncRNA</td>\n",
       "      <td>LOC107985703</td>\n",
       "      <td>CHS.58932</td>\n",
       "      <td>chrX:145843690-145875646:+</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>107985703</td>\n",
       "      <td>uncharacterized LOC107985703</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>lncRNA</td>\n",
       "      <td>LOC101929315,RP11-561E1.1</td>\n",
       "      <td>CHS.52999</td>\n",
       "      <td>chr8:24490149-24524864:-</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>101929315</td>\n",
       "      <td>uncharacterized LOC101929315</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>antisense_RNA</td>\n",
       "      <td>-</td>\n",
       "      <td>CHS.23074</td>\n",
       "      <td>chr17:68824421-68845942:-</td>\n",
       "      <td>CHESS</td>\n",
       "      <td>-</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>lncRNA</td>\n",
       "      <td>TCONS_00029157</td>\n",
       "      <td>CHS.34878</td>\n",
       "      <td>chr21:43358055-43362349:-</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>101928399</td>\n",
       "      <td>uncharacterized LOC101928399</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>protein_coding</td>\n",
       "      <td>ERCC5</td>\n",
       "      <td>CHS.14618</td>\n",
       "      <td>chr13:102845841-102876001:+</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>2073</td>\n",
       "      <td>excision repair cross-complementation group 5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>lncRNA</td>\n",
       "      <td>RP11-96K19.5</td>\n",
       "      <td>CHS.2516</td>\n",
       "      <td>chr1:110963302-110964649:+</td>\n",
       "      <td>GENCODE</td>\n",
       "      <td>-</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>protein_coding</td>\n",
       "      <td>APBB2</td>\n",
       "      <td>CHS.40262</td>\n",
       "      <td>chr4:40804188-41216714:-</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>323</td>\n",
       "      <td>amyloid beta precursor protein binding family ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>protein_coding</td>\n",
       "      <td>KLHL13</td>\n",
       "      <td>CHS.58619</td>\n",
       "      <td>chrX:117896464-118117340:-</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>90293</td>\n",
       "      <td>kelch like family member 13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>lncRNA</td>\n",
       "      <td>-</td>\n",
       "      <td>CHS.42771</td>\n",
       "      <td>chr5:49609715-49611301:+</td>\n",
       "      <td>CHESS</td>\n",
       "      <td>-</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>lncRNA</td>\n",
       "      <td>LOC105378980</td>\n",
       "      <td>CHS.42888</td>\n",
       "      <td>chr5:56895350-56900402:-</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>105378980</td>\n",
       "      <td>uncharacterized LOC105378980</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>lncRNA</td>\n",
       "      <td>LOC102723842</td>\n",
       "      <td>CHS.50710</td>\n",
       "      <td>chr7:57866002-57868134:-</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>102723842</td>\n",
       "      <td>uncharacterized LOC102723842</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>antisense_RNA</td>\n",
       "      <td>-</td>\n",
       "      <td>CHS.58611</td>\n",
       "      <td>chrX:115844876-115884730:+</td>\n",
       "      <td>CHESS</td>\n",
       "      <td>-</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>lncRNA</td>\n",
       "      <td>LOC107984029</td>\n",
       "      <td>CHS.7037</td>\n",
       "      <td>chr10:101844640-101849768:+</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>107984029</td>\n",
       "      <td>uncharacterized LOC107984029</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>protein_coding</td>\n",
       "      <td>CADM4</td>\n",
       "      <td>CHS.26900</td>\n",
       "      <td>chr19:43622368-43641984:-</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>199731</td>\n",
       "      <td>cell adhesion molecule 4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>antisense_RNA</td>\n",
       "      <td>-</td>\n",
       "      <td>CHS.28909</td>\n",
       "      <td>chr2:20596470-20598686:+</td>\n",
       "      <td>FANTOM</td>\n",
       "      <td>-</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>lncRNA</td>\n",
       "      <td>-</td>\n",
       "      <td>CHS.7761</td>\n",
       "      <td>chr11:1451420-1454484:-</td>\n",
       "      <td>CHESS</td>\n",
       "      <td>-</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>lncRNA</td>\n",
       "      <td>LOC107985073</td>\n",
       "      <td>CHS.23091</td>\n",
       "      <td>chr17:69572642-69582123:-</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>107985073</td>\n",
       "      <td>uncharacterized LOC107985073</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>protein_coding</td>\n",
       "      <td>SLC10A7</td>\n",
       "      <td>CHS.41522</td>\n",
       "      <td>chr4:146253981-146522351:-</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>84068</td>\n",
       "      <td>solute carrier family 10 member 7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>protein_coding</td>\n",
       "      <td>CFHR5</td>\n",
       "      <td>CHS.4075</td>\n",
       "      <td>chr1:196975022-197009725:+</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>81494</td>\n",
       "      <td>complement factor H related 5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>protein_coding</td>\n",
       "      <td>OR2K2</td>\n",
       "      <td>CHS.56459</td>\n",
       "      <td>chr9:111326884-111330224:-</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>26248</td>\n",
       "      <td>olfactory receptor family 2 subfamily K member 2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42581</th>\n",
       "      <td>protein_coding</td>\n",
       "      <td>FAM49B</td>\n",
       "      <td>CHS.54502</td>\n",
       "      <td>chr8:129839593-130016697:-</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>51571</td>\n",
       "      <td>family with sequence similarity 49 member B</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42582</th>\n",
       "      <td>lncRNA</td>\n",
       "      <td>LOC105375363</td>\n",
       "      <td>CHS.51026</td>\n",
       "      <td>chr7:77653341-77682570:+</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>105375363</td>\n",
       "      <td>uncharacterized LOC105375363</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42583</th>\n",
       "      <td>lncRNA</td>\n",
       "      <td>LOC105379597</td>\n",
       "      <td>CHS.18527</td>\n",
       "      <td>chr15_KI270850v1_alt:74397-80341:-</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>105379597</td>\n",
       "      <td>uncharacterized LOC105379597</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42584</th>\n",
       "      <td>lncRNA</td>\n",
       "      <td>-</td>\n",
       "      <td>CHS.42776</td>\n",
       "      <td>chr5:49643907-49644683:+</td>\n",
       "      <td>CHESS</td>\n",
       "      <td>-</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42585</th>\n",
       "      <td>antisense_RNA</td>\n",
       "      <td>-</td>\n",
       "      <td>CHS.14002</td>\n",
       "      <td>chr13:46387367-46392782:+</td>\n",
       "      <td>FANTOM</td>\n",
       "      <td>-</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42586</th>\n",
       "      <td>lncRNA</td>\n",
       "      <td>LOC105379086</td>\n",
       "      <td>CHS.43435</td>\n",
       "      <td>chr5:94575781-94583069:+</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>105379086</td>\n",
       "      <td>uncharacterized LOC105379086</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42587</th>\n",
       "      <td>protein_coding</td>\n",
       "      <td>EAF2</td>\n",
       "      <td>CHS.38150</td>\n",
       "      <td>chr3:121835180-121886526:+</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>55840</td>\n",
       "      <td>ELL associated factor 2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42588</th>\n",
       "      <td>lncRNA</td>\n",
       "      <td>LOC107986415</td>\n",
       "      <td>CHS.42844</td>\n",
       "      <td>chr5:55377324-55378208:+</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>107986415</td>\n",
       "      <td>uncharacterized LOC107986415</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42589</th>\n",
       "      <td>protein_coding</td>\n",
       "      <td>VPS13B</td>\n",
       "      <td>CHS.54046</td>\n",
       "      <td>chr8:99007014-99877586:+</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>157680</td>\n",
       "      <td>vacuolar protein sorting 13 homolog B (yeast)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42590</th>\n",
       "      <td>protein_coding</td>\n",
       "      <td>TP53TG3B</td>\n",
       "      <td>CHS.19681</td>\n",
       "      <td>chr16:33358385-33363478:+</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>729355</td>\n",
       "      <td>TP53 target 3B</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42591</th>\n",
       "      <td>lncRNA</td>\n",
       "      <td>FAM106B</td>\n",
       "      <td>CHS.21578</td>\n",
       "      <td>chr17:20417292-20419050:+</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>100996259</td>\n",
       "      <td>family with sequence similarity 106 member B</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42592</th>\n",
       "      <td>lncRNA</td>\n",
       "      <td>LOC105378768</td>\n",
       "      <td>CHS.1801</td>\n",
       "      <td>chr1:62423816-62432939:+</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>105378768</td>\n",
       "      <td>uncharacterized LOC105378768</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42593</th>\n",
       "      <td>protein_coding</td>\n",
       "      <td>GPR15</td>\n",
       "      <td>CHS.37811</td>\n",
       "      <td>chr3:98531899-98534679:+</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>2838</td>\n",
       "      <td>G protein-coupled receptor 15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42594</th>\n",
       "      <td>lncRNA</td>\n",
       "      <td>TTTY17A</td>\n",
       "      <td>CHS.59338</td>\n",
       "      <td>chrY:22851584-22852715:+</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>252949</td>\n",
       "      <td>testis-specific transcript%2C Y-linked 17A (no...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42595</th>\n",
       "      <td>protein_coding</td>\n",
       "      <td>CSF2</td>\n",
       "      <td>CHS.43864</td>\n",
       "      <td>chr5:132073792-132079241:+</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>1437</td>\n",
       "      <td>colony stimulating factor 2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42596</th>\n",
       "      <td>antisense_RNA</td>\n",
       "      <td>-</td>\n",
       "      <td>CHS.51443</td>\n",
       "      <td>chr7:102376259-102396117:-</td>\n",
       "      <td>FANTOM</td>\n",
       "      <td>-</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42597</th>\n",
       "      <td>protein_coding</td>\n",
       "      <td>SELO</td>\n",
       "      <td>CHS.36236</td>\n",
       "      <td>chr22:50200979-50217640:+</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>83642</td>\n",
       "      <td>selenoprotein O</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42598</th>\n",
       "      <td>protein_coding</td>\n",
       "      <td>GRIP2</td>\n",
       "      <td>CHS.36660</td>\n",
       "      <td>chr3:14489111-14602996:-</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>80852</td>\n",
       "      <td>glutamate receptor interacting protein 2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42599</th>\n",
       "      <td>lncRNA</td>\n",
       "      <td>LOC101927613</td>\n",
       "      <td>CHS.41465</td>\n",
       "      <td>chr4:142565058-142662044:+</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>101927613</td>\n",
       "      <td>uncharacterized LOC101927613</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42600</th>\n",
       "      <td>protein_coding</td>\n",
       "      <td>GPLD1</td>\n",
       "      <td>CHS.45614</td>\n",
       "      <td>chr6:24418371-24495163:-</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>2822</td>\n",
       "      <td>glycosylphosphatidylinositol specific phosphol...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42601</th>\n",
       "      <td>protein_coding</td>\n",
       "      <td>LOC107986873</td>\n",
       "      <td>CHS.54132</td>\n",
       "      <td>chr8:102810302-102863067:+</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>107986873</td>\n",
       "      <td>forkhead box protein L2-like</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42602</th>\n",
       "      <td>protein_coding</td>\n",
       "      <td>RXFP2</td>\n",
       "      <td>CHS.13741</td>\n",
       "      <td>chr13:31739526-31802872:+</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>122042</td>\n",
       "      <td>relaxin/insulin like family peptide receptor 2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42603</th>\n",
       "      <td>protein_coding</td>\n",
       "      <td>PCIF1</td>\n",
       "      <td>CHS.33658</td>\n",
       "      <td>chr20:45934674-45948023:+</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>63935</td>\n",
       "      <td>PDX1 C-terminal inhibiting factor 1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42604</th>\n",
       "      <td>protein_coding</td>\n",
       "      <td>SNRPD2</td>\n",
       "      <td>CHS.27026</td>\n",
       "      <td>chr19:45687454-45692333:-</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>6633</td>\n",
       "      <td>small nuclear ribonucleoprotein D2 polypeptide</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42605</th>\n",
       "      <td>protein_coding</td>\n",
       "      <td>AQP7</td>\n",
       "      <td>CHS.55319</td>\n",
       "      <td>chr9:33383135-33402682:-</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>364</td>\n",
       "      <td>aquaporin 7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42606</th>\n",
       "      <td>lncRNA</td>\n",
       "      <td>LOC105378347</td>\n",
       "      <td>CHS.6394</td>\n",
       "      <td>chr10:69806962-69822323:-</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>105378347</td>\n",
       "      <td>uncharacterized LOC105378347</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42607</th>\n",
       "      <td>lncRNA</td>\n",
       "      <td>LOC105375541</td>\n",
       "      <td>CHS.52065</td>\n",
       "      <td>chr7:142812612-142822982:+</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>105375541</td>\n",
       "      <td>uncharacterized LOC105375541</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42608</th>\n",
       "      <td>protein_coding</td>\n",
       "      <td>AC011043.2</td>\n",
       "      <td>CHS.57331</td>\n",
       "      <td>chrUn_GL000195v1:173872-179372:+</td>\n",
       "      <td>GENCODE</td>\n",
       "      <td>-</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42609</th>\n",
       "      <td>protein_coding</td>\n",
       "      <td>AC011841.1</td>\n",
       "      <td>CHS.23667</td>\n",
       "      <td>chr17_GL000205v2_random:99351-104855:-</td>\n",
       "      <td>GENCODE</td>\n",
       "      <td>-</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42610</th>\n",
       "      <td>protein_coding</td>\n",
       "      <td>AP000351.1</td>\n",
       "      <td>CHS.36369</td>\n",
       "      <td>chr22_KI270879v1_alt:259292-259749:-</td>\n",
       "      <td>GENCODE</td>\n",
       "      <td>-</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>42611 rows Ã— 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            Gene_Type                  Gene_Name     GFF_ID  \\\n",
       "0      protein_coding               LOC105371921  CHS.23496   \n",
       "1              lncRNA               LOC107986127  CHS.38280   \n",
       "2      protein_coding                      RPS11  CHS.27254   \n",
       "3      protein_coding                    CREB3L1   CHS.8579   \n",
       "4              lncRNA               LOC101930053  CHS.54970   \n",
       "5      protein_coding                      PNMA1  CHS.15903   \n",
       "6              lncRNA              RP11-798K23.4  CHS.44935   \n",
       "7      protein_coding                    TMEM216   CHS.8887   \n",
       "8              lncRNA               TRAF3IP2-AS1  CHS.47202   \n",
       "9      protein_coding                   C10orf90   CHS.7496   \n",
       "10             lncRNA               LOC107985703  CHS.58932   \n",
       "11             lncRNA  LOC101929315,RP11-561E1.1  CHS.52999   \n",
       "12      antisense_RNA                          -  CHS.23074   \n",
       "13             lncRNA             TCONS_00029157  CHS.34878   \n",
       "14     protein_coding                      ERCC5  CHS.14618   \n",
       "15             lncRNA               RP11-96K19.5   CHS.2516   \n",
       "16     protein_coding                      APBB2  CHS.40262   \n",
       "17     protein_coding                     KLHL13  CHS.58619   \n",
       "18             lncRNA                          -  CHS.42771   \n",
       "19             lncRNA               LOC105378980  CHS.42888   \n",
       "20             lncRNA               LOC102723842  CHS.50710   \n",
       "21      antisense_RNA                          -  CHS.58611   \n",
       "22             lncRNA               LOC107984029   CHS.7037   \n",
       "23     protein_coding                      CADM4  CHS.26900   \n",
       "24      antisense_RNA                          -  CHS.28909   \n",
       "25             lncRNA                          -   CHS.7761   \n",
       "26             lncRNA               LOC107985073  CHS.23091   \n",
       "27     protein_coding                    SLC10A7  CHS.41522   \n",
       "28     protein_coding                      CFHR5   CHS.4075   \n",
       "29     protein_coding                      OR2K2  CHS.56459   \n",
       "...               ...                        ...        ...   \n",
       "42581  protein_coding                     FAM49B  CHS.54502   \n",
       "42582          lncRNA               LOC105375363  CHS.51026   \n",
       "42583          lncRNA               LOC105379597  CHS.18527   \n",
       "42584          lncRNA                          -  CHS.42776   \n",
       "42585   antisense_RNA                          -  CHS.14002   \n",
       "42586          lncRNA               LOC105379086  CHS.43435   \n",
       "42587  protein_coding                       EAF2  CHS.38150   \n",
       "42588          lncRNA               LOC107986415  CHS.42844   \n",
       "42589  protein_coding                     VPS13B  CHS.54046   \n",
       "42590  protein_coding                   TP53TG3B  CHS.19681   \n",
       "42591          lncRNA                    FAM106B  CHS.21578   \n",
       "42592          lncRNA               LOC105378768   CHS.1801   \n",
       "42593  protein_coding                      GPR15  CHS.37811   \n",
       "42594          lncRNA                    TTTY17A  CHS.59338   \n",
       "42595  protein_coding                       CSF2  CHS.43864   \n",
       "42596   antisense_RNA                          -  CHS.51443   \n",
       "42597  protein_coding                       SELO  CHS.36236   \n",
       "42598  protein_coding                      GRIP2  CHS.36660   \n",
       "42599          lncRNA               LOC101927613  CHS.41465   \n",
       "42600  protein_coding                      GPLD1  CHS.45614   \n",
       "42601  protein_coding               LOC107986873  CHS.54132   \n",
       "42602  protein_coding                      RXFP2  CHS.13741   \n",
       "42603  protein_coding                      PCIF1  CHS.33658   \n",
       "42604  protein_coding                     SNRPD2  CHS.27026   \n",
       "42605  protein_coding                       AQP7  CHS.55319   \n",
       "42606          lncRNA               LOC105378347   CHS.6394   \n",
       "42607          lncRNA               LOC105375541  CHS.52065   \n",
       "42608  protein_coding                 AC011043.2  CHS.57331   \n",
       "42609  protein_coding                 AC011841.1  CHS.23667   \n",
       "42610  protein_coding                 AP000351.1  CHS.36369   \n",
       "\n",
       "                                     Location Database RefSeq_GeneID  \\\n",
       "0                   chr17:80885776-80892131:-   RefSeq     105371921   \n",
       "1                  chr3:127744219-127751505:-   RefSeq     107986127   \n",
       "2                   chr19:49496365-49499712:+   RefSeq          6205   \n",
       "3                   chr11:46276634-46321430:+   RefSeq         90993   \n",
       "4                      chr9:2385141-2494223:-   RefSeq     101930053   \n",
       "5                   chr14:73711783-73714425:-   RefSeq          9240   \n",
       "6                  chr5:179503322-179515579:-  GENCODE             -   \n",
       "7                   chr11:61391687-61402809:+   RefSeq         51259   \n",
       "8                  chr6:111482713-111602930:+   RefSeq        643749   \n",
       "9                 chr10:126421500-126671075:-   RefSeq        118611   \n",
       "10                 chrX:145843690-145875646:+   RefSeq     107985703   \n",
       "11                   chr8:24490149-24524864:-   RefSeq     101929315   \n",
       "12                  chr17:68824421-68845942:-    CHESS             -   \n",
       "13                  chr21:43358055-43362349:-   RefSeq     101928399   \n",
       "14                chr13:102845841-102876001:+   RefSeq          2073   \n",
       "15                 chr1:110963302-110964649:+  GENCODE             -   \n",
       "16                   chr4:40804188-41216714:-   RefSeq           323   \n",
       "17                 chrX:117896464-118117340:-   RefSeq         90293   \n",
       "18                   chr5:49609715-49611301:+    CHESS             -   \n",
       "19                   chr5:56895350-56900402:-   RefSeq     105378980   \n",
       "20                   chr7:57866002-57868134:-   RefSeq     102723842   \n",
       "21                 chrX:115844876-115884730:+    CHESS             -   \n",
       "22                chr10:101844640-101849768:+   RefSeq     107984029   \n",
       "23                  chr19:43622368-43641984:-   RefSeq        199731   \n",
       "24                   chr2:20596470-20598686:+   FANTOM             -   \n",
       "25                    chr11:1451420-1454484:-    CHESS             -   \n",
       "26                  chr17:69572642-69582123:-   RefSeq     107985073   \n",
       "27                 chr4:146253981-146522351:-   RefSeq         84068   \n",
       "28                 chr1:196975022-197009725:+   RefSeq         81494   \n",
       "29                 chr9:111326884-111330224:-   RefSeq         26248   \n",
       "...                                       ...      ...           ...   \n",
       "42581              chr8:129839593-130016697:-   RefSeq         51571   \n",
       "42582                chr7:77653341-77682570:+   RefSeq     105375363   \n",
       "42583      chr15_KI270850v1_alt:74397-80341:-   RefSeq     105379597   \n",
       "42584                chr5:49643907-49644683:+    CHESS             -   \n",
       "42585               chr13:46387367-46392782:+   FANTOM             -   \n",
       "42586                chr5:94575781-94583069:+   RefSeq     105379086   \n",
       "42587              chr3:121835180-121886526:+   RefSeq         55840   \n",
       "42588                chr5:55377324-55378208:+   RefSeq     107986415   \n",
       "42589                chr8:99007014-99877586:+   RefSeq        157680   \n",
       "42590               chr16:33358385-33363478:+   RefSeq        729355   \n",
       "42591               chr17:20417292-20419050:+   RefSeq     100996259   \n",
       "42592                chr1:62423816-62432939:+   RefSeq     105378768   \n",
       "42593                chr3:98531899-98534679:+   RefSeq          2838   \n",
       "42594                chrY:22851584-22852715:+   RefSeq        252949   \n",
       "42595              chr5:132073792-132079241:+   RefSeq          1437   \n",
       "42596              chr7:102376259-102396117:-   FANTOM             -   \n",
       "42597               chr22:50200979-50217640:+   RefSeq         83642   \n",
       "42598                chr3:14489111-14602996:-   RefSeq         80852   \n",
       "42599              chr4:142565058-142662044:+   RefSeq     101927613   \n",
       "42600                chr6:24418371-24495163:-   RefSeq          2822   \n",
       "42601              chr8:102810302-102863067:+   RefSeq     107986873   \n",
       "42602               chr13:31739526-31802872:+   RefSeq        122042   \n",
       "42603               chr20:45934674-45948023:+   RefSeq         63935   \n",
       "42604               chr19:45687454-45692333:-   RefSeq          6633   \n",
       "42605                chr9:33383135-33402682:-   RefSeq           364   \n",
       "42606               chr10:69806962-69822323:-   RefSeq     105378347   \n",
       "42607              chr7:142812612-142822982:+   RefSeq     105375541   \n",
       "42608        chrUn_GL000195v1:173872-179372:+  GENCODE             -   \n",
       "42609  chr17_GL000205v2_random:99351-104855:-  GENCODE             -   \n",
       "42610    chr22_KI270879v1_alt:259292-259749:-  GENCODE             -   \n",
       "\n",
       "                                             Description  \n",
       "0                           uncharacterized LOC105371921  \n",
       "1                           uncharacterized LOC107986127  \n",
       "2                                  ribosomal protein S11  \n",
       "3       cAMP responsive element binding protein 3-like 1  \n",
       "4                           uncharacterized LOC101930053  \n",
       "5                            paraneoplastic Ma antigen 1  \n",
       "6                                                    NaN  \n",
       "7                              transmembrane protein 216  \n",
       "8                               TRAF3IP2 antisense RNA 1  \n",
       "9                    chromosome 10 open reading frame 90  \n",
       "10                          uncharacterized LOC107985703  \n",
       "11                          uncharacterized LOC101929315  \n",
       "12                                                   NaN  \n",
       "13                          uncharacterized LOC101928399  \n",
       "14         excision repair cross-complementation group 5  \n",
       "15                                                   NaN  \n",
       "16     amyloid beta precursor protein binding family ...  \n",
       "17                           kelch like family member 13  \n",
       "18                                                   NaN  \n",
       "19                          uncharacterized LOC105378980  \n",
       "20                          uncharacterized LOC102723842  \n",
       "21                                                   NaN  \n",
       "22                          uncharacterized LOC107984029  \n",
       "23                              cell adhesion molecule 4  \n",
       "24                                                   NaN  \n",
       "25                                                   NaN  \n",
       "26                          uncharacterized LOC107985073  \n",
       "27                     solute carrier family 10 member 7  \n",
       "28                         complement factor H related 5  \n",
       "29      olfactory receptor family 2 subfamily K member 2  \n",
       "...                                                  ...  \n",
       "42581        family with sequence similarity 49 member B  \n",
       "42582                       uncharacterized LOC105375363  \n",
       "42583                       uncharacterized LOC105379597  \n",
       "42584                                                NaN  \n",
       "42585                                                NaN  \n",
       "42586                       uncharacterized LOC105379086  \n",
       "42587                            ELL associated factor 2  \n",
       "42588                       uncharacterized LOC107986415  \n",
       "42589      vacuolar protein sorting 13 homolog B (yeast)  \n",
       "42590                                     TP53 target 3B  \n",
       "42591       family with sequence similarity 106 member B  \n",
       "42592                       uncharacterized LOC105378768  \n",
       "42593                      G protein-coupled receptor 15  \n",
       "42594  testis-specific transcript%2C Y-linked 17A (no...  \n",
       "42595                        colony stimulating factor 2  \n",
       "42596                                                NaN  \n",
       "42597                                    selenoprotein O  \n",
       "42598           glutamate receptor interacting protein 2  \n",
       "42599                       uncharacterized LOC101927613  \n",
       "42600  glycosylphosphatidylinositol specific phosphol...  \n",
       "42601                       forkhead box protein L2-like  \n",
       "42602     relaxin/insulin like family peptide receptor 2  \n",
       "42603                PDX1 C-terminal inhibiting factor 1  \n",
       "42604     small nuclear ribonucleoprotein D2 polypeptide  \n",
       "42605                                        aquaporin 7  \n",
       "42606                       uncharacterized LOC105378347  \n",
       "42607                       uncharacterized LOC105375541  \n",
       "42608                                                NaN  \n",
       "42609                                                NaN  \n",
       "42610                                                NaN  \n",
       "\n",
       "[42611 rows x 7 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df1_genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "number of RefSeq and Gencode protein_coding genes: 20128\n",
      "number of RefSeq and Gencode protein_coding genes: 22656\n",
      "number of RefSeq and Gencode lncRNA genes: 16216\n",
      "number of RefSeq and Gencode lncRNA genes: 17236\n",
      "number of RefSeq and Gencode misc_RNA genes: 1227\n",
      "number of RefSeq and Gencode misc_RNA genes: 1488\n",
      "number of RefSeq and Gencode antisense_RNA genes: 598\n",
      "number of RefSeq and Gencode antisense_RNA genes: 598\n"
     ]
    }
   ],
   "source": [
    "print(\"number of RefSeq and Gencode protein_coding genes: \"+str(len(df1_genes[(df1_genes[\"Gene_Type\"]==\"protein_coding\")&(df1_genes[\"Database\"].isin(['GENCODE', 'RefSeq', 'RefSeq,GENCODE']))])))\n",
    "listGenes=[]\n",
    "df1_genes[(df1_genes[\"Gene_Type\"]==\"protein_coding\")&(df1_genes[\"Database\"].isin(['GENCODE', 'RefSeq', 'RefSeq,GENCODE']))].apply(lambda row: addGene(row),axis=1)\n",
    "setProteinCodingGenesREFGEN=set(listGenes)\n",
    "print(\"number of RefSeq and Gencode protein_coding genes: \"+str(len(setProteinCodingGenesREFGEN)))\n",
    "\n",
    "print(\"number of RefSeq and Gencode lncRNA genes: \"+str(len(df1_genes[(df1_genes[\"Gene_Type\"]==\"lncRNA\")&(df1_genes[\"Database\"].isin(['GENCODE', 'RefSeq', 'RefSeq,GENCODE']))])))\n",
    "listGenes=[]\n",
    "df1_genes[(df1_genes[\"Gene_Type\"]==\"lncRNA\")&(df1_genes[\"Database\"].isin(['GENCODE', 'RefSeq', 'RefSeq,GENCODE']))].apply(lambda row: addGene(row),axis=1)\n",
    "setlncRNAGenesREFGEN=set(listGenes)\n",
    "print(\"number of RefSeq and Gencode lncRNA genes: \"+str(len(setlncRNAGenesREFGEN)))\n",
    "\n",
    "print(\"number of RefSeq and Gencode misc_RNA genes: \"+str(len(df1_genes[(df1_genes[\"Gene_Type\"]==\"misc_RNA\")&(df1_genes[\"Database\"].isin(['GENCODE', 'RefSeq', 'RefSeq,GENCODE']))])))\n",
    "listGenes=[]\n",
    "df1_genes[(df1_genes[\"Gene_Type\"]==\"misc_RNA\")&(df1_genes[\"Database\"].isin(['GENCODE', 'RefSeq', 'RefSeq,GENCODE']))].apply(lambda row: addGene(row),axis=1)\n",
    "setmiscRNAGenesREFGEN=set(listGenes)\n",
    "print(\"number of RefSeq and Gencode misc_RNA genes: \"+str(len(setmiscRNAGenesREFGEN)))\n",
    "\n",
    "print(\"number of RefSeq and Gencode antisense_RNA genes: \"+str(len(df1_genes[(df1_genes[\"Gene_Type\"]==\"antisense_RNA\")&(df1_genes[\"Database\"].isin(['GENCODE', 'RefSeq', 'RefSeq,GENCODE']))])))\n",
    "listGenes=[]\n",
    "df1_genes[(df1_genes[\"Gene_Type\"]==\"antisense_RNA\")&(df1_genes[\"Database\"].isin(['GENCODE', 'RefSeq', 'RefSeq,GENCODE']))].apply(lambda row: addGene(row),axis=1)\n",
    "setantisenseRNAGenesREFGEN=set(listGenes)\n",
    "print(\"number of RefSeq and Gencode antisense_RNA genes: \"+str(len(setantisenseRNAGenesREFGEN)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "number of FANTOM-CONFIRMED NOVEL protein_coding genes: 26\n",
      "number of FANTOM-CONFIRMED NOVEL protein_coding genes: 26\n",
      "number of FANTOM-CONFIRMED NOVEL lncRNA genes: 1407\n",
      "number of FANTOM-CONFIRMED NOVEL lncRNA genes: 1407\n",
      "number of FANTOM-CONFIRMED NOVEL misc_RNA genes: 1\n",
      "number of FANTOM-CONFIRMED NOVEL misc_RNA genes: 1\n",
      "number of FANTOM-CONFIRMED NOVEL antisense_RNA genes: 494\n",
      "number of FANTOM-CONFIRMED NOVEL antisense_RNA genes: 494\n"
     ]
    }
   ],
   "source": [
    "print(\"number of FANTOM-CONFIRMED NOVEL protein_coding genes: \"+str(len(df1_genes[(df1_genes[\"Gene_Type\"]==\"protein_coding\")&(df1_genes[\"Database\"].isin(['FANTOM']))])))\n",
    "listGenes=[]\n",
    "df1_genes[(df1_genes[\"Gene_Type\"]==\"protein_coding\")&(df1_genes[\"Database\"].isin(['FANTOM']))].apply(lambda row: addGene(row),axis=1)\n",
    "setProteinCodingGenesFANTOM=set(listGenes)\n",
    "print(\"number of FANTOM-CONFIRMED NOVEL protein_coding genes: \"+str(len(setProteinCodingGenesFANTOM)))\n",
    "\n",
    "print(\"number of FANTOM-CONFIRMED NOVEL lncRNA genes: \"+str(len(df1_genes[(df1_genes[\"Gene_Type\"]==\"lncRNA\")&(df1_genes[\"Database\"].isin(['FANTOM']))])))\n",
    "listGenes=[]\n",
    "df1_genes[(df1_genes[\"Gene_Type\"]==\"lncRNA\")&(df1_genes[\"Database\"].isin(['FANTOM']))].apply(lambda row: addGene(row),axis=1)\n",
    "setlncRNAGenesFANTOM=set(listGenes)\n",
    "print(\"number of FANTOM-CONFIRMED NOVEL lncRNA genes: \"+str(len(setlncRNAGenesFANTOM)))\n",
    "\n",
    "print(\"number of FANTOM-CONFIRMED NOVEL misc_RNA genes: \"+str(len(df1_genes[(df1_genes[\"Gene_Type\"]==\"misc_RNA\")&(df1_genes[\"Database\"].isin(['FANTOM']))])))\n",
    "listGenes=[]\n",
    "df1_genes[(df1_genes[\"Gene_Type\"]==\"misc_RNA\")&(df1_genes[\"Database\"].isin(['FANTOM']))].apply(lambda row: addGene(row),axis=1)\n",
    "setmiscRNAGenesFANTOM=set(listGenes)\n",
    "print(\"number of FANTOM-CONFIRMED NOVEL misc_RNA genes: \"+str(len(setmiscRNAGenesFANTOM)))\n",
    "\n",
    "print(\"number of FANTOM-CONFIRMED NOVEL antisense_RNA genes: \"+str(len(df1_genes[(df1_genes[\"Gene_Type\"]==\"antisense_RNA\")&(df1_genes[\"Database\"].isin(['FANTOM']))])))\n",
    "listGenes=[]\n",
    "df1_genes[(df1_genes[\"Gene_Type\"]==\"antisense_RNA\")&(df1_genes[\"Database\"].isin(['FANTOM']))].apply(lambda row: addGene(row),axis=1)\n",
    "setantisenseRNAGenesFANTOM=set(listGenes)\n",
    "print(\"number of FANTOM-CONFIRMED NOVEL antisense_RNA genes: \"+str(len(setantisenseRNAGenesFANTOM)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "number of FANTOM-CONFIRMED AND FANTOM-UNCONFIRMED NOVEL protein_coding genes: 224\n",
      "number of FANTOM-CONFIRMED AND FANTOM-UNCONFIRMED NOVEL protein_coding genes: 224\n",
      "number of FANTOM-CONFIRMED AND FANTOM-UNCONFIRMED NOVEL lncRNA genes: 2671\n",
      "number of FANTOM-CONFIRMED AND FANTOM-UNCONFIRMED NOVEL lncRNA genes: 2671\n",
      "number of FANTOM-CONFIRMED AND FANTOM-UNCONFIRMED NOVEL misc_RNA genes: 1\n",
      "number of FANTOM-CONFIRMED AND FANTOM-UNCONFIRMED NOVEL misc_RNA genes: 1\n",
      "number of FANTOM-CONFIRMED AND FANTOM-UNCONFIRMED NOVEL antisense_RNA genes: 1546\n",
      "number of FANTOM-CONFIRMED AND FANTOM-UNCONFIRMED NOVEL antisense_RNA genes: 1547\n"
     ]
    }
   ],
   "source": [
    "print(\"number of FANTOM-CONFIRMED AND FANTOM-UNCONFIRMED NOVEL protein_coding genes: \"+str(len(df1_genes[(df1_genes[\"Gene_Type\"]==\"protein_coding\")&(df1_genes[\"Database\"].isin(['CHESS','FANTOM']))])))\n",
    "listGenes=[]\n",
    "df1_genes[(df1_genes[\"Gene_Type\"]==\"protein_coding\")&(df1_genes[\"Database\"].isin(['CHESS',\"FANTOM\"]))].apply(lambda row: addGene(row),axis=1)\n",
    "setProteinCodingGenesNOVEL=set(listGenes)\n",
    "print(\"number of FANTOM-CONFIRMED AND FANTOM-UNCONFIRMED NOVEL protein_coding genes: \"+str(len(setProteinCodingGenesNOVEL)))\n",
    "\n",
    "print(\"number of FANTOM-CONFIRMED AND FANTOM-UNCONFIRMED NOVEL lncRNA genes: \"+str(len(df1_genes[(df1_genes[\"Gene_Type\"]==\"lncRNA\")&(df1_genes[\"Database\"].isin(['CHESS','FANTOM']))])))\n",
    "listGenes=[]\n",
    "df1_genes[(df1_genes[\"Gene_Type\"]==\"lncRNA\")&(df1_genes[\"Database\"].isin(['CHESS',\"FANTOM\"]))].apply(lambda row: addGene(row),axis=1)\n",
    "setlncRNAGenesNOVEL=set(listGenes)\n",
    "print(\"number of FANTOM-CONFIRMED AND FANTOM-UNCONFIRMED NOVEL lncRNA genes: \"+str(len(setlncRNAGenesNOVEL)))\n",
    "\n",
    "print(\"number of FANTOM-CONFIRMED AND FANTOM-UNCONFIRMED NOVEL misc_RNA genes: \"+str(len(df1_genes[(df1_genes[\"Gene_Type\"]==\"misc_RNA\")&(df1_genes[\"Database\"].isin(['CHESS','FANTOM']))])))\n",
    "listGenes=[]\n",
    "df1_genes[(df1_genes[\"Gene_Type\"]==\"misc_RNA\")&(df1_genes[\"Database\"].isin(['CHESS',\"FANTOM\"]))].apply(lambda row: addGene(row),axis=1)\n",
    "setmiscRNAGenesNOVEL=set(listGenes)\n",
    "print(\"number of FANTOM-CONFIRMED AND FANTOM-UNCONFIRMED NOVEL misc_RNA genes: \"+str(len(setmiscRNAGenesNOVEL)))\n",
    "\n",
    "print(\"number of FANTOM-CONFIRMED AND FANTOM-UNCONFIRMED NOVEL antisense_RNA genes: \"+str(len(df1_genes[(df1_genes[\"Gene_Type\"]==\"antisense_RNA\")&(df1_genes[\"Database\"].isin(['CHESS','FANTOM']))])))\n",
    "listGenes=[]\n",
    "df1_genes[(df1_genes[\"Gene_Type\"]==\"antisense_RNA\")&(df1_genes[\"Database\"].isin(['CHESS',\"FANTOM\"]))].apply(lambda row: addGene(row),axis=1)\n",
    "setantisenseRNAGenesNOVEL=set(listGenes)\n",
    "print(\"number of FANTOM-CONFIRMED AND FANTOM-UNCONFIRMED NOVEL antisense_RNA genes: \"+str(len(setantisenseRNAGenesNOVEL)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "set(['RefSeq', 'Novel', 'GENCODE', 'FANTOM'])\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>TranscriptID</th>\n",
       "      <th>CHESS_gene</th>\n",
       "      <th>Gene_name</th>\n",
       "      <th>Gene_type</th>\n",
       "      <th>Gene_Database</th>\n",
       "      <th>GeneID</th>\n",
       "      <th>Transcript_Database</th>\n",
       "      <th>Location</th>\n",
       "      <th>No_of_exons</th>\n",
       "      <th>Transcript_Length</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CHS.45778.5</td>\n",
       "      <td>CHS.45778</td>\n",
       "      <td>ZSCAN16</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>80345</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>chr6:28124556-28130086:+</td>\n",
       "      <td>4</td>\n",
       "      <td>1336</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CHS.45778.2</td>\n",
       "      <td>CHS.45778</td>\n",
       "      <td>ZSCAN16</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>80345</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>chr6:28124556-28127572:+</td>\n",
       "      <td>3</td>\n",
       "      <td>1330</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CHS.45778.4</td>\n",
       "      <td>CHS.45778</td>\n",
       "      <td>ZSCAN16</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>80345</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>chr6:28124556-28130086:+</td>\n",
       "      <td>4</td>\n",
       "      <td>1394</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CHS.45778.3</td>\n",
       "      <td>CHS.45778</td>\n",
       "      <td>ZSCAN16</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>80345</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>chr6:28124556-28130086:+</td>\n",
       "      <td>4</td>\n",
       "      <td>1386</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CHS.45778.8</td>\n",
       "      <td>CHS.45778</td>\n",
       "      <td>-</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>-</td>\n",
       "      <td>FANTOM</td>\n",
       "      <td>chr6:28124622-28129805:+</td>\n",
       "      <td>3</td>\n",
       "      <td>1724</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>CHS.45778.1</td>\n",
       "      <td>CHS.45778</td>\n",
       "      <td>ZSCAN16</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>80345</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>chr6:28123753-28130082:+</td>\n",
       "      <td>4</td>\n",
       "      <td>2301</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>CHS.45778.6</td>\n",
       "      <td>CHS.45778</td>\n",
       "      <td>ZSCAN16</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>80345</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>chr6:28124556-28130086:+</td>\n",
       "      <td>4</td>\n",
       "      <td>1249</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>CHS.45778.7</td>\n",
       "      <td>CHS.45778</td>\n",
       "      <td>-</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>-</td>\n",
       "      <td>FANTOM</td>\n",
       "      <td>chr6:28124622-28129631:+</td>\n",
       "      <td>3</td>\n",
       "      <td>1608</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>CHS.53500.1</td>\n",
       "      <td>CHS.53500</td>\n",
       "      <td>UBXN2B</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>137886</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>chr8:58411264-58451501:+</td>\n",
       "      <td>8</td>\n",
       "      <td>5068</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>CHS.53500.6</td>\n",
       "      <td>CHS.53500</td>\n",
       "      <td>UBXN2B</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>137886</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>chr8:58411361-58451508:+</td>\n",
       "      <td>9</td>\n",
       "      <td>5069</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>CHS.53500.5</td>\n",
       "      <td>CHS.53500</td>\n",
       "      <td>UBXN2B</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>137886</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>chr8:58411361-58435106:+</td>\n",
       "      <td>6</td>\n",
       "      <td>807</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>CHS.53500.7</td>\n",
       "      <td>CHS.53500</td>\n",
       "      <td>UBXN2B</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>137886</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>chr8:58411361-58451508:+</td>\n",
       "      <td>7</td>\n",
       "      <td>4840</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>CHS.53500.10</td>\n",
       "      <td>CHS.53500</td>\n",
       "      <td>-</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>-</td>\n",
       "      <td>FANTOM</td>\n",
       "      <td>chr8:58416888-58448190:+</td>\n",
       "      <td>7</td>\n",
       "      <td>1513</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>CHS.53500.16</td>\n",
       "      <td>CHS.53500</td>\n",
       "      <td>-</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>-</td>\n",
       "      <td>FANTOM</td>\n",
       "      <td>chr8:58433191-58448190:+</td>\n",
       "      <td>5</td>\n",
       "      <td>1265</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>CHS.53500.18</td>\n",
       "      <td>CHS.53500</td>\n",
       "      <td>-</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>-</td>\n",
       "      <td>FANTOM</td>\n",
       "      <td>chr8:58433191-58449840:+</td>\n",
       "      <td>4</td>\n",
       "      <td>4066</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>CHS.53500.4</td>\n",
       "      <td>CHS.53500</td>\n",
       "      <td>UBXN2B</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>-</td>\n",
       "      <td>GENCODE</td>\n",
       "      <td>chr8:58411359-58434941:+</td>\n",
       "      <td>7</td>\n",
       "      <td>717</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>CHS.53500.3</td>\n",
       "      <td>CHS.53500</td>\n",
       "      <td>-</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>-</td>\n",
       "      <td>Novel</td>\n",
       "      <td>chr8:58411356-58451507:+</td>\n",
       "      <td>3</td>\n",
       "      <td>4337</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>CHS.53500.8</td>\n",
       "      <td>CHS.53500</td>\n",
       "      <td>-</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>-</td>\n",
       "      <td>Novel</td>\n",
       "      <td>chr8:58411381-58451507:+</td>\n",
       "      <td>7</td>\n",
       "      <td>4806</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>CHS.53500.12</td>\n",
       "      <td>CHS.53500</td>\n",
       "      <td>-</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>-</td>\n",
       "      <td>FANTOM</td>\n",
       "      <td>chr8:58429515-58435972:+</td>\n",
       "      <td>3</td>\n",
       "      <td>2817</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>CHS.53500.2</td>\n",
       "      <td>CHS.53500</td>\n",
       "      <td>-</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>-</td>\n",
       "      <td>Novel</td>\n",
       "      <td>chr8:58411351-58451507:+</td>\n",
       "      <td>11</td>\n",
       "      <td>5376</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>CHS.53500.15</td>\n",
       "      <td>CHS.53500</td>\n",
       "      <td>-</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>-</td>\n",
       "      <td>FANTOM</td>\n",
       "      <td>chr8:58430622-58448190:+</td>\n",
       "      <td>6</td>\n",
       "      <td>1788</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>CHS.53500.17</td>\n",
       "      <td>CHS.53500</td>\n",
       "      <td>-</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>-</td>\n",
       "      <td>FANTOM</td>\n",
       "      <td>chr8:58433191-58448190:+</td>\n",
       "      <td>5</td>\n",
       "      <td>1709</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>CHS.42607.4</td>\n",
       "      <td>CHS.42607</td>\n",
       "      <td>SKP2</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>6502</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>chr5:36152043-36184041:+</td>\n",
       "      <td>10</td>\n",
       "      <td>3505</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>CHS.42607.5</td>\n",
       "      <td>CHS.42607</td>\n",
       "      <td>SKP2</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>6502</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>chr5:36152043-36184041:+</td>\n",
       "      <td>10</td>\n",
       "      <td>1483</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>CHS.42607.1</td>\n",
       "      <td>CHS.42607</td>\n",
       "      <td>SKP2</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>6502</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>chr5:36151989-36181154:+</td>\n",
       "      <td>10</td>\n",
       "      <td>3892</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>CHS.42607.9</td>\n",
       "      <td>CHS.42607</td>\n",
       "      <td>SKP2</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>6502</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>chr5:36152128-36184036:+</td>\n",
       "      <td>9</td>\n",
       "      <td>3303</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>CHS.42607.8</td>\n",
       "      <td>CHS.42607</td>\n",
       "      <td>SKP2</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>6502</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>chr5:36152128-36184032:+</td>\n",
       "      <td>9</td>\n",
       "      <td>1277</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>CHS.42607.3</td>\n",
       "      <td>CHS.42607</td>\n",
       "      <td>SKP2</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>6502</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>chr5:36151989-36192792:+</td>\n",
       "      <td>10</td>\n",
       "      <td>1507</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>CHS.42607.11</td>\n",
       "      <td>CHS.42607</td>\n",
       "      <td>-</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>-</td>\n",
       "      <td>FANTOM</td>\n",
       "      <td>chr5:36166607-36183274:+</td>\n",
       "      <td>7</td>\n",
       "      <td>2038</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>CHS.42607.12</td>\n",
       "      <td>CHS.42607</td>\n",
       "      <td>-</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>-</td>\n",
       "      <td>FANTOM</td>\n",
       "      <td>chr5:36168355-36183274:+</td>\n",
       "      <td>6</td>\n",
       "      <td>1940</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>323786</th>\n",
       "      <td>CHS.29258.7</td>\n",
       "      <td>CHS.29258</td>\n",
       "      <td>-</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>-</td>\n",
       "      <td>Novel</td>\n",
       "      <td>chr2:38797730-38875882:-</td>\n",
       "      <td>24</td>\n",
       "      <td>4699</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>323787</th>\n",
       "      <td>CHS.29258.14</td>\n",
       "      <td>CHS.29258</td>\n",
       "      <td>-</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>-</td>\n",
       "      <td>FANTOM</td>\n",
       "      <td>chr2:38819049-38825962:-</td>\n",
       "      <td>3</td>\n",
       "      <td>489</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>323788</th>\n",
       "      <td>CHS.29258.9</td>\n",
       "      <td>CHS.29258</td>\n",
       "      <td>-</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>-</td>\n",
       "      <td>Novel</td>\n",
       "      <td>chr2:38797743-38867236:-</td>\n",
       "      <td>25</td>\n",
       "      <td>4880</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>323789</th>\n",
       "      <td>CHS.29258.1</td>\n",
       "      <td>CHS.29258</td>\n",
       "      <td>-</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>-</td>\n",
       "      <td>Novel</td>\n",
       "      <td>chr2:38797729-38867214:-</td>\n",
       "      <td>24</td>\n",
       "      <td>4676</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>323790</th>\n",
       "      <td>CHS.29258.15</td>\n",
       "      <td>CHS.29258</td>\n",
       "      <td>DHX57</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>-</td>\n",
       "      <td>GENCODE</td>\n",
       "      <td>chr2:38853045-38875879:-</td>\n",
       "      <td>8</td>\n",
       "      <td>4016</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>323791</th>\n",
       "      <td>CHS.28135.1</td>\n",
       "      <td>CHS.28135</td>\n",
       "      <td>KIR2DS2</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>100132285</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>chr19_KI270882v1_alt:194387-208698:-</td>\n",
       "      <td>6</td>\n",
       "      <td>979</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>323792</th>\n",
       "      <td>CHS.28135.2</td>\n",
       "      <td>CHS.28135</td>\n",
       "      <td>KIR2DS2</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>100132285</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>chr19_KI270882v1_alt:194387-208698:-</td>\n",
       "      <td>7</td>\n",
       "      <td>1273</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>323793</th>\n",
       "      <td>CHS.28135.3</td>\n",
       "      <td>CHS.28135</td>\n",
       "      <td>KIR2DS2</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>100132285</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>chr19_KI270882v1_alt:194387-208698:-</td>\n",
       "      <td>7</td>\n",
       "      <td>1522</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>323794</th>\n",
       "      <td>CHS.28135.4</td>\n",
       "      <td>CHS.28135</td>\n",
       "      <td>KIR2DS2</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>100132285</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>chr19_KI270882v1_alt:194387-208698:-</td>\n",
       "      <td>8</td>\n",
       "      <td>1507</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>323795</th>\n",
       "      <td>CHS.28135.5</td>\n",
       "      <td>CHS.28135</td>\n",
       "      <td>KIR2DS2</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>100132285</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>chr19_KI270882v1_alt:194387-208698:-</td>\n",
       "      <td>8</td>\n",
       "      <td>1573</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>323796</th>\n",
       "      <td>CHS.8802.1</td>\n",
       "      <td>CHS.8802</td>\n",
       "      <td>OR5A1</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>219982</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>chr11:59443169-59444116:+</td>\n",
       "      <td>1</td>\n",
       "      <td>948</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>323797</th>\n",
       "      <td>CHS.26655.15</td>\n",
       "      <td>CHS.26655</td>\n",
       "      <td>RINL</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>-</td>\n",
       "      <td>GENCODE</td>\n",
       "      <td>chr19:38869231-38869925:-</td>\n",
       "      <td>2</td>\n",
       "      <td>533</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>323798</th>\n",
       "      <td>CHS.26655.11</td>\n",
       "      <td>CHS.26655</td>\n",
       "      <td>RINL</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>-</td>\n",
       "      <td>GENCODE</td>\n",
       "      <td>chr19:38867834-38870043:-</td>\n",
       "      <td>3</td>\n",
       "      <td>1810</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>323799</th>\n",
       "      <td>CHS.26655.10</td>\n",
       "      <td>CHS.26655</td>\n",
       "      <td>RINL</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>126432</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>chr19:38867830-38878279:-</td>\n",
       "      <td>12</td>\n",
       "      <td>3062</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>323800</th>\n",
       "      <td>CHS.26655.8</td>\n",
       "      <td>CHS.26655</td>\n",
       "      <td>RINL</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>126432</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>chr19:38867829-38878273:-</td>\n",
       "      <td>11</td>\n",
       "      <td>3137</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>323801</th>\n",
       "      <td>CHS.26655.2</td>\n",
       "      <td>CHS.26655</td>\n",
       "      <td>-</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>-</td>\n",
       "      <td>FANTOM</td>\n",
       "      <td>chr19:38867674-38870070:-</td>\n",
       "      <td>4</td>\n",
       "      <td>2025</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>323802</th>\n",
       "      <td>CHS.26655.7</td>\n",
       "      <td>CHS.26655</td>\n",
       "      <td>RINL</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>126432</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>chr19:38867829-38878273:-</td>\n",
       "      <td>12</td>\n",
       "      <td>3041</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>323803</th>\n",
       "      <td>CHS.26655.16</td>\n",
       "      <td>CHS.26655</td>\n",
       "      <td>RINL</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>-</td>\n",
       "      <td>GENCODE</td>\n",
       "      <td>chr19:38869284-38869921:-</td>\n",
       "      <td>3</td>\n",
       "      <td>346</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>323804</th>\n",
       "      <td>CHS.26655.9</td>\n",
       "      <td>CHS.26655</td>\n",
       "      <td>RINL</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>126432</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>chr19:38867830-38878279:-</td>\n",
       "      <td>12</td>\n",
       "      <td>3046</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>323805</th>\n",
       "      <td>CHS.26655.6</td>\n",
       "      <td>CHS.26655</td>\n",
       "      <td>RINL</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>126432</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>chr19:38867829-38878257:-</td>\n",
       "      <td>11</td>\n",
       "      <td>3105</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>323806</th>\n",
       "      <td>CHS.26655.5</td>\n",
       "      <td>CHS.26655</td>\n",
       "      <td>RINL</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>126432</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>chr19:38867829-38878223:-</td>\n",
       "      <td>11</td>\n",
       "      <td>3207</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>323807</th>\n",
       "      <td>CHS.26655.13</td>\n",
       "      <td>CHS.26655</td>\n",
       "      <td>-</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>-</td>\n",
       "      <td>FANTOM</td>\n",
       "      <td>chr19:38867964-38870070:-</td>\n",
       "      <td>4</td>\n",
       "      <td>1762</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>323808</th>\n",
       "      <td>CHS.26655.12</td>\n",
       "      <td>CHS.26655</td>\n",
       "      <td>-</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>-</td>\n",
       "      <td>Novel</td>\n",
       "      <td>chr19:38867866-38870997:-</td>\n",
       "      <td>5</td>\n",
       "      <td>2576</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>323809</th>\n",
       "      <td>CHS.26655.17</td>\n",
       "      <td>CHS.26655</td>\n",
       "      <td>RINL</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>-</td>\n",
       "      <td>GENCODE</td>\n",
       "      <td>chr19:38873886-38878273:-</td>\n",
       "      <td>3</td>\n",
       "      <td>596</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>323810</th>\n",
       "      <td>CHS.26655.14</td>\n",
       "      <td>CHS.26655</td>\n",
       "      <td>RINL</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>-</td>\n",
       "      <td>GENCODE</td>\n",
       "      <td>chr19:38868982-38876781:-</td>\n",
       "      <td>11</td>\n",
       "      <td>1846</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>323812</th>\n",
       "      <td>CHS.57331.1</td>\n",
       "      <td>CHS.57331</td>\n",
       "      <td>AC011043.2</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>GENCODE</td>\n",
       "      <td>-</td>\n",
       "      <td>GENCODE</td>\n",
       "      <td>chrUn_GL000195v1:173872-179372:+</td>\n",
       "      <td>4</td>\n",
       "      <td>1468</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>323813</th>\n",
       "      <td>CHS.23667.1</td>\n",
       "      <td>CHS.23667</td>\n",
       "      <td>AC011841.1</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>GENCODE</td>\n",
       "      <td>-</td>\n",
       "      <td>GENCODE</td>\n",
       "      <td>chr17_GL000205v2_random:99351-104855:-</td>\n",
       "      <td>4</td>\n",
       "      <td>1468</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>323814</th>\n",
       "      <td>CHS.36369.1</td>\n",
       "      <td>CHS.36369</td>\n",
       "      <td>AP000351.1</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>GENCODE</td>\n",
       "      <td>-</td>\n",
       "      <td>GENCODE</td>\n",
       "      <td>chr22_KI270879v1_alt:259292-259749:-</td>\n",
       "      <td>2</td>\n",
       "      <td>111</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>323815</th>\n",
       "      <td>CHS.56219.1</td>\n",
       "      <td>CHS.56219</td>\n",
       "      <td>-</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>FANTOM</td>\n",
       "      <td>-</td>\n",
       "      <td>Novel</td>\n",
       "      <td>chr9:97799806-97804974:-</td>\n",
       "      <td>2</td>\n",
       "      <td>4860</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>323816</th>\n",
       "      <td>CHS.56219.2</td>\n",
       "      <td>CHS.56219</td>\n",
       "      <td>-</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>FANTOM</td>\n",
       "      <td>-</td>\n",
       "      <td>FANTOM</td>\n",
       "      <td>chr9:97802361-97803464:-</td>\n",
       "      <td>1</td>\n",
       "      <td>1104</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>266331 rows Ã— 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        TranscriptID CHESS_gene   Gene_name       Gene_type Gene_Database  \\\n",
       "0        CHS.45778.5  CHS.45778     ZSCAN16  protein_coding        RefSeq   \n",
       "1        CHS.45778.2  CHS.45778     ZSCAN16  protein_coding        RefSeq   \n",
       "2        CHS.45778.4  CHS.45778     ZSCAN16  protein_coding        RefSeq   \n",
       "3        CHS.45778.3  CHS.45778     ZSCAN16  protein_coding        RefSeq   \n",
       "4        CHS.45778.8  CHS.45778           -  protein_coding        RefSeq   \n",
       "5        CHS.45778.1  CHS.45778     ZSCAN16  protein_coding        RefSeq   \n",
       "6        CHS.45778.6  CHS.45778     ZSCAN16  protein_coding        RefSeq   \n",
       "7        CHS.45778.7  CHS.45778           -  protein_coding        RefSeq   \n",
       "9        CHS.53500.1  CHS.53500      UBXN2B  protein_coding        RefSeq   \n",
       "10       CHS.53500.6  CHS.53500      UBXN2B  protein_coding        RefSeq   \n",
       "11       CHS.53500.5  CHS.53500      UBXN2B  protein_coding        RefSeq   \n",
       "12       CHS.53500.7  CHS.53500      UBXN2B  protein_coding        RefSeq   \n",
       "13      CHS.53500.10  CHS.53500           -  protein_coding        RefSeq   \n",
       "14      CHS.53500.16  CHS.53500           -  protein_coding        RefSeq   \n",
       "15      CHS.53500.18  CHS.53500           -  protein_coding        RefSeq   \n",
       "16       CHS.53500.4  CHS.53500      UBXN2B  protein_coding        RefSeq   \n",
       "17       CHS.53500.3  CHS.53500           -  protein_coding        RefSeq   \n",
       "18       CHS.53500.8  CHS.53500           -  protein_coding        RefSeq   \n",
       "19      CHS.53500.12  CHS.53500           -  protein_coding        RefSeq   \n",
       "20       CHS.53500.2  CHS.53500           -  protein_coding        RefSeq   \n",
       "21      CHS.53500.15  CHS.53500           -  protein_coding        RefSeq   \n",
       "22      CHS.53500.17  CHS.53500           -  protein_coding        RefSeq   \n",
       "25       CHS.42607.4  CHS.42607        SKP2  protein_coding        RefSeq   \n",
       "26       CHS.42607.5  CHS.42607        SKP2  protein_coding        RefSeq   \n",
       "27       CHS.42607.1  CHS.42607        SKP2  protein_coding        RefSeq   \n",
       "28       CHS.42607.9  CHS.42607        SKP2  protein_coding        RefSeq   \n",
       "29       CHS.42607.8  CHS.42607        SKP2  protein_coding        RefSeq   \n",
       "30       CHS.42607.3  CHS.42607        SKP2  protein_coding        RefSeq   \n",
       "31      CHS.42607.11  CHS.42607           -  protein_coding        RefSeq   \n",
       "32      CHS.42607.12  CHS.42607           -  protein_coding        RefSeq   \n",
       "...              ...        ...         ...             ...           ...   \n",
       "323786   CHS.29258.7  CHS.29258           -  protein_coding        RefSeq   \n",
       "323787  CHS.29258.14  CHS.29258           -  protein_coding        RefSeq   \n",
       "323788   CHS.29258.9  CHS.29258           -  protein_coding        RefSeq   \n",
       "323789   CHS.29258.1  CHS.29258           -  protein_coding        RefSeq   \n",
       "323790  CHS.29258.15  CHS.29258       DHX57  protein_coding        RefSeq   \n",
       "323791   CHS.28135.1  CHS.28135     KIR2DS2  protein_coding        RefSeq   \n",
       "323792   CHS.28135.2  CHS.28135     KIR2DS2  protein_coding        RefSeq   \n",
       "323793   CHS.28135.3  CHS.28135     KIR2DS2  protein_coding        RefSeq   \n",
       "323794   CHS.28135.4  CHS.28135     KIR2DS2  protein_coding        RefSeq   \n",
       "323795   CHS.28135.5  CHS.28135     KIR2DS2  protein_coding        RefSeq   \n",
       "323796    CHS.8802.1   CHS.8802       OR5A1  protein_coding        RefSeq   \n",
       "323797  CHS.26655.15  CHS.26655        RINL  protein_coding        RefSeq   \n",
       "323798  CHS.26655.11  CHS.26655        RINL  protein_coding        RefSeq   \n",
       "323799  CHS.26655.10  CHS.26655        RINL  protein_coding        RefSeq   \n",
       "323800   CHS.26655.8  CHS.26655        RINL  protein_coding        RefSeq   \n",
       "323801   CHS.26655.2  CHS.26655           -  protein_coding        RefSeq   \n",
       "323802   CHS.26655.7  CHS.26655        RINL  protein_coding        RefSeq   \n",
       "323803  CHS.26655.16  CHS.26655        RINL  protein_coding        RefSeq   \n",
       "323804   CHS.26655.9  CHS.26655        RINL  protein_coding        RefSeq   \n",
       "323805   CHS.26655.6  CHS.26655        RINL  protein_coding        RefSeq   \n",
       "323806   CHS.26655.5  CHS.26655        RINL  protein_coding        RefSeq   \n",
       "323807  CHS.26655.13  CHS.26655           -  protein_coding        RefSeq   \n",
       "323808  CHS.26655.12  CHS.26655           -  protein_coding        RefSeq   \n",
       "323809  CHS.26655.17  CHS.26655        RINL  protein_coding        RefSeq   \n",
       "323810  CHS.26655.14  CHS.26655        RINL  protein_coding        RefSeq   \n",
       "323812   CHS.57331.1  CHS.57331  AC011043.2  protein_coding       GENCODE   \n",
       "323813   CHS.23667.1  CHS.23667  AC011841.1  protein_coding       GENCODE   \n",
       "323814   CHS.36369.1  CHS.36369  AP000351.1  protein_coding       GENCODE   \n",
       "323815   CHS.56219.1  CHS.56219           -  protein_coding        FANTOM   \n",
       "323816   CHS.56219.2  CHS.56219           -  protein_coding        FANTOM   \n",
       "\n",
       "           GeneID Transcript_Database                                Location  \\\n",
       "0           80345              RefSeq                chr6:28124556-28130086:+   \n",
       "1           80345              RefSeq                chr6:28124556-28127572:+   \n",
       "2           80345              RefSeq                chr6:28124556-28130086:+   \n",
       "3           80345              RefSeq                chr6:28124556-28130086:+   \n",
       "4               -              FANTOM                chr6:28124622-28129805:+   \n",
       "5           80345              RefSeq                chr6:28123753-28130082:+   \n",
       "6           80345              RefSeq                chr6:28124556-28130086:+   \n",
       "7               -              FANTOM                chr6:28124622-28129631:+   \n",
       "9          137886              RefSeq                chr8:58411264-58451501:+   \n",
       "10         137886              RefSeq                chr8:58411361-58451508:+   \n",
       "11         137886              RefSeq                chr8:58411361-58435106:+   \n",
       "12         137886              RefSeq                chr8:58411361-58451508:+   \n",
       "13              -              FANTOM                chr8:58416888-58448190:+   \n",
       "14              -              FANTOM                chr8:58433191-58448190:+   \n",
       "15              -              FANTOM                chr8:58433191-58449840:+   \n",
       "16              -             GENCODE                chr8:58411359-58434941:+   \n",
       "17              -               Novel                chr8:58411356-58451507:+   \n",
       "18              -               Novel                chr8:58411381-58451507:+   \n",
       "19              -              FANTOM                chr8:58429515-58435972:+   \n",
       "20              -               Novel                chr8:58411351-58451507:+   \n",
       "21              -              FANTOM                chr8:58430622-58448190:+   \n",
       "22              -              FANTOM                chr8:58433191-58448190:+   \n",
       "25           6502              RefSeq                chr5:36152043-36184041:+   \n",
       "26           6502              RefSeq                chr5:36152043-36184041:+   \n",
       "27           6502              RefSeq                chr5:36151989-36181154:+   \n",
       "28           6502              RefSeq                chr5:36152128-36184036:+   \n",
       "29           6502              RefSeq                chr5:36152128-36184032:+   \n",
       "30           6502              RefSeq                chr5:36151989-36192792:+   \n",
       "31              -              FANTOM                chr5:36166607-36183274:+   \n",
       "32              -              FANTOM                chr5:36168355-36183274:+   \n",
       "...           ...                 ...                                     ...   \n",
       "323786          -               Novel                chr2:38797730-38875882:-   \n",
       "323787          -              FANTOM                chr2:38819049-38825962:-   \n",
       "323788          -               Novel                chr2:38797743-38867236:-   \n",
       "323789          -               Novel                chr2:38797729-38867214:-   \n",
       "323790          -             GENCODE                chr2:38853045-38875879:-   \n",
       "323791  100132285              RefSeq    chr19_KI270882v1_alt:194387-208698:-   \n",
       "323792  100132285              RefSeq    chr19_KI270882v1_alt:194387-208698:-   \n",
       "323793  100132285              RefSeq    chr19_KI270882v1_alt:194387-208698:-   \n",
       "323794  100132285              RefSeq    chr19_KI270882v1_alt:194387-208698:-   \n",
       "323795  100132285              RefSeq    chr19_KI270882v1_alt:194387-208698:-   \n",
       "323796     219982              RefSeq               chr11:59443169-59444116:+   \n",
       "323797          -             GENCODE               chr19:38869231-38869925:-   \n",
       "323798          -             GENCODE               chr19:38867834-38870043:-   \n",
       "323799     126432              RefSeq               chr19:38867830-38878279:-   \n",
       "323800     126432              RefSeq               chr19:38867829-38878273:-   \n",
       "323801          -              FANTOM               chr19:38867674-38870070:-   \n",
       "323802     126432              RefSeq               chr19:38867829-38878273:-   \n",
       "323803          -             GENCODE               chr19:38869284-38869921:-   \n",
       "323804     126432              RefSeq               chr19:38867830-38878279:-   \n",
       "323805     126432              RefSeq               chr19:38867829-38878257:-   \n",
       "323806     126432              RefSeq               chr19:38867829-38878223:-   \n",
       "323807          -              FANTOM               chr19:38867964-38870070:-   \n",
       "323808          -               Novel               chr19:38867866-38870997:-   \n",
       "323809          -             GENCODE               chr19:38873886-38878273:-   \n",
       "323810          -             GENCODE               chr19:38868982-38876781:-   \n",
       "323812          -             GENCODE        chrUn_GL000195v1:173872-179372:+   \n",
       "323813          -             GENCODE  chr17_GL000205v2_random:99351-104855:-   \n",
       "323814          -             GENCODE    chr22_KI270879v1_alt:259292-259749:-   \n",
       "323815          -               Novel                chr9:97799806-97804974:-   \n",
       "323816          -              FANTOM                chr9:97802361-97803464:-   \n",
       "\n",
       "        No_of_exons  Transcript_Length  \n",
       "0                 4               1336  \n",
       "1                 3               1330  \n",
       "2                 4               1394  \n",
       "3                 4               1386  \n",
       "4                 3               1724  \n",
       "5                 4               2301  \n",
       "6                 4               1249  \n",
       "7                 3               1608  \n",
       "9                 8               5068  \n",
       "10                9               5069  \n",
       "11                6                807  \n",
       "12                7               4840  \n",
       "13                7               1513  \n",
       "14                5               1265  \n",
       "15                4               4066  \n",
       "16                7                717  \n",
       "17                3               4337  \n",
       "18                7               4806  \n",
       "19                3               2817  \n",
       "20               11               5376  \n",
       "21                6               1788  \n",
       "22                5               1709  \n",
       "25               10               3505  \n",
       "26               10               1483  \n",
       "27               10               3892  \n",
       "28                9               3303  \n",
       "29                9               1277  \n",
       "30               10               1507  \n",
       "31                7               2038  \n",
       "32                6               1940  \n",
       "...             ...                ...  \n",
       "323786           24               4699  \n",
       "323787            3                489  \n",
       "323788           25               4880  \n",
       "323789           24               4676  \n",
       "323790            8               4016  \n",
       "323791            6                979  \n",
       "323792            7               1273  \n",
       "323793            7               1522  \n",
       "323794            8               1507  \n",
       "323795            8               1573  \n",
       "323796            1                948  \n",
       "323797            2                533  \n",
       "323798            3               1810  \n",
       "323799           12               3062  \n",
       "323800           11               3137  \n",
       "323801            4               2025  \n",
       "323802           12               3041  \n",
       "323803            3                346  \n",
       "323804           12               3046  \n",
       "323805           11               3105  \n",
       "323806           11               3207  \n",
       "323807            4               1762  \n",
       "323808            5               2576  \n",
       "323809            3                596  \n",
       "323810           11               1846  \n",
       "323812            4               1468  \n",
       "323813            4               1468  \n",
       "323814            2                111  \n",
       "323815            2               4860  \n",
       "323816            1               1104  \n",
       "\n",
       "[266331 rows x 10 columns]"
      ]
     },
     "execution_count": 39,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print(set(dft_protein_coding[\"Transcript_Database\"]))\n",
    "dft_protein_coding"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "number of RefSeq and Gencode protein_coding transcripts: 169959\n",
      "number of RefSeq and Gencode protein_coding transcripts: 169959\n",
      "number of RefSeq and Gencode lncRNA transcripts: 34222\n",
      "number of RefSeq and Gencode lncRNA transcripts: 34222\n",
      "number of RefSeq and Gencode misc_RNA transcripts: 2284\n",
      "number of RefSeq and Gencode misc_RNA transcripts: 2284\n",
      "number of RefSeq and Gencode antisense_RNA transcripts: 637\n",
      "number of RefSeq and Gencode antisense_RNA transcripts: 637\n"
     ]
    }
   ],
   "source": [
    "print(\"number of RefSeq and Gencode protein_coding transcripts: \"+str(len(df1t[(df1t[\"tID\"].isin(set(dft_protein_coding[\"TranscriptID\"])))&(df1t[\"source\"].isin([\"BestRefSeq\",\"Curated Genomic\",\"ENSEMBL\",\"Gnomon\",\"HAVANA\",\"RefSeq\"]))])))\n",
    "print(\"number of RefSeq and Gencode protein_coding transcripts: \"+str(len(dft_protein_coding[(dft_protein_coding[\"Transcript_Database\"].isin([\"RefSeq\",\"GENCODE\"]))&~(dft_protein_coding[\"TranscriptID\"].isin(dontCount))])))\n",
    "print(\"number of RefSeq and Gencode lncRNA transcripts: \"+str(len(df1t[(df1t[\"tID\"].isin(set(dft_lncRNA[\"TranscriptID\"])))&(df1t[\"source\"].isin([\"BestRefSeq\",\"Curated Genomic\",\"ENSEMBL\",\"Gnomon\",\"HAVANA\",\"RefSeq\"]))])))\n",
    "print(\"number of RefSeq and Gencode lncRNA transcripts: \"+str(len(dft_lncRNA[(dft_lncRNA[\"Transcript_Database\"].isin([\"RefSeq\",\"GENCODE\"]))&~(dft_lncRNA[\"TranscriptID\"].isin(dontCount))])))\n",
    "print(\"number of RefSeq and Gencode misc_RNA transcripts: \"+str(len(df1t[(df1t[\"tID\"].isin(set(dft_misc_RNA[\"TranscriptID\"])))&(df1t[\"source\"].isin([\"BestRefSeq\",\"Curated Genomic\",\"ENSEMBL\",\"Gnomon\",\"HAVANA\",\"RefSeq\"]))])))\n",
    "print(\"number of RefSeq and Gencode misc_RNA transcripts: \"+str(len(dft_misc_RNA[(dft_misc_RNA[\"Transcript_Database\"].isin([\"RefSeq\",\"GENCODE\"]))&~(dft_misc_RNA[\"TranscriptID\"].isin(dontCount))])))\n",
    "print(\"number of RefSeq and Gencode antisense_RNA transcripts: \"+str(len(df1t[(df1t[\"tID\"].isin(set(dft_antisense_RNA[\"TranscriptID\"])))&(df1t[\"source\"].isin([\"BestRefSeq\",\"Curated Genomic\",\"ENSEMBL\",\"Gnomon\",\"HAVANA\",\"RefSeq\"]))])))\n",
    "print(\"number of RefSeq and Gencode antisense_RNA transcripts: \"+str(len(dft_antisense_RNA[(dft_antisense_RNA[\"Transcript_Database\"].isin([\"RefSeq\",\"GENCODE\"]))&~(dft_antisense_RNA[\"TranscriptID\"].isin(dontCount))])))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "number of StringTie and FANTOM protein_coding transcripts: 96372\n",
      "number of RefSeq and Gencode antisense_RNA transcripts: 96372\n",
      "number of StringTie and FANTOM lncRNA transcripts: 15670\n",
      "number of RefSeq and Gencode antisense_RNA transcripts: 15670\n",
      "number of StringTie and FANTOM misc_RNA transcripts: 2063\n",
      "number of RefSeq and Gencode antisense_RNA transcripts: 2063\n",
      "number of StringTie and FANTOM antisense_RNA transcripts: 2051\n",
      "number of RefSeq and Gencode antisense_RNA transcripts: 2051\n"
     ]
    }
   ],
   "source": [
    "print(\"number of StringTie and FANTOM protein_coding transcripts: \"+str(len(df1t[(df1t[\"tID\"].isin(set(dft_protein_coding[\"TranscriptID\"])))&(df1t[\"source\"].isin([\"StringTie\",\"FANTOM\"]))])))\n",
    "print(\"number of RefSeq and Gencode antisense_RNA transcripts: \"+str(len(dft_protein_coding[(dft_protein_coding[\"Transcript_Database\"].isin([\"FANTOM\",\"Novel\"]))&~(dft_protein_coding[\"TranscriptID\"].isin(dontCount))])))\n",
    "print(\"number of StringTie and FANTOM lncRNA transcripts: \"+str(len(df1t[(df1t[\"tID\"].isin(set(dft_lncRNA[\"TranscriptID\"])))&(df1t[\"source\"].isin([\"StringTie\",\"FANTOM\"]))])))\n",
    "print(\"number of RefSeq and Gencode antisense_RNA transcripts: \"+str(len(dft_lncRNA[(dft_lncRNA[\"Transcript_Database\"].isin([\"FANTOM\",\"Novel\"]))&~(dft_lncRNA[\"TranscriptID\"].isin(dontCount))])))\n",
    "print(\"number of StringTie and FANTOM misc_RNA transcripts: \"+str(len(df1t[(df1t[\"tID\"].isin(set(dft_misc_RNA[\"TranscriptID\"])))&(df1t[\"source\"].isin([\"StringTie\",\"FANTOM\"]))])))\n",
    "print(\"number of RefSeq and Gencode antisense_RNA transcripts: \"+str(len(dft_misc_RNA[(dft_misc_RNA[\"Transcript_Database\"].isin([\"FANTOM\",\"Novel\"]))&~(dft_misc_RNA[\"TranscriptID\"].isin(dontCount))])))\n",
    "print(\"number of StringTie and FANTOM antisense_RNA transcripts: \"+str(len(df1t[(df1t[\"tID\"].isin(set(dft_antisense_RNA[\"TranscriptID\"])))&(df1t[\"source\"].isin([\"StringTie\",\"FANTOM\"]))])))\n",
    "print(\"number of RefSeq and Gencode antisense_RNA transcripts: \"+str(len(dft_antisense_RNA[(dft_antisense_RNA[\"Transcript_Database\"].isin([\"FANTOM\",\"Novel\"]))&~(dft_antisense_RNA[\"TranscriptID\"].isin(dontCount))])))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "number of FANTOM protein_coding transcripts: 23102\n",
      "number of FANTOM protein_coding transcripts: 23102\n",
      "number of FANTOM lncRNA transcripts: 5832\n",
      "number of FANTOM lncRNA transcripts: 5840\n",
      "number of FANTOM misc_RNA transcripts: 476\n",
      "number of FANTOM misc_RNA transcripts: 476\n",
      "number of FANTOM antisense_RNA transcripts: 606\n",
      "number of FANTOM antisense_RNA transcripts: 606\n"
     ]
    }
   ],
   "source": [
    "print(\"number of FANTOM protein_coding transcripts: \"+str(len(df1t[(df1t[\"tID\"].isin(set(dft_protein_coding[\"TranscriptID\"])))&(df1t[\"source\"].isin([\"FANTOM\"]))])))\n",
    "print(\"number of FANTOM protein_coding transcripts: \"+str(len(dft_protein_coding[(dft_protein_coding[\"Transcript_Database\"].isin([\"FANTOM\"]))&~(dft_protein_coding[\"TranscriptID\"].isin(dontCount))])))\n",
    "print(\"number of FANTOM lncRNA transcripts: \"+str(len(df1t[(df1t[\"tID\"].isin(set(dft_lncRNA[\"TranscriptID\"])))&(df1t[\"source\"].isin([\"FANTOM\"]))])))\n",
    "print(\"number of FANTOM lncRNA transcripts: \"+str(len(dft_lncRNA[(dft_lncRNA[\"Transcript_Database\"].isin([\"FANTOM\"]))&~(dft_lncRNA[\"TranscriptID\"].isin(dontCount))])))\n",
    "print(\"number of FANTOM misc_RNA transcripts: \"+str(len(df1t[(df1t[\"tID\"].isin(set(dft_misc_RNA[\"TranscriptID\"])))&(df1t[\"source\"].isin([\"FANTOM\"]))])))\n",
    "print(\"number of FANTOM misc_RNA transcripts: \"+str(len(dft_misc_RNA[(dft_misc_RNA[\"Transcript_Database\"].isin([\"FANTOM\"]))&~(dft_misc_RNA[\"TranscriptID\"].isin(dontCount))])))\n",
    "print(\"number of FANTOM antisense_RNA transcripts: \"+str(len(df1t[(df1t[\"tID\"].isin(set(dft_antisense_RNA[\"TranscriptID\"])))&(df1t[\"source\"].isin([\"FANTOM\"]))])))\n",
    "print(\"number of FANTOM antisense_RNA transcripts: \"+str(len(dft_antisense_RNA[(dft_antisense_RNA[\"Transcript_Database\"].isin([\"FANTOM\"]))&~(dft_antisense_RNA[\"TranscriptID\"].isin(dontCount))])))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'CHS.16819.1',\n",
       " 'CHS.1933.1',\n",
       " 'CHS.1934.1',\n",
       " 'CHS.2571.2',\n",
       " 'CHS.54408.1',\n",
       " 'CHS.55723.1'}"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "set(dft_antisense_RNA[\"TranscriptID\"]).symmetric_difference(set(dft[(dft[\"Gene_type\"]==\"antisense_RNA\")&~(dft[\"TranscriptID\"].isin(dontCount))]['TranscriptID']))"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.15"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
