{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import sys\n",
    "import os\n",
    "\n",
    "annFP=\"./chess2.0.gff\"\n",
    "genesFP=\"./chess2.0.genes\"\n",
    "transcriptsFP=\"./chess2.0.transcripts\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "total number of genes: 43162\n",
      "total number of genes: 46972\n",
      "total number of transcripts: 323827\n",
      "number of protein_coding genes: 21306\n",
      "number of protein_coding transcripts: 267478\n",
      "number of lncRNA genes: 18484\n",
      "number of lncRNA transcripts: 49314\n",
      "number of antisense_RNA genes: 2144\n",
      "number of antisense_RNA transcripts: 2688\n",
      "number of misc_RNA genes: 1228\n",
      "number of misc_RNA transcripts: 4347\n",
      "total number of novel genes: 4993\n",
      "number of novel protein_coding genes: 1178\n",
      "number of novel protein_coding transcripts: 1446\n",
      "number of novel lncRNA genes: 2268\n",
      "number of novel lncRNA transcripts: 2755\n",
      "number of novel antisense genes: 1546\n",
      "number of novel antisense transcripts: 1917\n",
      "number of novel misc genes: 1\n",
      "number of novel misc transcripts: 1\n"
     ]
    }
   ],
   "source": [
    "listGenes=[]\n",
    "def addGene(row):\n",
    "    global listGenes\n",
    "    ids=row[\"GFF_ID\"].split(\",\")\n",
    "    listGenes=listGenes+ids\n",
    "\n",
    "gffCols=[\"seqid\",\"source\",\"type\",\"start\",\"end\",\"score\",\"strand\",\"phase\",\"attributes\"]\n",
    "df1=pd.read_csv(annFP,sep=\"\\t\",comment=\"#\",names=gffCols)\n",
    "df1t=df1[df1[\"type\"]==\"transcript\"].reset_index(drop=True)\n",
    "df1t[\"tID\"]=df1t[\"attributes\"].str.split(\"ID=\",expand=True)[1].str.split(\";\",expand=True)[0]\n",
    "df1t[\"i2\"]=df1t[\"tID\"].str.extract('(CHS.(\\d)*.)',expand=True)[0]\n",
    "df1t[\"gID\"]=df1t[\"i2\"].str.extract('(CHS.(\\d)*)',expand=True)[0]\n",
    "df1t.drop([\"i2\"],axis=1,inplace=True)\n",
    "\n",
    "df1_genes=pd.read_csv(genesFP,sep=\"\\t\")\n",
    "print(\"total number of genes: \"+str(len(df1_genes)))\n",
    "listGenes=[]\n",
    "df1_genes.apply(lambda row: addGene(row),axis=1)\n",
    "setTotalGenes=set(listGenes)\n",
    "print(\"total number of genes: \"+str(len(setTotalGenes)))\n",
    "print(\"total number of transcripts: \"+str(len(set(df1t[\"tID\"]))))\n",
    "\n",
    "print(\"number of protein_coding genes: \"+str(len(df1_genes[df1_genes[\"Gene_Type\"]==\"protein_coding\"])))\n",
    "listGenes=[]\n",
    "df1_genes[df1_genes[\"Gene_Type\"]==\"protein_coding\"].apply(lambda row: addGene(row),axis=1)\n",
    "setProteinCodingGenes=set(listGenes)\n",
    "df1t_prot=df1t[df1t[\"gID\"].isin(setProteinCodingGenes)].reset_index(drop=True)\n",
    "print(\"number of protein_coding transcripts: \"+str(len(df1t_prot)))\n",
    "\n",
    "print(\"number of lncRNA genes: \"+str(len(df1_genes[df1_genes[\"Gene_Type\"]==\"lncRNA\"])))\n",
    "listGenes=[]\n",
    "df1_genes[df1_genes[\"Gene_Type\"]==\"lncRNA\"].apply(lambda row: addGene(row),axis=1)\n",
    "setlncRNAGenes=set(listGenes)\n",
    "df1t_lncRNA=df1t[df1t[\"gID\"].isin(setlncRNAGenes)].reset_index(drop=True)\n",
    "print(\"number of lncRNA transcripts: \"+str(len(df1t_lncRNA)))\n",
    "\n",
    "print(\"number of antisense_RNA genes: \"+str(len(df1_genes[df1_genes[\"Gene_Type\"]==\"antisense_RNA\"])))\n",
    "listGenes=[]\n",
    "df1_genes[df1_genes[\"Gene_Type\"]==\"antisense_RNA\"].apply(lambda row: addGene(row),axis=1)\n",
    "setantisenseRNAGenes=set(listGenes)\n",
    "df1t_antisenseRNA=df1t[df1t[\"gID\"].isin(setantisenseRNAGenes)].reset_index(drop=True)\n",
    "print(\"number of antisense_RNA transcripts: \"+str(len(df1t_antisenseRNA)))\n",
    "\n",
    "print(\"number of misc_RNA genes: \"+str(len(df1_genes[df1_genes[\"Gene_Type\"]==\"misc_RNA\"])))\n",
    "listGenes=[]\n",
    "df1_genes[df1_genes[\"Gene_Type\"]==\"misc_RNA\"].apply(lambda row: addGene(row),axis=1)\n",
    "setmiscRNAGenes=set(listGenes)\n",
    "df1t_miscRNA=df1t[df1t[\"gID\"].isin(setmiscRNAGenes)].reset_index(drop=True)\n",
    "print(\"number of misc_RNA transcripts: \"+str(len(df1t_miscRNA)))\n",
    "\n",
    "# print(\"number of other genes: \"+str(len(df1_genes)-(len(df1_genes[df1_genes[\"Gene_Type\"]==\"protein_coding\"])+len(df1_genes[df1_genes[\"Gene_Type\"]==\"lncRNA\"]))))\n",
    "# print(\"number of other transcripts: \"+str(len(set(df1t[\"tID\"]))-(len(df1t_prot)+len(df1t_lncRNA))))\n",
    "\n",
    "novel_genes=df1_genes[(df1_genes[\"Database\"]==\"CHESS\")|(df1_genes[\"Database\"]==\"FANTOM\")]\n",
    "print(\"total number of novel genes: \"+str(len(novel_genes)))\n",
    "print(\"number of novel protein_coding genes: \"+str(len(novel_genes[novel_genes[\"Gene_Type\"]==\"protein_coding\"])))\n",
    "listGenes=[]\n",
    "novel_genes[novel_genes[\"Gene_Type\"]==\"protein_coding\"].apply(lambda row: addGene(row),axis=1)\n",
    "setNovelprotGenes=set(listGenes)\n",
    "novel_transc_prot=df1t[df1t[\"gID\"].isin(setNovelprotGenes)].reset_index(drop=True)\n",
    "print(\"number of novel protein_coding transcripts: \"+str(len(novel_transc_prot)))\n",
    "\n",
    "print(\"number of novel lncRNA genes: \"+str(len(novel_genes[novel_genes[\"Gene_Type\"]==\"lncRNA\"])))\n",
    "listGenes=[]\n",
    "novel_genes[novel_genes[\"Gene_Type\"]==\"lncRNA\"].apply(lambda row: addGene(row),axis=1)\n",
    "setNovellncRNAGenes=set(listGenes)\n",
    "novel_transc_lncRNA=df1t[df1t[\"gID\"].isin(setNovellncRNAGenes)].reset_index(drop=True)\n",
    "print(\"number of novel lncRNA transcripts: \"+str(len(novel_transc_lncRNA)))\n",
    "\n",
    "print(\"number of novel antisense genes: \"+str(len(novel_genes[novel_genes[\"Gene_Type\"]==\"antisense_RNA\"])))\n",
    "listGenes=[]\n",
    "novel_genes[novel_genes[\"Gene_Type\"]==\"antisense_RNA\"].apply(lambda row: addGene(row),axis=1)\n",
    "setNovelantiRNAGenes=set(listGenes)\n",
    "novel_transc_antiRNA=df1t[df1t[\"gID\"].isin(setNovelantiRNAGenes)].reset_index(drop=True)\n",
    "print(\"number of novel antisense transcripts: \"+str(len(novel_transc_antiRNA)))\n",
    "\n",
    "print(\"number of novel misc genes: \"+str(len(novel_genes[novel_genes[\"Gene_Type\"]==\"misc_RNA\"])))\n",
    "listGenes=[]\n",
    "novel_genes[novel_genes[\"Gene_Type\"]==\"misc_RNA\"].apply(lambda row: addGene(row),axis=1)\n",
    "setNovelmiscRNAGenes=set(listGenes)\n",
    "novel_transc_miscRNA=df1t[df1t[\"gID\"].isin(setNovelmiscRNAGenes)].reset_index(drop=True)\n",
    "print(\"number of novel misc transcripts: \"+str(len(novel_transc_miscRNA)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "total number of genes: 43162\n",
      "total number of genes: 46972\n",
      "total number of transcripts: 323819\n",
      "number of protein_coding genes: 21306\n",
      "number of protein_coding genes: 23834\n",
      "number of lncRNA genes: 18484\n",
      "number of lncRNA genes: 19504\n",
      "number of antisense_RNA genes: 2144\n",
      "number of antisense_RNA genes: 2145\n",
      "number of misc_RNA genes: 1228\n",
      "number of misc_RNA genes: 1489\n"
     ]
    }
   ],
   "source": [
    "# need to find polycistronic genes/transcripts\n",
    "# needs to be done based on refseq\n",
    "dontCount=[\"CHS.59388.1\",\n",
    "            \"CHS.51943.4\",\n",
    "            \"CHS.10180.8\",\n",
    "            \"CHS.26116.2\",\n",
    "            \"CHS.48289.2\",\n",
    "            \"CHS.48289.3\",\n",
    "            \"CHS.48749.2\",\n",
    "            \"CHS.48928.2\"]\n",
    "\n",
    "listGenes=[]\n",
    "def addGene(row):\n",
    "    global listGenes\n",
    "    ids=row[\"GFF_ID\"].split(\",\")\n",
    "    listGenes=listGenes+ids\n",
    "\n",
    "gffCols=[\"seqid\",\"source\",\"type\",\"start\",\"end\",\"score\",\"strand\",\"phase\",\"attributes\"]\n",
    "df1=pd.read_csv(annFP,sep=\"\\t\",comment=\"#\",names=gffCols)\n",
    "df1t=df1[df1[\"type\"]==\"transcript\"].reset_index(drop=True)\n",
    "df1t[\"tID\"]=df1t[\"attributes\"].str.split(\"ID=\",expand=True)[1].str.split(\";\",expand=True)[0]\n",
    "df1t[\"i2\"]=df1t[\"tID\"].str.extract('(CHS.(\\d)*.)',expand=True)[0]\n",
    "df1t[\"gID\"]=df1t[\"i2\"].str.extract('(CHS.(\\d)*)',expand=True)[0]\n",
    "df1t.drop([\"i2\"],axis=1,inplace=True)\n",
    "\n",
    "df1_genes=pd.read_csv(genesFP,sep=\"\\t\")\n",
    "print(\"total number of genes: \"+str(len(df1_genes)))\n",
    "listGenes=[]\n",
    "df1_genes.apply(lambda row: addGene(row),axis=1)\n",
    "setTotalGenes=set(listGenes)\n",
    "print(\"total number of genes: \"+str(len(setTotalGenes)))\n",
    "print(\"total number of transcripts: \"+str(len(set(df1t[~df1t[\"tID\"].isin(dontCount)][\"tID\"]))))\n",
    "\n",
    "print(\"number of protein_coding genes: \"+str(len(df1_genes[df1_genes[\"Gene_Type\"]==\"protein_coding\"])))\n",
    "listGenes=[]\n",
    "df1_genes[df1_genes[\"Gene_Type\"]==\"protein_coding\"].apply(lambda row: addGene(row),axis=1)\n",
    "setProteinCodingGenes=set(listGenes)\n",
    "print(\"number of protein_coding genes: \"+str(len(setProteinCodingGenes)))\n",
    "# df1t_prot=df1t[(df1t[\"gID\"].isin(setProteinCodingGenes))&~(df1t[\"tID\"].isin(dontCount))].reset_index(drop=True)\n",
    "# print(\"number of protein_coding transcripts: \"+str(len(df1t_prot)))\n",
    "\n",
    "print(\"number of lncRNA genes: \"+str(len(df1_genes[df1_genes[\"Gene_Type\"]==\"lncRNA\"])))\n",
    "listGenes=[]\n",
    "df1_genes[df1_genes[\"Gene_Type\"]==\"lncRNA\"].apply(lambda row: addGene(row),axis=1)\n",
    "setlncRNAGenes=set(listGenes)\n",
    "print(\"number of lncRNA genes: \"+str(len(setlncRNAGenes)))\n",
    "\n",
    "print(\"number of antisense_RNA genes: \"+str(len(df1_genes[df1_genes[\"Gene_Type\"]==\"antisense_RNA\"])))\n",
    "listGenes=[]\n",
    "df1_genes[df1_genes[\"Gene_Type\"]==\"antisense_RNA\"].apply(lambda row: addGene(row),axis=1)\n",
    "setantisenseRNAGenes=set(listGenes)\n",
    "print(\"number of antisense_RNA genes: \"+str(len(setantisenseRNAGenes)))\n",
    "\n",
    "print(\"number of misc_RNA genes: \"+str(len(df1_genes[df1_genes[\"Gene_Type\"]==\"misc_RNA\"])))\n",
    "listGenes=[]\n",
    "df1_genes[df1_genes[\"Gene_Type\"]==\"misc_RNA\"].apply(lambda row: addGene(row),axis=1)\n",
    "setmiscRNAGenes=set(listGenes)\n",
    "print(\"number of misc_RNA genes: \"+str(len(setmiscRNAGenes)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "number of protein-coding transcripts: 267470\n",
      "number of protein-coding transcripts: 267470\n"
     ]
    }
   ],
   "source": [
    "dft=pd.read_csv(transcriptsFP,sep=\"\\t\")\n",
    "print(\"number of protein-coding transcripts: \"+str(len(dft[(dft[\"Gene_type\"]==\"protein_coding\")&~(dft[\"TranscriptID\"].isin(dontCount))])))\n",
    "dft_protein_coding=dft[(dft[\"CHESS_gene\"].isin(setProteinCodingGenes))&(dft[\"Gene_type\"]==\"protein_coding\")&~(dft[\"TranscriptID\"].isin(dontCount))]\n",
    "print(\"number of protein-coding transcripts: \"+str(len(dft_protein_coding)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "number of misc_RNA transcripts: 4347\n",
      "number of misc_RNA transcripts: 4347\n"
     ]
    }
   ],
   "source": [
    "print(\"number of misc_RNA transcripts: \"+str(len(dft[(dft[\"Gene_type\"]==\"misc_RNA\")&~(dft[\"TranscriptID\"].isin(dontCount))])))\n",
    "dft_misc_RNA=dft[(dft[\"CHESS_gene\"].isin(setmiscRNAGenes))&(dft[\"Gene_type\"]==\"misc_RNA\")&~(dft[\"TranscriptID\"].isin(dontCount))]\n",
    "print(\"number of misc_RNA transcripts: \"+str(len(dft_misc_RNA)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "number of antisense_RNA transcripts: 2694\n",
      "number of antisense_RNA transcripts: 2688\n"
     ]
    }
   ],
   "source": [
    "print(\"number of antisense_RNA transcripts: \"+str(len(dft[(dft[\"Gene_type\"]==\"antisense_RNA\")&~(dft[\"TranscriptID\"].isin(dontCount))])))\n",
    "dft_antisense_RNA=dft[(dft[\"CHESS_gene\"].isin(setantisenseRNAGenes))&(dft[\"Gene_type\"]==\"antisense_RNA\")&~(dft[\"TranscriptID\"].isin(dontCount))]\n",
    "print(\"number of antisense_RNA transcripts: \"+str(len(dft_antisense_RNA)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "number of lncRNA transcripts: 49308\n",
      "number of lncRNA transcripts: 49314\n"
     ]
    }
   ],
   "source": [
    "print(\"number of lncRNA transcripts: \"+str(len(dft[(dft[\"Gene_type\"]==\"lncRNA\")&~(dft[\"TranscriptID\"].isin(dontCount))])))\n",
    "dft_lncRNA=dft[(dft[\"CHESS_gene\"].isin(setlncRNAGenes))&~(dft[\"TranscriptID\"].isin(dontCount))]\n",
    "print(\"number of lncRNA transcripts: \"+str(len(dft_lncRNA)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene_Type</th>\n",
       "      <th>Gene_Name</th>\n",
       "      <th>GFF_ID</th>\n",
       "      <th>Location</th>\n",
       "      <th>Database</th>\n",
       "      <th>RefSeq_GeneID</th>\n",
       "      <th>Description</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>protein_coding</td>\n",
       "      <td>LOC105371921</td>\n",
       "      <td>CHS.23496</td>\n",
       "      <td>chr17:80885776-80892131:-</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>105371921</td>\n",
       "      <td>uncharacterized LOC105371921</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>lncRNA</td>\n",
       "      <td>LOC107986127</td>\n",
       "      <td>CHS.38280</td>\n",
       "      <td>chr3:127744219-127751505:-</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>107986127</td>\n",
       "      <td>uncharacterized LOC107986127</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>protein_coding</td>\n",
       "      <td>RPS11</td>\n",
       "      <td>CHS.27254</td>\n",
       "      <td>chr19:49496365-49499712:+</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>6205</td>\n",
       "      <td>ribosomal protein S11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>protein_coding</td>\n",
       "      <td>CREB3L1</td>\n",
       "      <td>CHS.8579</td>\n",
       "      <td>chr11:46276634-46321430:+</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>90993</td>\n",
       "      <td>cAMP responsive element binding protein 3-like 1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>lncRNA</td>\n",
       "      <td>LOC101930053</td>\n",
       "      <td>CHS.54970</td>\n",
       "      <td>chr9:2385141-2494223:-</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>101930053</td>\n",
       "      <td>uncharacterized LOC101930053</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>protein_coding</td>\n",
       "      <td>PNMA1</td>\n",
       "      <td>CHS.15903</td>\n",
       "      <td>chr14:73711783-73714425:-</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>9240</td>\n",
       "      <td>paraneoplastic Ma antigen 1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>lncRNA</td>\n",
       "      <td>RP11-798K23.4</td>\n",
       "      <td>CHS.44935</td>\n",
       "      <td>chr5:179503322-179515579:-</td>\n",
       "      <td>GENCODE</td>\n",
       "      <td>-</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>protein_coding</td>\n",
       "      <td>TMEM216</td>\n",
       "      <td>CHS.8887</td>\n",
       "      <td>chr11:61391687-61402809:+</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>51259</td>\n",
       "      <td>transmembrane protein 216</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>lncRNA</td>\n",
       "      <td>TRAF3IP2-AS1</td>\n",
       "      <td>CHS.47202</td>\n",
       "      <td>chr6:111482713-111602930:+</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>643749</td>\n",
       "      <td>TRAF3IP2 antisense RNA 1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>protein_coding</td>\n",
       "      <td>C10orf90</td>\n",
       "      <td>CHS.7496</td>\n",
       "      <td>chr10:126421500-126671075:-</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>118611</td>\n",
       "      <td>chromosome 10 open reading frame 90</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>lncRNA</td>\n",
       "      <td>LOC107985703</td>\n",
       "      <td>CHS.58932</td>\n",
       "      <td>chrX:145843690-145875646:+</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>107985703</td>\n",
       "      <td>uncharacterized LOC107985703</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>lncRNA</td>\n",
       "      <td>LOC101929315,RP11-561E1.1</td>\n",
       "      <td>CHS.52999</td>\n",
       "      <td>chr8:24490149-24524864:-</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>101929315</td>\n",
       "      <td>uncharacterized LOC101929315</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>antisense_RNA</td>\n",
       "      <td>-</td>\n",
       "      <td>CHS.23074</td>\n",
       "      <td>chr17:68824421-68845942:-</td>\n",
       "      <td>CHESS</td>\n",
       "      <td>-</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>lncRNA</td>\n",
       "      <td>TCONS_00029157</td>\n",
       "      <td>CHS.34878</td>\n",
       "      <td>chr21:43358055-43362349:-</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>101928399</td>\n",
       "      <td>uncharacterized LOC101928399</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>protein_coding</td>\n",
       "      <td>ERCC5</td>\n",
       "      <td>CHS.14618</td>\n",
       "      <td>chr13:102845841-102876001:+</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>2073</td>\n",
       "      <td>excision repair cross-complementation group 5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>lncRNA</td>\n",
       "      <td>RP11-96K19.5</td>\n",
       "      <td>CHS.2516</td>\n",
       "      <td>chr1:110963302-110964649:+</td>\n",
       "      <td>GENCODE</td>\n",
       "      <td>-</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>protein_coding</td>\n",
       "      <td>APBB2</td>\n",
       "      <td>CHS.40262</td>\n",
       "      <td>chr4:40804188-41216714:-</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>323</td>\n",
       "      <td>amyloid beta precursor protein binding family ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>protein_coding</td>\n",
       "      <td>KLHL13</td>\n",
       "      <td>CHS.58619</td>\n",
       "      <td>chrX:117896464-118117340:-</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>90293</td>\n",
       "      <td>kelch like family member 13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>lncRNA</td>\n",
       "      <td>-</td>\n",
       "      <td>CHS.42771</td>\n",
       "      <td>chr5:49609715-49611301:+</td>\n",
       "      <td>CHESS</td>\n",
       "      <td>-</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>lncRNA</td>\n",
       "      <td>LOC105378980</td>\n",
       "      <td>CHS.42888</td>\n",
       "      <td>chr5:56895350-56900402:-</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>105378980</td>\n",
       "      <td>uncharacterized LOC105378980</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>lncRNA</td>\n",
       "      <td>LOC102723842</td>\n",
       "      <td>CHS.50710</td>\n",
       "      <td>chr7:57866002-57868134:-</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>102723842</td>\n",
       "      <td>uncharacterized LOC102723842</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>antisense_RNA</td>\n",
       "      <td>-</td>\n",
       "      <td>CHS.58611</td>\n",
       "      <td>chrX:115844876-115884730:+</td>\n",
       "      <td>CHESS</td>\n",
       "      <td>-</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>lncRNA</td>\n",
       "      <td>LOC107984029</td>\n",
       "      <td>CHS.7037</td>\n",
       "      <td>chr10:101844640-101849768:+</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>107984029</td>\n",
       "      <td>uncharacterized LOC107984029</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>protein_coding</td>\n",
       "      <td>CADM4</td>\n",
       "      <td>CHS.26900</td>\n",
       "      <td>chr19:43622368-43641984:-</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>199731</td>\n",
       "      <td>cell adhesion molecule 4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>antisense_RNA</td>\n",
       "      <td>-</td>\n",
       "      <td>CHS.28909</td>\n",
       "      <td>chr2:20596470-20598686:+</td>\n",
       "      <td>FANTOM</td>\n",
       "      <td>-</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>lncRNA</td>\n",
       "      <td>-</td>\n",
       "      <td>CHS.7761</td>\n",
       "      <td>chr11:1451420-1454484:-</td>\n",
       "      <td>CHESS</td>\n",
       "      <td>-</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>lncRNA</td>\n",
       "      <td>LOC107985073</td>\n",
       "      <td>CHS.23091</td>\n",
       "      <td>chr17:69572642-69582123:-</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>107985073</td>\n",
       "      <td>uncharacterized LOC107985073</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>protein_coding</td>\n",
       "      <td>SLC10A7</td>\n",
       "      <td>CHS.41522</td>\n",
       "      <td>chr4:146253981-146522351:-</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>84068</td>\n",
       "      <td>solute carrier family 10 member 7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>protein_coding</td>\n",
       "      <td>CFHR5</td>\n",
       "      <td>CHS.4075</td>\n",
       "      <td>chr1:196975022-197009725:+</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>81494</td>\n",
       "      <td>complement factor H related 5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>protein_coding</td>\n",
       "      <td>OR2K2</td>\n",
       "      <td>CHS.56459</td>\n",
       "      <td>chr9:111326884-111330224:-</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>26248</td>\n",
       "      <td>olfactory receptor family 2 subfamily K member 2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43132</th>\n",
       "      <td>protein_coding</td>\n",
       "      <td>FAM49B</td>\n",
       "      <td>CHS.54502</td>\n",
       "      <td>chr8:129839593-130016697:-</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>51571</td>\n",
       "      <td>family with sequence similarity 49 member B</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43133</th>\n",
       "      <td>lncRNA</td>\n",
       "      <td>LOC105375363</td>\n",
       "      <td>CHS.51026</td>\n",
       "      <td>chr7:77653341-77682570:+</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>105375363</td>\n",
       "      <td>uncharacterized LOC105375363</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43134</th>\n",
       "      <td>lncRNA</td>\n",
       "      <td>LOC105379597</td>\n",
       "      <td>CHS.18527</td>\n",
       "      <td>chr15_KI270850v1_alt:74397-80341:-</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>105379597</td>\n",
       "      <td>uncharacterized LOC105379597</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43135</th>\n",
       "      <td>lncRNA</td>\n",
       "      <td>-</td>\n",
       "      <td>CHS.42776</td>\n",
       "      <td>chr5:49643907-49644683:+</td>\n",
       "      <td>CHESS</td>\n",
       "      <td>-</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43136</th>\n",
       "      <td>antisense_RNA</td>\n",
       "      <td>-</td>\n",
       "      <td>CHS.14002</td>\n",
       "      <td>chr13:46387367-46392782:+</td>\n",
       "      <td>FANTOM</td>\n",
       "      <td>-</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43137</th>\n",
       "      <td>lncRNA</td>\n",
       "      <td>LOC105379086</td>\n",
       "      <td>CHS.43435</td>\n",
       "      <td>chr5:94575781-94583069:+</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>105379086</td>\n",
       "      <td>uncharacterized LOC105379086</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43138</th>\n",
       "      <td>protein_coding</td>\n",
       "      <td>EAF2</td>\n",
       "      <td>CHS.38150</td>\n",
       "      <td>chr3:121835180-121886526:+</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>55840</td>\n",
       "      <td>ELL associated factor 2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43139</th>\n",
       "      <td>lncRNA</td>\n",
       "      <td>LOC107986415</td>\n",
       "      <td>CHS.42844</td>\n",
       "      <td>chr5:55377324-55378208:+</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>107986415</td>\n",
       "      <td>uncharacterized LOC107986415</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43140</th>\n",
       "      <td>protein_coding</td>\n",
       "      <td>VPS13B</td>\n",
       "      <td>CHS.54046</td>\n",
       "      <td>chr8:99007014-99877586:+</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>157680</td>\n",
       "      <td>vacuolar protein sorting 13 homolog B (yeast)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43141</th>\n",
       "      <td>protein_coding</td>\n",
       "      <td>TP53TG3B</td>\n",
       "      <td>CHS.19681</td>\n",
       "      <td>chr16:33358385-33363478:+</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>729355</td>\n",
       "      <td>TP53 target 3B</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43142</th>\n",
       "      <td>lncRNA</td>\n",
       "      <td>FAM106B</td>\n",
       "      <td>CHS.21578</td>\n",
       "      <td>chr17:20417292-20419050:+</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>100996259</td>\n",
       "      <td>family with sequence similarity 106 member B</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43143</th>\n",
       "      <td>lncRNA</td>\n",
       "      <td>LOC105378768</td>\n",
       "      <td>CHS.1801</td>\n",
       "      <td>chr1:62423816-62432939:+</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>105378768</td>\n",
       "      <td>uncharacterized LOC105378768</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43144</th>\n",
       "      <td>protein_coding</td>\n",
       "      <td>GPR15</td>\n",
       "      <td>CHS.37811</td>\n",
       "      <td>chr3:98531899-98534679:+</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>2838</td>\n",
       "      <td>G protein-coupled receptor 15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43145</th>\n",
       "      <td>lncRNA</td>\n",
       "      <td>TTTY17A</td>\n",
       "      <td>CHS.59338</td>\n",
       "      <td>chrY:22851584-22852715:+</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>252949</td>\n",
       "      <td>testis-specific transcript%2C Y-linked 17A (no...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43146</th>\n",
       "      <td>protein_coding</td>\n",
       "      <td>CSF2</td>\n",
       "      <td>CHS.43864</td>\n",
       "      <td>chr5:132073792-132079241:+</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>1437</td>\n",
       "      <td>colony stimulating factor 2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43147</th>\n",
       "      <td>antisense_RNA</td>\n",
       "      <td>-</td>\n",
       "      <td>CHS.51443</td>\n",
       "      <td>chr7:102376259-102396117:-</td>\n",
       "      <td>FANTOM</td>\n",
       "      <td>-</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43148</th>\n",
       "      <td>protein_coding</td>\n",
       "      <td>SELO</td>\n",
       "      <td>CHS.36236</td>\n",
       "      <td>chr22:50200979-50217640:+</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>83642</td>\n",
       "      <td>selenoprotein O</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43149</th>\n",
       "      <td>protein_coding</td>\n",
       "      <td>GRIP2</td>\n",
       "      <td>CHS.36660</td>\n",
       "      <td>chr3:14489111-14602996:-</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>80852</td>\n",
       "      <td>glutamate receptor interacting protein 2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43150</th>\n",
       "      <td>lncRNA</td>\n",
       "      <td>LOC101927613</td>\n",
       "      <td>CHS.41465</td>\n",
       "      <td>chr4:142565058-142662044:+</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>101927613</td>\n",
       "      <td>uncharacterized LOC101927613</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43151</th>\n",
       "      <td>protein_coding</td>\n",
       "      <td>GPLD1</td>\n",
       "      <td>CHS.45614</td>\n",
       "      <td>chr6:24418371-24495163:-</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>2822</td>\n",
       "      <td>glycosylphosphatidylinositol specific phosphol...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43152</th>\n",
       "      <td>protein_coding</td>\n",
       "      <td>LOC107986873</td>\n",
       "      <td>CHS.54132</td>\n",
       "      <td>chr8:102810302-102863067:+</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>107986873</td>\n",
       "      <td>forkhead box protein L2-like</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43153</th>\n",
       "      <td>protein_coding</td>\n",
       "      <td>RXFP2</td>\n",
       "      <td>CHS.13741</td>\n",
       "      <td>chr13:31739526-31802872:+</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>122042</td>\n",
       "      <td>relaxin/insulin like family peptide receptor 2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43154</th>\n",
       "      <td>protein_coding</td>\n",
       "      <td>PCIF1</td>\n",
       "      <td>CHS.33658</td>\n",
       "      <td>chr20:45934674-45948023:+</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>63935</td>\n",
       "      <td>PDX1 C-terminal inhibiting factor 1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43155</th>\n",
       "      <td>protein_coding</td>\n",
       "      <td>SNRPD2</td>\n",
       "      <td>CHS.27026</td>\n",
       "      <td>chr19:45687454-45692333:-</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>6633</td>\n",
       "      <td>small nuclear ribonucleoprotein D2 polypeptide</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43156</th>\n",
       "      <td>protein_coding</td>\n",
       "      <td>AQP7</td>\n",
       "      <td>CHS.55319</td>\n",
       "      <td>chr9:33383135-33402682:-</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>364</td>\n",
       "      <td>aquaporin 7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43157</th>\n",
       "      <td>lncRNA</td>\n",
       "      <td>LOC105378347</td>\n",
       "      <td>CHS.6394</td>\n",
       "      <td>chr10:69806962-69822323:-</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>105378347</td>\n",
       "      <td>uncharacterized LOC105378347</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43158</th>\n",
       "      <td>lncRNA</td>\n",
       "      <td>LOC105375541</td>\n",
       "      <td>CHS.52065</td>\n",
       "      <td>chr7:142812612-142822982:+</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>105375541</td>\n",
       "      <td>uncharacterized LOC105375541</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43159</th>\n",
       "      <td>protein_coding</td>\n",
       "      <td>AC011043.2</td>\n",
       "      <td>CHS.57331</td>\n",
       "      <td>chrUn_GL000195v1:173872-179372:+</td>\n",
       "      <td>GENCODE</td>\n",
       "      <td>-</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43160</th>\n",
       "      <td>protein_coding</td>\n",
       "      <td>AC011841.1</td>\n",
       "      <td>CHS.23667</td>\n",
       "      <td>chr17_GL000205v2_random:99351-104855:-</td>\n",
       "      <td>GENCODE</td>\n",
       "      <td>-</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43161</th>\n",
       "      <td>protein_coding</td>\n",
       "      <td>AP000351.1</td>\n",
       "      <td>CHS.36369</td>\n",
       "      <td>chr22_KI270879v1_alt:259292-259749:-</td>\n",
       "      <td>GENCODE</td>\n",
       "      <td>-</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>43162 rows Ã— 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            Gene_Type                  Gene_Name     GFF_ID  \\\n",
       "0      protein_coding               LOC105371921  CHS.23496   \n",
       "1              lncRNA               LOC107986127  CHS.38280   \n",
       "2      protein_coding                      RPS11  CHS.27254   \n",
       "3      protein_coding                    CREB3L1   CHS.8579   \n",
       "4              lncRNA               LOC101930053  CHS.54970   \n",
       "5      protein_coding                      PNMA1  CHS.15903   \n",
       "6              lncRNA              RP11-798K23.4  CHS.44935   \n",
       "7      protein_coding                    TMEM216   CHS.8887   \n",
       "8              lncRNA               TRAF3IP2-AS1  CHS.47202   \n",
       "9      protein_coding                   C10orf90   CHS.7496   \n",
       "10             lncRNA               LOC107985703  CHS.58932   \n",
       "11             lncRNA  LOC101929315,RP11-561E1.1  CHS.52999   \n",
       "12      antisense_RNA                          -  CHS.23074   \n",
       "13             lncRNA             TCONS_00029157  CHS.34878   \n",
       "14     protein_coding                      ERCC5  CHS.14618   \n",
       "15             lncRNA               RP11-96K19.5   CHS.2516   \n",
       "16     protein_coding                      APBB2  CHS.40262   \n",
       "17     protein_coding                     KLHL13  CHS.58619   \n",
       "18             lncRNA                          -  CHS.42771   \n",
       "19             lncRNA               LOC105378980  CHS.42888   \n",
       "20             lncRNA               LOC102723842  CHS.50710   \n",
       "21      antisense_RNA                          -  CHS.58611   \n",
       "22             lncRNA               LOC107984029   CHS.7037   \n",
       "23     protein_coding                      CADM4  CHS.26900   \n",
       "24      antisense_RNA                          -  CHS.28909   \n",
       "25             lncRNA                          -   CHS.7761   \n",
       "26             lncRNA               LOC107985073  CHS.23091   \n",
       "27     protein_coding                    SLC10A7  CHS.41522   \n",
       "28     protein_coding                      CFHR5   CHS.4075   \n",
       "29     protein_coding                      OR2K2  CHS.56459   \n",
       "...               ...                        ...        ...   \n",
       "43132  protein_coding                     FAM49B  CHS.54502   \n",
       "43133          lncRNA               LOC105375363  CHS.51026   \n",
       "43134          lncRNA               LOC105379597  CHS.18527   \n",
       "43135          lncRNA                          -  CHS.42776   \n",
       "43136   antisense_RNA                          -  CHS.14002   \n",
       "43137          lncRNA               LOC105379086  CHS.43435   \n",
       "43138  protein_coding                       EAF2  CHS.38150   \n",
       "43139          lncRNA               LOC107986415  CHS.42844   \n",
       "43140  protein_coding                     VPS13B  CHS.54046   \n",
       "43141  protein_coding                   TP53TG3B  CHS.19681   \n",
       "43142          lncRNA                    FAM106B  CHS.21578   \n",
       "43143          lncRNA               LOC105378768   CHS.1801   \n",
       "43144  protein_coding                      GPR15  CHS.37811   \n",
       "43145          lncRNA                    TTTY17A  CHS.59338   \n",
       "43146  protein_coding                       CSF2  CHS.43864   \n",
       "43147   antisense_RNA                          -  CHS.51443   \n",
       "43148  protein_coding                       SELO  CHS.36236   \n",
       "43149  protein_coding                      GRIP2  CHS.36660   \n",
       "43150          lncRNA               LOC101927613  CHS.41465   \n",
       "43151  protein_coding                      GPLD1  CHS.45614   \n",
       "43152  protein_coding               LOC107986873  CHS.54132   \n",
       "43153  protein_coding                      RXFP2  CHS.13741   \n",
       "43154  protein_coding                      PCIF1  CHS.33658   \n",
       "43155  protein_coding                     SNRPD2  CHS.27026   \n",
       "43156  protein_coding                       AQP7  CHS.55319   \n",
       "43157          lncRNA               LOC105378347   CHS.6394   \n",
       "43158          lncRNA               LOC105375541  CHS.52065   \n",
       "43159  protein_coding                 AC011043.2  CHS.57331   \n",
       "43160  protein_coding                 AC011841.1  CHS.23667   \n",
       "43161  protein_coding                 AP000351.1  CHS.36369   \n",
       "\n",
       "                                     Location Database RefSeq_GeneID  \\\n",
       "0                   chr17:80885776-80892131:-   RefSeq     105371921   \n",
       "1                  chr3:127744219-127751505:-   RefSeq     107986127   \n",
       "2                   chr19:49496365-49499712:+   RefSeq          6205   \n",
       "3                   chr11:46276634-46321430:+   RefSeq         90993   \n",
       "4                      chr9:2385141-2494223:-   RefSeq     101930053   \n",
       "5                   chr14:73711783-73714425:-   RefSeq          9240   \n",
       "6                  chr5:179503322-179515579:-  GENCODE             -   \n",
       "7                   chr11:61391687-61402809:+   RefSeq         51259   \n",
       "8                  chr6:111482713-111602930:+   RefSeq        643749   \n",
       "9                 chr10:126421500-126671075:-   RefSeq        118611   \n",
       "10                 chrX:145843690-145875646:+   RefSeq     107985703   \n",
       "11                   chr8:24490149-24524864:-   RefSeq     101929315   \n",
       "12                  chr17:68824421-68845942:-    CHESS             -   \n",
       "13                  chr21:43358055-43362349:-   RefSeq     101928399   \n",
       "14                chr13:102845841-102876001:+   RefSeq          2073   \n",
       "15                 chr1:110963302-110964649:+  GENCODE             -   \n",
       "16                   chr4:40804188-41216714:-   RefSeq           323   \n",
       "17                 chrX:117896464-118117340:-   RefSeq         90293   \n",
       "18                   chr5:49609715-49611301:+    CHESS             -   \n",
       "19                   chr5:56895350-56900402:-   RefSeq     105378980   \n",
       "20                   chr7:57866002-57868134:-   RefSeq     102723842   \n",
       "21                 chrX:115844876-115884730:+    CHESS             -   \n",
       "22                chr10:101844640-101849768:+   RefSeq     107984029   \n",
       "23                  chr19:43622368-43641984:-   RefSeq        199731   \n",
       "24                   chr2:20596470-20598686:+   FANTOM             -   \n",
       "25                    chr11:1451420-1454484:-    CHESS             -   \n",
       "26                  chr17:69572642-69582123:-   RefSeq     107985073   \n",
       "27                 chr4:146253981-146522351:-   RefSeq         84068   \n",
       "28                 chr1:196975022-197009725:+   RefSeq         81494   \n",
       "29                 chr9:111326884-111330224:-   RefSeq         26248   \n",
       "...                                       ...      ...           ...   \n",
       "43132              chr8:129839593-130016697:-   RefSeq         51571   \n",
       "43133                chr7:77653341-77682570:+   RefSeq     105375363   \n",
       "43134      chr15_KI270850v1_alt:74397-80341:-   RefSeq     105379597   \n",
       "43135                chr5:49643907-49644683:+    CHESS             -   \n",
       "43136               chr13:46387367-46392782:+   FANTOM             -   \n",
       "43137                chr5:94575781-94583069:+   RefSeq     105379086   \n",
       "43138              chr3:121835180-121886526:+   RefSeq         55840   \n",
       "43139                chr5:55377324-55378208:+   RefSeq     107986415   \n",
       "43140                chr8:99007014-99877586:+   RefSeq        157680   \n",
       "43141               chr16:33358385-33363478:+   RefSeq        729355   \n",
       "43142               chr17:20417292-20419050:+   RefSeq     100996259   \n",
       "43143                chr1:62423816-62432939:+   RefSeq     105378768   \n",
       "43144                chr3:98531899-98534679:+   RefSeq          2838   \n",
       "43145                chrY:22851584-22852715:+   RefSeq        252949   \n",
       "43146              chr5:132073792-132079241:+   RefSeq          1437   \n",
       "43147              chr7:102376259-102396117:-   FANTOM             -   \n",
       "43148               chr22:50200979-50217640:+   RefSeq         83642   \n",
       "43149                chr3:14489111-14602996:-   RefSeq         80852   \n",
       "43150              chr4:142565058-142662044:+   RefSeq     101927613   \n",
       "43151                chr6:24418371-24495163:-   RefSeq          2822   \n",
       "43152              chr8:102810302-102863067:+   RefSeq     107986873   \n",
       "43153               chr13:31739526-31802872:+   RefSeq        122042   \n",
       "43154               chr20:45934674-45948023:+   RefSeq         63935   \n",
       "43155               chr19:45687454-45692333:-   RefSeq          6633   \n",
       "43156                chr9:33383135-33402682:-   RefSeq           364   \n",
       "43157               chr10:69806962-69822323:-   RefSeq     105378347   \n",
       "43158              chr7:142812612-142822982:+   RefSeq     105375541   \n",
       "43159        chrUn_GL000195v1:173872-179372:+  GENCODE             -   \n",
       "43160  chr17_GL000205v2_random:99351-104855:-  GENCODE             -   \n",
       "43161    chr22_KI270879v1_alt:259292-259749:-  GENCODE             -   \n",
       "\n",
       "                                             Description  \n",
       "0                           uncharacterized LOC105371921  \n",
       "1                           uncharacterized LOC107986127  \n",
       "2                                  ribosomal protein S11  \n",
       "3       cAMP responsive element binding protein 3-like 1  \n",
       "4                           uncharacterized LOC101930053  \n",
       "5                            paraneoplastic Ma antigen 1  \n",
       "6                                                    NaN  \n",
       "7                              transmembrane protein 216  \n",
       "8                               TRAF3IP2 antisense RNA 1  \n",
       "9                    chromosome 10 open reading frame 90  \n",
       "10                          uncharacterized LOC107985703  \n",
       "11                          uncharacterized LOC101929315  \n",
       "12                                                   NaN  \n",
       "13                          uncharacterized LOC101928399  \n",
       "14         excision repair cross-complementation group 5  \n",
       "15                                                   NaN  \n",
       "16     amyloid beta precursor protein binding family ...  \n",
       "17                           kelch like family member 13  \n",
       "18                                                   NaN  \n",
       "19                          uncharacterized LOC105378980  \n",
       "20                          uncharacterized LOC102723842  \n",
       "21                                                   NaN  \n",
       "22                          uncharacterized LOC107984029  \n",
       "23                              cell adhesion molecule 4  \n",
       "24                                                   NaN  \n",
       "25                                                   NaN  \n",
       "26                          uncharacterized LOC107985073  \n",
       "27                     solute carrier family 10 member 7  \n",
       "28                         complement factor H related 5  \n",
       "29      olfactory receptor family 2 subfamily K member 2  \n",
       "...                                                  ...  \n",
       "43132        family with sequence similarity 49 member B  \n",
       "43133                       uncharacterized LOC105375363  \n",
       "43134                       uncharacterized LOC105379597  \n",
       "43135                                                NaN  \n",
       "43136                                                NaN  \n",
       "43137                       uncharacterized LOC105379086  \n",
       "43138                            ELL associated factor 2  \n",
       "43139                       uncharacterized LOC107986415  \n",
       "43140      vacuolar protein sorting 13 homolog B (yeast)  \n",
       "43141                                     TP53 target 3B  \n",
       "43142       family with sequence similarity 106 member B  \n",
       "43143                       uncharacterized LOC105378768  \n",
       "43144                      G protein-coupled receptor 15  \n",
       "43145  testis-specific transcript%2C Y-linked 17A (no...  \n",
       "43146                        colony stimulating factor 2  \n",
       "43147                                                NaN  \n",
       "43148                                    selenoprotein O  \n",
       "43149           glutamate receptor interacting protein 2  \n",
       "43150                       uncharacterized LOC101927613  \n",
       "43151  glycosylphosphatidylinositol specific phosphol...  \n",
       "43152                       forkhead box protein L2-like  \n",
       "43153     relaxin/insulin like family peptide receptor 2  \n",
       "43154                PDX1 C-terminal inhibiting factor 1  \n",
       "43155     small nuclear ribonucleoprotein D2 polypeptide  \n",
       "43156                                        aquaporin 7  \n",
       "43157                       uncharacterized LOC105378347  \n",
       "43158                       uncharacterized LOC105375541  \n",
       "43159                                                NaN  \n",
       "43160                                                NaN  \n",
       "43161                                                NaN  \n",
       "\n",
       "[43162 rows x 7 columns]"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df1_genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "number of RefSeq and Gencode protein_coding genes: 20128\n",
      "number of RefSeq and Gencode protein_coding genes: 22656\n",
      "number of RefSeq and Gencode lncRNA genes: 16216\n",
      "number of RefSeq and Gencode lncRNA genes: 17236\n",
      "number of RefSeq and Gencode misc_RNA genes: 1227\n",
      "number of RefSeq and Gencode misc_RNA genes: 1488\n",
      "number of RefSeq and Gencode antisense_RNA genes: 598\n",
      "number of RefSeq and Gencode antisense_RNA genes: 598\n"
     ]
    }
   ],
   "source": [
    "print(\"number of RefSeq and Gencode protein_coding genes: \"+str(len(df1_genes[(df1_genes[\"Gene_Type\"]==\"protein_coding\")&(df1_genes[\"Database\"].isin(['GENCODE', 'RefSeq', 'RefSeq,GENCODE']))])))\n",
    "listGenes=[]\n",
    "df1_genes[(df1_genes[\"Gene_Type\"]==\"protein_coding\")&(df1_genes[\"Database\"].isin(['GENCODE', 'RefSeq', 'RefSeq,GENCODE']))].apply(lambda row: addGene(row),axis=1)\n",
    "setProteinCodingGenesREFGEN=set(listGenes)\n",
    "print(\"number of RefSeq and Gencode protein_coding genes: \"+str(len(setProteinCodingGenesREFGEN)))\n",
    "\n",
    "print(\"number of RefSeq and Gencode lncRNA genes: \"+str(len(df1_genes[(df1_genes[\"Gene_Type\"]==\"lncRNA\")&(df1_genes[\"Database\"].isin(['GENCODE', 'RefSeq', 'RefSeq,GENCODE']))])))\n",
    "listGenes=[]\n",
    "df1_genes[(df1_genes[\"Gene_Type\"]==\"lncRNA\")&(df1_genes[\"Database\"].isin(['GENCODE', 'RefSeq', 'RefSeq,GENCODE']))].apply(lambda row: addGene(row),axis=1)\n",
    "setlncRNAGenesREFGEN=set(listGenes)\n",
    "print(\"number of RefSeq and Gencode lncRNA genes: \"+str(len(setlncRNAGenesREFGEN)))\n",
    "\n",
    "print(\"number of RefSeq and Gencode misc_RNA genes: \"+str(len(df1_genes[(df1_genes[\"Gene_Type\"]==\"misc_RNA\")&(df1_genes[\"Database\"].isin(['GENCODE', 'RefSeq', 'RefSeq,GENCODE']))])))\n",
    "listGenes=[]\n",
    "df1_genes[(df1_genes[\"Gene_Type\"]==\"misc_RNA\")&(df1_genes[\"Database\"].isin(['GENCODE', 'RefSeq', 'RefSeq,GENCODE']))].apply(lambda row: addGene(row),axis=1)\n",
    "setmiscRNAGenesREFGEN=set(listGenes)\n",
    "print(\"number of RefSeq and Gencode misc_RNA genes: \"+str(len(setmiscRNAGenesREFGEN)))\n",
    "\n",
    "print(\"number of RefSeq and Gencode antisense_RNA genes: \"+str(len(df1_genes[(df1_genes[\"Gene_Type\"]==\"antisense_RNA\")&(df1_genes[\"Database\"].isin(['GENCODE', 'RefSeq', 'RefSeq,GENCODE']))])))\n",
    "listGenes=[]\n",
    "df1_genes[(df1_genes[\"Gene_Type\"]==\"antisense_RNA\")&(df1_genes[\"Database\"].isin(['GENCODE', 'RefSeq', 'RefSeq,GENCODE']))].apply(lambda row: addGene(row),axis=1)\n",
    "setantisenseRNAGenesREFGEN=set(listGenes)\n",
    "print(\"number of RefSeq and Gencode antisense_RNA genes: \"+str(len(setantisenseRNAGenesREFGEN)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "number of RefSeq and Gencode protein_coding genes: 98\n",
      "number of RefSeq and Gencode protein_coding genes: 98\n",
      "number of RefSeq and Gencode lncRNA genes: 1352\n",
      "number of RefSeq and Gencode lncRNA genes: 1352\n",
      "number of RefSeq and Gencode misc_RNA genes: 1\n",
      "number of RefSeq and Gencode misc_RNA genes: 1\n",
      "number of RefSeq and Gencode antisense_RNA genes: 494\n",
      "number of RefSeq and Gencode antisense_RNA genes: 494\n"
     ]
    }
   ],
   "source": [
    "print(\"number of RefSeq and Gencode protein_coding genes: \"+str(len(df1_genes[(df1_genes[\"Gene_Type\"]==\"protein_coding\")&(df1_genes[\"Database\"].isin(['FANTOM']))])))\n",
    "listGenes=[]\n",
    "df1_genes[(df1_genes[\"Gene_Type\"]==\"protein_coding\")&(df1_genes[\"Database\"].isin(['FANTOM']))].apply(lambda row: addGene(row),axis=1)\n",
    "setProteinCodingGenesFANTOM=set(listGenes)\n",
    "print(\"number of RefSeq and Gencode protein_coding genes: \"+str(len(setProteinCodingGenesFANTOM)))\n",
    "\n",
    "print(\"number of RefSeq and Gencode lncRNA genes: \"+str(len(df1_genes[(df1_genes[\"Gene_Type\"]==\"lncRNA\")&(df1_genes[\"Database\"].isin(['FANTOM']))])))\n",
    "listGenes=[]\n",
    "df1_genes[(df1_genes[\"Gene_Type\"]==\"lncRNA\")&(df1_genes[\"Database\"].isin(['FANTOM']))].apply(lambda row: addGene(row),axis=1)\n",
    "setlncRNAGenesFANTOM=set(listGenes)\n",
    "print(\"number of RefSeq and Gencode lncRNA genes: \"+str(len(setlncRNAGenesFANTOM)))\n",
    "\n",
    "print(\"number of RefSeq and Gencode misc_RNA genes: \"+str(len(df1_genes[(df1_genes[\"Gene_Type\"]==\"misc_RNA\")&(df1_genes[\"Database\"].isin(['FANTOM']))])))\n",
    "listGenes=[]\n",
    "df1_genes[(df1_genes[\"Gene_Type\"]==\"misc_RNA\")&(df1_genes[\"Database\"].isin(['FANTOM']))].apply(lambda row: addGene(row),axis=1)\n",
    "setmiscRNAGenesFANTOM=set(listGenes)\n",
    "print(\"number of RefSeq and Gencode misc_RNA genes: \"+str(len(setmiscRNAGenesFANTOM)))\n",
    "\n",
    "print(\"number of RefSeq and Gencode antisense_RNA genes: \"+str(len(df1_genes[(df1_genes[\"Gene_Type\"]==\"antisense_RNA\")&(df1_genes[\"Database\"].isin(['FANTOM']))])))\n",
    "listGenes=[]\n",
    "df1_genes[(df1_genes[\"Gene_Type\"]==\"antisense_RNA\")&(df1_genes[\"Database\"].isin(['FANTOM']))].apply(lambda row: addGene(row),axis=1)\n",
    "setantisenseRNAGenesFANTOM=set(listGenes)\n",
    "print(\"number of RefSeq and Gencode antisense_RNA genes: \"+str(len(setantisenseRNAGenesFANTOM)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "number of RefSeq and Gencode protein_coding genes: 1178\n",
      "number of RefSeq and Gencode protein_coding genes: 1178\n",
      "number of RefSeq and Gencode lncRNA genes: 2268\n",
      "number of RefSeq and Gencode lncRNA genes: 2268\n",
      "number of RefSeq and Gencode misc_RNA genes: 1\n",
      "number of RefSeq and Gencode misc_RNA genes: 1\n",
      "number of RefSeq and Gencode antisense_RNA genes: 1546\n",
      "number of RefSeq and Gencode antisense_RNA genes: 1547\n"
     ]
    }
   ],
   "source": [
    "print(\"number of RefSeq and Gencode protein_coding genes: \"+str(len(df1_genes[(df1_genes[\"Gene_Type\"]==\"protein_coding\")&(df1_genes[\"Database\"].isin(['CHESS','FANTOM']))])))\n",
    "listGenes=[]\n",
    "df1_genes[(df1_genes[\"Gene_Type\"]==\"protein_coding\")&(df1_genes[\"Database\"].isin(['CHESS',\"FANTOM\"]))].apply(lambda row: addGene(row),axis=1)\n",
    "setProteinCodingGenesNOVEL=set(listGenes)\n",
    "print(\"number of RefSeq and Gencode protein_coding genes: \"+str(len(setProteinCodingGenesNOVEL)))\n",
    "\n",
    "print(\"number of RefSeq and Gencode lncRNA genes: \"+str(len(df1_genes[(df1_genes[\"Gene_Type\"]==\"lncRNA\")&(df1_genes[\"Database\"].isin(['CHESS','FANTOM']))])))\n",
    "listGenes=[]\n",
    "df1_genes[(df1_genes[\"Gene_Type\"]==\"lncRNA\")&(df1_genes[\"Database\"].isin(['CHESS',\"FANTOM\"]))].apply(lambda row: addGene(row),axis=1)\n",
    "setlncRNAGenesNOVEL=set(listGenes)\n",
    "print(\"number of RefSeq and Gencode lncRNA genes: \"+str(len(setlncRNAGenesNOVEL)))\n",
    "\n",
    "print(\"number of RefSeq and Gencode misc_RNA genes: \"+str(len(df1_genes[(df1_genes[\"Gene_Type\"]==\"misc_RNA\")&(df1_genes[\"Database\"].isin(['CHESS','FANTOM']))])))\n",
    "listGenes=[]\n",
    "df1_genes[(df1_genes[\"Gene_Type\"]==\"misc_RNA\")&(df1_genes[\"Database\"].isin(['CHESS',\"FANTOM\"]))].apply(lambda row: addGene(row),axis=1)\n",
    "setmiscRNAGenesNOVEL=set(listGenes)\n",
    "print(\"number of RefSeq and Gencode misc_RNA genes: \"+str(len(setmiscRNAGenesNOVEL)))\n",
    "\n",
    "print(\"number of RefSeq and Gencode antisense_RNA genes: \"+str(len(df1_genes[(df1_genes[\"Gene_Type\"]==\"antisense_RNA\")&(df1_genes[\"Database\"].isin(['CHESS','FANTOM']))])))\n",
    "listGenes=[]\n",
    "df1_genes[(df1_genes[\"Gene_Type\"]==\"antisense_RNA\")&(df1_genes[\"Database\"].isin(['CHESS',\"FANTOM\"]))].apply(lambda row: addGene(row),axis=1)\n",
    "setantisenseRNAGenesNOVEL=set(listGenes)\n",
    "print(\"number of RefSeq and Gencode antisense_RNA genes: \"+str(len(setantisenseRNAGenesNOVEL)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "number of RefSeq and Gencode protein_coding transcripts: 169959\n",
      "number of RefSeq and Gencode lncRNA transcripts: 34222\n",
      "number of RefSeq and Gencode misc_RNA transcripts: 2284\n",
      "number of RefSeq and Gencode antisense_RNA transcripts: 637\n"
     ]
    }
   ],
   "source": [
    "print(\"number of RefSeq and Gencode protein_coding transcripts: \"+str(len(df1t[(df1t[\"tID\"].isin(set(dft_protein_coding[\"TranscriptID\"])))&(df1t[\"source\"].isin([\"BestRefSeq\",\"Curated Genomic\",\"ENSEMBL\",\"Gnomon\",\"HAVANA\",\"RefSeq\"]))])))\n",
    "print(\"number of RefSeq and Gencode lncRNA transcripts: \"+str(len(df1t[(df1t[\"tID\"].isin(set(dft_lncRNA[\"TranscriptID\"])))&(df1t[\"source\"].isin([\"BestRefSeq\",\"Curated Genomic\",\"ENSEMBL\",\"Gnomon\",\"HAVANA\",\"RefSeq\"]))])))\n",
    "print(\"number of RefSeq and Gencode misc_RNA transcripts: \"+str(len(df1t[(df1t[\"tID\"].isin(set(dft_misc_RNA[\"TranscriptID\"])))&(df1t[\"source\"].isin([\"BestRefSeq\",\"Curated Genomic\",\"ENSEMBL\",\"Gnomon\",\"HAVANA\",\"RefSeq\"]))])))\n",
    "print(\"number of RefSeq and Gencode antisense_RNA transcripts: \"+str(len(df1t[(df1t[\"tID\"].isin(set(dft_antisense_RNA[\"TranscriptID\"])))&(df1t[\"source\"].isin([\"BestRefSeq\",\"Curated Genomic\",\"ENSEMBL\",\"Gnomon\",\"HAVANA\",\"RefSeq\"]))])))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "number of StringTie and FANTOM protein_coding transcripts: 97511\n",
      "number of StringTie and FANTOM lncRNA transcripts: 15092\n",
      "number of StringTie and FANTOM misc_RNA transcripts: 2063\n",
      "number of StringTie and FANTOM antisense_RNA transcripts: 2051\n"
     ]
    }
   ],
   "source": [
    "print(\"number of StringTie and FANTOM protein_coding transcripts: \"+str(len(df1t[(df1t[\"tID\"].isin(set(dft_protein_coding[\"TranscriptID\"])))&(df1t[\"source\"].isin([\"StringTie\",\"FANTOM\"]))])))\n",
    "print(\"number of StringTie and FANTOM lncRNA transcripts: \"+str(len(df1t[(df1t[\"tID\"].isin(set(dft_lncRNA[\"TranscriptID\"])))&(df1t[\"source\"].isin([\"StringTie\",\"FANTOM\"]))])))\n",
    "print(\"number of StringTie and FANTOM misc_RNA transcripts: \"+str(len(df1t[(df1t[\"tID\"].isin(set(dft_misc_RNA[\"TranscriptID\"])))&(df1t[\"source\"].isin([\"StringTie\",\"FANTOM\"]))])))\n",
    "print(\"number of StringTie and FANTOM antisense_RNA transcripts: \"+str(len(df1t[(df1t[\"tID\"].isin(set(dft_antisense_RNA[\"TranscriptID\"])))&(df1t[\"source\"].isin([\"StringTie\",\"FANTOM\"]))])))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "number of FANTOM protein_coding transcripts: 23185\n",
      "number of FANTOM lncRNA transcripts: 5766\n",
      "number of FANTOM misc_RNA transcripts: 476\n",
      "number of FANTOM antisense_RNA transcripts: 606\n"
     ]
    }
   ],
   "source": [
    "print(\"number of FANTOM protein_coding transcripts: \"+str(len(df1t[(df1t[\"tID\"].isin(set(dft_protein_coding[\"TranscriptID\"])))&(df1t[\"source\"].isin([\"FANTOM\"]))])))\n",
    "print(\"number of FANTOM lncRNA transcripts: \"+str(len(df1t[(df1t[\"tID\"].isin(set(dft_lncRNA[\"TranscriptID\"])))&(df1t[\"source\"].isin([\"FANTOM\"]))])))\n",
    "print(\"number of FANTOM misc_RNA transcripts: \"+str(len(df1t[(df1t[\"tID\"].isin(set(dft_misc_RNA[\"TranscriptID\"])))&(df1t[\"source\"].isin([\"FANTOM\"]))])))\n",
    "print(\"number of FANTOM antisense_RNA transcripts: \"+str(len(df1t[(df1t[\"tID\"].isin(set(dft_antisense_RNA[\"TranscriptID\"])))&(df1t[\"source\"].isin([\"FANTOM\"]))])))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'CHS.16819.1',\n",
       " 'CHS.1933.1',\n",
       " 'CHS.1934.1',\n",
       " 'CHS.2571.2',\n",
       " 'CHS.54408.1',\n",
       " 'CHS.55723.1'}"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "set(dft_antisense_RNA[\"TranscriptID\"]).symmetric_difference(set(dft[(dft[\"Gene_type\"]==\"antisense_RNA\")&~(dft[\"TranscriptID\"].isin(dontCount))]['TranscriptID']))"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.15"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
