{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import subprocess\n",
    "import sys\n",
    "import os\n",
    "import csv\n",
    "import matplotlib.pyplot as plt\n",
    "from matplotlib import animation\n",
    "import seaborn as sns\n",
    "plt.rcParams['figure.figsize'] = (20.0, 10.0)\n",
    "plt.rcParams['font.family'] = \"serif\"\n",
    "%matplotlib inline\n",
    "\n",
    "gff3Cols=[\"seqid\",\"source\",\"type\",\"start\",\"end\",\"score\",\"strand\",\"phase\",\"attributes\"]\n",
    "novelSources=['CHESS','StringTie','FANTOM']"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "we shall investigate whether novel transcripts to known (RefSeq and Gencode genes) tend to preserve the original ORF\n",
    "or instead would results in a novel protein\n",
    "\n",
    "Later we shall also investigate whether in novel genes there are multiple ORFs or whether different isoforms\n",
    "tend to be ORF-compatible\n",
    "\n",
    "to achieve this the following must be done\n",
    "1. we need to extract all genes and the corresponding transcript, CDS and exon entries\n",
    "    - this must be done only for protein-coding genes\n",
    "\n",
    "\n",
    "eventually we need to look at whether novel ORFs introduce a frameshift or not"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "def formBlocks(series):\n",
    "    return \",\".join([str(x) for x in sorted(series.tolist())])\n",
    "\n",
    "def getUniqueOrfCount(series):\n",
    "    return len(set(series.tolist()))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'Gnomon', 'HAVANA', 'CHESS', 'StringTie', 'GENCODE', 'BestRefSeq', 'Curated Genomic', 'RefSeq', 'FANTOM', 'ENSEMBL'}\n",
      "set of all gene_types in CHESS {'lncRNA', 'misc_RNA', 'antisense_RNA', 'protein_coding'}\n",
      "number of known protein-coding genes is: 22659\n",
      "the number of transcripts in protein-coding known genes is: 169965\n",
      "the number of CDSs associated with transcripts in protein-coding known genes is 129847\n"
     ]
    }
   ],
   "source": [
    "# first build a dataframe of all known transcripts only\n",
    "df=pd.read_csv(\"./chess2.02.gff\",sep=\"\\t\",names=gff3Cols)\n",
    "df.dropna(inplace=True,axis=0)\n",
    "df.reset_index(inplace=True,drop=True)\n",
    "df[\"start\"]=df[\"start\"].astype(int)\n",
    "df[\"end\"]=df[\"end\"].astype(int)\n",
    "\n",
    "# next we shall create the subset of all non-novel genes\n",
    "print(set(df[\"source\"]))\n",
    "df=df[~(df['source'].isin(novelSources))].reset_index(drop=True)\n",
    "\n",
    "# now let's isolate the protein-coding genes and their transcripts, exons and CDS\n",
    "\n",
    "# first let's extract IDs\n",
    "df[\"id\"]=df.attributes.str.split(\"ID=\",expand=True)[1].str.split(\";\",expand=True)[0]\n",
    "df[\"parent\"]=df.attributes.str.split(\"Parent=\",expand=True)[1].str.split(\";\",expand=True)[0]\n",
    "df[\"geneID\"]=np.where(df[\"type\"].isin(['transcript','exon','CDS']),df.parent.str.extract('(CHS.(\\d)*)',expand=True)[0],df['id'])\n",
    "# first create a tmpdf (to be removed right after) of just the genes, so we can get their coding potential\n",
    "tmp=df[df[\"type\"]==\"gene\"].reset_index(drop=True)\n",
    "# next let's extract information about the gene type\n",
    "tmp[\"gene_type\"]=tmp.attributes.str.split(\"GENE_TYPE=\",expand=True)[1].str.split(\";\",expand=True)[0].str.strip(\"\\n\")\n",
    "print(\"set of all gene_types in CHESS\",set(tmp[\"gene_type\"]))\n",
    "\n",
    "# now we can get a subset of all known protein_coding genes\n",
    "tmp=tmp[tmp[\"gene_type\"]==\"protein_coding\"].reset_index(drop=True)\n",
    "print(\"number of known protein-coding genes is: %d\"%len(tmp))\n",
    "# now we just need to get geneIDs for the protein_coding sequences\n",
    "setProtGenes=set(tmp[\"geneID\"])\n",
    "del tmp\n",
    "# df.drop([\"id\",\"parent\"],inplace=True,axis=1)\n",
    "df=df[df[\"geneID\"].isin(setProtGenes)].reset_index(drop=True)\n",
    "\n",
    "# next we would like to test whether all transcripts in these known protein-coding genes contain a CDS\n",
    "\n",
    "# first get a set of parent IDs for all CDSs\n",
    "cdsParents=set(df[df[\"type\"]==\"CDS\"][\"parent\"])\n",
    "# now form a set of IDs for all transcripts\n",
    "transIDs=set(df[df[\"type\"]==\"transcript\"][\"id\"])\n",
    "print(\"the number of transcripts in protein-coding known genes is: %d\\nthe number of CDSs associated with transcripts in protein-coding known genes is %d\"%(len(transIDs),len(cdsParents)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "500\n",
      "1000\n",
      "1500\n",
      "2000\n",
      "2500\n",
      "3000\n",
      "3500\n",
      "4000\n",
      "4500\n",
      "5000\n",
      "5500\n",
      "6000\n",
      "6500\n",
      "7000\n",
      "7500\n",
      "8000\n",
      "8500\n",
      "9000\n",
      "9500\n",
      "10000\n",
      "10500\n",
      "11000\n",
      "11500\n",
      "12000\n",
      "12500\n",
      "13000\n",
      "13500\n",
      "14000\n",
      "14500\n",
      "15000\n",
      "15500\n",
      "16000\n",
      "16500\n",
      "17000\n",
      "17500\n",
      "18000\n",
      "18500\n",
      "19000\n",
      "19500\n",
      "20000\n",
      "20500\n",
      "21000\n",
      "21500\n",
      "22000\n",
      "22500\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gID</th>\n",
       "      <th>uniqueORFs_x</th>\n",
       "      <th>numTranscripts_x</th>\n",
       "      <th>numProteins</th>\n",
       "      <th>numUniqueProteins</th>\n",
       "      <th>uniqueORFs_y</th>\n",
       "      <th>numTranscripts_y</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CHS.10000</td>\n",
       "      <td>8</td>\n",
       "      <td>8</td>\n",
       "      <td>8</td>\n",
       "      <td>8</td>\n",
       "      <td>8</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CHS.10003</td>\n",
       "      <td>10</td>\n",
       "      <td>14</td>\n",
       "      <td>14</td>\n",
       "      <td>10</td>\n",
       "      <td>10</td>\n",
       "      <td>14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CHS.10004</td>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CHS.10005</td>\n",
       "      <td>6</td>\n",
       "      <td>7</td>\n",
       "      <td>7</td>\n",
       "      <td>5</td>\n",
       "      <td>6</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CHS.10008</td>\n",
       "      <td>6</td>\n",
       "      <td>7</td>\n",
       "      <td>7</td>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>CHS.10009</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>CHS.1001</td>\n",
       "      <td>3</td>\n",
       "      <td>8</td>\n",
       "      <td>8</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>CHS.10010</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>CHS.10011</td>\n",
       "      <td>28</td>\n",
       "      <td>30</td>\n",
       "      <td>30</td>\n",
       "      <td>28</td>\n",
       "      <td>28</td>\n",
       "      <td>30</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>CHS.10013</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>CHS.10014</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>CHS.10019</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>CHS.1002</td>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>CHS.10020</td>\n",
       "      <td>2</td>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>CHS.10026</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>CHS.10027</td>\n",
       "      <td>7</td>\n",
       "      <td>7</td>\n",
       "      <td>7</td>\n",
       "      <td>7</td>\n",
       "      <td>7</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>CHS.10028</td>\n",
       "      <td>4</td>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>CHS.10029</td>\n",
       "      <td>5</td>\n",
       "      <td>9</td>\n",
       "      <td>5</td>\n",
       "      <td>5</td>\n",
       "      <td>5</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>CHS.1003</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>CHS.10030</td>\n",
       "      <td>4</td>\n",
       "      <td>9</td>\n",
       "      <td>9</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>CHS.10031</td>\n",
       "      <td>5</td>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "      <td>5</td>\n",
       "      <td>5</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>CHS.10036</td>\n",
       "      <td>13</td>\n",
       "      <td>13</td>\n",
       "      <td>13</td>\n",
       "      <td>13</td>\n",
       "      <td>13</td>\n",
       "      <td>13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>CHS.1005</td>\n",
       "      <td>53</td>\n",
       "      <td>57</td>\n",
       "      <td>57</td>\n",
       "      <td>52</td>\n",
       "      <td>53</td>\n",
       "      <td>57</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>CHS.10064</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>CHS.10066</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>CHS.10068</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>CHS.1007</td>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>CHS.10070</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>CHS.10072</td>\n",
       "      <td>3</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>CHS.10073</td>\n",
       "      <td>4</td>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "      <td>3</td>\n",
       "      <td>4</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22514</th>\n",
       "      <td>CHS.994</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22515</th>\n",
       "      <td>CHS.9940</td>\n",
       "      <td>5</td>\n",
       "      <td>7</td>\n",
       "      <td>7</td>\n",
       "      <td>5</td>\n",
       "      <td>5</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22516</th>\n",
       "      <td>CHS.9946</td>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22517</th>\n",
       "      <td>CHS.9947</td>\n",
       "      <td>1</td>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22518</th>\n",
       "      <td>CHS.9948</td>\n",
       "      <td>4</td>\n",
       "      <td>8</td>\n",
       "      <td>8</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22519</th>\n",
       "      <td>CHS.9950</td>\n",
       "      <td>5</td>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "      <td>5</td>\n",
       "      <td>5</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22520</th>\n",
       "      <td>CHS.9951</td>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22521</th>\n",
       "      <td>CHS.9957</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22522</th>\n",
       "      <td>CHS.9958</td>\n",
       "      <td>6</td>\n",
       "      <td>8</td>\n",
       "      <td>8</td>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22523</th>\n",
       "      <td>CHS.9960</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22524</th>\n",
       "      <td>CHS.9961</td>\n",
       "      <td>9</td>\n",
       "      <td>9</td>\n",
       "      <td>9</td>\n",
       "      <td>9</td>\n",
       "      <td>9</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22525</th>\n",
       "      <td>CHS.9963</td>\n",
       "      <td>18</td>\n",
       "      <td>23</td>\n",
       "      <td>23</td>\n",
       "      <td>18</td>\n",
       "      <td>18</td>\n",
       "      <td>23</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22526</th>\n",
       "      <td>CHS.9964</td>\n",
       "      <td>5</td>\n",
       "      <td>5</td>\n",
       "      <td>5</td>\n",
       "      <td>5</td>\n",
       "      <td>5</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22527</th>\n",
       "      <td>CHS.9965</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22528</th>\n",
       "      <td>CHS.9966</td>\n",
       "      <td>9</td>\n",
       "      <td>12</td>\n",
       "      <td>12</td>\n",
       "      <td>7</td>\n",
       "      <td>9</td>\n",
       "      <td>12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22529</th>\n",
       "      <td>CHS.9967</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22530</th>\n",
       "      <td>CHS.9968</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22531</th>\n",
       "      <td>CHS.997</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22532</th>\n",
       "      <td>CHS.9971</td>\n",
       "      <td>6</td>\n",
       "      <td>7</td>\n",
       "      <td>7</td>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22533</th>\n",
       "      <td>CHS.9972</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22534</th>\n",
       "      <td>CHS.9973</td>\n",
       "      <td>11</td>\n",
       "      <td>12</td>\n",
       "      <td>12</td>\n",
       "      <td>9</td>\n",
       "      <td>11</td>\n",
       "      <td>12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22535</th>\n",
       "      <td>CHS.9974</td>\n",
       "      <td>6</td>\n",
       "      <td>8</td>\n",
       "      <td>8</td>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22536</th>\n",
       "      <td>CHS.9975</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22537</th>\n",
       "      <td>CHS.9976</td>\n",
       "      <td>7</td>\n",
       "      <td>7</td>\n",
       "      <td>7</td>\n",
       "      <td>6</td>\n",
       "      <td>7</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22538</th>\n",
       "      <td>CHS.998</td>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22539</th>\n",
       "      <td>CHS.9980</td>\n",
       "      <td>2</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22540</th>\n",
       "      <td>CHS.9983</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22541</th>\n",
       "      <td>CHS.9984</td>\n",
       "      <td>7</td>\n",
       "      <td>7</td>\n",
       "      <td>7</td>\n",
       "      <td>7</td>\n",
       "      <td>7</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22542</th>\n",
       "      <td>CHS.9985</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22543</th>\n",
       "      <td>CHS.9986</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>22544 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             gID  uniqueORFs_x  numTranscripts_x  numProteins  \\\n",
       "0      CHS.10000             8                 8            8   \n",
       "1      CHS.10003            10                14           14   \n",
       "2      CHS.10004             6                 6            6   \n",
       "3      CHS.10005             6                 7            7   \n",
       "4      CHS.10008             6                 7            7   \n",
       "5      CHS.10009             3                 3            3   \n",
       "6       CHS.1001             3                 8            8   \n",
       "7      CHS.10010             1                 1            1   \n",
       "8      CHS.10011            28                30           30   \n",
       "9      CHS.10013             3                 3            3   \n",
       "10     CHS.10014             3                 3            3   \n",
       "11     CHS.10019             2                 2            2   \n",
       "12      CHS.1002             2                 3            3   \n",
       "13     CHS.10020             2                 6            6   \n",
       "14     CHS.10026             2                 2            2   \n",
       "15     CHS.10027             7                 7            7   \n",
       "16     CHS.10028             4                 6            6   \n",
       "17     CHS.10029             5                 9            5   \n",
       "18      CHS.1003             1                 1            1   \n",
       "19     CHS.10030             4                 9            9   \n",
       "20     CHS.10031             5                 6            6   \n",
       "21     CHS.10036            13                13           13   \n",
       "22      CHS.1005            53                57           57   \n",
       "23     CHS.10064             3                 3            3   \n",
       "24     CHS.10066             2                 2            2   \n",
       "25     CHS.10068             1                 3            3   \n",
       "26      CHS.1007             2                 3            3   \n",
       "27     CHS.10070             1                 1            1   \n",
       "28     CHS.10072             3                 4            4   \n",
       "29     CHS.10073             4                 6            6   \n",
       "...          ...           ...               ...          ...   \n",
       "22514    CHS.994             2                 2            2   \n",
       "22515   CHS.9940             5                 7            7   \n",
       "22516   CHS.9946             2                 3            3   \n",
       "22517   CHS.9947             1                 6            6   \n",
       "22518   CHS.9948             4                 8            8   \n",
       "22519   CHS.9950             5                 6            6   \n",
       "22520   CHS.9951             6                 6            6   \n",
       "22521   CHS.9957             2                 2            2   \n",
       "22522   CHS.9958             6                 8            8   \n",
       "22523   CHS.9960             3                 3            3   \n",
       "22524   CHS.9961             9                 9            9   \n",
       "22525   CHS.9963            18                23           23   \n",
       "22526   CHS.9964             5                 5            5   \n",
       "22527   CHS.9965             1                 1            1   \n",
       "22528   CHS.9966             9                12           12   \n",
       "22529   CHS.9967             1                 1            1   \n",
       "22530   CHS.9968             1                 1            1   \n",
       "22531    CHS.997             3                 3            3   \n",
       "22532   CHS.9971             6                 7            7   \n",
       "22533   CHS.9972             4                 4            4   \n",
       "22534   CHS.9973            11                12           12   \n",
       "22535   CHS.9974             6                 8            8   \n",
       "22536   CHS.9975             1                 1            1   \n",
       "22537   CHS.9976             7                 7            7   \n",
       "22538    CHS.998             2                 3            3   \n",
       "22539   CHS.9980             2                 4            4   \n",
       "22540   CHS.9983             2                 2            2   \n",
       "22541   CHS.9984             7                 7            7   \n",
       "22542   CHS.9985             4                 4            4   \n",
       "22543   CHS.9986             1                 1            1   \n",
       "\n",
       "       numUniqueProteins  uniqueORFs_y  numTranscripts_y  \n",
       "0                      8             8                 8  \n",
       "1                     10            10                14  \n",
       "2                      6             6                 6  \n",
       "3                      5             6                 7  \n",
       "4                      6             6                 7  \n",
       "5                      3             3                 3  \n",
       "6                      3             3                 8  \n",
       "7                      1             1                 1  \n",
       "8                     28            28                30  \n",
       "9                      3             3                 3  \n",
       "10                     3             3                 3  \n",
       "11                     1             2                 2  \n",
       "12                     2             2                 3  \n",
       "13                     2             2                 6  \n",
       "14                     2             2                 2  \n",
       "15                     7             7                 7  \n",
       "16                     4             4                 6  \n",
       "17                     5             5                 9  \n",
       "18                     1             1                 1  \n",
       "19                     4             4                 9  \n",
       "20                     5             5                 6  \n",
       "21                    13            13                13  \n",
       "22                    52            53                57  \n",
       "23                     3             3                 3  \n",
       "24                     2             2                 2  \n",
       "25                     1             1                 3  \n",
       "26                     2             2                 3  \n",
       "27                     1             1                 1  \n",
       "28                     3             3                 4  \n",
       "29                     3             4                 6  \n",
       "...                  ...           ...               ...  \n",
       "22514                  2             2                 2  \n",
       "22515                  5             5                 7  \n",
       "22516                  2             2                 3  \n",
       "22517                  1             1                 6  \n",
       "22518                  4             4                 8  \n",
       "22519                  5             5                 6  \n",
       "22520                  6             6                 6  \n",
       "22521                  2             2                 2  \n",
       "22522                  6             6                 8  \n",
       "22523                  3             3                 3  \n",
       "22524                  9             9                 9  \n",
       "22525                 18            18                23  \n",
       "22526                  5             5                 5  \n",
       "22527                  1             1                 1  \n",
       "22528                  7             9                12  \n",
       "22529                  1             1                 1  \n",
       "22530                  1             1                 1  \n",
       "22531                  3             3                 3  \n",
       "22532                  6             6                 7  \n",
       "22533                  4             4                 4  \n",
       "22534                  9            11                12  \n",
       "22535                  6             6                 8  \n",
       "22536                  1             1                 1  \n",
       "22537                  6             7                 7  \n",
       "22538                  2             2                 3  \n",
       "22539                  2             2                 4  \n",
       "22540                  1             2                 2  \n",
       "22541                  7             7                 7  \n",
       "22542                  4             4                 4  \n",
       "22543                  1             1                 1  \n",
       "\n",
       "[22544 rows x 7 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# First order of business is to find out how to get ORFs out of the gffs\n",
    "\n",
    "# one way would be to group all CDS-sequences from same transcript\n",
    "# and to compile all starts and ends into format similar to tlst\n",
    "# the tlst block format can then be used as a unique identifier of the intron chain as well as the orf in general\n",
    "# if multiple transcripts exist for the same gene with the same CDS - those can later be merged as well.\n",
    "# Otherwise a gene has isoforms coding different proteins\n",
    "cdsDF=df[df['type']=='CDS'].reset_index(drop=True)\n",
    "\n",
    "cdsGDF=cdsDF.groupby(\"parent\").agg({'start':formBlocks,'end':formBlocks}).reset_index()\n",
    "mergedDF=cdsDF[['seqid','parent','strand']].merge(cdsGDF,on='parent',how='outer',indicator=True)\n",
    "assert len(mergedDF[mergedDF[\"_merge\"]==\"both\"])==len(mergedDF), \"ids don't match\"\n",
    "mergedDF.drop_duplicates(inplace=True)\n",
    "mergedDF.reset_index(drop=True,inplace=True)\n",
    "mergedDF.drop('_merge',inplace=True,axis=1)\n",
    "\n",
    "mergedDF['uid']=mergedDF['seqid']+mergedDF['strand']+mergedDF[\"start\"]+\"-\"+mergedDF[\"end\"]\n",
    "mergedDF=mergedDF[['parent','uid']]\n",
    "\n",
    "# now we can extract gene IDs and groupby them\n",
    "# we should count the total number of isoforms included\n",
    "# as well as the number of unique ORFs present for that gene\n",
    "\n",
    "# first get the gene ID\n",
    "mergedDF.reset_index(drop=True,inplace=True)\n",
    "mergedDF[\"gID\"]=mergedDF[\"parent\"].reset_index(drop=True).str.extract('(CHS.(\\d)*)',expand=True)[0]\n",
    "\n",
    "# now we shall group these sequences by the gene ID and compute things of interest\n",
    "mergedGDF=mergedDF.groupby('gID').agg({'uid':[getUniqueOrfCount,'count']}).reset_index()\n",
    "mergedGDF.columns=['gID','uniqueORFs','numTranscripts']\n",
    "\n",
    "# now perform a protein search from gffread output\n",
    "mergedGDF[\"numProteins\"]=np.nan\n",
    "mergedGDF[\"numUniqueProteins\"]=np.nan\n",
    "\n",
    "prots={}\n",
    "setProts=set()\n",
    "chsID=\"\"\n",
    "geneID=\"\"\n",
    "tmp=\"\"\n",
    "curGeneID=\"\"\n",
    "skipWrite=True\n",
    "skipGene=True\n",
    "allKnownIso=set(mergedDF[\"parent\"].tolist())\n",
    "\n",
    "firstGeneID=\"\"\n",
    "c=0\n",
    "\n",
    "with open(\"/home/sparrow/JHU/chess/www/data_update/chess2.02.protein.fa\",\"r\") as inFP:\n",
    "    for line in inFP.readlines():\n",
    "        if line[0]==\">\":\n",
    "            if not skipWrite:\n",
    "                setProts.add(prots[chsID])\n",
    "            # first check if given isoform ID is in known isoforms\n",
    "            tmpID=line[1:].split(\"\\t\")[0].strip(\"\\n\")\n",
    "            if tmpID in allKnownIso:\n",
    "                # proceed safely\n",
    "                skipWrite=False\n",
    "                chsID=tmpID\n",
    "                curGeneID=\"CHS.\"+chsID.split(\".\")[1]\n",
    "                if c==0:\n",
    "                    c+=1\n",
    "                    geneID=curGeneID\n",
    "                if not curGeneID==geneID and not skipGene:\n",
    "                    # we are done with the current gene - should compare isoforms\n",
    "                    mergedGDF.loc[mergedGDF[\"gID\"]==geneID,\"numProteins\"]=len(prots)\n",
    "                    mergedGDF.loc[mergedGDF[\"gID\"]==geneID,\"numUniqueProteins\"]=len(setProts)\n",
    "                    geneID=curGeneID\n",
    "                    prots={}\n",
    "                    setProts=set()\n",
    "                assert chsID not in prots,\"duplicate isoform\"\n",
    "                prots[chsID]=\"\"\n",
    "                skipGene=False # on next entry into this section we can safely save the information into the\n",
    "            else:\n",
    "                skipWrite=True\n",
    "        elif not skipWrite:\n",
    "            prots[chsID]+=line.strip(\"\\n\")\n",
    "        else:\n",
    "            continue\n",
    "            \n",
    "    if not skipGene:\n",
    "        mergedGDF.loc[mergedGDF[\"gID\"]==geneID,\"numProteins\"]=len(prots)\n",
    "        mergedGDF.loc[mergedGDF[\"gID\"]==geneID,\"numUniqueProteins\"]=len(setProts)\n",
    "        \n",
    "assert len(mergedGDF[mergedGDF[\"numUniqueProteins\"].isnull()])==0,\"something is wrong with the proteins - null entries present\"\n",
    "# now here is an interesting observation\n",
    "# Some of the proteins are the same despite having been translated from different exon chains\n",
    "mergedGDF[\"numProteins\"]=mergedGDF[\"numProteins\"].astype(int)\n",
    "mergedGDF[\"numUniqueProteins\"]=mergedGDF[\"numUniqueProteins\"].astype(int)\n",
    "mergedGDF[mergedGDF[\"uniqueORFs\"]!=mergedGDF[\"numUniqueProteins\"]]\n",
    "\n",
    "# important to verify that the number of isoforms present in the fasta file is the same as the ones deduced\n",
    "mergedGDF[mergedGDF[\"numProteins\"]!=mergedGDF[\"numTranscripts\"]]\n",
    "\n",
    "# First order of business is to find out how to get ORFs out of the gffs\n",
    "\n",
    "# one way would be to group all CDS-sequences from same transcript\n",
    "# and to compile all starts and ends into format similar to tlst\n",
    "# the tlst block format can then be used as a unique identifier of the intron chain as well as the orf in general\n",
    "# if multiple transcripts exist for the same gene with the same CDS - those can later be merged as well.\n",
    "# Otherwise a gene has isoforms coding different proteins\n",
    "cdsDF=df[df['type']=='CDS'].reset_index(drop=True)\n",
    "\n",
    "cdsGDF=cdsDF.groupby(\"parent\").agg({'start':formBlocks,'end':formBlocks}).reset_index()\n",
    "mergedDF_2=cdsDF[['seqid','parent','strand']].merge(cdsGDF,on='parent',how='outer',indicator=True)\n",
    "assert len(mergedDF_2[mergedDF_2[\"_merge\"]==\"both\"])==len(mergedDF_2), \"ids don't match\"\n",
    "mergedDF_2.drop_duplicates(inplace=True)\n",
    "mergedDF_2.reset_index(drop=True,inplace=True)\n",
    "mergedDF_2.drop('_merge',inplace=True,axis=1)\n",
    "\n",
    "mergedDF_2['uid']=mergedDF_2['seqid']+mergedDF_2['strand']+mergedDF_2[\"start\"]+\"-\"+mergedDF_2[\"end\"]\n",
    "mergedDF_2=mergedDF_2[['parent','uid']]\n",
    "\n",
    "# now we can extract gene IDs and groupby them\n",
    "# we should count the total number of isoforms included\n",
    "# as well as the number of unique ORFs present for that gene\n",
    "\n",
    "# first get the gene ID\n",
    "mergedDF_2[\"gID\"]=mergedDF_2[\"parent\"].reset_index(drop=True).str.extract('(CHS.(\\d)*)',expand=True)[0]\n",
    "\n",
    "# now we shall group these sequences by the gene ID and compute things of interest\n",
    "mergedGDF_2=mergedDF_2.groupby('gID').agg({'uid':[getUniqueOrfCount,'count']}).reset_index()\n",
    "mergedGDF_2.columns=['gID','uniqueORFs','numTranscripts']\n",
    "mergedGDF_2\n",
    "\n",
    "resDF=mergedGDF.merge(right=mergedGDF_2,how=\"left\",on=\"gID\")\n",
    "resDF"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "resDF.to_csv(\"./revision/ORFpreservation.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# one of the issues is that potentially, there could be two transcripts with different exon/intron-chains\n",
    "# however the ORFs can potentially stay the same.\n",
    "# perhaps we should translate all ORFs and compare them directly instead of relying on the intron chain\n",
    "# in order to learn whether there are any cases where two different chains result in the same ORF"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# next, let's see if all transcripts within protein-coding genes contain CDS sequences\n",
    "# if so - no questions\n",
    "# if yes - is that also the case for novel protein-coding genes?"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
