{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import os\n",
    "import sys\n",
    "\n",
    "version=\"2.01\"\n",
    "\n",
    "gff3Cols=[\"seqid\",\"source\",\"type\",\"start\",\"end\",\"score\",\"strand\",\"phase\",\"attributes\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqid</th>\n",
       "      <th>source</th>\n",
       "      <th>type</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>phase</th>\n",
       "      <th>attributes</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chrM</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>gene</td>\n",
       "      <td>3307.0</td>\n",
       "      <td>4262.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=CHS.57315;GENE_TYPE=protein_coding;STATUS=k...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chrM</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>transcript</td>\n",
       "      <td>3307.0</td>\n",
       "      <td>4262.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=CHS.57315.1;Parent=CHS.57315;gene_name=ND1;...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chrM</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>exon</td>\n",
       "      <td>3307.0</td>\n",
       "      <td>4262.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>Parent=CHS.57315.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chrM</td>\n",
       "      <td>StringTie</td>\n",
       "      <td>transcript</td>\n",
       "      <td>3308.0</td>\n",
       "      <td>4261.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=CHS.57315.2;Parent=CHS.57315;STATUS=novel;A...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chrM</td>\n",
       "      <td>StringTie</td>\n",
       "      <td>exon</td>\n",
       "      <td>3308.0</td>\n",
       "      <td>3934.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>Parent=CHS.57315.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>chrM</td>\n",
       "      <td>StringTie</td>\n",
       "      <td>exon</td>\n",
       "      <td>3959.0</td>\n",
       "      <td>4261.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>Parent=CHS.57315.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>chrM</td>\n",
       "      <td>StringTie</td>\n",
       "      <td>transcript</td>\n",
       "      <td>3308.0</td>\n",
       "      <td>4261.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=CHS.57315.3;Parent=CHS.57315;STATUS=novel;A...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>chrM</td>\n",
       "      <td>StringTie</td>\n",
       "      <td>exon</td>\n",
       "      <td>3308.0</td>\n",
       "      <td>3391.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>Parent=CHS.57315.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>chrM</td>\n",
       "      <td>StringTie</td>\n",
       "      <td>exon</td>\n",
       "      <td>3428.0</td>\n",
       "      <td>4261.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>Parent=CHS.57315.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>chrM</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>gene</td>\n",
       "      <td>4470.0</td>\n",
       "      <td>5511.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=CHS.57316;GENE_TYPE=protein_coding;STATUS=k...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>chrM</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>transcript</td>\n",
       "      <td>4470.0</td>\n",
       "      <td>5511.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=CHS.57316.1;Parent=CHS.57316;gene_name=ND2;...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>chrM</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>exon</td>\n",
       "      <td>4470.0</td>\n",
       "      <td>5511.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>Parent=CHS.57316.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>chrM</td>\n",
       "      <td>StringTie</td>\n",
       "      <td>transcript</td>\n",
       "      <td>4471.0</td>\n",
       "      <td>5493.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=CHS.57316.2;Parent=CHS.57316;STATUS=novel;A...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>chrM</td>\n",
       "      <td>StringTie</td>\n",
       "      <td>exon</td>\n",
       "      <td>4471.0</td>\n",
       "      <td>5066.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>Parent=CHS.57316.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>chrM</td>\n",
       "      <td>StringTie</td>\n",
       "      <td>exon</td>\n",
       "      <td>5183.0</td>\n",
       "      <td>5493.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>Parent=CHS.57316.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>chrM</td>\n",
       "      <td>StringTie</td>\n",
       "      <td>transcript</td>\n",
       "      <td>4471.0</td>\n",
       "      <td>5510.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=CHS.57316.3;Parent=CHS.57316;STATUS=novel;A...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>chrM</td>\n",
       "      <td>StringTie</td>\n",
       "      <td>exon</td>\n",
       "      <td>4471.0</td>\n",
       "      <td>4864.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>Parent=CHS.57316.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>chrM</td>\n",
       "      <td>StringTie</td>\n",
       "      <td>exon</td>\n",
       "      <td>4937.0</td>\n",
       "      <td>5510.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>Parent=CHS.57316.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>chrM</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>gene</td>\n",
       "      <td>5904.0</td>\n",
       "      <td>7445.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=CHS.57317;GENE_TYPE=protein_coding;STATUS=k...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>chrM</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>transcript</td>\n",
       "      <td>5904.0</td>\n",
       "      <td>7445.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=CHS.57317.1;Parent=CHS.57317;gene_name=COX1...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>chrM</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>exon</td>\n",
       "      <td>5904.0</td>\n",
       "      <td>7445.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>Parent=CHS.57317.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>chrM</td>\n",
       "      <td>StringTie</td>\n",
       "      <td>transcript</td>\n",
       "      <td>5905.0</td>\n",
       "      <td>7444.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=CHS.57317.2;Parent=CHS.57317;STATUS=novel;A...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>chrM</td>\n",
       "      <td>StringTie</td>\n",
       "      <td>exon</td>\n",
       "      <td>5905.0</td>\n",
       "      <td>6518.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>Parent=CHS.57317.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>chrM</td>\n",
       "      <td>StringTie</td>\n",
       "      <td>exon</td>\n",
       "      <td>6552.0</td>\n",
       "      <td>7444.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>Parent=CHS.57317.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>chrM</td>\n",
       "      <td>StringTie</td>\n",
       "      <td>transcript</td>\n",
       "      <td>5905.0</td>\n",
       "      <td>7444.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=CHS.57317.3;Parent=CHS.57317;STATUS=novel;A...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>chrM</td>\n",
       "      <td>StringTie</td>\n",
       "      <td>exon</td>\n",
       "      <td>5905.0</td>\n",
       "      <td>6717.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>Parent=CHS.57317.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>chrM</td>\n",
       "      <td>StringTie</td>\n",
       "      <td>exon</td>\n",
       "      <td>6955.0</td>\n",
       "      <td>7444.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>Parent=CHS.57317.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>chrM</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>gene</td>\n",
       "      <td>7586.0</td>\n",
       "      <td>8269.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=CHS.57318;GENE_TYPE=protein_coding;STATUS=k...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>chrM</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>transcript</td>\n",
       "      <td>7586.0</td>\n",
       "      <td>8269.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=CHS.57318.1;Parent=CHS.57318;gene_name=COX2...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>chrM</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>exon</td>\n",
       "      <td>7586.0</td>\n",
       "      <td>8269.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>Parent=CHS.57318.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>chrM</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>gene</td>\n",
       "      <td>8527.0</td>\n",
       "      <td>9207.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=CHS.57321;GENE_TYPE=protein_coding;STATUS=k...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>chrM</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>transcript</td>\n",
       "      <td>8527.0</td>\n",
       "      <td>9207.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=CHS.57321.1;Parent=CHS.57321;gene_name=ATP6...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41</th>\n",
       "      <td>chrM</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>exon</td>\n",
       "      <td>8527.0</td>\n",
       "      <td>9207.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>Parent=CHS.57321.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42</th>\n",
       "      <td>chrM</td>\n",
       "      <td>StringTie</td>\n",
       "      <td>transcript</td>\n",
       "      <td>8573.0</td>\n",
       "      <td>9206.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=CHS.57321.2;Parent=CHS.57321;STATUS=novel;A...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43</th>\n",
       "      <td>chrM</td>\n",
       "      <td>StringTie</td>\n",
       "      <td>exon</td>\n",
       "      <td>8573.0</td>\n",
       "      <td>8857.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>Parent=CHS.57321.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44</th>\n",
       "      <td>chrM</td>\n",
       "      <td>StringTie</td>\n",
       "      <td>exon</td>\n",
       "      <td>8921.0</td>\n",
       "      <td>9206.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>Parent=CHS.57321.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>45</th>\n",
       "      <td>chrM</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>gene</td>\n",
       "      <td>10059.0</td>\n",
       "      <td>10404.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=CHS.57322;GENE_TYPE=protein_coding;STATUS=k...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46</th>\n",
       "      <td>chrM</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>transcript</td>\n",
       "      <td>10059.0</td>\n",
       "      <td>10404.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=CHS.57322.1;Parent=CHS.57322;gene_name=ND3;...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>47</th>\n",
       "      <td>chrM</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>exon</td>\n",
       "      <td>10059.0</td>\n",
       "      <td>10404.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>Parent=CHS.57322.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48</th>\n",
       "      <td>chrM</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>gene</td>\n",
       "      <td>10470.0</td>\n",
       "      <td>10766.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=CHS.57323;GENE_TYPE=protein_coding;STATUS=k...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>49</th>\n",
       "      <td>chrM</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>transcript</td>\n",
       "      <td>10470.0</td>\n",
       "      <td>10766.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=CHS.57323.1;Parent=CHS.57323;gene_name=ND4L...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50</th>\n",
       "      <td>chrM</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>exon</td>\n",
       "      <td>10470.0</td>\n",
       "      <td>10766.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>Parent=CHS.57323.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>51</th>\n",
       "      <td>chrM</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>gene</td>\n",
       "      <td>10760.0</td>\n",
       "      <td>12137.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=CHS.57324;GENE_TYPE=protein_coding;STATUS=k...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>52</th>\n",
       "      <td>chrM</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>transcript</td>\n",
       "      <td>10760.0</td>\n",
       "      <td>12137.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=CHS.57324.1;Parent=CHS.57324;gene_name=ND4;...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>53</th>\n",
       "      <td>chrM</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>exon</td>\n",
       "      <td>10760.0</td>\n",
       "      <td>12137.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>Parent=CHS.57324.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54</th>\n",
       "      <td>chrM</td>\n",
       "      <td>StringTie</td>\n",
       "      <td>transcript</td>\n",
       "      <td>10767.0</td>\n",
       "      <td>12136.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=CHS.57324.2;Parent=CHS.57324;STATUS=novel;A...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>55</th>\n",
       "      <td>chrM</td>\n",
       "      <td>StringTie</td>\n",
       "      <td>exon</td>\n",
       "      <td>10767.0</td>\n",
       "      <td>11614.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>Parent=CHS.57324.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>56</th>\n",
       "      <td>chrM</td>\n",
       "      <td>StringTie</td>\n",
       "      <td>exon</td>\n",
       "      <td>11639.0</td>\n",
       "      <td>12136.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>Parent=CHS.57324.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>57</th>\n",
       "      <td>chrM</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>gene</td>\n",
       "      <td>12337.0</td>\n",
       "      <td>14148.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=CHS.57325;GENE_TYPE=protein_coding;STATUS=k...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58</th>\n",
       "      <td>chrM</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>transcript</td>\n",
       "      <td>12337.0</td>\n",
       "      <td>14148.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=CHS.57325.1;Parent=CHS.57325;gene_name=ND5;...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>59</th>\n",
       "      <td>chrM</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>exon</td>\n",
       "      <td>12337.0</td>\n",
       "      <td>14148.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>Parent=CHS.57325.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60</th>\n",
       "      <td>chrM</td>\n",
       "      <td>StringTie</td>\n",
       "      <td>transcript</td>\n",
       "      <td>12349.0</td>\n",
       "      <td>14147.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=CHS.57325.2;Parent=CHS.57325;STATUS=novel;A...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>61</th>\n",
       "      <td>chrM</td>\n",
       "      <td>StringTie</td>\n",
       "      <td>exon</td>\n",
       "      <td>12349.0</td>\n",
       "      <td>13018.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>Parent=CHS.57325.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>62</th>\n",
       "      <td>chrM</td>\n",
       "      <td>StringTie</td>\n",
       "      <td>exon</td>\n",
       "      <td>13052.0</td>\n",
       "      <td>14147.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>Parent=CHS.57325.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>63</th>\n",
       "      <td>chrM</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>gene</td>\n",
       "      <td>14149.0</td>\n",
       "      <td>14673.0</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=CHS.57326;GENE_TYPE=protein_coding;STATUS=k...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>64</th>\n",
       "      <td>chrM</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>transcript</td>\n",
       "      <td>14149.0</td>\n",
       "      <td>14673.0</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=CHS.57326.1;Parent=CHS.57326;gene_name=ND6;...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>65</th>\n",
       "      <td>chrM</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>exon</td>\n",
       "      <td>14149.0</td>\n",
       "      <td>14673.0</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>Parent=CHS.57326.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>66</th>\n",
       "      <td>chrM</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>gene</td>\n",
       "      <td>14747.0</td>\n",
       "      <td>15887.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=CHS.57327;GENE_TYPE=protein_coding;STATUS=k...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>67</th>\n",
       "      <td>chrM</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>transcript</td>\n",
       "      <td>14747.0</td>\n",
       "      <td>15887.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=CHS.57327.1;Parent=CHS.57327;gene_name=CYTB...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>68</th>\n",
       "      <td>chrM</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>exon</td>\n",
       "      <td>14747.0</td>\n",
       "      <td>15887.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>Parent=CHS.57327.1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>69 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   seqid     source        type    start      end score strand phase  \\\n",
       "0   chrM     RefSeq        gene   3307.0   4262.0     .      +     .   \n",
       "1   chrM     RefSeq  transcript   3307.0   4262.0     .      +     .   \n",
       "2   chrM     RefSeq        exon   3307.0   4262.0     .      +     .   \n",
       "3   chrM  StringTie  transcript   3308.0   4261.0     .      +     .   \n",
       "4   chrM  StringTie        exon   3308.0   3934.0     .      +     .   \n",
       "5   chrM  StringTie        exon   3959.0   4261.0     .      +     .   \n",
       "6   chrM  StringTie  transcript   3308.0   4261.0     .      +     .   \n",
       "7   chrM  StringTie        exon   3308.0   3391.0     .      +     .   \n",
       "8   chrM  StringTie        exon   3428.0   4261.0     .      +     .   \n",
       "9   chrM     RefSeq        gene   4470.0   5511.0     .      +     .   \n",
       "10  chrM     RefSeq  transcript   4470.0   5511.0     .      +     .   \n",
       "11  chrM     RefSeq        exon   4470.0   5511.0     .      +     .   \n",
       "12  chrM  StringTie  transcript   4471.0   5493.0     .      +     .   \n",
       "13  chrM  StringTie        exon   4471.0   5066.0     .      +     .   \n",
       "14  chrM  StringTie        exon   5183.0   5493.0     .      +     .   \n",
       "15  chrM  StringTie  transcript   4471.0   5510.0     .      +     .   \n",
       "16  chrM  StringTie        exon   4471.0   4864.0     .      +     .   \n",
       "17  chrM  StringTie        exon   4937.0   5510.0     .      +     .   \n",
       "18  chrM     RefSeq        gene   5904.0   7445.0     .      +     .   \n",
       "19  chrM     RefSeq  transcript   5904.0   7445.0     .      +     .   \n",
       "20  chrM     RefSeq        exon   5904.0   7445.0     .      +     .   \n",
       "21  chrM  StringTie  transcript   5905.0   7444.0     .      +     .   \n",
       "22  chrM  StringTie        exon   5905.0   6518.0     .      +     .   \n",
       "23  chrM  StringTie        exon   6552.0   7444.0     .      +     .   \n",
       "24  chrM  StringTie  transcript   5905.0   7444.0     .      +     .   \n",
       "25  chrM  StringTie        exon   5905.0   6717.0     .      +     .   \n",
       "26  chrM  StringTie        exon   6955.0   7444.0     .      +     .   \n",
       "27  chrM     RefSeq        gene   7586.0   8269.0     .      +     .   \n",
       "28  chrM     RefSeq  transcript   7586.0   8269.0     .      +     .   \n",
       "29  chrM     RefSeq        exon   7586.0   8269.0     .      +     .   \n",
       "..   ...        ...         ...      ...      ...   ...    ...   ...   \n",
       "39  chrM     RefSeq        gene   8527.0   9207.0     .      +     .   \n",
       "40  chrM     RefSeq  transcript   8527.0   9207.0     .      +     .   \n",
       "41  chrM     RefSeq        exon   8527.0   9207.0     .      +     .   \n",
       "42  chrM  StringTie  transcript   8573.0   9206.0     .      +     .   \n",
       "43  chrM  StringTie        exon   8573.0   8857.0     .      +     .   \n",
       "44  chrM  StringTie        exon   8921.0   9206.0     .      +     .   \n",
       "45  chrM     RefSeq        gene  10059.0  10404.0     .      +     .   \n",
       "46  chrM     RefSeq  transcript  10059.0  10404.0     .      +     .   \n",
       "47  chrM     RefSeq        exon  10059.0  10404.0     .      +     .   \n",
       "48  chrM     RefSeq        gene  10470.0  10766.0     .      +     .   \n",
       "49  chrM     RefSeq  transcript  10470.0  10766.0     .      +     .   \n",
       "50  chrM     RefSeq        exon  10470.0  10766.0     .      +     .   \n",
       "51  chrM     RefSeq        gene  10760.0  12137.0     .      +     .   \n",
       "52  chrM     RefSeq  transcript  10760.0  12137.0     .      +     .   \n",
       "53  chrM     RefSeq        exon  10760.0  12137.0     .      +     .   \n",
       "54  chrM  StringTie  transcript  10767.0  12136.0     .      +     .   \n",
       "55  chrM  StringTie        exon  10767.0  11614.0     .      +     .   \n",
       "56  chrM  StringTie        exon  11639.0  12136.0     .      +     .   \n",
       "57  chrM     RefSeq        gene  12337.0  14148.0     .      +     .   \n",
       "58  chrM     RefSeq  transcript  12337.0  14148.0     .      +     .   \n",
       "59  chrM     RefSeq        exon  12337.0  14148.0     .      +     .   \n",
       "60  chrM  StringTie  transcript  12349.0  14147.0     .      +     .   \n",
       "61  chrM  StringTie        exon  12349.0  13018.0     .      +     .   \n",
       "62  chrM  StringTie        exon  13052.0  14147.0     .      +     .   \n",
       "63  chrM     RefSeq        gene  14149.0  14673.0     .      -     .   \n",
       "64  chrM     RefSeq  transcript  14149.0  14673.0     .      -     .   \n",
       "65  chrM     RefSeq        exon  14149.0  14673.0     .      -     .   \n",
       "66  chrM     RefSeq        gene  14747.0  15887.0     .      +     .   \n",
       "67  chrM     RefSeq  transcript  14747.0  15887.0     .      +     .   \n",
       "68  chrM     RefSeq        exon  14747.0  15887.0     .      +     .   \n",
       "\n",
       "                                           attributes  \n",
       "0   ID=CHS.57315;GENE_TYPE=protein_coding;STATUS=k...  \n",
       "1   ID=CHS.57315.1;Parent=CHS.57315;gene_name=ND1;...  \n",
       "2                                  Parent=CHS.57315.1  \n",
       "3   ID=CHS.57315.2;Parent=CHS.57315;STATUS=novel;A...  \n",
       "4                                  Parent=CHS.57315.2  \n",
       "5                                  Parent=CHS.57315.2  \n",
       "6   ID=CHS.57315.3;Parent=CHS.57315;STATUS=novel;A...  \n",
       "7                                  Parent=CHS.57315.3  \n",
       "8                                  Parent=CHS.57315.3  \n",
       "9   ID=CHS.57316;GENE_TYPE=protein_coding;STATUS=k...  \n",
       "10  ID=CHS.57316.1;Parent=CHS.57316;gene_name=ND2;...  \n",
       "11                                 Parent=CHS.57316.1  \n",
       "12  ID=CHS.57316.2;Parent=CHS.57316;STATUS=novel;A...  \n",
       "13                                 Parent=CHS.57316.2  \n",
       "14                                 Parent=CHS.57316.2  \n",
       "15  ID=CHS.57316.3;Parent=CHS.57316;STATUS=novel;A...  \n",
       "16                                 Parent=CHS.57316.3  \n",
       "17                                 Parent=CHS.57316.3  \n",
       "18  ID=CHS.57317;GENE_TYPE=protein_coding;STATUS=k...  \n",
       "19  ID=CHS.57317.1;Parent=CHS.57317;gene_name=COX1...  \n",
       "20                                 Parent=CHS.57317.1  \n",
       "21  ID=CHS.57317.2;Parent=CHS.57317;STATUS=novel;A...  \n",
       "22                                 Parent=CHS.57317.2  \n",
       "23                                 Parent=CHS.57317.2  \n",
       "24  ID=CHS.57317.3;Parent=CHS.57317;STATUS=novel;A...  \n",
       "25                                 Parent=CHS.57317.3  \n",
       "26                                 Parent=CHS.57317.3  \n",
       "27  ID=CHS.57318;GENE_TYPE=protein_coding;STATUS=k...  \n",
       "28  ID=CHS.57318.1;Parent=CHS.57318;gene_name=COX2...  \n",
       "29                                 Parent=CHS.57318.1  \n",
       "..                                                ...  \n",
       "39  ID=CHS.57321;GENE_TYPE=protein_coding;STATUS=k...  \n",
       "40  ID=CHS.57321.1;Parent=CHS.57321;gene_name=ATP6...  \n",
       "41                                 Parent=CHS.57321.1  \n",
       "42  ID=CHS.57321.2;Parent=CHS.57321;STATUS=novel;A...  \n",
       "43                                 Parent=CHS.57321.2  \n",
       "44                                 Parent=CHS.57321.2  \n",
       "45  ID=CHS.57322;GENE_TYPE=protein_coding;STATUS=k...  \n",
       "46  ID=CHS.57322.1;Parent=CHS.57322;gene_name=ND3;...  \n",
       "47                                 Parent=CHS.57322.1  \n",
       "48  ID=CHS.57323;GENE_TYPE=protein_coding;STATUS=k...  \n",
       "49  ID=CHS.57323.1;Parent=CHS.57323;gene_name=ND4L...  \n",
       "50                                 Parent=CHS.57323.1  \n",
       "51  ID=CHS.57324;GENE_TYPE=protein_coding;STATUS=k...  \n",
       "52  ID=CHS.57324.1;Parent=CHS.57324;gene_name=ND4;...  \n",
       "53                                 Parent=CHS.57324.1  \n",
       "54  ID=CHS.57324.2;Parent=CHS.57324;STATUS=novel;A...  \n",
       "55                                 Parent=CHS.57324.2  \n",
       "56                                 Parent=CHS.57324.2  \n",
       "57  ID=CHS.57325;GENE_TYPE=protein_coding;STATUS=k...  \n",
       "58  ID=CHS.57325.1;Parent=CHS.57325;gene_name=ND5;...  \n",
       "59                                 Parent=CHS.57325.1  \n",
       "60  ID=CHS.57325.2;Parent=CHS.57325;STATUS=novel;A...  \n",
       "61                                 Parent=CHS.57325.2  \n",
       "62                                 Parent=CHS.57325.2  \n",
       "63  ID=CHS.57326;GENE_TYPE=protein_coding;STATUS=k...  \n",
       "64  ID=CHS.57326.1;Parent=CHS.57326;gene_name=ND6;...  \n",
       "65                                 Parent=CHS.57326.1  \n",
       "66  ID=CHS.57327;GENE_TYPE=protein_coding;STATUS=k...  \n",
       "67  ID=CHS.57327.1;Parent=CHS.57327;gene_name=CYTB...  \n",
       "68                                 Parent=CHS.57327.1  \n",
       "\n",
       "[69 rows x 9 columns]"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df=pd.read_csv(\"./chess2.0.gff\",sep=\"\\t\",skiprows=3,names=gff3Cols)\n",
    "df=df[df[\"seqid\"]==\"chrM\"].reset_index(drop=True)\n",
    "df=df[df[\"type\"].isin([\"gene\",\"exon\",\"transcript\"])].reset_index(drop=True)\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "number of mitochondrial genes in CHESS is: 13\n"
     ]
    }
   ],
   "source": [
    "print(\"number of mitochondrial genes in CHESS is: \"+str(len(df[df[\"type\"]==\"gene\"].reset_index(drop=True))))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "# first test on gencode to see if there is anything with multiple exons. Only take those annotated as either gene,transcript or exon\n",
    "# next proceed with the same procedure onto refseq\n",
    "# how many genes/transcripts are there in total?\n",
    "# is the number the same for gencode and refseq?\n",
    "# next look at chess and decide what to do"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "number of mitochondrial genes in GENCODE is: 0\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqid</th>\n",
       "      <th>source</th>\n",
       "      <th>type</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>phase</th>\n",
       "      <th>attributes</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chrM</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>transcript</td>\n",
       "      <td>577</td>\n",
       "      <td>647</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=ENST00000387314.1;geneID=ENSG00000210049.1;...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chrM</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>exon</td>\n",
       "      <td>577</td>\n",
       "      <td>647</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>Parent=ENST00000387314.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chrM</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>transcript</td>\n",
       "      <td>648</td>\n",
       "      <td>1601</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=ENST00000389680.2;geneID=ENSG00000211459.2;...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chrM</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>exon</td>\n",
       "      <td>648</td>\n",
       "      <td>1601</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>Parent=ENST00000389680.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chrM</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>transcript</td>\n",
       "      <td>1602</td>\n",
       "      <td>1670</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=ENST00000387342.1;geneID=ENSG00000210077.1;...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>chrM</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>exon</td>\n",
       "      <td>1602</td>\n",
       "      <td>1670</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>Parent=ENST00000387342.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>chrM</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>transcript</td>\n",
       "      <td>1671</td>\n",
       "      <td>3229</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=ENST00000387347.2;geneID=ENSG00000210082.2;...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>chrM</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>exon</td>\n",
       "      <td>1671</td>\n",
       "      <td>3229</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>Parent=ENST00000387347.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>chrM</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>transcript</td>\n",
       "      <td>3230</td>\n",
       "      <td>3304</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=ENST00000386347.1;geneID=ENSG00000209082.1;...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>chrM</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>exon</td>\n",
       "      <td>3230</td>\n",
       "      <td>3304</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>Parent=ENST00000386347.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>chrM</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>mRNA</td>\n",
       "      <td>3307</td>\n",
       "      <td>4262</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=ENST00000361390.2;geneID=ENSG00000198888.2;...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>chrM</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>exon</td>\n",
       "      <td>3307</td>\n",
       "      <td>4262</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>Parent=ENST00000361390.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>chrM</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>transcript</td>\n",
       "      <td>4263</td>\n",
       "      <td>4331</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=ENST00000387365.1;geneID=ENSG00000210100.1;...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>chrM</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>exon</td>\n",
       "      <td>4263</td>\n",
       "      <td>4331</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>Parent=ENST00000387365.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>chrM</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>transcript</td>\n",
       "      <td>4329</td>\n",
       "      <td>4400</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=ENST00000387372.1;geneID=ENSG00000210107.1;...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>chrM</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>exon</td>\n",
       "      <td>4329</td>\n",
       "      <td>4400</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>Parent=ENST00000387372.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>chrM</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>transcript</td>\n",
       "      <td>4402</td>\n",
       "      <td>4469</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=ENST00000387377.1;geneID=ENSG00000210112.1;...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>chrM</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>exon</td>\n",
       "      <td>4402</td>\n",
       "      <td>4469</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>Parent=ENST00000387377.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>chrM</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>mRNA</td>\n",
       "      <td>4470</td>\n",
       "      <td>5511</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=ENST00000361453.3;geneID=ENSG00000198763.3;...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>chrM</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>exon</td>\n",
       "      <td>4470</td>\n",
       "      <td>5511</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>Parent=ENST00000361453.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>chrM</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>transcript</td>\n",
       "      <td>5512</td>\n",
       "      <td>5579</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=ENST00000387382.1;geneID=ENSG00000210117.1;...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>chrM</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>exon</td>\n",
       "      <td>5512</td>\n",
       "      <td>5579</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>Parent=ENST00000387382.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>chrM</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>transcript</td>\n",
       "      <td>5587</td>\n",
       "      <td>5655</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=ENST00000387392.1;geneID=ENSG00000210127.1;...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>chrM</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>exon</td>\n",
       "      <td>5587</td>\n",
       "      <td>5655</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>Parent=ENST00000387392.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>chrM</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>transcript</td>\n",
       "      <td>5657</td>\n",
       "      <td>5729</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=ENST00000387400.1;geneID=ENSG00000210135.1;...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>chrM</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>exon</td>\n",
       "      <td>5657</td>\n",
       "      <td>5729</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>Parent=ENST00000387400.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>chrM</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>transcript</td>\n",
       "      <td>5761</td>\n",
       "      <td>5826</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=ENST00000387405.1;geneID=ENSG00000210140.1;...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>chrM</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>exon</td>\n",
       "      <td>5761</td>\n",
       "      <td>5826</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>Parent=ENST00000387405.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>chrM</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>transcript</td>\n",
       "      <td>5826</td>\n",
       "      <td>5891</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=ENST00000387409.1;geneID=ENSG00000210144.1;...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>chrM</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>exon</td>\n",
       "      <td>5826</td>\n",
       "      <td>5891</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>Parent=ENST00000387409.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44</th>\n",
       "      <td>chrM</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>mRNA</td>\n",
       "      <td>9207</td>\n",
       "      <td>9990</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=ENST00000362079.2;geneID=ENSG00000198938.2;...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>45</th>\n",
       "      <td>chrM</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>exon</td>\n",
       "      <td>9207</td>\n",
       "      <td>9990</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>Parent=ENST00000362079.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46</th>\n",
       "      <td>chrM</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>transcript</td>\n",
       "      <td>9991</td>\n",
       "      <td>10058</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=ENST00000387429.1;geneID=ENSG00000210164.1;...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>47</th>\n",
       "      <td>chrM</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>exon</td>\n",
       "      <td>9991</td>\n",
       "      <td>10058</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>Parent=ENST00000387429.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48</th>\n",
       "      <td>chrM</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>mRNA</td>\n",
       "      <td>10059</td>\n",
       "      <td>10404</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=ENST00000361227.2;geneID=ENSG00000198840.2;...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>49</th>\n",
       "      <td>chrM</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>exon</td>\n",
       "      <td>10059</td>\n",
       "      <td>10404</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>Parent=ENST00000361227.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50</th>\n",
       "      <td>chrM</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>transcript</td>\n",
       "      <td>10405</td>\n",
       "      <td>10469</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=ENST00000387439.1;geneID=ENSG00000210174.1;...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>51</th>\n",
       "      <td>chrM</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>exon</td>\n",
       "      <td>10405</td>\n",
       "      <td>10469</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>Parent=ENST00000387439.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>52</th>\n",
       "      <td>chrM</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>mRNA</td>\n",
       "      <td>10470</td>\n",
       "      <td>10766</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=ENST00000361335.1;geneID=ENSG00000212907.2;...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>53</th>\n",
       "      <td>chrM</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>exon</td>\n",
       "      <td>10470</td>\n",
       "      <td>10766</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>Parent=ENST00000361335.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54</th>\n",
       "      <td>chrM</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>mRNA</td>\n",
       "      <td>10760</td>\n",
       "      <td>12137</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=ENST00000361381.2;geneID=ENSG00000198886.2;...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>55</th>\n",
       "      <td>chrM</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>exon</td>\n",
       "      <td>10760</td>\n",
       "      <td>12137</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>Parent=ENST00000361381.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>56</th>\n",
       "      <td>chrM</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>transcript</td>\n",
       "      <td>12138</td>\n",
       "      <td>12206</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=ENST00000387441.1;geneID=ENSG00000210176.1;...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>57</th>\n",
       "      <td>chrM</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>exon</td>\n",
       "      <td>12138</td>\n",
       "      <td>12206</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>Parent=ENST00000387441.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58</th>\n",
       "      <td>chrM</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>transcript</td>\n",
       "      <td>12207</td>\n",
       "      <td>12265</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=ENST00000387449.1;geneID=ENSG00000210184.1;...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>59</th>\n",
       "      <td>chrM</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>exon</td>\n",
       "      <td>12207</td>\n",
       "      <td>12265</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>Parent=ENST00000387449.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60</th>\n",
       "      <td>chrM</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>transcript</td>\n",
       "      <td>12266</td>\n",
       "      <td>12336</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=ENST00000387456.1;geneID=ENSG00000210191.1;...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>61</th>\n",
       "      <td>chrM</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>exon</td>\n",
       "      <td>12266</td>\n",
       "      <td>12336</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>Parent=ENST00000387456.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>62</th>\n",
       "      <td>chrM</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>mRNA</td>\n",
       "      <td>12337</td>\n",
       "      <td>14148</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=ENST00000361567.2;geneID=ENSG00000198786.2;...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>63</th>\n",
       "      <td>chrM</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>exon</td>\n",
       "      <td>12337</td>\n",
       "      <td>14148</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>Parent=ENST00000361567.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>64</th>\n",
       "      <td>chrM</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>mRNA</td>\n",
       "      <td>14149</td>\n",
       "      <td>14673</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=ENST00000361681.2;geneID=ENSG00000198695.2;...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>65</th>\n",
       "      <td>chrM</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>exon</td>\n",
       "      <td>14149</td>\n",
       "      <td>14673</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>Parent=ENST00000361681.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>66</th>\n",
       "      <td>chrM</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>transcript</td>\n",
       "      <td>14674</td>\n",
       "      <td>14742</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=ENST00000387459.1;geneID=ENSG00000210194.1;...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>67</th>\n",
       "      <td>chrM</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>exon</td>\n",
       "      <td>14674</td>\n",
       "      <td>14742</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>Parent=ENST00000387459.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>68</th>\n",
       "      <td>chrM</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>mRNA</td>\n",
       "      <td>14747</td>\n",
       "      <td>15887</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=ENST00000361789.2;geneID=ENSG00000198727.2;...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>69</th>\n",
       "      <td>chrM</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>exon</td>\n",
       "      <td>14747</td>\n",
       "      <td>15887</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>Parent=ENST00000361789.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>70</th>\n",
       "      <td>chrM</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>transcript</td>\n",
       "      <td>15888</td>\n",
       "      <td>15953</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=ENST00000387460.2;geneID=ENSG00000210195.2;...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>71</th>\n",
       "      <td>chrM</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>exon</td>\n",
       "      <td>15888</td>\n",
       "      <td>15953</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>Parent=ENST00000387460.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>72</th>\n",
       "      <td>chrM</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>transcript</td>\n",
       "      <td>15956</td>\n",
       "      <td>16023</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=ENST00000387461.2;geneID=ENSG00000210196.2;...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>73</th>\n",
       "      <td>chrM</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>exon</td>\n",
       "      <td>15956</td>\n",
       "      <td>16023</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>Parent=ENST00000387461.2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>74 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   seqid   source        type  start    end score strand phase  \\\n",
       "0   chrM  ENSEMBL  transcript    577    647     .      +     .   \n",
       "1   chrM  ENSEMBL        exon    577    647     .      +     .   \n",
       "2   chrM  ENSEMBL  transcript    648   1601     .      +     .   \n",
       "3   chrM  ENSEMBL        exon    648   1601     .      +     .   \n",
       "4   chrM  ENSEMBL  transcript   1602   1670     .      +     .   \n",
       "5   chrM  ENSEMBL        exon   1602   1670     .      +     .   \n",
       "6   chrM  ENSEMBL  transcript   1671   3229     .      +     .   \n",
       "7   chrM  ENSEMBL        exon   1671   3229     .      +     .   \n",
       "8   chrM  ENSEMBL  transcript   3230   3304     .      +     .   \n",
       "9   chrM  ENSEMBL        exon   3230   3304     .      +     .   \n",
       "10  chrM  ENSEMBL        mRNA   3307   4262     .      +     .   \n",
       "11  chrM  ENSEMBL        exon   3307   4262     .      +     .   \n",
       "12  chrM  ENSEMBL  transcript   4263   4331     .      +     .   \n",
       "13  chrM  ENSEMBL        exon   4263   4331     .      +     .   \n",
       "14  chrM  ENSEMBL  transcript   4329   4400     .      -     .   \n",
       "15  chrM  ENSEMBL        exon   4329   4400     .      -     .   \n",
       "16  chrM  ENSEMBL  transcript   4402   4469     .      +     .   \n",
       "17  chrM  ENSEMBL        exon   4402   4469     .      +     .   \n",
       "18  chrM  ENSEMBL        mRNA   4470   5511     .      +     .   \n",
       "19  chrM  ENSEMBL        exon   4470   5511     .      +     .   \n",
       "20  chrM  ENSEMBL  transcript   5512   5579     .      +     .   \n",
       "21  chrM  ENSEMBL        exon   5512   5579     .      +     .   \n",
       "22  chrM  ENSEMBL  transcript   5587   5655     .      -     .   \n",
       "23  chrM  ENSEMBL        exon   5587   5655     .      -     .   \n",
       "24  chrM  ENSEMBL  transcript   5657   5729     .      -     .   \n",
       "25  chrM  ENSEMBL        exon   5657   5729     .      -     .   \n",
       "26  chrM  ENSEMBL  transcript   5761   5826     .      -     .   \n",
       "27  chrM  ENSEMBL        exon   5761   5826     .      -     .   \n",
       "28  chrM  ENSEMBL  transcript   5826   5891     .      -     .   \n",
       "29  chrM  ENSEMBL        exon   5826   5891     .      -     .   \n",
       "..   ...      ...         ...    ...    ...   ...    ...   ...   \n",
       "44  chrM  ENSEMBL        mRNA   9207   9990     .      +     .   \n",
       "45  chrM  ENSEMBL        exon   9207   9990     .      +     .   \n",
       "46  chrM  ENSEMBL  transcript   9991  10058     .      +     .   \n",
       "47  chrM  ENSEMBL        exon   9991  10058     .      +     .   \n",
       "48  chrM  ENSEMBL        mRNA  10059  10404     .      +     .   \n",
       "49  chrM  ENSEMBL        exon  10059  10404     .      +     .   \n",
       "50  chrM  ENSEMBL  transcript  10405  10469     .      +     .   \n",
       "51  chrM  ENSEMBL        exon  10405  10469     .      +     .   \n",
       "52  chrM  ENSEMBL        mRNA  10470  10766     .      +     .   \n",
       "53  chrM  ENSEMBL        exon  10470  10766     .      +     .   \n",
       "54  chrM  ENSEMBL        mRNA  10760  12137     .      +     .   \n",
       "55  chrM  ENSEMBL        exon  10760  12137     .      +     .   \n",
       "56  chrM  ENSEMBL  transcript  12138  12206     .      +     .   \n",
       "57  chrM  ENSEMBL        exon  12138  12206     .      +     .   \n",
       "58  chrM  ENSEMBL  transcript  12207  12265     .      +     .   \n",
       "59  chrM  ENSEMBL        exon  12207  12265     .      +     .   \n",
       "60  chrM  ENSEMBL  transcript  12266  12336     .      +     .   \n",
       "61  chrM  ENSEMBL        exon  12266  12336     .      +     .   \n",
       "62  chrM  ENSEMBL        mRNA  12337  14148     .      +     .   \n",
       "63  chrM  ENSEMBL        exon  12337  14148     .      +     .   \n",
       "64  chrM  ENSEMBL        mRNA  14149  14673     .      -     .   \n",
       "65  chrM  ENSEMBL        exon  14149  14673     .      -     .   \n",
       "66  chrM  ENSEMBL  transcript  14674  14742     .      -     .   \n",
       "67  chrM  ENSEMBL        exon  14674  14742     .      -     .   \n",
       "68  chrM  ENSEMBL        mRNA  14747  15887     .      +     .   \n",
       "69  chrM  ENSEMBL        exon  14747  15887     .      +     .   \n",
       "70  chrM  ENSEMBL  transcript  15888  15953     .      +     .   \n",
       "71  chrM  ENSEMBL        exon  15888  15953     .      +     .   \n",
       "72  chrM  ENSEMBL  transcript  15956  16023     .      -     .   \n",
       "73  chrM  ENSEMBL        exon  15956  16023     .      -     .   \n",
       "\n",
       "                                           attributes  \n",
       "0   ID=ENST00000387314.1;geneID=ENSG00000210049.1;...  \n",
       "1                            Parent=ENST00000387314.1  \n",
       "2   ID=ENST00000389680.2;geneID=ENSG00000211459.2;...  \n",
       "3                            Parent=ENST00000389680.2  \n",
       "4   ID=ENST00000387342.1;geneID=ENSG00000210077.1;...  \n",
       "5                            Parent=ENST00000387342.1  \n",
       "6   ID=ENST00000387347.2;geneID=ENSG00000210082.2;...  \n",
       "7                            Parent=ENST00000387347.2  \n",
       "8   ID=ENST00000386347.1;geneID=ENSG00000209082.1;...  \n",
       "9                            Parent=ENST00000386347.1  \n",
       "10  ID=ENST00000361390.2;geneID=ENSG00000198888.2;...  \n",
       "11                           Parent=ENST00000361390.2  \n",
       "12  ID=ENST00000387365.1;geneID=ENSG00000210100.1;...  \n",
       "13                           Parent=ENST00000387365.1  \n",
       "14  ID=ENST00000387372.1;geneID=ENSG00000210107.1;...  \n",
       "15                           Parent=ENST00000387372.1  \n",
       "16  ID=ENST00000387377.1;geneID=ENSG00000210112.1;...  \n",
       "17                           Parent=ENST00000387377.1  \n",
       "18  ID=ENST00000361453.3;geneID=ENSG00000198763.3;...  \n",
       "19                           Parent=ENST00000361453.3  \n",
       "20  ID=ENST00000387382.1;geneID=ENSG00000210117.1;...  \n",
       "21                           Parent=ENST00000387382.1  \n",
       "22  ID=ENST00000387392.1;geneID=ENSG00000210127.1;...  \n",
       "23                           Parent=ENST00000387392.1  \n",
       "24  ID=ENST00000387400.1;geneID=ENSG00000210135.1;...  \n",
       "25                           Parent=ENST00000387400.1  \n",
       "26  ID=ENST00000387405.1;geneID=ENSG00000210140.1;...  \n",
       "27                           Parent=ENST00000387405.1  \n",
       "28  ID=ENST00000387409.1;geneID=ENSG00000210144.1;...  \n",
       "29                           Parent=ENST00000387409.1  \n",
       "..                                                ...  \n",
       "44  ID=ENST00000362079.2;geneID=ENSG00000198938.2;...  \n",
       "45                           Parent=ENST00000362079.2  \n",
       "46  ID=ENST00000387429.1;geneID=ENSG00000210164.1;...  \n",
       "47                           Parent=ENST00000387429.1  \n",
       "48  ID=ENST00000361227.2;geneID=ENSG00000198840.2;...  \n",
       "49                           Parent=ENST00000361227.2  \n",
       "50  ID=ENST00000387439.1;geneID=ENSG00000210174.1;...  \n",
       "51                           Parent=ENST00000387439.1  \n",
       "52  ID=ENST00000361335.1;geneID=ENSG00000212907.2;...  \n",
       "53                           Parent=ENST00000361335.1  \n",
       "54  ID=ENST00000361381.2;geneID=ENSG00000198886.2;...  \n",
       "55                           Parent=ENST00000361381.2  \n",
       "56  ID=ENST00000387441.1;geneID=ENSG00000210176.1;...  \n",
       "57                           Parent=ENST00000387441.1  \n",
       "58  ID=ENST00000387449.1;geneID=ENSG00000210184.1;...  \n",
       "59                           Parent=ENST00000387449.1  \n",
       "60  ID=ENST00000387456.1;geneID=ENSG00000210191.1;...  \n",
       "61                           Parent=ENST00000387456.1  \n",
       "62  ID=ENST00000361567.2;geneID=ENSG00000198786.2;...  \n",
       "63                           Parent=ENST00000361567.2  \n",
       "64  ID=ENST00000361681.2;geneID=ENSG00000198695.2;...  \n",
       "65                           Parent=ENST00000361681.2  \n",
       "66  ID=ENST00000387459.1;geneID=ENSG00000210194.1;...  \n",
       "67                           Parent=ENST00000387459.1  \n",
       "68  ID=ENST00000361789.2;geneID=ENSG00000198727.2;...  \n",
       "69                           Parent=ENST00000361789.2  \n",
       "70  ID=ENST00000387460.2;geneID=ENSG00000210195.2;...  \n",
       "71                           Parent=ENST00000387460.2  \n",
       "72  ID=ENST00000387461.2;geneID=ENSG00000210196.2;...  \n",
       "73                           Parent=ENST00000387461.2  \n",
       "\n",
       "[74 rows x 9 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gencode=pd.read_csv(\"/home/sparrow/genomicData/hg38/annotations/hg38_p8_gencode25.gff3\",skiprows=5,names=gff3Cols,sep=\"\\t\")\n",
    "gencode=gencode[gencode[\"seqid\"]==\"chrM\"].reset_index(drop=True)\n",
    "gencode=gencode[gencode[\"type\"].isin([\"gene\",\"exon\",\"transcript\",\"mRNA\"])].reset_index(drop=True)\n",
    "print(\"number of mitochondrial genes in GENCODE is: \"+str(len(gencode[gencode[\"type\"]==\"gene\"])))\n",
    "gencode"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqid</th>\n",
       "      <th>source</th>\n",
       "      <th>type</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>phase</th>\n",
       "      <th>attributes</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chrM</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>tRNA</td>\n",
       "      <td>577</td>\n",
       "      <td>647</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=rna164024;geneID=gene60051;gene_name=TRNF;D...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chrM</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>exon</td>\n",
       "      <td>577</td>\n",
       "      <td>647</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>Parent=rna164024</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chrM</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>rRNA</td>\n",
       "      <td>648</td>\n",
       "      <td>1601</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=rna164025;geneID=gene60052;gene_name=RNR1;D...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chrM</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>exon</td>\n",
       "      <td>648</td>\n",
       "      <td>1601</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>Parent=rna164025</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chrM</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>tRNA</td>\n",
       "      <td>1602</td>\n",
       "      <td>1670</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=rna164026;geneID=gene60053;gene_name=TRNV;D...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>chrM</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>exon</td>\n",
       "      <td>1602</td>\n",
       "      <td>1670</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>Parent=rna164026</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>chrM</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>rRNA</td>\n",
       "      <td>1671</td>\n",
       "      <td>3229</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=rna164027;geneID=gene60054;gene_name=RNR2;D...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>chrM</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>exon</td>\n",
       "      <td>1671</td>\n",
       "      <td>3229</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>Parent=rna164027</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>chrM</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>tRNA</td>\n",
       "      <td>3230</td>\n",
       "      <td>3304</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=rna164028;geneID=gene60055;gene_name=TRNL1;...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>chrM</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>exon</td>\n",
       "      <td>3230</td>\n",
       "      <td>3304</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>Parent=rna164028</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>chrM</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>exon</td>\n",
       "      <td>3307</td>\n",
       "      <td>4262</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>Parent=gene60056</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>chrM</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>tRNA</td>\n",
       "      <td>4263</td>\n",
       "      <td>4331</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=rna164029;geneID=gene60057;gene_name=TRNI;D...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>chrM</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>exon</td>\n",
       "      <td>4263</td>\n",
       "      <td>4331</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>Parent=rna164029</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>chrM</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>tRNA</td>\n",
       "      <td>4329</td>\n",
       "      <td>4400</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=rna164030;geneID=gene60058;gene_name=TRNQ;D...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>chrM</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>exon</td>\n",
       "      <td>4329</td>\n",
       "      <td>4400</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>Parent=rna164030</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>chrM</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>tRNA</td>\n",
       "      <td>4402</td>\n",
       "      <td>4469</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=rna164031;geneID=gene60059;gene_name=TRNM;D...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>chrM</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>exon</td>\n",
       "      <td>4402</td>\n",
       "      <td>4469</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>Parent=rna164031</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>chrM</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>exon</td>\n",
       "      <td>4470</td>\n",
       "      <td>5511</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>Parent=gene60060</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>chrM</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>tRNA</td>\n",
       "      <td>5512</td>\n",
       "      <td>5579</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=rna164032;geneID=gene60061;gene_name=TRNW;D...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>chrM</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>exon</td>\n",
       "      <td>5512</td>\n",
       "      <td>5579</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>Parent=rna164032</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>chrM</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>tRNA</td>\n",
       "      <td>5587</td>\n",
       "      <td>5655</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=rna164033;geneID=gene60062;gene_name=TRNA;D...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>chrM</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>exon</td>\n",
       "      <td>5587</td>\n",
       "      <td>5655</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>Parent=rna164033</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>chrM</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>tRNA</td>\n",
       "      <td>5657</td>\n",
       "      <td>5729</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=rna164034;geneID=gene60063;gene_name=TRNN;D...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>chrM</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>exon</td>\n",
       "      <td>5657</td>\n",
       "      <td>5729</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>Parent=rna164034</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>chrM</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>tRNA</td>\n",
       "      <td>5761</td>\n",
       "      <td>5826</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=rna164035;geneID=gene60064;gene_name=TRNC;D...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>chrM</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>exon</td>\n",
       "      <td>5761</td>\n",
       "      <td>5826</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>Parent=rna164035</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>chrM</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>tRNA</td>\n",
       "      <td>5826</td>\n",
       "      <td>5891</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=rna164036;geneID=gene60065;gene_name=TRNY;D...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>chrM</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>exon</td>\n",
       "      <td>5826</td>\n",
       "      <td>5891</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>Parent=rna164036</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>chrM</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>exon</td>\n",
       "      <td>5904</td>\n",
       "      <td>7445</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>Parent=gene60066</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>chrM</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>tRNA</td>\n",
       "      <td>7446</td>\n",
       "      <td>7514</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=rna164037;geneID=gene60067;gene_name=TRNS1;...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>chrM</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>tRNA</td>\n",
       "      <td>7518</td>\n",
       "      <td>7585</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=rna164038;geneID=gene60068;gene_name=TRND;D...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>chrM</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>exon</td>\n",
       "      <td>7518</td>\n",
       "      <td>7585</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>Parent=rna164038</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>chrM</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>exon</td>\n",
       "      <td>7586</td>\n",
       "      <td>8269</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>Parent=gene60069</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>chrM</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>tRNA</td>\n",
       "      <td>8295</td>\n",
       "      <td>8364</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=rna164039;geneID=gene60070;gene_name=TRNK;D...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>chrM</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>exon</td>\n",
       "      <td>8295</td>\n",
       "      <td>8364</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>Parent=rna164039</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>chrM</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>exon</td>\n",
       "      <td>8366</td>\n",
       "      <td>8572</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>Parent=gene60071</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>chrM</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>exon</td>\n",
       "      <td>8527</td>\n",
       "      <td>9207</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>Parent=gene60072</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>chrM</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>exon</td>\n",
       "      <td>9207</td>\n",
       "      <td>9990</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>Parent=gene60073</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>chrM</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>tRNA</td>\n",
       "      <td>9991</td>\n",
       "      <td>10058</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=rna164040;geneID=gene60074;gene_name=TRNG;D...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>chrM</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>exon</td>\n",
       "      <td>9991</td>\n",
       "      <td>10058</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>Parent=rna164040</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41</th>\n",
       "      <td>chrM</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>exon</td>\n",
       "      <td>10059</td>\n",
       "      <td>10404</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>Parent=gene60075</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42</th>\n",
       "      <td>chrM</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>tRNA</td>\n",
       "      <td>10405</td>\n",
       "      <td>10469</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=rna164041;geneID=gene60076;gene_name=TRNR;D...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43</th>\n",
       "      <td>chrM</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>exon</td>\n",
       "      <td>10405</td>\n",
       "      <td>10469</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>Parent=rna164041</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44</th>\n",
       "      <td>chrM</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>exon</td>\n",
       "      <td>10470</td>\n",
       "      <td>10766</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>Parent=gene60077</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>45</th>\n",
       "      <td>chrM</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>exon</td>\n",
       "      <td>10760</td>\n",
       "      <td>12137</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>Parent=gene60078</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46</th>\n",
       "      <td>chrM</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>tRNA</td>\n",
       "      <td>12138</td>\n",
       "      <td>12206</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=rna164042;geneID=gene60079;gene_name=TRNH;D...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>47</th>\n",
       "      <td>chrM</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>exon</td>\n",
       "      <td>12138</td>\n",
       "      <td>12206</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>Parent=rna164042</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48</th>\n",
       "      <td>chrM</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>tRNA</td>\n",
       "      <td>12207</td>\n",
       "      <td>12265</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=rna164043;geneID=gene60080;gene_name=TRNS2;...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>49</th>\n",
       "      <td>chrM</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>exon</td>\n",
       "      <td>12207</td>\n",
       "      <td>12265</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>Parent=rna164043</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50</th>\n",
       "      <td>chrM</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>tRNA</td>\n",
       "      <td>12266</td>\n",
       "      <td>12336</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=rna164044;geneID=gene60081;gene_name=TRNL2;...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>51</th>\n",
       "      <td>chrM</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>exon</td>\n",
       "      <td>12266</td>\n",
       "      <td>12336</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>Parent=rna164044</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>52</th>\n",
       "      <td>chrM</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>exon</td>\n",
       "      <td>12337</td>\n",
       "      <td>14148</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>Parent=gene60082</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>53</th>\n",
       "      <td>chrM</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>exon</td>\n",
       "      <td>14149</td>\n",
       "      <td>14673</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>Parent=gene60083</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54</th>\n",
       "      <td>chrM</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>tRNA</td>\n",
       "      <td>14674</td>\n",
       "      <td>14742</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=rna164045;geneID=gene60084;gene_name=TRNE;D...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>55</th>\n",
       "      <td>chrM</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>exon</td>\n",
       "      <td>14674</td>\n",
       "      <td>14742</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>Parent=rna164045</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>56</th>\n",
       "      <td>chrM</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>exon</td>\n",
       "      <td>14747</td>\n",
       "      <td>15887</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>Parent=gene60085</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>57</th>\n",
       "      <td>chrM</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>tRNA</td>\n",
       "      <td>15888</td>\n",
       "      <td>15953</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=rna164046;geneID=gene60086;gene_name=TRNT;D...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58</th>\n",
       "      <td>chrM</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>exon</td>\n",
       "      <td>15888</td>\n",
       "      <td>15953</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>Parent=rna164046</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>59</th>\n",
       "      <td>chrM</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>tRNA</td>\n",
       "      <td>15956</td>\n",
       "      <td>16023</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=rna164047;geneID=gene60087;gene_name=TRNP;D...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60</th>\n",
       "      <td>chrM</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>exon</td>\n",
       "      <td>15956</td>\n",
       "      <td>16023</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>Parent=rna164047</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>61 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   seqid  source  type  start    end score strand phase  \\\n",
       "0   chrM  RefSeq  tRNA    577    647     .      +     .   \n",
       "1   chrM  RefSeq  exon    577    647     .      +     .   \n",
       "2   chrM  RefSeq  rRNA    648   1601     .      +     .   \n",
       "3   chrM  RefSeq  exon    648   1601     .      +     .   \n",
       "4   chrM  RefSeq  tRNA   1602   1670     .      +     .   \n",
       "5   chrM  RefSeq  exon   1602   1670     .      +     .   \n",
       "6   chrM  RefSeq  rRNA   1671   3229     .      +     .   \n",
       "7   chrM  RefSeq  exon   1671   3229     .      +     .   \n",
       "8   chrM  RefSeq  tRNA   3230   3304     .      +     .   \n",
       "9   chrM  RefSeq  exon   3230   3304     .      +     .   \n",
       "10  chrM  RefSeq  exon   3307   4262     .      +     .   \n",
       "11  chrM  RefSeq  tRNA   4263   4331     .      +     .   \n",
       "12  chrM  RefSeq  exon   4263   4331     .      +     .   \n",
       "13  chrM  RefSeq  tRNA   4329   4400     .      -     .   \n",
       "14  chrM  RefSeq  exon   4329   4400     .      -     .   \n",
       "15  chrM  RefSeq  tRNA   4402   4469     .      +     .   \n",
       "16  chrM  RefSeq  exon   4402   4469     .      +     .   \n",
       "17  chrM  RefSeq  exon   4470   5511     .      +     .   \n",
       "18  chrM  RefSeq  tRNA   5512   5579     .      +     .   \n",
       "19  chrM  RefSeq  exon   5512   5579     .      +     .   \n",
       "20  chrM  RefSeq  tRNA   5587   5655     .      -     .   \n",
       "21  chrM  RefSeq  exon   5587   5655     .      -     .   \n",
       "22  chrM  RefSeq  tRNA   5657   5729     .      -     .   \n",
       "23  chrM  RefSeq  exon   5657   5729     .      -     .   \n",
       "24  chrM  RefSeq  tRNA   5761   5826     .      -     .   \n",
       "25  chrM  RefSeq  exon   5761   5826     .      -     .   \n",
       "26  chrM  RefSeq  tRNA   5826   5891     .      -     .   \n",
       "27  chrM  RefSeq  exon   5826   5891     .      -     .   \n",
       "28  chrM  RefSeq  exon   5904   7445     .      +     .   \n",
       "29  chrM  RefSeq  tRNA   7446   7514     .      -     .   \n",
       "..   ...     ...   ...    ...    ...   ...    ...   ...   \n",
       "31  chrM  RefSeq  tRNA   7518   7585     .      +     .   \n",
       "32  chrM  RefSeq  exon   7518   7585     .      +     .   \n",
       "33  chrM  RefSeq  exon   7586   8269     .      +     .   \n",
       "34  chrM  RefSeq  tRNA   8295   8364     .      +     .   \n",
       "35  chrM  RefSeq  exon   8295   8364     .      +     .   \n",
       "36  chrM  RefSeq  exon   8366   8572     .      +     .   \n",
       "37  chrM  RefSeq  exon   8527   9207     .      +     .   \n",
       "38  chrM  RefSeq  exon   9207   9990     .      +     .   \n",
       "39  chrM  RefSeq  tRNA   9991  10058     .      +     .   \n",
       "40  chrM  RefSeq  exon   9991  10058     .      +     .   \n",
       "41  chrM  RefSeq  exon  10059  10404     .      +     .   \n",
       "42  chrM  RefSeq  tRNA  10405  10469     .      +     .   \n",
       "43  chrM  RefSeq  exon  10405  10469     .      +     .   \n",
       "44  chrM  RefSeq  exon  10470  10766     .      +     .   \n",
       "45  chrM  RefSeq  exon  10760  12137     .      +     .   \n",
       "46  chrM  RefSeq  tRNA  12138  12206     .      +     .   \n",
       "47  chrM  RefSeq  exon  12138  12206     .      +     .   \n",
       "48  chrM  RefSeq  tRNA  12207  12265     .      +     .   \n",
       "49  chrM  RefSeq  exon  12207  12265     .      +     .   \n",
       "50  chrM  RefSeq  tRNA  12266  12336     .      +     .   \n",
       "51  chrM  RefSeq  exon  12266  12336     .      +     .   \n",
       "52  chrM  RefSeq  exon  12337  14148     .      +     .   \n",
       "53  chrM  RefSeq  exon  14149  14673     .      -     .   \n",
       "54  chrM  RefSeq  tRNA  14674  14742     .      -     .   \n",
       "55  chrM  RefSeq  exon  14674  14742     .      -     .   \n",
       "56  chrM  RefSeq  exon  14747  15887     .      +     .   \n",
       "57  chrM  RefSeq  tRNA  15888  15953     .      +     .   \n",
       "58  chrM  RefSeq  exon  15888  15953     .      +     .   \n",
       "59  chrM  RefSeq  tRNA  15956  16023     .      -     .   \n",
       "60  chrM  RefSeq  exon  15956  16023     .      -     .   \n",
       "\n",
       "                                           attributes  \n",
       "0   ID=rna164024;geneID=gene60051;gene_name=TRNF;D...  \n",
       "1                                    Parent=rna164024  \n",
       "2   ID=rna164025;geneID=gene60052;gene_name=RNR1;D...  \n",
       "3                                    Parent=rna164025  \n",
       "4   ID=rna164026;geneID=gene60053;gene_name=TRNV;D...  \n",
       "5                                    Parent=rna164026  \n",
       "6   ID=rna164027;geneID=gene60054;gene_name=RNR2;D...  \n",
       "7                                    Parent=rna164027  \n",
       "8   ID=rna164028;geneID=gene60055;gene_name=TRNL1;...  \n",
       "9                                    Parent=rna164028  \n",
       "10                                   Parent=gene60056  \n",
       "11  ID=rna164029;geneID=gene60057;gene_name=TRNI;D...  \n",
       "12                                   Parent=rna164029  \n",
       "13  ID=rna164030;geneID=gene60058;gene_name=TRNQ;D...  \n",
       "14                                   Parent=rna164030  \n",
       "15  ID=rna164031;geneID=gene60059;gene_name=TRNM;D...  \n",
       "16                                   Parent=rna164031  \n",
       "17                                   Parent=gene60060  \n",
       "18  ID=rna164032;geneID=gene60061;gene_name=TRNW;D...  \n",
       "19                                   Parent=rna164032  \n",
       "20  ID=rna164033;geneID=gene60062;gene_name=TRNA;D...  \n",
       "21                                   Parent=rna164033  \n",
       "22  ID=rna164034;geneID=gene60063;gene_name=TRNN;D...  \n",
       "23                                   Parent=rna164034  \n",
       "24  ID=rna164035;geneID=gene60064;gene_name=TRNC;D...  \n",
       "25                                   Parent=rna164035  \n",
       "26  ID=rna164036;geneID=gene60065;gene_name=TRNY;D...  \n",
       "27                                   Parent=rna164036  \n",
       "28                                   Parent=gene60066  \n",
       "29  ID=rna164037;geneID=gene60067;gene_name=TRNS1;...  \n",
       "..                                                ...  \n",
       "31  ID=rna164038;geneID=gene60068;gene_name=TRND;D...  \n",
       "32                                   Parent=rna164038  \n",
       "33                                   Parent=gene60069  \n",
       "34  ID=rna164039;geneID=gene60070;gene_name=TRNK;D...  \n",
       "35                                   Parent=rna164039  \n",
       "36                                   Parent=gene60071  \n",
       "37                                   Parent=gene60072  \n",
       "38                                   Parent=gene60073  \n",
       "39  ID=rna164040;geneID=gene60074;gene_name=TRNG;D...  \n",
       "40                                   Parent=rna164040  \n",
       "41                                   Parent=gene60075  \n",
       "42  ID=rna164041;geneID=gene60076;gene_name=TRNR;D...  \n",
       "43                                   Parent=rna164041  \n",
       "44                                   Parent=gene60077  \n",
       "45                                   Parent=gene60078  \n",
       "46  ID=rna164042;geneID=gene60079;gene_name=TRNH;D...  \n",
       "47                                   Parent=rna164042  \n",
       "48  ID=rna164043;geneID=gene60080;gene_name=TRNS2;...  \n",
       "49                                   Parent=rna164043  \n",
       "50  ID=rna164044;geneID=gene60081;gene_name=TRNL2;...  \n",
       "51                                   Parent=rna164044  \n",
       "52                                   Parent=gene60082  \n",
       "53                                   Parent=gene60083  \n",
       "54  ID=rna164045;geneID=gene60084;gene_name=TRNE;D...  \n",
       "55                                   Parent=rna164045  \n",
       "56                                   Parent=gene60085  \n",
       "57  ID=rna164046;geneID=gene60086;gene_name=TRNT;D...  \n",
       "58                                   Parent=rna164046  \n",
       "59  ID=rna164047;geneID=gene60087;gene_name=TRNP;D...  \n",
       "60                                   Parent=rna164047  \n",
       "\n",
       "[61 rows x 9 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "refseq=pd.read_csv(\"/home/sparrow/genomicData/hg38/annotations/hg38_p8_refseq.gff3\",skiprows=5,names=gff3Cols,sep=\"\\t\")\n",
    "refseq=refseq[refseq[\"seqid\"]==\"chrM\"].reset_index(drop=True)\n",
    "refseq=refseq[refseq[\"type\"].isin([\"exon\",\"tRNA\",\"rRNA\"])].reset_index(drop=True)\n",
    "refseq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gID</th>\n",
       "      <th>count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CHS.57315</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CHS.57316</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CHS.57317</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CHS.57318</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CHS.57319</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>CHS.57320</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>CHS.57321</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>CHS.57322</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>CHS.57323</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>CHS.57324</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>CHS.57325</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>CHS.57326</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>CHS.57327</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          gID  count\n",
       "0   CHS.57315      3\n",
       "1   CHS.57316      3\n",
       "2   CHS.57317      3\n",
       "3   CHS.57318      2\n",
       "4   CHS.57319      1\n",
       "5   CHS.57320      1\n",
       "6   CHS.57321      2\n",
       "7   CHS.57322      1\n",
       "8   CHS.57323      1\n",
       "9   CHS.57324      2\n",
       "10  CHS.57325      2\n",
       "11  CHS.57326      1\n",
       "12  CHS.57327      1"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>tID</th>\n",
       "      <th>count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CHS.57315.1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CHS.57315.2</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CHS.57315.3</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CHS.57316.1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CHS.57316.2</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>CHS.57316.3</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>CHS.57317.1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>CHS.57317.2</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>CHS.57317.3</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>CHS.57318.1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>CHS.57318.2</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>CHS.57319.1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>CHS.57320.1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>CHS.57321.1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>CHS.57321.2</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>CHS.57322.1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>CHS.57323.1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>CHS.57324.1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>CHS.57324.2</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>CHS.57325.1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>CHS.57325.2</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>CHS.57326.1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>CHS.57327.1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            tID  count\n",
       "0   CHS.57315.1      1\n",
       "1   CHS.57315.2      2\n",
       "2   CHS.57315.3      2\n",
       "3   CHS.57316.1      1\n",
       "4   CHS.57316.2      2\n",
       "5   CHS.57316.3      2\n",
       "6   CHS.57317.1      1\n",
       "7   CHS.57317.2      2\n",
       "8   CHS.57317.3      2\n",
       "9   CHS.57318.1      1\n",
       "10  CHS.57318.2      2\n",
       "11  CHS.57319.1      1\n",
       "12  CHS.57320.1      1\n",
       "13  CHS.57321.1      1\n",
       "14  CHS.57321.2      2\n",
       "15  CHS.57322.1      1\n",
       "16  CHS.57323.1      1\n",
       "17  CHS.57324.1      1\n",
       "18  CHS.57324.2      2\n",
       "19  CHS.57325.1      1\n",
       "20  CHS.57325.2      2\n",
       "21  CHS.57326.1      1\n",
       "22  CHS.57327.1      1"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# now need to count the number of exons per given transcript\n",
    "# as well as the number of transcripts per given gene\n",
    "# this should ultimately be 1-1 ratio in both cases\n",
    "\n",
    "# to get the number of transcripts per gene\n",
    "chessGT=df[df[\"type\"].isin([\"transcript\"])].reset_index(drop=True)\n",
    "chessTE=df[df[\"type\"].isin([\"exon\"])].reset_index(drop=True)\n",
    "\n",
    "chessGT[\"gID\"]=chessGT[\"attributes\"].str.split(\"ID=\",expand=True)[1].str.split(\";\",expand=True)[0]\n",
    "chessGT[\"gID\"]=np.where(chessGT[\"type\"]==\"transcript\",\"CHS.\"+chessGT[\"gID\"].str.split(\".\",expand=True)[1],chessGT[\"gID\"])\n",
    "chessGT=chessGT[[\"gID\",\"type\"]].groupby(\"gID\").count().reset_index()\n",
    "chessGT.columns=[\"gID\",\"count\"]\n",
    "display(chessGT)\n",
    "\n",
    "chessTE[\"tID\"]=chessTE[\"attributes\"].str.split(\"Parent=\",expand=True)[1].str.strip(\"\\n\")\n",
    "chessTE=chessTE[[\"tID\",\"type\"]].groupby(\"tID\").count().reset_index()\n",
    "chessTE.columns=[\"tID\",\"count\"]\n",
    "display(chessTE)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gID</th>\n",
       "      <th>count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000198695.2</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000198712.1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000198727.2</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000198763.3</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000198786.2</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>ENSG00000198804.2</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>ENSG00000198840.2</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>ENSG00000198886.2</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>ENSG00000198888.2</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>ENSG00000198899.2</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>ENSG00000198938.2</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>ENSG00000209082.1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>ENSG00000210049.1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>ENSG00000210077.1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>ENSG00000210082.2</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>ENSG00000210100.1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>ENSG00000210107.1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>ENSG00000210112.1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>ENSG00000210117.1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>ENSG00000210127.1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>ENSG00000210135.1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>ENSG00000210140.1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>ENSG00000210144.1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>ENSG00000210151.2</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>ENSG00000210154.1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>ENSG00000210156.1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>ENSG00000210164.1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>ENSG00000210174.1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>ENSG00000210176.1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>ENSG00000210184.1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>ENSG00000210191.1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>ENSG00000210194.1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>ENSG00000210195.2</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>ENSG00000210196.2</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>ENSG00000211459.2</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>ENSG00000212907.2</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>ENSG00000228253.1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                  gID  count\n",
       "0   ENSG00000198695.2      1\n",
       "1   ENSG00000198712.1      1\n",
       "2   ENSG00000198727.2      1\n",
       "3   ENSG00000198763.3      1\n",
       "4   ENSG00000198786.2      1\n",
       "5   ENSG00000198804.2      1\n",
       "6   ENSG00000198840.2      1\n",
       "7   ENSG00000198886.2      1\n",
       "8   ENSG00000198888.2      1\n",
       "9   ENSG00000198899.2      1\n",
       "10  ENSG00000198938.2      1\n",
       "11  ENSG00000209082.1      1\n",
       "12  ENSG00000210049.1      1\n",
       "13  ENSG00000210077.1      1\n",
       "14  ENSG00000210082.2      1\n",
       "15  ENSG00000210100.1      1\n",
       "16  ENSG00000210107.1      1\n",
       "17  ENSG00000210112.1      1\n",
       "18  ENSG00000210117.1      1\n",
       "19  ENSG00000210127.1      1\n",
       "20  ENSG00000210135.1      1\n",
       "21  ENSG00000210140.1      1\n",
       "22  ENSG00000210144.1      1\n",
       "23  ENSG00000210151.2      1\n",
       "24  ENSG00000210154.1      1\n",
       "25  ENSG00000210156.1      1\n",
       "26  ENSG00000210164.1      1\n",
       "27  ENSG00000210174.1      1\n",
       "28  ENSG00000210176.1      1\n",
       "29  ENSG00000210184.1      1\n",
       "30  ENSG00000210191.1      1\n",
       "31  ENSG00000210194.1      1\n",
       "32  ENSG00000210195.2      1\n",
       "33  ENSG00000210196.2      1\n",
       "34  ENSG00000211459.2      1\n",
       "35  ENSG00000212907.2      1\n",
       "36  ENSG00000228253.1      1"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>tID</th>\n",
       "      <th>count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENST00000361227.2</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENST00000361335.1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENST00000361381.2</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENST00000361390.2</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENST00000361453.3</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>ENST00000361567.2</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>ENST00000361624.2</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>ENST00000361681.2</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>ENST00000361739.1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>ENST00000361789.2</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>ENST00000361851.1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>ENST00000361899.2</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>ENST00000362079.2</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>ENST00000386347.1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>ENST00000387314.1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>ENST00000387342.1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>ENST00000387347.2</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>ENST00000387365.1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>ENST00000387372.1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>ENST00000387377.1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>ENST00000387382.1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>ENST00000387392.1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>ENST00000387400.1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>ENST00000387405.1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>ENST00000387409.1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>ENST00000387416.2</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>ENST00000387419.1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>ENST00000387421.1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>ENST00000387429.1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>ENST00000387439.1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>ENST00000387441.1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>ENST00000387449.1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>ENST00000387456.1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>ENST00000387459.1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>ENST00000387460.2</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>ENST00000387461.2</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>ENST00000389680.2</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                  tID  count\n",
       "0   ENST00000361227.2      1\n",
       "1   ENST00000361335.1      1\n",
       "2   ENST00000361381.2      1\n",
       "3   ENST00000361390.2      1\n",
       "4   ENST00000361453.3      1\n",
       "5   ENST00000361567.2      1\n",
       "6   ENST00000361624.2      1\n",
       "7   ENST00000361681.2      1\n",
       "8   ENST00000361739.1      1\n",
       "9   ENST00000361789.2      1\n",
       "10  ENST00000361851.1      1\n",
       "11  ENST00000361899.2      1\n",
       "12  ENST00000362079.2      1\n",
       "13  ENST00000386347.1      1\n",
       "14  ENST00000387314.1      1\n",
       "15  ENST00000387342.1      1\n",
       "16  ENST00000387347.2      1\n",
       "17  ENST00000387365.1      1\n",
       "18  ENST00000387372.1      1\n",
       "19  ENST00000387377.1      1\n",
       "20  ENST00000387382.1      1\n",
       "21  ENST00000387392.1      1\n",
       "22  ENST00000387400.1      1\n",
       "23  ENST00000387405.1      1\n",
       "24  ENST00000387409.1      1\n",
       "25  ENST00000387416.2      1\n",
       "26  ENST00000387419.1      1\n",
       "27  ENST00000387421.1      1\n",
       "28  ENST00000387429.1      1\n",
       "29  ENST00000387439.1      1\n",
       "30  ENST00000387441.1      1\n",
       "31  ENST00000387449.1      1\n",
       "32  ENST00000387456.1      1\n",
       "33  ENST00000387459.1      1\n",
       "34  ENST00000387460.2      1\n",
       "35  ENST00000387461.2      1\n",
       "36  ENST00000389680.2      1"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# same now for gencode\n",
    "gencodeGT=gencode[gencode[\"type\"].isin([\"transcript\",\"mRNA\"])].reset_index(drop=True)\n",
    "gencodeTE=gencode[gencode[\"type\"].isin([\"exon\"])].reset_index(drop=True)\n",
    "\n",
    "gencodeGT[\"gID\"]=gencodeGT[\"attributes\"].str.split(\"geneID=\",expand=True)[1].str.split(\";\",expand=True)[0]\n",
    "gencodeGT=gencodeGT[[\"gID\",\"type\"]].groupby(\"gID\").count().reset_index()\n",
    "gencodeGT.columns=[\"gID\",\"count\"]\n",
    "display(gencodeGT)\n",
    "\n",
    "gencodeTE[\"tID\"]=gencodeTE[\"attributes\"].str.split(\"Parent=\",expand=True)[1].str.strip(\"\\n\")\n",
    "gencodeTE=gencodeTE[[\"tID\",\"type\"]].groupby(\"tID\").count().reset_index()\n",
    "gencodeTE.columns=[\"tID\",\"count\"]\n",
    "display(gencodeTE)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gID</th>\n",
       "      <th>count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>gene60051</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>gene60052</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>gene60053</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>gene60054</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>gene60055</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>gene60057</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>gene60058</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>gene60059</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>gene60061</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>gene60062</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>gene60063</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>gene60064</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>gene60065</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>gene60067</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>gene60068</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>gene60070</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>gene60074</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>gene60076</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>gene60079</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>gene60080</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>gene60081</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>gene60084</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>gene60086</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>gene60087</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          gID  count\n",
       "0   gene60051      1\n",
       "1   gene60052      1\n",
       "2   gene60053      1\n",
       "3   gene60054      1\n",
       "4   gene60055      1\n",
       "5   gene60057      1\n",
       "6   gene60058      1\n",
       "7   gene60059      1\n",
       "8   gene60061      1\n",
       "9   gene60062      1\n",
       "10  gene60063      1\n",
       "11  gene60064      1\n",
       "12  gene60065      1\n",
       "13  gene60067      1\n",
       "14  gene60068      1\n",
       "15  gene60070      1\n",
       "16  gene60074      1\n",
       "17  gene60076      1\n",
       "18  gene60079      1\n",
       "19  gene60080      1\n",
       "20  gene60081      1\n",
       "21  gene60084      1\n",
       "22  gene60086      1\n",
       "23  gene60087      1"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>tID</th>\n",
       "      <th>count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>gene60056</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>gene60060</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>gene60066</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>gene60069</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>gene60071</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>gene60072</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>gene60073</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>gene60075</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>gene60077</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>gene60078</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>gene60082</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>gene60083</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>gene60085</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>rna164024</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>rna164025</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>rna164026</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>rna164027</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>rna164028</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>rna164029</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>rna164030</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>rna164031</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>rna164032</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>rna164033</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>rna164034</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>rna164035</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>rna164036</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>rna164037</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>rna164038</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>rna164039</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>rna164040</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>rna164041</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>rna164042</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>rna164043</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>rna164044</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>rna164045</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>rna164046</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>rna164047</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          tID  count\n",
       "0   gene60056      1\n",
       "1   gene60060      1\n",
       "2   gene60066      1\n",
       "3   gene60069      1\n",
       "4   gene60071      1\n",
       "5   gene60072      1\n",
       "6   gene60073      1\n",
       "7   gene60075      1\n",
       "8   gene60077      1\n",
       "9   gene60078      1\n",
       "10  gene60082      1\n",
       "11  gene60083      1\n",
       "12  gene60085      1\n",
       "13  rna164024      1\n",
       "14  rna164025      1\n",
       "15  rna164026      1\n",
       "16  rna164027      1\n",
       "17  rna164028      1\n",
       "18  rna164029      1\n",
       "19  rna164030      1\n",
       "20  rna164031      1\n",
       "21  rna164032      1\n",
       "22  rna164033      1\n",
       "23  rna164034      1\n",
       "24  rna164035      1\n",
       "25  rna164036      1\n",
       "26  rna164037      1\n",
       "27  rna164038      1\n",
       "28  rna164039      1\n",
       "29  rna164040      1\n",
       "30  rna164041      1\n",
       "31  rna164042      1\n",
       "32  rna164043      1\n",
       "33  rna164044      1\n",
       "34  rna164045      1\n",
       "35  rna164046      1\n",
       "36  rna164047      1"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# same now for gencode\n",
    "\n",
    "refseqGT=refseq[refseq[\"type\"].isin([\"transcript\",\"tRNA\",\"rRNA\"])].reset_index(drop=True)\n",
    "refseqTE=refseq[refseq[\"type\"].isin([\"exon\"])].reset_index(drop=True)\n",
    "\n",
    "refseqGT[\"gID\"]=refseqGT[\"attributes\"].str.split(\"geneID=\",expand=True)[1].str.split(\";\",expand=True)[0]\n",
    "refseqGT=refseqGT[[\"gID\",\"type\"]].groupby(\"gID\").count().reset_index()\n",
    "refseqGT.columns=[\"gID\",\"count\"]\n",
    "display(refseqGT)\n",
    "\n",
    "refseqTE[\"tID\"]=refseqTE[\"attributes\"].str.split(\"Parent=\",expand=True)[1].str.strip(\"\\n\")\n",
    "refseqTE=refseqTE[[\"tID\",\"type\"]].groupby(\"tID\").count().reset_index()\n",
    "refseqTE.columns=[\"tID\",\"count\"]\n",
    "display(refseqTE)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "there were 37 out of 37 exons with matching coordinates between RefSeq and GENCODE found\n",
      "13 out of 13 exons with an ambiguous transcript from RefSeq match ambiguous exons from GENCODE\n"
     ]
    }
   ],
   "source": [
    "# Seems like both refseq and gencode 25 have the same number of genes and transcripts\n",
    "# let us now compare the two to see if the coordinates also match between the two\n",
    "ge=gencode[gencode[\"type\"].isin([\"exon\"])].reset_index(drop=True)\n",
    "re=refseq[refseq[\"type\"].isin([\"exon\"])].reset_index(drop=True)\n",
    "merged=ge.merge(right=re,how=\"outer\",on=[\"start\",\"end\"],indicator=True)\n",
    "# display(merged)\n",
    "print(\"there were %d out of %d exons with matching coordinates between RefSeq and GENCODE found\"%(len(merged[merged[\"_merge\"]==\"both\"]),len(merged)))\n",
    "# identical numbers imply that RefSeq and gencode have identical exon annotations of the mitochondrial genome\n",
    "\n",
    "# now to investigate the missing transcripts from both gencode and refseq\n",
    "# in gencode the missing transcripts are due to the \"mRNA\" type present instead of the \"transcript\" type\n",
    "# The question is whether refseq and gencode have the same transcripts missing\n",
    "\n",
    "gt=gencode[gencode[\"type\"]==\"mRNA\"].reset_index(drop=True) # collection of GENCODE mRNA transcripts\n",
    "ge[\"tID\"]=ge[\"attributes\"].str.split(\"Parent=\",expand=True)[1].str.strip(\"\\n\")\n",
    "setMRNA=set(gt[\"attributes\"].str.split(\"ID=\",expand=True)[1].str.split(\";\",expand=True)[0])\n",
    "gte=ge[ge[\"tID\"].isin(setMRNA)].reset_index(drop=True)\n",
    "\n",
    "# now to do the sme for refseq\n",
    "re[\"tID\"]=re[\"attributes\"].str.split(\"Parent=\",expand=True)[1].str.strip(\"\\n\")\n",
    "rt=refseq[~(refseq[\"type\"]==\"exon\")].reset_index(drop=True)\n",
    "setRT=set(rt[\"attributes\"].str.split(\"ID=\",expand=True)[1].str.split(\";\",expand=True)[0])\n",
    "re=re[~(re[\"tID\"].isin(setRT))].reset_index(drop=True)\n",
    "\n",
    "# now to figure out whether the two are the same we shall merge the two coordinate frames\n",
    "regte=re.merge(right=gte,how=\"outer\",on=[\"start\",\"end\"],indicator=True)\n",
    "print(\"%d out of %d exons with an ambiguous transcript from RefSeq match ambiguous exons from GENCODE\"%(len(regte[regte[\"_merge\"]==\"both\"]),len(regte)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "13 genes have a transcript with a single exon out of a total of 13 mitochondrial genes present in chess\n"
     ]
    }
   ],
   "source": [
    "# now just need to ask Ela/Steven about it\n",
    "# for now let's get back to the CHESS annotation and see what we can do to improve it further\n",
    "# 1. why not see how many single-exon transcripts there are in total first\n",
    "# 2. second we shall see how many of these transcripts/exons match those defined in the gencode and refseq annotations\n",
    "\n",
    "\n",
    "# first we shall confirm there are no stray exons (such as found in gencode and refseq)\n",
    "ce=df[df[\"type\"]==\"exon\"].reset_index(drop=True)\n",
    "ce[\"tID\"]=ce[\"attributes\"].str.split(\"Parent=\",expand=True)[1].str.strip(\"\\n\")\n",
    "ct=df[df[\"type\"]==\"transcript\"].reset_index(drop=True)\n",
    "ct[\"gID\"]=ct[\"attributes\"].str.split(\"Parent=\",expand=True)[1].str.split(\";\",expand=True)[0]\n",
    "\n",
    "ceg=ce[[\"tID\",\"type\"]].groupby(\"tID\").count()\n",
    "ceg.reset_index(inplace=True)\n",
    "ctg=ct[[\"gID\",\"type\"]].groupby(\"gID\").count()\n",
    "ctg.reset_index(inplace=True)\n",
    "\n",
    "good_tIDs=ceg[ceg[\"type\"]==1][\"tID\"].tolist() # get single-exon only\n",
    "bad_tIDs=ceg[ceg[\"type\"]>1][\"tID\"].tolist()\n",
    "good_gIDs=[\"CHS.\"+x.split(\".\")[1] for x in good_tIDs]\n",
    "bad_gIDs=[\"CHS.\"+x.split(\".\")[1] for x in bad_tIDs]\n",
    "\n",
    "# next we need ot see whether these geneIDs account for all genes currently present in chess\n",
    "# basically, the question is whether all genes have at one single-exon transcript annotated for them\n",
    "all_gIDs=ctg[\"gID\"].tolist()\n",
    "all_gIDs\n",
    "print(\"%d genes have a transcript with a single exon out of a total of %d mitochondrial genes present in chess\"%(len(good_gIDs),len(all_gIDs)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Out of a total of 37 single-exon transcripts:\n",
      "\t13 are in CHESS\n",
      "\t37 are in GENCODE/RefSeq\n",
      "\t13 are shared between two annotations\n",
      "\t0 are found exclusively in CHESS\n",
      "\t24 are found exclusively in GENCODE/RefSeq\n",
      "24\n",
      "ID=ENST00000387314.1;geneID=ENSG00000210049.1;gene_name=MT-TF;gene_type=Mt_tRNA;gene_status=KNOWN;transcript_type=Mt_tRNA;transcript_status=KNOWN;level=3;transcript_support_level=NA;tag=basic;level=3;Name=MT-TF-201;locus=RLOC_00054904\n",
      "ID=ENST00000389680.2;geneID=ENSG00000211459.2;gene_name=MT-RNR1;gene_type=Mt_rRNA;gene_status=KNOWN;transcript_type=Mt_rRNA;transcript_status=KNOWN;level=3;transcript_support_level=NA;tag=basic;level=3;Name=MT-RNR1-201;locus=RLOC_00054905\n",
      "ID=ENST00000387342.1;geneID=ENSG00000210077.1;gene_name=MT-TV;gene_type=Mt_tRNA;gene_status=KNOWN;transcript_type=Mt_tRNA;transcript_status=KNOWN;level=3;transcript_support_level=NA;tag=basic;level=3;Name=MT-TV-201;locus=RLOC_00054906\n",
      "ID=ENST00000387347.2;geneID=ENSG00000210082.2;gene_name=MT-RNR2;gene_type=Mt_rRNA;gene_status=KNOWN;transcript_type=Mt_rRNA;transcript_status=KNOWN;level=3;transcript_support_level=NA;tag=basic;level=3;Name=MT-RNR2-201;locus=RLOC_00054907\n",
      "ID=ENST00000386347.1;geneID=ENSG00000209082.1;gene_name=MT-TL1;gene_type=Mt_tRNA;gene_status=KNOWN;transcript_type=Mt_tRNA;transcript_status=KNOWN;level=3;transcript_support_level=NA;tag=basic;level=3;Name=MT-TL1-201;locus=RLOC_00054908\n",
      "ID=ENST00000387365.1;geneID=ENSG00000210100.1;gene_name=MT-TI;gene_type=Mt_tRNA;gene_status=KNOWN;transcript_type=Mt_tRNA;transcript_status=KNOWN;level=3;transcript_support_level=NA;tag=basic;level=3;Name=MT-TI-201;locus=RLOC_00054910\n",
      "ID=ENST00000387372.1;geneID=ENSG00000210107.1;gene_name=MT-TQ;gene_type=Mt_tRNA;gene_status=KNOWN;transcript_type=Mt_tRNA;transcript_status=KNOWN;level=3;transcript_support_level=NA;tag=basic;level=3;Name=MT-TQ-201;locus=RLOC_00054911\n",
      "ID=ENST00000387377.1;geneID=ENSG00000210112.1;gene_name=MT-TM;gene_type=Mt_tRNA;gene_status=KNOWN;transcript_type=Mt_tRNA;transcript_status=KNOWN;level=3;transcript_support_level=NA;tag=basic;level=3;Name=MT-TM-201;locus=RLOC_00054912\n",
      "ID=ENST00000387382.1;geneID=ENSG00000210117.1;gene_name=MT-TW;gene_type=Mt_tRNA;gene_status=KNOWN;transcript_type=Mt_tRNA;transcript_status=KNOWN;level=3;transcript_support_level=NA;tag=basic;level=3;Name=MT-TW-201;locus=RLOC_00054914\n",
      "ID=ENST00000387392.1;geneID=ENSG00000210127.1;gene_name=MT-TA;gene_type=Mt_tRNA;gene_status=KNOWN;transcript_type=Mt_tRNA;transcript_status=KNOWN;level=3;transcript_support_level=NA;tag=basic;level=3;Name=MT-TA-201;locus=RLOC_00054915\n",
      "ID=ENST00000387400.1;geneID=ENSG00000210135.1;gene_name=MT-TN;gene_type=Mt_tRNA;gene_status=KNOWN;transcript_type=Mt_tRNA;transcript_status=KNOWN;level=3;transcript_support_level=NA;tag=basic;level=3;Name=MT-TN-201;locus=RLOC_00054916\n",
      "ID=ENST00000387405.1;geneID=ENSG00000210140.1;gene_name=MT-TC;gene_type=Mt_tRNA;gene_status=KNOWN;transcript_type=Mt_tRNA;transcript_status=KNOWN;level=3;transcript_support_level=NA;tag=basic;level=3;Name=MT-TC-201;locus=RLOC_00054917\n",
      "ID=ENST00000387409.1;geneID=ENSG00000210144.1;gene_name=MT-TY;gene_type=Mt_tRNA;gene_status=KNOWN;transcript_type=Mt_tRNA;transcript_status=KNOWN;level=3;transcript_support_level=NA;tag=basic;level=3;Name=MT-TY-201;locus=RLOC_00054917\n",
      "ID=ENST00000387416.2;geneID=ENSG00000210151.2;gene_name=MT-TS1;gene_type=Mt_tRNA;gene_status=KNOWN;transcript_type=Mt_tRNA;transcript_status=KNOWN;level=3;transcript_support_level=NA;tag=basic;level=3;Name=MT-TS1-201;locus=RLOC_00054919\n",
      "ID=ENST00000387419.1;geneID=ENSG00000210154.1;gene_name=MT-TD;gene_type=Mt_tRNA;gene_status=KNOWN;transcript_type=Mt_tRNA;transcript_status=KNOWN;level=3;transcript_support_level=NA;tag=basic;level=3;Name=MT-TD-201;locus=RLOC_00054920\n",
      "ID=ENST00000387421.1;geneID=ENSG00000210156.1;gene_name=MT-TK;gene_type=Mt_tRNA;gene_status=KNOWN;transcript_type=Mt_tRNA;transcript_status=KNOWN;level=3;transcript_support_level=NA;tag=basic;level=3;Name=MT-TK-201;locus=RLOC_00054922\n",
      "ID=ENST00000387429.1;geneID=ENSG00000210164.1;gene_name=MT-TG;gene_type=Mt_tRNA;gene_status=KNOWN;transcript_type=Mt_tRNA;transcript_status=KNOWN;level=3;transcript_support_level=NA;tag=basic;level=3;Name=MT-TG-201;locus=RLOC_00054924\n",
      "ID=ENST00000387439.1;geneID=ENSG00000210174.1;gene_name=MT-TR;gene_type=Mt_tRNA;gene_status=KNOWN;transcript_type=Mt_tRNA;transcript_status=KNOWN;level=3;transcript_support_level=NA;tag=basic;level=3;Name=MT-TR-201;locus=RLOC_00054926\n",
      "ID=ENST00000387441.1;geneID=ENSG00000210176.1;gene_name=MT-TH;gene_type=Mt_tRNA;gene_status=KNOWN;transcript_type=Mt_tRNA;transcript_status=KNOWN;level=3;transcript_support_level=NA;tag=basic;level=3;Name=MT-TH-201;locus=RLOC_00054928\n",
      "ID=ENST00000387449.1;geneID=ENSG00000210184.1;gene_name=MT-TS2;gene_type=Mt_tRNA;gene_status=KNOWN;transcript_type=Mt_tRNA;transcript_status=KNOWN;level=3;transcript_support_level=NA;tag=basic;level=3;Name=MT-TS2-201;locus=RLOC_00054929\n",
      "ID=ENST00000387456.1;geneID=ENSG00000210191.1;gene_name=MT-TL2;gene_type=Mt_tRNA;gene_status=KNOWN;transcript_type=Mt_tRNA;transcript_status=KNOWN;level=3;transcript_support_level=NA;tag=basic;level=3;Name=MT-TL2-201;locus=RLOC_00054930\n",
      "ID=ENST00000387459.1;geneID=ENSG00000210194.1;gene_name=MT-TE;gene_type=Mt_tRNA;gene_status=KNOWN;transcript_type=Mt_tRNA;transcript_status=KNOWN;level=3;transcript_support_level=NA;tag=basic;level=3;Name=MT-TE-201;locus=RLOC_00054933\n",
      "ID=ENST00000387460.2;geneID=ENSG00000210195.2;gene_name=MT-TT;gene_type=Mt_tRNA;gene_status=KNOWN;transcript_type=Mt_tRNA;transcript_status=KNOWN;level=3;transcript_support_level=NA;tag=basic;level=3;Name=MT-TT-201;locus=RLOC_00054935\n",
      "ID=ENST00000387461.2;geneID=ENSG00000210196.2;gene_name=MT-TP;gene_type=Mt_tRNA;gene_status=KNOWN;transcript_type=Mt_tRNA;transcript_status=KNOWN;level=3;transcript_support_level=NA;tag=basic;level=3;Name=MT-TP-201;locus=RLOC_00054936\n"
     ]
    }
   ],
   "source": [
    "# Now we need to find out which ones from gencode and refseq are not included in chess\n",
    "\n",
    "chess_se=ce[ce[\"tID\"].isin(good_tIDs)].reset_index(drop=True)\n",
    "gencode_merge_chess=chess_se.merge(right=gencode[gencode[\"type\"]==\"exon\"].reset_index(drop=True),how=\"outer\",on=[\"start\",\"end\"],indicator=True)\n",
    "\n",
    "print(\"Out of a total of %d single-exon transcripts:\\n\\t%d are in CHESS\\n\\t%d are in GENCODE/RefSeq\\n\\t%d are shared between two annotations\\n\\t%d are found exclusively in CHESS\\n\\t%d are found exclusively in GENCODE/RefSeq\"%(len(gencode_merge_chess),len(chess_se),len(gencode[gencode[\"type\"]==\"exon\"]),len(gencode_merge_chess[gencode_merge_chess[\"_merge\"]==\"both\"]),len(gencode_merge_chess[gencode_merge_chess[\"_merge\"]==\"left_only\"]),len(gencode_merge_chess[gencode_merge_chess[\"_merge\"]==\"right_only\"])))\n",
    "\n",
    "# now let's have a look at the attirbutes of those genes/transcripts that are not included in CHESS\n",
    "notInChess=gencode_merge_chess[gencode_merge_chess[\"_merge\"]==\"right_only\"]\n",
    "notInChess_genIDs=notInChess[\"attributes_y\"].str.split(\"Parent=\",expand=True)[1].str.strip(\"\\n\")\n",
    "print(len(notInChess_genIDs))\n",
    "\n",
    "notInChessTrans=gencode[~(gencode[\"type\"].isin([\"exon\"]))].reset_index(drop=True)\n",
    "notInChessTrans[\"ID\"]=notInChessTrans[\"attributes\"].str.split(\"ID=\",expand=True)[1].str.split(\";\",expand=True)[0]\n",
    "notInChessTrans=notInChessTrans[notInChessTrans[\"ID\"].isin(notInChess_genIDs)].reset_index(drop=True)\n",
    "notInChessTrans\n",
    "for i in notInChess_genIDs:\n",
    "    print(notInChessTrans[notInChessTrans[\"ID\"]==i][\"attributes\"].iloc[0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqid</th>\n",
       "      <th>source</th>\n",
       "      <th>type</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>phase</th>\n",
       "      <th>attributes</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chrM</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>mRNA</td>\n",
       "      <td>3307</td>\n",
       "      <td>4262</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=ENST00000361390.2;geneID=ENSG00000198888.2;...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chrM</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>mRNA</td>\n",
       "      <td>4470</td>\n",
       "      <td>5511</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=ENST00000361453.3;geneID=ENSG00000198763.3;...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chrM</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>mRNA</td>\n",
       "      <td>5904</td>\n",
       "      <td>7445</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=ENST00000361624.2;geneID=ENSG00000198804.2;...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chrM</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>mRNA</td>\n",
       "      <td>7586</td>\n",
       "      <td>8269</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=ENST00000361739.1;geneID=ENSG00000198712.1;...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chrM</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>mRNA</td>\n",
       "      <td>8366</td>\n",
       "      <td>8572</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=ENST00000361851.1;geneID=ENSG00000228253.1;...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>chrM</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>mRNA</td>\n",
       "      <td>8527</td>\n",
       "      <td>9207</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=ENST00000361899.2;geneID=ENSG00000198899.2;...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>chrM</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>mRNA</td>\n",
       "      <td>9207</td>\n",
       "      <td>9990</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=ENST00000362079.2;geneID=ENSG00000198938.2;...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>chrM</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>mRNA</td>\n",
       "      <td>10059</td>\n",
       "      <td>10404</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=ENST00000361227.2;geneID=ENSG00000198840.2;...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>chrM</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>mRNA</td>\n",
       "      <td>10470</td>\n",
       "      <td>10766</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=ENST00000361335.1;geneID=ENSG00000212907.2;...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>chrM</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>mRNA</td>\n",
       "      <td>10760</td>\n",
       "      <td>12137</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=ENST00000361381.2;geneID=ENSG00000198886.2;...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>chrM</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>mRNA</td>\n",
       "      <td>12337</td>\n",
       "      <td>14148</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=ENST00000361567.2;geneID=ENSG00000198786.2;...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>chrM</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>mRNA</td>\n",
       "      <td>14149</td>\n",
       "      <td>14673</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=ENST00000361681.2;geneID=ENSG00000198695.2;...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>chrM</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>mRNA</td>\n",
       "      <td>14747</td>\n",
       "      <td>15887</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=ENST00000361789.2;geneID=ENSG00000198727.2;...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   seqid   source  type  start    end score strand phase  \\\n",
       "0   chrM  ENSEMBL  mRNA   3307   4262     .      +     .   \n",
       "1   chrM  ENSEMBL  mRNA   4470   5511     .      +     .   \n",
       "2   chrM  ENSEMBL  mRNA   5904   7445     .      +     .   \n",
       "3   chrM  ENSEMBL  mRNA   7586   8269     .      +     .   \n",
       "4   chrM  ENSEMBL  mRNA   8366   8572     .      +     .   \n",
       "5   chrM  ENSEMBL  mRNA   8527   9207     .      +     .   \n",
       "6   chrM  ENSEMBL  mRNA   9207   9990     .      +     .   \n",
       "7   chrM  ENSEMBL  mRNA  10059  10404     .      +     .   \n",
       "8   chrM  ENSEMBL  mRNA  10470  10766     .      +     .   \n",
       "9   chrM  ENSEMBL  mRNA  10760  12137     .      +     .   \n",
       "10  chrM  ENSEMBL  mRNA  12337  14148     .      +     .   \n",
       "11  chrM  ENSEMBL  mRNA  14149  14673     .      -     .   \n",
       "12  chrM  ENSEMBL  mRNA  14747  15887     .      +     .   \n",
       "\n",
       "                                           attributes  \n",
       "0   ID=ENST00000361390.2;geneID=ENSG00000198888.2;...  \n",
       "1   ID=ENST00000361453.3;geneID=ENSG00000198763.3;...  \n",
       "2   ID=ENST00000361624.2;geneID=ENSG00000198804.2;...  \n",
       "3   ID=ENST00000361739.1;geneID=ENSG00000198712.1;...  \n",
       "4   ID=ENST00000361851.1;geneID=ENSG00000228253.1;...  \n",
       "5   ID=ENST00000361899.2;geneID=ENSG00000198899.2;...  \n",
       "6   ID=ENST00000362079.2;geneID=ENSG00000198938.2;...  \n",
       "7   ID=ENST00000361227.2;geneID=ENSG00000198840.2;...  \n",
       "8   ID=ENST00000361335.1;geneID=ENSG00000212907.2;...  \n",
       "9   ID=ENST00000361381.2;geneID=ENSG00000198886.2;...  \n",
       "10  ID=ENST00000361567.2;geneID=ENSG00000198786.2;...  \n",
       "11  ID=ENST00000361681.2;geneID=ENSG00000198695.2;...  \n",
       "12  ID=ENST00000361789.2;geneID=ENSG00000198727.2;...  "
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# notInChessTrans[~(notInChessTrans[\"ID\"].isin(notInChess_genIDs))]\n",
    "gencode[(gencode[\"type\"].isin([\"transcript\",\"mRNA\"]))&~(gencode[\"attributes\"].str.split(\"ID=\",expand=True)[1].str.split(\";\",expand=True)[0].isin(notInChess_genIDs))].reset_index(drop=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "7\n"
     ]
    }
   ],
   "source": [
    "# now onto actually producing the new gff\n",
    "\n",
    "# first need to get the list of transcript and gene IDs to remove\n",
    "outFP=open(\"./chess2.01.gff\",\"w+\")\n",
    "\n",
    "count=0\n",
    "with open(\"chess2.0.gff\",\"r\") as inFP:\n",
    "    for line in inFP.readlines():\n",
    "        if not line[0]==\"#\": # skip comments, otherwise proceed safely\n",
    "            lineCols=line.split(\"\\t\")\n",
    "            if lineCols[0]==\"chrM\":\n",
    "                if lineCols[2]==\"exon\" or lineCols[2]==\"CDS\": # extract parent\n",
    "                    parentID=lineCols[8].split(\"=\")[-1].strip()\n",
    "                    if not parentID in bad_tIDs:\n",
    "                        outFP.write(line)\n",
    "                elif lineCols[2]==\"gene\": # extract and compare geneID\n",
    "                    geneID=lineCols[8].split(\"=\")[1].split(\";\")[0]\n",
    "                    if geneID in bad_gIDs:\n",
    "                        count+=1\n",
    "                    if not geneID in bad_gIDs or geneID in good_gIDs:\n",
    "                        outFP.write(line)\n",
    "                else: # extract transcript ID and compare\n",
    "                    transID=lineCols[8].split(\"=\")[1].split(\";\")[0]\n",
    "                    if not transID in bad_tIDs:\n",
    "                        outFP.write(line)\n",
    "            else:\n",
    "                outFP.write(line)\n",
    "        else:\n",
    "            if line[:7]==\"##CHESS\":\n",
    "                outFP.write(\"##CHESS\"+version+\"\\n\")\n",
    "            else:\n",
    "                outFP.write(line)\n",
    "            continue\n",
    "print(count)\n",
    "inFP.close()\n",
    "outFP.close()\n",
    "\n",
    "# then repeat for the other files:\n",
    "# 1. chess_and_refseq\n",
    "# 2. chess_assembly\n",
    "# 3. recreate proteins\n",
    "# 4. recreate genes\n",
    "# 5. recreate transcripts\n",
    "# 6. mapfile.txt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "chrM\tStringTie\ttranscript\t3308\t4261\t.\t+\t.\tID=CHS.57315.2;Parent=CHS.57315;STATUS=novel;ASSEMBLED=yes\n",
      " CHS.57315.2\n",
      "chrM\tStringTie\texon\t3308\t3934\t.\t+\t.\tParent=CHS.57315.2\n",
      " CHS.57315.2\n",
      "chrM\tStringTie\texon\t3959\t4261\t.\t+\t.\tParent=CHS.57315.2\n",
      " CHS.57315.2\n",
      "chrM\tStringTie\ttranscript\t3308\t4261\t.\t+\t.\tID=CHS.57315.3;Parent=CHS.57315;STATUS=novel;ASSEMBLED=yes\n",
      " CHS.57315.3\n",
      "chrM\tStringTie\texon\t3308\t3391\t.\t+\t.\tParent=CHS.57315.3\n",
      " CHS.57315.3\n",
      "chrM\tStringTie\texon\t3428\t4261\t.\t+\t.\tParent=CHS.57315.3\n",
      " CHS.57315.3\n",
      "chrM\tStringTie\ttranscript\t4471\t5493\t.\t+\t.\tID=CHS.57316.2;Parent=CHS.57316;STATUS=novel;ASSEMBLED=yes\n",
      " CHS.57316.2\n",
      "chrM\tStringTie\texon\t4471\t5066\t.\t+\t.\tParent=CHS.57316.2\n",
      " CHS.57316.2\n",
      "chrM\tStringTie\texon\t5183\t5493\t.\t+\t.\tParent=CHS.57316.2\n",
      " CHS.57316.2\n",
      "chrM\tStringTie\ttranscript\t4471\t5510\t.\t+\t.\tID=CHS.57316.3;Parent=CHS.57316;STATUS=novel;ASSEMBLED=yes\n",
      " CHS.57316.3\n",
      "chrM\tStringTie\texon\t4471\t4864\t.\t+\t.\tParent=CHS.57316.3\n",
      " CHS.57316.3\n",
      "chrM\tStringTie\texon\t4937\t5510\t.\t+\t.\tParent=CHS.57316.3\n",
      " CHS.57316.3\n",
      "chrM\tStringTie\ttranscript\t5905\t7444\t.\t+\t.\tID=CHS.57317.2;Parent=CHS.57317;STATUS=novel;ASSEMBLED=yes\n",
      " CHS.57317.2\n",
      "chrM\tStringTie\texon\t5905\t6518\t.\t+\t.\tParent=CHS.57317.2\n",
      " CHS.57317.2\n",
      "chrM\tStringTie\texon\t6552\t7444\t.\t+\t.\tParent=CHS.57317.2\n",
      " CHS.57317.2\n",
      "chrM\tStringTie\ttranscript\t5905\t7444\t.\t+\t.\tID=CHS.57317.3;Parent=CHS.57317;STATUS=novel;ASSEMBLED=yes\n",
      " CHS.57317.3\n",
      "chrM\tStringTie\texon\t5905\t6717\t.\t+\t.\tParent=CHS.57317.3\n",
      " CHS.57317.3\n",
      "chrM\tStringTie\texon\t6955\t7444\t.\t+\t.\tParent=CHS.57317.3\n",
      " CHS.57317.3\n",
      "chrM\tStringTie\ttranscript\t7600\t8268\t.\t+\t.\tID=CHS.57318.2;Parent=CHS.57318;STATUS=novel;ASSEMBLED=yes\n",
      " CHS.57318.2\n",
      "chrM\tStringTie\texon\t7600\t7852\t.\t+\t.\tParent=CHS.57318.2\n",
      " CHS.57318.2\n",
      "chrM\tStringTie\texon\t7913\t8268\t.\t+\t.\tParent=CHS.57318.2\n",
      " CHS.57318.2\n",
      "chrM\tStringTie\ttranscript\t8573\t9206\t.\t+\t.\tID=CHS.57321.2;Parent=CHS.57321;STATUS=novel;ASSEMBLED=yes\n",
      " CHS.57321.2\n",
      "chrM\tStringTie\texon\t8573\t8857\t.\t+\t.\tParent=CHS.57321.2\n",
      " CHS.57321.2\n",
      "chrM\tStringTie\texon\t8921\t9206\t.\t+\t.\tParent=CHS.57321.2\n",
      " CHS.57321.2\n",
      "chrM\tStringTie\ttranscript\t10767\t12136\t.\t+\t.\tID=CHS.57324.2;Parent=CHS.57324;STATUS=novel;ASSEMBLED=yes\n",
      " CHS.57324.2\n",
      "chrM\tStringTie\texon\t10767\t11614\t.\t+\t.\tParent=CHS.57324.2\n",
      " CHS.57324.2\n",
      "chrM\tStringTie\texon\t11639\t12136\t.\t+\t.\tParent=CHS.57324.2\n",
      " CHS.57324.2\n",
      "chrM\tStringTie\ttranscript\t12349\t14147\t.\t+\t.\tID=CHS.57325.2;Parent=CHS.57325;STATUS=novel;ASSEMBLED=yes\n",
      " CHS.57325.2\n",
      "chrM\tStringTie\texon\t12349\t13018\t.\t+\t.\tParent=CHS.57325.2\n",
      " CHS.57325.2\n",
      "chrM\tStringTie\texon\t13052\t14147\t.\t+\t.\tParent=CHS.57325.2\n",
      " CHS.57325.2\n"
     ]
    }
   ],
   "source": [
    "# chess_assembly\n",
    "\n",
    "outFP=open(\"./chess2.01_assembly.gff\",\"w+\")\n",
    "\n",
    "count=0\n",
    "with open(\"chess2.0_assembly.gff\",\"r\") as inFP:\n",
    "    for line in inFP.readlines():\n",
    "        if not line[0]==\"#\": # skip comments, otherwise proceed safely\n",
    "            lineCols=line.split(\"\\t\")\n",
    "            if lineCols[0]==\"chrM\":\n",
    "                if lineCols[2]==\"exon\" or lineCols[2]==\"CDS\": # extract parent\n",
    "                    parentID=lineCols[8].split(\"=\")[-1].strip()\n",
    "                    if not parentID in bad_tIDs:\n",
    "                        outFP.write(line)\n",
    "                    else:\n",
    "                        print(line,parentID)\n",
    "                elif lineCols[2]==\"gene\": # extract and compare geneID\n",
    "                    geneID=lineCols[8].split(\"=\")[1].split(\";\")[0]\n",
    "                    if not geneID in bad_gIDs or geneID in good_gIDs:\n",
    "                        outFP.write(line)\n",
    "                    else:\n",
    "                        print(line,geneID)\n",
    "                else: # extract transcript ID and compare\n",
    "                    transID=lineCols[8].split(\"=\")[1].split(\";\")[0]\n",
    "                    if not transID in bad_tIDs:\n",
    "                        outFP.write(line)\n",
    "                    else:\n",
    "                        print(line,transID)\n",
    "            else:\n",
    "                outFP.write(line)\n",
    "        else:\n",
    "            if line[:7]==\"##CHESS\":\n",
    "                outFP.write(\"##CHESS\"+version+\"\\n\")\n",
    "            else:\n",
    "                outFP.write(line)\n",
    "            continue\n",
    "\n",
    "inFP.close()\n",
    "outFP.close()\n",
    "\n",
    "# then repeat for the other files:\n",
    "# 2. chess_assembly\n",
    "# 3. recreate proteins\n",
    "# 4. recreate genes\n",
    "# 5. recreate transcripts\n",
    "# 6. mapfile.txt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "# chess_and_refseq\n",
    "\n",
    "outFP=open(\"./chess2.01_and_refseq.gff\",\"w+\")\n",
    "\n",
    "count=0\n",
    "with open(\"chess2.0_and_refseq.gff\",\"r\") as inFP:\n",
    "    for line in inFP.readlines():\n",
    "        if not line[0]==\"#\": # skip comments, otherwise proceed safely\n",
    "            lineCols=line.split(\"\\t\")\n",
    "            if lineCols[0]==\"chrM\":\n",
    "                if lineCols[2]==\"exon\" or lineCols[2]==\"CDS\": # extract parent\n",
    "                    parentID=lineCols[8].split(\"=\")[-1].strip()\n",
    "                    if not parentID in bad_tIDs:\n",
    "                        outFP.write(line)\n",
    "                elif lineCols[2]==\"gene\": # extract and compare geneID\n",
    "                    geneID=lineCols[8].split(\"=\")[1].split(\";\")[0]\n",
    "                    if not geneID in bad_gIDs or geneID in good_gIDs:\n",
    "                        outFP.write(line)\n",
    "                else: # extract transcript ID and compare\n",
    "                    transID=lineCols[8].split(\"=\")[1].split(\";\")[0]\n",
    "                    if not transID in bad_tIDs:\n",
    "                        outFP.write(line)\n",
    "            else:\n",
    "                outFP.write(line)\n",
    "        else:\n",
    "            if line[:7]==\"##CHESS\":\n",
    "                outFP.write(\"##CHESS\"+version+\"\\n\")\n",
    "            else:\n",
    "                outFP.write(line)\n",
    "            continue\n",
    "\n",
    "inFP.close()\n",
    "outFP.close()\n",
    "\n",
    "# then repeat for the other files:\n",
    "# 3. recreate proteins\n",
    "# 4. recreate genes\n",
    "# 5. recreate transcripts\n",
    "# 6. mapfile.txt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "# now let's deal with the proteins file\n",
    "\n",
    "outFP=open(\"./chess2.01.protein.fa\",\"w+\")\n",
    "\n",
    "count=0\n",
    "allGood=False\n",
    "with open(\"./chess2.0.protein.fa\",\"r\") as inFP:\n",
    "    for line in inFP.readlines():\n",
    "        if line[0]==\">\":\n",
    "            chsID=line[1:].split(\" \")[0]\n",
    "            count+=1\n",
    "            if not chsID in bad_tIDs:\n",
    "                outFP.write(line)\n",
    "                allGood=True\n",
    "            else:\n",
    "                allGood=False\n",
    "                print(chsID)\n",
    "        elif allGood:\n",
    "            outFP.write(line)\n",
    "        else:\n",
    "            continue\n",
    "        \n",
    "inFP.close()\n",
    "outFP.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "# now moving onto the mapfile.txt as the last stop\n",
    "\n",
    "outFP=open(\"./mapfile2.01.txt\",\"w+\")\n",
    "\n",
    "count=0\n",
    "with open(\"./mapfile.txt\",\"r\") as inFP:\n",
    "    for line in inFP.readlines():\n",
    "        lineCols=line.split(\"\\t\")\n",
    "        chsID=lineCols[1].strip()\n",
    "        if chsID in bad_tIDs or not (not chsID in bad_gIDs or chsID in good_gIDs):\n",
    "            continue\n",
    "        outFP.write(line)\n",
    "    \n",
    "outFP.close()\n",
    "inFP.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0\n",
      "0\n",
      "[]\n"
     ]
    }
   ],
   "source": [
    "# and now to fix the transcripts and genes files\n",
    "outFP=open(\"./chess2.01.genes\",\"w+\")\n",
    "\n",
    "count=0\n",
    "removed=[]\n",
    "countRemovedLines=0\n",
    "with open(\"./chess2.0.genes\",\"r\") as inFP:\n",
    "    for line in inFP.readlines():\n",
    "        lineCols=line.split(\"\\t\")\n",
    "        ids=lineCols[2].split(\",\")\n",
    "        newIDS=[]\n",
    "        for gid in ids:\n",
    "            if gid in bad_gIDs and not gid in good_gIDs:\n",
    "                count+=1\n",
    "                removed.append(gid)\n",
    "            else:\n",
    "                newIDS.append(gid)\n",
    "        if len(newIDS)==0: # remove line\n",
    "            countRemovedLines+=1\n",
    "            continue\n",
    "        else:\n",
    "            lineCols[2]=\",\".join(newIDS)\n",
    "            outFP.write(\"\\t\".join(lineCols))\n",
    "\n",
    "print(count)\n",
    "print(countRemovedLines)\n",
    "print(removed)\n",
    "outFP.close()\n",
    "inFP.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "10\n",
      "10\n",
      "10\n",
      "['CHS.57324.2', 'CHS.57321.2', 'CHS.57315.3', 'CHS.57315.2', 'CHS.57317.2', 'CHS.57317.3', 'CHS.57316.2', 'CHS.57316.3', 'CHS.57325.2', 'CHS.57318.2']\n"
     ]
    }
   ],
   "source": [
    "print(len(set(bad_tIDs)))\n",
    "\n",
    "outFP=open(\"./chess2.01.transcripts\",\"w+\")\n",
    "count=0\n",
    "removed=[]\n",
    "countRemovedLines=0\n",
    "with open(\"./chess2.0.transcripts\",\"r\") as inFP:\n",
    "    for line in inFP.readlines():\n",
    "        lineCols=line.split(\"\\t\")\n",
    "        ids=lineCols[0].split(\",\")\n",
    "        newIDS=[]\n",
    "        for tid in ids:\n",
    "            if tid in bad_tIDs:\n",
    "                count+=1\n",
    "                removed.append(tid)\n",
    "            else:\n",
    "                newIDS.append(tid)\n",
    "        if len(newIDS)==0:\n",
    "            countRemovedLines+=1\n",
    "        else:\n",
    "            lineCols[0]=\",\".join(newIDS)\n",
    "            outFP.write(\"\\t\".join(lineCols))\n",
    "\n",
    "print(count)\n",
    "print(countRemovedLines)\n",
    "print(removed)\n",
    "outFP.close()\n",
    "inFP.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "number of mitiochondrial genes in CHESS is: 13\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gID</th>\n",
       "      <th>count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CHS.57315</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CHS.57316</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CHS.57317</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CHS.57318</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CHS.57319</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>CHS.57320</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>CHS.57321</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>CHS.57322</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>CHS.57323</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>CHS.57324</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>CHS.57325</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>CHS.57326</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>CHS.57327</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          gID  count\n",
       "0   CHS.57315      1\n",
       "1   CHS.57316      1\n",
       "2   CHS.57317      1\n",
       "3   CHS.57318      1\n",
       "4   CHS.57319      1\n",
       "5   CHS.57320      1\n",
       "6   CHS.57321      1\n",
       "7   CHS.57322      1\n",
       "8   CHS.57323      1\n",
       "9   CHS.57324      1\n",
       "10  CHS.57325      1\n",
       "11  CHS.57326      1\n",
       "12  CHS.57327      1"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>tID</th>\n",
       "      <th>count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CHS.57315.1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CHS.57316.1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CHS.57317.1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CHS.57318.1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CHS.57319.1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>CHS.57320.1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>CHS.57321.1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>CHS.57322.1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>CHS.57323.1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>CHS.57324.1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>CHS.57325.1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>CHS.57326.1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>CHS.57327.1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            tID  count\n",
       "0   CHS.57315.1      1\n",
       "1   CHS.57316.1      1\n",
       "2   CHS.57317.1      1\n",
       "3   CHS.57318.1      1\n",
       "4   CHS.57319.1      1\n",
       "5   CHS.57320.1      1\n",
       "6   CHS.57321.1      1\n",
       "7   CHS.57322.1      1\n",
       "8   CHS.57323.1      1\n",
       "9   CHS.57324.1      1\n",
       "10  CHS.57325.1      1\n",
       "11  CHS.57326.1      1\n",
       "12  CHS.57327.1      1"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# lastly let's test the newly generated files for correctness in ways similar to how we initially tested GENCODE and RefSeq\n",
    "\n",
    "chess=pd.read_csv(\"./chess2.01.gff\",skiprows=3,names=gff3Cols,sep=\"\\t\")\n",
    "chess=chess[chess[\"seqid\"]==\"chrM\"].reset_index(drop=True)\n",
    "chess=chess[chess[\"type\"].isin([\"gene\",\"exon\",\"transcript\"])].reset_index(drop=True)\n",
    "print(\"number of mitiochondrial genes in CHESS is: \"+str(len(chess[chess[\"type\"]==\"gene\"])))\n",
    "\n",
    "# same now for gencode\n",
    "chessGT=chess[chess[\"type\"].isin([\"transcript\"])].reset_index(drop=True)\n",
    "chessTE=chess[chess[\"type\"].isin([\"exon\"])].reset_index(drop=True)\n",
    "\n",
    "chessGT[\"gID\"]=chessGT[\"attributes\"].str.split(\"Parent=\",expand=True)[1].str.split(\";\",expand=True)[0]\n",
    "chessGT=chessGT[[\"gID\",\"type\"]].groupby(\"gID\").count().reset_index()\n",
    "chessGT.columns=[\"gID\",\"count\"]\n",
    "display(chessGT)\n",
    "\n",
    "chessTE[\"tID\"]=chessTE[\"attributes\"].str.split(\"Parent=\",expand=True)[1].str.strip(\"\\n\")\n",
    "chessTE=chessTE[[\"tID\",\"type\"]].groupby(\"tID\").count().reset_index()\n",
    "chessTE.columns=[\"tID\",\"count\"]\n",
    "display(chessTE)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
