{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import sys\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "from shutil import copyfile\n",
    "\n",
    "version=\"2.04\"\n",
    "gff3Cols=[\"seqid\",\"source\",\"type\",\"start\",\"end\",\"score\",\"strand\",\"phase\",\"attributes\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# this notebook is dedicated to parsing the chess2.02.gff files, and introducing several improvements/corrections\n",
    "# the corrections include:\n",
    "# 1. Extension of the transcript start based on a suspected trimmed ORF (as inferred from the known isoforms)\n",
    "# 2. Extension of the transctipt end based on a suspected trimming og the ORF (inferred from the known isoforms)\n",
    "# 3. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "number of transcripts with extremely short or non-fitting ORFs is:  0\n",
      "total number of transcripts considered:  94337\n"
     ]
    }
   ],
   "source": [
    "# first need to load the frame\n",
    "df=pd.read_csv(\"./revision/mergedExonDF.csv\")\n",
    "df.drop(\"Unnamed: 0\",axis=1,inplace=True)\n",
    "\n",
    "# case 99999 should never exist for novel however, since all novel should have a corresponding known transcript\n",
    "# assert len(df[df[\"match\"]>2])==0, 'something is wrong witth the known and novel transcripts. Some novel transcript does not have a known pair'\n",
    "df=df[df[\"match\"].isin([0,1,2])].reset_index(drop=True)\n",
    "# let's isolate suspicious cases such as 777777\n",
    "setStartTrimShort=set(df[df[\"startTrim\"]==7777777][\"parent\"])\n",
    "setEndTrimShort=set(df[df[\"nonesenseTrim\"]==7777777][\"parent\"])\n",
    "assert setStartTrimShort==setEndTrimShort, \"something is wrong with the ORF fitting inside the novel transcript\"\n",
    "\n",
    "print(\"number of transcripts with extremely short or non-fitting ORFs is: \",len(setStartTrimShort))\n",
    "print(\"total number of transcripts considered: \",len(df))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "numnber of transcripts with trimmed start AND end is  38\n",
      "total number of transcripts that require adjustment 2971\n"
     ]
    }
   ],
   "source": [
    "# now we need to work with the rest of the transcripts and correct their annotation in the CHESS.gff\n",
    "\n",
    "# first let's isolate those transcripts for which the start is missing:\n",
    "startDF=df[(df[\"startTrim\"]<=100)&(df[\"startTrim\"]>0)&(df[\"numExonsStart\"]==0)].reset_index(drop=True)\n",
    "endDF=df[(df[\"nonesenseTrim\"]<=100)&(df[\"nonesenseTrim\"]>0)&(df[\"numExonsEnd\"]==0)].reset_index(drop=True)\n",
    "print('numnber of transcripts with trimmed start AND end is ',len(set(startDF['parent']).intersection(set(endDF['parent']))))\n",
    "print('total number of transcripts that require adjustment',len(set(startDF['parent']).union(set(endDF['parent']))))\n",
    "# now need to build dictionaries linking each parent to the number of bases trimmed of either start OR end (separate dataframes)\n",
    "startDict=pd.Series(startDF.startTrim.values,index=startDF.parent).to_dict()\n",
    "endDict=pd.Series(endDF.nonesenseTrim.values,index=endDF.parent).to_dict()\n",
    "setStarts=set(startDict) # to be used for quick lookup when traversing the GFF\n",
    "setEnds=set(endDict) # to be used for quick lookup when traversing the GFF"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "278"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(startDF[startDF[\"startTrim\"]<5])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "outFP=open(\"./chess2.04.gff\",\"w+\")\n",
    "tmpLine=None\n",
    "with open(\"./chess2.03.gff\",\"r\") as fp:\n",
    "    ifp=iter(fp)\n",
    "    while True:\n",
    "        try:\n",
    "            line=\"\"\n",
    "            if tmpLine==None:\n",
    "                line=next(ifp)\n",
    "            else:\n",
    "                line=tmpLine\n",
    "                tmpLine=None\n",
    "                \n",
    "            if line[0]==\"#\":\n",
    "                if line[:7]==\"##CHESS\":\n",
    "                    outFP.write(\"##CHESS\"+version+\"\\n\")\n",
    "                else:\n",
    "                    outFP.write(line)\n",
    "                continue\n",
    "            else:\n",
    "                lineCols=line.split(\"\\t\")\n",
    "                if lineCols[2]==\"transcript\":\n",
    "                    chsID=lineCols[8].split(\"ID=\")[-1].split(\";\")[0].strip()\n",
    "                    if chsID in setStarts:\n",
    "                        # now we shall collect all the necessary parts (exons) for the given novel transcript\n",
    "                        if lineCols[6]==\"+\":\n",
    "                            # change right-most coordinate of the transcript\n",
    "                            transStart=lineCols[3]\n",
    "                            lineCols[3]=str(int(transStart)-startDict[chsID])\n",
    "                            outFP.write(\"\\t\".join(lineCols))\n",
    "                            foundExon=False\n",
    "                            while True:\n",
    "                                line=next(ifp)\n",
    "                                if line[0]==\"#\":\n",
    "                                    tmpLine=line\n",
    "                                    assert foundExon==True,\"- wrong element, something is wrong with the gff format \"+chsID+\" \"+transStart\n",
    "                                    break\n",
    "                                lineCols=line.split(\"\\t\")\n",
    "                                if not lineCols[2]==\"exon\":\n",
    "                                    tmpLine=line\n",
    "                                    assert foundExon==True,\"+ wrong element, something is wrong with the gff format \"+chsID+\" \"+transStart\n",
    "                                    break\n",
    "                                else:\n",
    "                                    parentID=lineCols[8].split(\"Parent=\")[-1].split(\";\")[-1].strip()\n",
    "                                    assert parentID in setStarts and parentID==chsID, \"wrong exon identified \"+parentID+\" \"+chsID\n",
    "                                    if lineCols[3]==transStart:\n",
    "                                        lineCols[3]=str(int(lineCols[3])-startDict[chsID])\n",
    "                                        foundExon=True\n",
    "                                        outFP.write(\"\\t\".join(lineCols))\n",
    "                                    else:\n",
    "                                        outFP.write(line)\n",
    "                        else:\n",
    "                            # change left-most coordinate of the transcript\n",
    "                            transEnd=lineCols[4]\n",
    "                            lineCols[4]=str(int(transEnd)+startDict[chsID])\n",
    "                            outFP.write(\"\\t\".join(lineCols))\n",
    "                            foundExon=False\n",
    "                            while True:\n",
    "                                line=next(ifp)\n",
    "                                if line[0]==\"#\":\n",
    "                                    tmpLine=line\n",
    "                                    assert foundExon==True,\"- wrong element, something is wrong with the gff format \"+chsID+\" \"+transEnd\n",
    "                                    break\n",
    "                                lineCols=line.split(\"\\t\")\n",
    "                                if not lineCols[2]==\"exon\":\n",
    "                                    tmpLine=line\n",
    "                                    assert foundExon==True,\"- wrong element, something is wrong with the gff format \"+chsID+\" \"+transEnd\n",
    "                                    break\n",
    "                                else:\n",
    "                                    parentID=lineCols[8].split(\"Parent=\")[-1].split(\";\")[-1].strip()\n",
    "                                    assert parentID in setStarts and parentID==chsID, \"wrong exon identified \"+parentID+\" \"+chsID\n",
    "                                    if lineCols[4]==transEnd:\n",
    "                                        lineCols[4]=str(int(lineCols[4])+startDict[chsID])\n",
    "                                        foundExon=True\n",
    "                                        outFP.write(\"\\t\".join(lineCols))\n",
    "                                    else:\n",
    "                                        outFP.write(line)\n",
    "                                        \n",
    "                    elif chsID in setEnds:\n",
    "                        # now we shall collect all the necessary parts (exons) for the given novel transcript\n",
    "                        if lineCols[6]==\"+\":\n",
    "                            # change right-most coordinate of the transcript\n",
    "                            transEnd=lineCols[4]\n",
    "                            lineCols[4]=str(int(transEnd)+endDict[chsID])\n",
    "                            outFP.write(\"\\t\".join(lineCols))\n",
    "                            foundExon=False\n",
    "                            while True:\n",
    "                                line=next(ifp)\n",
    "                                if line[0]==\"#\":\n",
    "                                    tmpLine=line\n",
    "                                    assert foundExon==True,\"- wrong element, something is wrong with the gff format \"+chsID+\" \"+transEnd\n",
    "                                    break\n",
    "                                lineCols=line.split(\"\\t\")\n",
    "                                if not lineCols[2]==\"exon\":\n",
    "                                    tmpLine=line\n",
    "                                    assert foundExon==True,\"+ wrong element, something is wrong with the gff format \"+chsID+\" \"+transEnd\n",
    "                                    break\n",
    "                                else:\n",
    "                                    parentID=lineCols[8].split(\"Parent=\")[-1].split(\";\")[-1].strip()\n",
    "                                    assert parentID in setEnds and parentID==chsID, \"wrong exon identified \"+parentID+\" \"+chsID\n",
    "                                    if lineCols[4]==transEnd:\n",
    "                                        lineCols[4]=str(int(lineCols[4])-endDict[chsID])\n",
    "                                        foundExon=True\n",
    "                                        outFP.write(\"\\t\".join(lineCols))\n",
    "                                    else:\n",
    "                                        outFP.write(line)\n",
    "                        else:\n",
    "                            # change left-most coordinate of the transcript\n",
    "                            transStart=lineCols[3]\n",
    "                            lineCols[3]=str(int(transStart)+endDict[chsID])\n",
    "                            outFP.write(\"\\t\".join(lineCols))\n",
    "                            foundExon=False\n",
    "                            while True:\n",
    "                                line=next(ifp)\n",
    "                                if line[0]==\"#\":\n",
    "                                    tmpLine=line\n",
    "                                    assert foundExon==True,\"- wrong element, something is wrong with the gff format \"+chsID+\" \"+transStart\n",
    "                                    break\n",
    "                                lineCols=line.split(\"\\t\")\n",
    "                                if not lineCols[2]==\"exon\":\n",
    "                                    tmpLine=line\n",
    "                                    assert foundExon==True,\"- wrong element, something is wrong with the gff format \"+chsID+\" \"+transStart\n",
    "                                    break\n",
    "                                else:\n",
    "                                    parentID=lineCols[8].split(\"Parent=\")[-1].split(\";\")[-1].strip()\n",
    "                                    assert parentID in setEnds and parentID==chsID, \"wrong exon identified \"+parentID+\" \"+chsID\n",
    "                                    if lineCols[3]==transStart:\n",
    "                                        lineCols[3]=str(int(lineCols[3])+endDict[chsID])\n",
    "                                        foundExon=True\n",
    "                                        outFP.write(\"\\t\".join(lineCols))\n",
    "                                    else:\n",
    "                                        outFP.write(line)\n",
    "                    else:\n",
    "                        outFP.write(line)\n",
    "                else:\n",
    "                    outFP.write(line)\n",
    "        except StopIteration:\n",
    "            break\n",
    "            \n",
    "outFP.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "# now to do the same with the chess2.02_assembly.gff version\n",
    "\n",
    "outFP=open(\"./chess2.04_assembly.gff\",\"w+\")\n",
    "tmpLine=None\n",
    "with open(\"./chess2.03_assembly.gff\",\"r\") as fp:\n",
    "    ifp=iter(fp)\n",
    "    while True:\n",
    "        try:\n",
    "            line=\"\"\n",
    "            if tmpLine==None:\n",
    "                line=next(ifp)\n",
    "            else:\n",
    "                line=tmpLine\n",
    "                tmpLine=None\n",
    "                \n",
    "            if line[0]==\"#\":\n",
    "                if line[:7]==\"##CHESS\":\n",
    "                    outFP.write(\"##CHESS\"+version+\"\\n\")\n",
    "                else:\n",
    "                    outFP.write(line)\n",
    "                continue\n",
    "            else:\n",
    "                lineCols=line.split(\"\\t\")\n",
    "                if lineCols[2]==\"transcript\":\n",
    "                    chsID=lineCols[8].split(\"ID=\")[-1].split(\";\")[0].strip()\n",
    "                    if chsID in setStarts:\n",
    "                        # now we shall collect all the necessary parts (exons) for the given novel transcript\n",
    "                        if lineCols[6]==\"+\":\n",
    "                            # change right-most coordinate of the transcript\n",
    "                            transStart=lineCols[3]\n",
    "                            lineCols[3]=str(int(transStart)-startDict[chsID])\n",
    "                            outFP.write(\"\\t\".join(lineCols))\n",
    "                            foundExon=False\n",
    "                            while True:\n",
    "                                line=next(ifp)\n",
    "                                if line[0]==\"#\":\n",
    "                                    tmpLine=line\n",
    "                                    assert foundExon==True,\"- wrong element, something is wrong with the gff format \"+chsID+\" \"+transStart\n",
    "                                    break\n",
    "                                lineCols=line.split(\"\\t\")\n",
    "                                if not lineCols[2]==\"exon\":\n",
    "                                    tmpLine=line\n",
    "                                    assert foundExon==True,\"+ wrong element, something is wrong with the gff format \"+chsID+\" \"+transStart\n",
    "                                    break\n",
    "                                else:\n",
    "                                    parentID=lineCols[8].split(\"Parent=\")[-1].split(\";\")[-1].strip()\n",
    "                                    assert parentID in setStarts and parentID==chsID, \"wrong exon identified \"+parentID+\" \"+chsID\n",
    "                                    if lineCols[3]==transStart:\n",
    "                                        lineCols[3]=str(int(lineCols[3])-startDict[chsID])\n",
    "                                        foundExon=True\n",
    "                                        outFP.write(\"\\t\".join(lineCols))\n",
    "                                    else:\n",
    "                                        outFP.write(line)\n",
    "                        else:\n",
    "                            # change left-most coordinate of the transcript\n",
    "                            transEnd=lineCols[4]\n",
    "                            lineCols[4]=str(int(transEnd)+startDict[chsID])\n",
    "                            outFP.write(\"\\t\".join(lineCols))\n",
    "                            foundExon=False\n",
    "                            while True:\n",
    "                                line=next(ifp)\n",
    "                                if line[0]==\"#\":\n",
    "                                    tmpLine=line\n",
    "                                    assert foundExon==True,\"- wrong element, something is wrong with the gff format \"+chsID+\" \"+transEnd\n",
    "                                    break\n",
    "                                lineCols=line.split(\"\\t\")\n",
    "                                if not lineCols[2]==\"exon\":\n",
    "                                    tmpLine=line\n",
    "                                    assert foundExon==True,\"- wrong element, something is wrong with the gff format \"+chsID+\" \"+transEnd\n",
    "                                    break\n",
    "                                else:\n",
    "                                    parentID=lineCols[8].split(\"Parent=\")[-1].split(\";\")[-1].strip()\n",
    "                                    assert parentID in setStarts and parentID==chsID, \"wrong exon identified \"+parentID+\" \"+chsID\n",
    "                                    if lineCols[4]==transEnd:\n",
    "                                        lineCols[4]=str(int(lineCols[4])+startDict[chsID])\n",
    "                                        foundExon=True\n",
    "                                        outFP.write(\"\\t\".join(lineCols))\n",
    "                                    else:\n",
    "                                        outFP.write(line)\n",
    "                                        \n",
    "                    elif chsID in setEnds:\n",
    "                        # now we shall collect all the necessary parts (exons) for the given novel transcript\n",
    "                        if lineCols[6]==\"+\":\n",
    "                            # change right-most coordinate of the transcript\n",
    "                            transEnd=lineCols[4]\n",
    "                            lineCols[4]=str(int(transEnd)+endDict[chsID])\n",
    "                            outFP.write(\"\\t\".join(lineCols))\n",
    "                            foundExon=False\n",
    "                            while True:\n",
    "                                line=next(ifp)\n",
    "                                if line[0]==\"#\":\n",
    "                                    tmpLine=line\n",
    "                                    assert foundExon==True,\"- wrong element, something is wrong with the gff format \"+chsID+\" \"+transEnd\n",
    "                                    break\n",
    "                                lineCols=line.split(\"\\t\")\n",
    "                                if not lineCols[2]==\"exon\":\n",
    "                                    tmpLine=line\n",
    "                                    assert foundExon==True,\"+ wrong element, something is wrong with the gff format \"+chsID+\" \"+transEnd\n",
    "                                    break\n",
    "                                else:\n",
    "                                    parentID=lineCols[8].split(\"Parent=\")[-1].split(\";\")[-1].strip()\n",
    "                                    assert parentID in setEnds and parentID==chsID, \"wrong exon identified \"+parentID+\" \"+chsID\n",
    "                                    if lineCols[4]==transEnd:\n",
    "                                        lineCols[4]=str(int(lineCols[4])-endDict[chsID])\n",
    "                                        foundExon=True\n",
    "                                        outFP.write(\"\\t\".join(lineCols))\n",
    "                                    else:\n",
    "                                        outFP.write(line)\n",
    "                        else:\n",
    "                            # change left-most coordinate of the transcript\n",
    "                            transStart=lineCols[3]\n",
    "                            lineCols[3]=str(int(transStart)+endDict[chsID])\n",
    "                            outFP.write(\"\\t\".join(lineCols))\n",
    "                            foundExon=False\n",
    "                            while True:\n",
    "                                line=next(ifp)\n",
    "                                if line[0]==\"#\":\n",
    "                                    tmpLine=line\n",
    "                                    assert foundExon==True,\"- wrong element, something is wrong with the gff format \"+chsID+\" \"+transStart\n",
    "                                    break\n",
    "                                lineCols=line.split(\"\\t\")\n",
    "                                if not lineCols[2]==\"exon\":\n",
    "                                    tmpLine=line\n",
    "                                    assert foundExon==True,\"- wrong element, something is wrong with the gff format \"+chsID+\" \"+transStart\n",
    "                                    break\n",
    "                                else:\n",
    "                                    parentID=lineCols[8].split(\"Parent=\")[-1].split(\";\")[-1].strip()\n",
    "                                    assert parentID in setEnds and parentID==chsID, \"wrong exon identified \"+parentID+\" \"+chsID\n",
    "                                    if lineCols[3]==transStart:\n",
    "                                        lineCols[3]=str(int(lineCols[3])+endDict[chsID])\n",
    "                                        foundExon=True\n",
    "                                        outFP.write(\"\\t\".join(lineCols))\n",
    "                                    else:\n",
    "                                        outFP.write(line)\n",
    "                    else:\n",
    "                        outFP.write(line)\n",
    "                else:\n",
    "                    outFP.write(line)\n",
    "        except StopIteration:\n",
    "            break\n",
    "            \n",
    "outFP.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "# now to do the same with the chess2.02_and_refseq.gff version\n",
    "\n",
    "outFP=open(\"./chess2.04_and_refseq.gff\",\"w+\")\n",
    "tmpLine=None\n",
    "with open(\"./chess2.03_and_refseq.gff\",\"r\") as fp:\n",
    "    ifp=iter(fp)\n",
    "    while True:\n",
    "        try:\n",
    "            line=\"\"\n",
    "            if tmpLine==None:\n",
    "                line=next(ifp)\n",
    "            else:\n",
    "                line=tmpLine\n",
    "                tmpLine=None\n",
    "                \n",
    "            if line[0]==\"#\":\n",
    "                if line[:7]==\"##CHESS\":\n",
    "                    outFP.write(\"##CHESS\"+version+\"\\n\")\n",
    "                else:\n",
    "                    outFP.write(line)\n",
    "                continue\n",
    "            else:\n",
    "                lineCols=line.split(\"\\t\")\n",
    "                if lineCols[2]==\"transcript\":\n",
    "                    chsID=lineCols[8].split(\"ID=\")[-1].split(\";\")[0].strip()\n",
    "                    if chsID in setStarts:\n",
    "                        # now we shall collect all the necessary parts (exons) for the given novel transcript\n",
    "                        if lineCols[6]==\"+\":\n",
    "                            # change right-most coordinate of the transcript\n",
    "                            transStart=lineCols[3]\n",
    "                            lineCols[3]=str(int(transStart)-startDict[chsID])\n",
    "                            outFP.write(\"\\t\".join(lineCols))\n",
    "                            foundExon=False\n",
    "                            while True:\n",
    "                                line=next(ifp)\n",
    "                                if line[0]==\"#\":\n",
    "                                    tmpLine=line\n",
    "                                    assert foundExon==True,\"- wrong element, something is wrong with the gff format \"+chsID+\" \"+transStart\n",
    "                                    break\n",
    "                                lineCols=line.split(\"\\t\")\n",
    "                                if not lineCols[2]==\"exon\":\n",
    "                                    tmpLine=line\n",
    "                                    assert foundExon==True,\"+ wrong element, something is wrong with the gff format \"+chsID+\" \"+transStart\n",
    "                                    break\n",
    "                                else:\n",
    "                                    parentID=lineCols[8].split(\"Parent=\")[-1].split(\";\")[-1].strip()\n",
    "                                    assert parentID in setStarts and parentID==chsID, \"wrong exon identified \"+parentID+\" \"+chsID\n",
    "                                    if lineCols[3]==transStart:\n",
    "                                        lineCols[3]=str(int(lineCols[3])-startDict[chsID])\n",
    "                                        foundExon=True\n",
    "                                        outFP.write(\"\\t\".join(lineCols))\n",
    "                                    else:\n",
    "                                        outFP.write(line)\n",
    "                        else:\n",
    "                            # change left-most coordinate of the transcript\n",
    "                            transEnd=lineCols[4]\n",
    "                            lineCols[4]=str(int(transEnd)+startDict[chsID])\n",
    "                            outFP.write(\"\\t\".join(lineCols))\n",
    "                            foundExon=False\n",
    "                            while True:\n",
    "                                line=next(ifp)\n",
    "                                if line[0]==\"#\":\n",
    "                                    tmpLine=line\n",
    "                                    assert foundExon==True,\"- wrong element, something is wrong with the gff format \"+chsID+\" \"+transEnd\n",
    "                                    break\n",
    "                                lineCols=line.split(\"\\t\")\n",
    "                                if not lineCols[2]==\"exon\":\n",
    "                                    tmpLine=line\n",
    "                                    assert foundExon==True,\"- wrong element, something is wrong with the gff format \"+chsID+\" \"+transEnd\n",
    "                                    break\n",
    "                                else:\n",
    "                                    parentID=lineCols[8].split(\"Parent=\")[-1].split(\";\")[-1].strip()\n",
    "                                    assert parentID in setStarts and parentID==chsID, \"wrong exon identified \"+parentID+\" \"+chsID\n",
    "                                    if lineCols[4]==transEnd:\n",
    "                                        lineCols[4]=str(int(lineCols[4])+startDict[chsID])\n",
    "                                        foundExon=True\n",
    "                                        outFP.write(\"\\t\".join(lineCols))\n",
    "                                    else:\n",
    "                                        outFP.write(line)\n",
    "                                        \n",
    "                    elif chsID in setEnds:\n",
    "                        # now we shall collect all the necessary parts (exons) for the given novel transcript\n",
    "                        if lineCols[6]==\"+\":\n",
    "                            # change right-most coordinate of the transcript\n",
    "                            transEnd=lineCols[4]\n",
    "                            lineCols[4]=str(int(transEnd)+endDict[chsID])\n",
    "                            outFP.write(\"\\t\".join(lineCols))\n",
    "                            foundExon=False\n",
    "                            while True:\n",
    "                                line=next(ifp)\n",
    "                                if line[0]==\"#\":\n",
    "                                    tmpLine=line\n",
    "                                    assert foundExon==True,\"- wrong element, something is wrong with the gff format \"+chsID+\" \"+transEnd\n",
    "                                    break\n",
    "                                lineCols=line.split(\"\\t\")\n",
    "                                if not lineCols[2]==\"exon\":\n",
    "                                    tmpLine=line\n",
    "                                    assert foundExon==True,\"+ wrong element, something is wrong with the gff format \"+chsID+\" \"+transEnd\n",
    "                                    break\n",
    "                                else:\n",
    "                                    parentID=lineCols[8].split(\"Parent=\")[-1].split(\";\")[-1].strip()\n",
    "                                    assert parentID in setEnds and parentID==chsID, \"wrong exon identified \"+parentID+\" \"+chsID\n",
    "                                    if lineCols[4]==transEnd:\n",
    "                                        lineCols[4]=str(int(lineCols[4])-endDict[chsID])\n",
    "                                        foundExon=True\n",
    "                                        outFP.write(\"\\t\".join(lineCols))\n",
    "                                    else:\n",
    "                                        outFP.write(line)\n",
    "                        else:\n",
    "                            # change left-most coordinate of the transcript\n",
    "                            transStart=lineCols[3]\n",
    "                            lineCols[3]=str(int(transStart)+endDict[chsID])\n",
    "                            outFP.write(\"\\t\".join(lineCols))\n",
    "                            foundExon=False\n",
    "                            while True:\n",
    "                                line=next(ifp)\n",
    "                                if line[0]==\"#\":\n",
    "                                    tmpLine=line\n",
    "                                    assert foundExon==True,\"- wrong element, something is wrong with the gff format \"+chsID+\" \"+transStart\n",
    "                                    break\n",
    "                                lineCols=line.split(\"\\t\")\n",
    "                                if not lineCols[2]==\"exon\":\n",
    "                                    tmpLine=line\n",
    "                                    assert foundExon==True,\"- wrong element, something is wrong with the gff format \"+chsID+\" \"+transStart\n",
    "                                    break\n",
    "                                else:\n",
    "                                    parentID=lineCols[8].split(\"Parent=\")[-1].split(\";\")[-1].strip()\n",
    "                                    assert parentID in setEnds and parentID==chsID, \"wrong exon identified \"+parentID+\" \"+chsID\n",
    "                                    if lineCols[3]==transStart:\n",
    "                                        lineCols[3]=str(int(lineCols[3])+endDict[chsID])\n",
    "                                        foundExon=True\n",
    "                                        outFP.write(\"\\t\".join(lineCols))\n",
    "                                    else:\n",
    "                                        outFP.write(line)\n",
    "                    else:\n",
    "                        outFP.write(line)\n",
    "                else:\n",
    "                    outFP.write(line)\n",
    "        except StopIteration:\n",
    "            break\n",
    "            \n",
    "outFP.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqid</th>\n",
       "      <th>source</th>\n",
       "      <th>type</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>phase</th>\n",
       "      <th>attributes</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>gene</td>\n",
       "      <td>11874.0</td>\n",
       "      <td>14409.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=CHS.1;GENE_TYPE=misc_RNA;STATUS=known_refse...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>gene</td>\n",
       "      <td>14362.0</td>\n",
       "      <td>29370.0</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=CHS.2;GENE_TYPE=misc_RNA;STATUS=known_refse...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>gene</td>\n",
       "      <td>29926.0</td>\n",
       "      <td>31295.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=CHS.3;GENE_TYPE=lncRNA;STATUS=known_refseq;...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>gene</td>\n",
       "      <td>34611.0</td>\n",
       "      <td>36081.0</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=CHS.4;GENE_TYPE=lncRNA;STATUS=known_refseq;...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>gene</td>\n",
       "      <td>51943.0</td>\n",
       "      <td>53959.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=CHS.5;GENE_TYPE=lncRNA;STATUS=known_refseq;...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  seqid  source  type    start      end score strand phase  \\\n",
       "0  chr1  RefSeq  gene  11874.0  14409.0     .      +     .   \n",
       "1  chr1  RefSeq  gene  14362.0  29370.0     .      -     .   \n",
       "2  chr1  RefSeq  gene  29926.0  31295.0     .      +     .   \n",
       "3  chr1  RefSeq  gene  34611.0  36081.0     .      -     .   \n",
       "4  chr1  RefSeq  gene  51943.0  53959.0     .      +     .   \n",
       "\n",
       "                                          attributes  \n",
       "0  ID=CHS.1;GENE_TYPE=misc_RNA;STATUS=known_refse...  \n",
       "1  ID=CHS.2;GENE_TYPE=misc_RNA;STATUS=known_refse...  \n",
       "2  ID=CHS.3;GENE_TYPE=lncRNA;STATUS=known_refseq;...  \n",
       "3  ID=CHS.4;GENE_TYPE=lncRNA;STATUS=known_refseq;...  \n",
       "4  ID=CHS.5;GENE_TYPE=lncRNA;STATUS=known_refseq;...  "
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# mapfile stays the same here\n",
    "copyfile(\"./mapfile2.03.txt\",\"mapfile2.04.txt\")\n",
    "# genes file should be just fine, although it would be best to verify that everything is indeed correct\n",
    "ann=pd.read_csv(\"./chess2.04.gff\",sep=\"\\t\",names=gff3Cols)\n",
    "ann.dropna(inplace=True,axis=0)\n",
    "ann=ann[ann[\"type\"]==\"gene\"].reset_index(drop=True)\n",
    "ann.head()\n",
    "# the real problem here is the transcripts file, since we need to provide updated coordinates for it."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqid</th>\n",
       "      <th>source</th>\n",
       "      <th>type</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>phase</th>\n",
       "      <th>attributes</th>\n",
       "      <th>id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>gene</td>\n",
       "      <td>11874.0</td>\n",
       "      <td>14409.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=CHS.1;GENE_TYPE=misc_RNA;STATUS=known_refse...</td>\n",
       "      <td>CHS.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>gene</td>\n",
       "      <td>14362.0</td>\n",
       "      <td>29370.0</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=CHS.2;GENE_TYPE=misc_RNA;STATUS=known_refse...</td>\n",
       "      <td>CHS.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>gene</td>\n",
       "      <td>29926.0</td>\n",
       "      <td>31295.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=CHS.3;GENE_TYPE=lncRNA;STATUS=known_refseq;...</td>\n",
       "      <td>CHS.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>gene</td>\n",
       "      <td>34611.0</td>\n",
       "      <td>36081.0</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=CHS.4;GENE_TYPE=lncRNA;STATUS=known_refseq;...</td>\n",
       "      <td>CHS.4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>gene</td>\n",
       "      <td>51943.0</td>\n",
       "      <td>53959.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=CHS.5;GENE_TYPE=lncRNA;STATUS=known_refseq;...</td>\n",
       "      <td>CHS.5</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  seqid  source  type    start      end score strand phase  \\\n",
       "0  chr1  RefSeq  gene  11874.0  14409.0     .      +     .   \n",
       "1  chr1  RefSeq  gene  14362.0  29370.0     .      -     .   \n",
       "2  chr1  RefSeq  gene  29926.0  31295.0     .      +     .   \n",
       "3  chr1  RefSeq  gene  34611.0  36081.0     .      -     .   \n",
       "4  chr1  RefSeq  gene  51943.0  53959.0     .      +     .   \n",
       "\n",
       "                                          attributes     id  \n",
       "0  ID=CHS.1;GENE_TYPE=misc_RNA;STATUS=known_refse...  CHS.1  \n",
       "1  ID=CHS.2;GENE_TYPE=misc_RNA;STATUS=known_refse...  CHS.2  \n",
       "2  ID=CHS.3;GENE_TYPE=lncRNA;STATUS=known_refseq;...  CHS.3  \n",
       "3  ID=CHS.4;GENE_TYPE=lncRNA;STATUS=known_refseq;...  CHS.4  \n",
       "4  ID=CHS.5;GENE_TYPE=lncRNA;STATUS=known_refseq;...  CHS.5  "
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ann[\"id\"]=ann.attributes.str.split(\"ID=\",expand=True)[1].str.split(\";\",expand=True)[0]\n",
    "ann.head()\n",
    "ann[\"start\"]==ann['start'].astype(int)\n",
    "ann[\"end\"]==ann['end'].astype(int)\n",
    "ann.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene_Type</th>\n",
       "      <th>Gene_Name</th>\n",
       "      <th>GFF_ID</th>\n",
       "      <th>Location</th>\n",
       "      <th>Database</th>\n",
       "      <th>RefSeq_GeneID</th>\n",
       "      <th>Description</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>protein_coding</td>\n",
       "      <td>LOC105371921</td>\n",
       "      <td>CHS.23496</td>\n",
       "      <td>chr17:80885776-80892131:-</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>105371921</td>\n",
       "      <td>uncharacterized LOC105371921</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>lncRNA</td>\n",
       "      <td>LOC107986127</td>\n",
       "      <td>CHS.38280</td>\n",
       "      <td>chr3:127744219-127751505:-</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>107986127</td>\n",
       "      <td>uncharacterized LOC107986127</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>protein_coding</td>\n",
       "      <td>RPS11</td>\n",
       "      <td>CHS.27254</td>\n",
       "      <td>chr19:49496365-49499712:+</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>6205</td>\n",
       "      <td>ribosomal protein S11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>protein_coding</td>\n",
       "      <td>CREB3L1</td>\n",
       "      <td>CHS.8579</td>\n",
       "      <td>chr11:46276634-46321430:+</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>90993</td>\n",
       "      <td>cAMP responsive element binding protein 3-like 1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>lncRNA</td>\n",
       "      <td>LOC101930053</td>\n",
       "      <td>CHS.54970</td>\n",
       "      <td>chr9:2385141-2494223:-</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>101930053</td>\n",
       "      <td>uncharacterized LOC101930053</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        Gene_Type     Gene_Name     GFF_ID                    Location  \\\n",
       "0  protein_coding  LOC105371921  CHS.23496   chr17:80885776-80892131:-   \n",
       "1          lncRNA  LOC107986127  CHS.38280  chr3:127744219-127751505:-   \n",
       "2  protein_coding         RPS11  CHS.27254   chr19:49496365-49499712:+   \n",
       "3  protein_coding       CREB3L1   CHS.8579   chr11:46276634-46321430:+   \n",
       "4          lncRNA  LOC101930053  CHS.54970      chr9:2385141-2494223:-   \n",
       "\n",
       "  Database RefSeq_GeneID                                       Description  \n",
       "0   RefSeq     105371921                      uncharacterized LOC105371921  \n",
       "1   RefSeq     107986127                      uncharacterized LOC107986127  \n",
       "2   RefSeq          6205                             ribosomal protein S11  \n",
       "3   RefSeq         90993  cAMP responsive element binding protein 3-like 1  \n",
       "4   RefSeq     101930053                      uncharacterized LOC101930053  "
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genes=pd.read_csv(\"./chess2.03.genes\",sep=\"\\t\")\n",
    "genes.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "# credit goes to https://stackoverflow.com/questions/12680754/split-explode-pandas-dataframe-string-entry-to-separate-rows/40449726#40449726\n",
    "def explode(df, lst_cols, fill_value=''):\n",
    "    # make sure `lst_cols` is a list\n",
    "    if lst_cols and not isinstance(lst_cols, list):\n",
    "        lst_cols = [lst_cols]\n",
    "    # all columns except `lst_cols`\n",
    "    idx_cols = df.columns.difference(lst_cols)\n",
    "\n",
    "    # calculate lengths of lists\n",
    "    lens = df[lst_cols[0]].str.len()\n",
    "\n",
    "    if (lens > 0).all():\n",
    "        # ALL lists in cells aren't empty\n",
    "        return pd.DataFrame({\n",
    "            col:np.repeat(df[col].values, lens)\n",
    "            for col in idx_cols\n",
    "        }).assign(**{col:np.concatenate(df[col].values) for col in lst_cols}) \\\n",
    "          .loc[:, df.columns]\n",
    "    else:\n",
    "        # at least one list in cells is empty\n",
    "        return pd.DataFrame({\n",
    "            col:np.repeat(df[col].values, lens)\n",
    "            for col in idx_cols\n",
    "        }).assign(**{col:np.concatenate(df[col].values) for col in lst_cols}) \\\n",
    "          .append(df.loc[lens==0, idx_cols]).fillna(fill_value) \\\n",
    "          .loc[:, df.columns]\n",
    "    \n",
    "genes=explode(genes.assign(GFF_ID=genes.GFF_ID.str.split(',')), 'GFF_ID')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene_Type</th>\n",
       "      <th>Gene_Name</th>\n",
       "      <th>GFF_ID</th>\n",
       "      <th>Location</th>\n",
       "      <th>Database</th>\n",
       "      <th>RefSeq_GeneID</th>\n",
       "      <th>Description</th>\n",
       "      <th>id</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>_merge</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>protein_coding</td>\n",
       "      <td>LOC105371921</td>\n",
       "      <td>CHS.23496</td>\n",
       "      <td>chr17:80885776-80892131:-</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>105371921</td>\n",
       "      <td>uncharacterized LOC105371921</td>\n",
       "      <td>CHS.23496</td>\n",
       "      <td>80885776.0</td>\n",
       "      <td>80892131.0</td>\n",
       "      <td>both</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>lncRNA</td>\n",
       "      <td>LOC107986127</td>\n",
       "      <td>CHS.38280</td>\n",
       "      <td>chr3:127744219-127751505:-</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>107986127</td>\n",
       "      <td>uncharacterized LOC107986127</td>\n",
       "      <td>CHS.38280</td>\n",
       "      <td>127744219.0</td>\n",
       "      <td>127751505.0</td>\n",
       "      <td>both</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>protein_coding</td>\n",
       "      <td>RPS11</td>\n",
       "      <td>CHS.27254</td>\n",
       "      <td>chr19:49496365-49499712:+</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>6205</td>\n",
       "      <td>ribosomal protein S11</td>\n",
       "      <td>CHS.27254</td>\n",
       "      <td>49496365.0</td>\n",
       "      <td>49499712.0</td>\n",
       "      <td>both</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>protein_coding</td>\n",
       "      <td>CREB3L1</td>\n",
       "      <td>CHS.8579</td>\n",
       "      <td>chr11:46276634-46321430:+</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>90993</td>\n",
       "      <td>cAMP responsive element binding protein 3-like 1</td>\n",
       "      <td>CHS.8579</td>\n",
       "      <td>46276634.0</td>\n",
       "      <td>46321430.0</td>\n",
       "      <td>both</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>lncRNA</td>\n",
       "      <td>LOC101930053</td>\n",
       "      <td>CHS.54970</td>\n",
       "      <td>chr9:2385141-2494223:-</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>101930053</td>\n",
       "      <td>uncharacterized LOC101930053</td>\n",
       "      <td>CHS.54970</td>\n",
       "      <td>2385141.0</td>\n",
       "      <td>2494223.0</td>\n",
       "      <td>both</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        Gene_Type     Gene_Name     GFF_ID                    Location  \\\n",
       "0  protein_coding  LOC105371921  CHS.23496   chr17:80885776-80892131:-   \n",
       "1          lncRNA  LOC107986127  CHS.38280  chr3:127744219-127751505:-   \n",
       "2  protein_coding         RPS11  CHS.27254   chr19:49496365-49499712:+   \n",
       "3  protein_coding       CREB3L1   CHS.8579   chr11:46276634-46321430:+   \n",
       "4          lncRNA  LOC101930053  CHS.54970      chr9:2385141-2494223:-   \n",
       "\n",
       "  Database RefSeq_GeneID                                       Description  \\\n",
       "0   RefSeq     105371921                      uncharacterized LOC105371921   \n",
       "1   RefSeq     107986127                      uncharacterized LOC107986127   \n",
       "2   RefSeq          6205                             ribosomal protein S11   \n",
       "3   RefSeq         90993  cAMP responsive element binding protein 3-like 1   \n",
       "4   RefSeq     101930053                      uncharacterized LOC101930053   \n",
       "\n",
       "          id        start          end _merge  \n",
       "0  CHS.23496   80885776.0   80892131.0   both  \n",
       "1  CHS.38280  127744219.0  127751505.0   both  \n",
       "2  CHS.27254   49496365.0   49499712.0   both  \n",
       "3   CHS.8579   46276634.0   46321430.0   both  \n",
       "4  CHS.54970    2385141.0    2494223.0   both  "
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mdf=genes.merge(ann[['id','start','end']],how=\"outer\",left_on='GFF_ID',right_on='id',indicator=True)\n",
    "mdf.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "assert len(mdf[mdf[\"_merge\"]==\"left_only\"])==0, \"there are genes present in the .genes file but not in the annotation: \"+\";\".join(mdf[mdf[\"_merge\"]==\"left_only\"][\"id\"])\n",
    "assert len(mdf[mdf[\"_merge\"]==\"right_only\"])==0, \" there are genes present in the annotation but not in the .genes file: \"+\";\".join(mdf[mdf[\"_merge\"]==\"right_only\"][\"id\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene_Type</th>\n",
       "      <th>Gene_Name</th>\n",
       "      <th>GFF_ID</th>\n",
       "      <th>Location</th>\n",
       "      <th>Database</th>\n",
       "      <th>RefSeq_GeneID</th>\n",
       "      <th>Description</th>\n",
       "      <th>id</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>_merge</th>\n",
       "      <th>start_old</th>\n",
       "      <th>end_old</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>48</th>\n",
       "      <td>protein_coding</td>\n",
       "      <td>ZBTB12</td>\n",
       "      <td>CHS.48511</td>\n",
       "      <td>chr6:31899194-31902331:-,chr6_GL000251v2_alt:3...</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>221527</td>\n",
       "      <td>zinc finger and BTB domain containing 12</td>\n",
       "      <td>CHS.48511</td>\n",
       "      <td>3377042</td>\n",
       "      <td>3379417</td>\n",
       "      <td>both</td>\n",
       "      <td>31899194</td>\n",
       "      <td>31902331</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>49</th>\n",
       "      <td>protein_coding</td>\n",
       "      <td>ZBTB12</td>\n",
       "      <td>CHS.48732</td>\n",
       "      <td>chr6:31899194-31902331:-,chr6_GL000251v2_alt:3...</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>221527</td>\n",
       "      <td>zinc finger and BTB domain containing 12</td>\n",
       "      <td>CHS.48732</td>\n",
       "      <td>3147391</td>\n",
       "      <td>3149766</td>\n",
       "      <td>both</td>\n",
       "      <td>31899194</td>\n",
       "      <td>31902331</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50</th>\n",
       "      <td>protein_coding</td>\n",
       "      <td>ZBTB12</td>\n",
       "      <td>CHS.49124</td>\n",
       "      <td>chr6:31899194-31902331:-,chr6_GL000251v2_alt:3...</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>221527</td>\n",
       "      <td>zinc finger and BTB domain containing 12</td>\n",
       "      <td>CHS.49124</td>\n",
       "      <td>3241692</td>\n",
       "      <td>3244452</td>\n",
       "      <td>both</td>\n",
       "      <td>31899194</td>\n",
       "      <td>31902331</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>51</th>\n",
       "      <td>protein_coding</td>\n",
       "      <td>ZBTB12</td>\n",
       "      <td>CHS.49345</td>\n",
       "      <td>chr6:31899194-31902331:-,chr6_GL000251v2_alt:3...</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>221527</td>\n",
       "      <td>zinc finger and BTB domain containing 12</td>\n",
       "      <td>CHS.49345</td>\n",
       "      <td>3155590</td>\n",
       "      <td>3157967</td>\n",
       "      <td>both</td>\n",
       "      <td>31899194</td>\n",
       "      <td>31902331</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75</th>\n",
       "      <td>lncRNA</td>\n",
       "      <td>LOC105375115</td>\n",
       "      <td>CHS.52474</td>\n",
       "      <td>chr7:174920-176013:+,chr7_KI270804v1_alt:13432...</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>105375115</td>\n",
       "      <td>uncharacterized LOC105375115</td>\n",
       "      <td>CHS.52474</td>\n",
       "      <td>134324</td>\n",
       "      <td>135417</td>\n",
       "      <td>both</td>\n",
       "      <td>174920</td>\n",
       "      <td>176013</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>76</th>\n",
       "      <td>lncRNA</td>\n",
       "      <td>LOC105375115</td>\n",
       "      <td>CHS.52499</td>\n",
       "      <td>chr7:174920-176013:+,chr7_KI270804v1_alt:13432...</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>105375115</td>\n",
       "      <td>uncharacterized LOC105375115</td>\n",
       "      <td>CHS.52499</td>\n",
       "      <td>167241</td>\n",
       "      <td>168334</td>\n",
       "      <td>both</td>\n",
       "      <td>174920</td>\n",
       "      <td>176013</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>85</th>\n",
       "      <td>protein_coding</td>\n",
       "      <td>SERPINF1</td>\n",
       "      <td>CHS.23883</td>\n",
       "      <td>chr17:1761965-1779392:+,chr17_KI270861v1_alt:1...</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>5176</td>\n",
       "      <td>serpin family F member 1</td>\n",
       "      <td>CHS.23883</td>\n",
       "      <td>190927</td>\n",
       "      <td>195947</td>\n",
       "      <td>both</td>\n",
       "      <td>1761965</td>\n",
       "      <td>1779392</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>166</th>\n",
       "      <td>protein_coding</td>\n",
       "      <td>ABR</td>\n",
       "      <td>CHS.23887</td>\n",
       "      <td>chr17:999953-1229713:-,chr17_KI270862v1_alt:42...</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>29</td>\n",
       "      <td>active BCR-related</td>\n",
       "      <td>CHS.23887</td>\n",
       "      <td>42765</td>\n",
       "      <td>231794</td>\n",
       "      <td>both</td>\n",
       "      <td>999953</td>\n",
       "      <td>1229713</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>167</th>\n",
       "      <td>protein_coding</td>\n",
       "      <td>ABR</td>\n",
       "      <td>CHS.23947</td>\n",
       "      <td>chr17:999953-1229713:-,chr17_KI270862v1_alt:42...</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>29</td>\n",
       "      <td>active BCR-related</td>\n",
       "      <td>CHS.23947</td>\n",
       "      <td>42765</td>\n",
       "      <td>153205</td>\n",
       "      <td>both</td>\n",
       "      <td>999953</td>\n",
       "      <td>1229713</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>169</th>\n",
       "      <td>protein_coding</td>\n",
       "      <td>OR9G1</td>\n",
       "      <td>CHS.10525</td>\n",
       "      <td>chr11:56700388-56701305:+,chr11_JH159137v1_alt...</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>390174</td>\n",
       "      <td>olfactory receptor family 9 subfamily G member 1</td>\n",
       "      <td>CHS.10525</td>\n",
       "      <td>7154</td>\n",
       "      <td>8071</td>\n",
       "      <td>both</td>\n",
       "      <td>56700388</td>\n",
       "      <td>56701305</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>175</th>\n",
       "      <td>lncRNA</td>\n",
       "      <td>LOC107984555</td>\n",
       "      <td>CHS.14862</td>\n",
       "      <td>chr13:112273139-112276195:+,chr13_KI270838v1_a...</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>107984555</td>\n",
       "      <td>uncharacterized LOC107984555</td>\n",
       "      <td>CHS.14862</td>\n",
       "      <td>90257</td>\n",
       "      <td>92661</td>\n",
       "      <td>both</td>\n",
       "      <td>112273139</td>\n",
       "      <td>112276195</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>280</th>\n",
       "      <td>misc_RNA</td>\n",
       "      <td>HLA-J</td>\n",
       "      <td>CHS.48258</td>\n",
       "      <td>chr6:30005971-30009956:+,chr6_GL000250v2_alt:1...</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>3137</td>\n",
       "      <td>major histocompatibility complex%2C class I%2C...</td>\n",
       "      <td>CHS.48258</td>\n",
       "      <td>1263680</td>\n",
       "      <td>1267665</td>\n",
       "      <td>both</td>\n",
       "      <td>30005971</td>\n",
       "      <td>30009956</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>281</th>\n",
       "      <td>misc_RNA</td>\n",
       "      <td>HLA-J</td>\n",
       "      <td>CHS.48395</td>\n",
       "      <td>chr6:30005971-30009956:+,chr6_GL000250v2_alt:1...</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>3137</td>\n",
       "      <td>major histocompatibility complex%2C class I%2C...</td>\n",
       "      <td>CHS.48395</td>\n",
       "      <td>1485356</td>\n",
       "      <td>1489341</td>\n",
       "      <td>both</td>\n",
       "      <td>30005971</td>\n",
       "      <td>30009956</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>282</th>\n",
       "      <td>misc_RNA</td>\n",
       "      <td>HLA-J</td>\n",
       "      <td>CHS.48620</td>\n",
       "      <td>chr6:30005971-30009956:+,chr6_GL000250v2_alt:1...</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>3137</td>\n",
       "      <td>major histocompatibility complex%2C class I%2C...</td>\n",
       "      <td>CHS.48620</td>\n",
       "      <td>1261506</td>\n",
       "      <td>1265491</td>\n",
       "      <td>both</td>\n",
       "      <td>30005971</td>\n",
       "      <td>30009956</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>283</th>\n",
       "      <td>misc_RNA</td>\n",
       "      <td>HLA-J</td>\n",
       "      <td>CHS.48837</td>\n",
       "      <td>chr6:30005971-30009956:+,chr6_GL000250v2_alt:1...</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>3137</td>\n",
       "      <td>major histocompatibility complex%2C class I%2C...</td>\n",
       "      <td>CHS.48837</td>\n",
       "      <td>1266702</td>\n",
       "      <td>1270687</td>\n",
       "      <td>both</td>\n",
       "      <td>30005971</td>\n",
       "      <td>30009956</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>284</th>\n",
       "      <td>misc_RNA</td>\n",
       "      <td>HLA-J</td>\n",
       "      <td>CHS.49023</td>\n",
       "      <td>chr6:30005971-30009956:+,chr6_GL000250v2_alt:1...</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>3137</td>\n",
       "      <td>major histocompatibility complex%2C class I%2C...</td>\n",
       "      <td>CHS.49023</td>\n",
       "      <td>1350027</td>\n",
       "      <td>1354012</td>\n",
       "      <td>both</td>\n",
       "      <td>30005971</td>\n",
       "      <td>30009956</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>285</th>\n",
       "      <td>misc_RNA</td>\n",
       "      <td>HLA-J</td>\n",
       "      <td>CHS.49230</td>\n",
       "      <td>chr6:30005971-30009956:+,chr6_GL000250v2_alt:1...</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>3137</td>\n",
       "      <td>major histocompatibility complex%2C class I%2C...</td>\n",
       "      <td>CHS.49230</td>\n",
       "      <td>1261068</td>\n",
       "      <td>1265040</td>\n",
       "      <td>both</td>\n",
       "      <td>30005971</td>\n",
       "      <td>30009956</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>286</th>\n",
       "      <td>misc_RNA</td>\n",
       "      <td>HLA-J</td>\n",
       "      <td>CHS.49451</td>\n",
       "      <td>chr6:30005971-30009956:+,chr6_GL000250v2_alt:1...</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>3137</td>\n",
       "      <td>major histocompatibility complex%2C class I%2C...</td>\n",
       "      <td>CHS.49451</td>\n",
       "      <td>1303778</td>\n",
       "      <td>1307763</td>\n",
       "      <td>both</td>\n",
       "      <td>30005971</td>\n",
       "      <td>30009956</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>333</th>\n",
       "      <td>lncRNA</td>\n",
       "      <td>LOC105378597</td>\n",
       "      <td>CHS.28495</td>\n",
       "      <td>chr1:2633111-2641043:+,chr1_KI270762v1_alt:184...</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>105378597</td>\n",
       "      <td>uncharacterized LOC105378597</td>\n",
       "      <td>CHS.28495</td>\n",
       "      <td>184276</td>\n",
       "      <td>187405</td>\n",
       "      <td>both</td>\n",
       "      <td>2633111</td>\n",
       "      <td>2641043</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>342</th>\n",
       "      <td>protein_coding</td>\n",
       "      <td>LILRA6</td>\n",
       "      <td>CHS.27857</td>\n",
       "      <td>chr19:54236592-54242768:-,chr19_GL949746v1_alt...</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>79168</td>\n",
       "      <td>leukocyte immunoglobulin like receptor A6</td>\n",
       "      <td>CHS.27857</td>\n",
       "      <td>212163</td>\n",
       "      <td>217575</td>\n",
       "      <td>both</td>\n",
       "      <td>54236592</td>\n",
       "      <td>54242768</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>343</th>\n",
       "      <td>protein_coding</td>\n",
       "      <td>LILRA6</td>\n",
       "      <td>CHS.28035</td>\n",
       "      <td>chr19:54236592-54242768:-,chr19_GL949746v1_alt...</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>79168</td>\n",
       "      <td>leukocyte immunoglobulin like receptor A6</td>\n",
       "      <td>CHS.28035</td>\n",
       "      <td>211581</td>\n",
       "      <td>217730</td>\n",
       "      <td>both</td>\n",
       "      <td>54236592</td>\n",
       "      <td>54242768</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>344</th>\n",
       "      <td>protein_coding</td>\n",
       "      <td>LILRA6</td>\n",
       "      <td>CHS.28085</td>\n",
       "      <td>chr19:54236592-54242768:-,chr19_GL949746v1_alt...</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>79168</td>\n",
       "      <td>leukocyte immunoglobulin like receptor A6</td>\n",
       "      <td>CHS.28085</td>\n",
       "      <td>213801</td>\n",
       "      <td>217756</td>\n",
       "      <td>both</td>\n",
       "      <td>54236592</td>\n",
       "      <td>54242768</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>345</th>\n",
       "      <td>protein_coding</td>\n",
       "      <td>LILRA6</td>\n",
       "      <td>CHS.28395</td>\n",
       "      <td>chr19:54236592-54242768:-,chr19_GL949746v1_alt...</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>79168</td>\n",
       "      <td>leukocyte immunoglobulin like receptor A6</td>\n",
       "      <td>CHS.28395</td>\n",
       "      <td>201519</td>\n",
       "      <td>218008</td>\n",
       "      <td>both</td>\n",
       "      <td>54236592</td>\n",
       "      <td>54242768</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>403</th>\n",
       "      <td>lncRNA</td>\n",
       "      <td>LOC107984593</td>\n",
       "      <td>CHS.59635</td>\n",
       "      <td>chr13:114126120-114127591:+,KN538371.1:185211-...</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>107984593</td>\n",
       "      <td>uncharacterized LOC107984593</td>\n",
       "      <td>CHS.59635</td>\n",
       "      <td>185211</td>\n",
       "      <td>186682</td>\n",
       "      <td>both</td>\n",
       "      <td>114126120</td>\n",
       "      <td>114127591</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>418</th>\n",
       "      <td>protein_coding</td>\n",
       "      <td>DNAJC9</td>\n",
       "      <td>CHS.6491</td>\n",
       "      <td>chr10:73241954-73247331:-,chr10:73247342-73248...</td>\n",
       "      <td>RefSeq,GENCODE</td>\n",
       "      <td>23234</td>\n",
       "      <td>DnaJ heat shock protein family (Hsp40) member C9</td>\n",
       "      <td>CHS.6491</td>\n",
       "      <td>73247342</td>\n",
       "      <td>73248268</td>\n",
       "      <td>both</td>\n",
       "      <td>73241954</td>\n",
       "      <td>73247331</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>440</th>\n",
       "      <td>protein_coding</td>\n",
       "      <td>TBC1D3B</td>\n",
       "      <td>CHS.23799</td>\n",
       "      <td>chr17:36165681-36176624:-,chr17_KI270857v1_alt...</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>414059</td>\n",
       "      <td>TBC1 domain family member 3B</td>\n",
       "      <td>CHS.23799</td>\n",
       "      <td>400589</td>\n",
       "      <td>411512</td>\n",
       "      <td>both</td>\n",
       "      <td>36165681</td>\n",
       "      <td>36176624</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>441</th>\n",
       "      <td>protein_coding</td>\n",
       "      <td>TBC1D3B</td>\n",
       "      <td>CHS.23933</td>\n",
       "      <td>chr17:36165681-36176624:-,chr17_KI270857v1_alt...</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>414059</td>\n",
       "      <td>TBC1 domain family member 3B</td>\n",
       "      <td>CHS.23933</td>\n",
       "      <td>119784</td>\n",
       "      <td>130727</td>\n",
       "      <td>both</td>\n",
       "      <td>36165681</td>\n",
       "      <td>36176624</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>477</th>\n",
       "      <td>lncRNA</td>\n",
       "      <td>LOC574538</td>\n",
       "      <td>CHS.13397</td>\n",
       "      <td>chr12:137411-149169:-,chr12_GL877875v1_alt:131...</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>574538</td>\n",
       "      <td>uncharacterized LOC574538</td>\n",
       "      <td>CHS.13397</td>\n",
       "      <td>131426</td>\n",
       "      <td>143184</td>\n",
       "      <td>both</td>\n",
       "      <td>137411</td>\n",
       "      <td>149169</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>481</th>\n",
       "      <td>lncRNA</td>\n",
       "      <td>FLJ13224</td>\n",
       "      <td>CHS.13427</td>\n",
       "      <td>chr12:31324316-31325945:+,chr12_KI270835v1_alt...</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>79857</td>\n",
       "      <td>uncharacterized LOC79857</td>\n",
       "      <td>CHS.13427</td>\n",
       "      <td>215110</td>\n",
       "      <td>216739</td>\n",
       "      <td>both</td>\n",
       "      <td>31324316</td>\n",
       "      <td>31325945</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>510</th>\n",
       "      <td>protein_coding</td>\n",
       "      <td>OR8K3</td>\n",
       "      <td>CHS.10520</td>\n",
       "      <td>chr11:56318307-56319245:+,chr11_JH159136v1_alt...</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>219473</td>\n",
       "      <td>olfactory receptor family 8 subfamily K member...</td>\n",
       "      <td>CHS.10520</td>\n",
       "      <td>129056</td>\n",
       "      <td>129994</td>\n",
       "      <td>both</td>\n",
       "      <td>56318307</td>\n",
       "      <td>56319245</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46586</th>\n",
       "      <td>protein_coding</td>\n",
       "      <td>PRRC2A</td>\n",
       "      <td>CHS.48475</td>\n",
       "      <td>chr6:31618007-31637780:+,chr6_GL000251v2_alt:3...</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>7916</td>\n",
       "      <td>proline rich coiled-coil 2A</td>\n",
       "      <td>CHS.48475</td>\n",
       "      <td>3097961</td>\n",
       "      <td>3115064</td>\n",
       "      <td>both</td>\n",
       "      <td>31618007</td>\n",
       "      <td>31637780</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46587</th>\n",
       "      <td>protein_coding</td>\n",
       "      <td>PRRC2A</td>\n",
       "      <td>CHS.48697</td>\n",
       "      <td>chr6:31618007-31637780:+,chr6_GL000251v2_alt:3...</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>7916</td>\n",
       "      <td>proline rich coiled-coil 2A</td>\n",
       "      <td>CHS.48697</td>\n",
       "      <td>2868437</td>\n",
       "      <td>2885541</td>\n",
       "      <td>both</td>\n",
       "      <td>31618007</td>\n",
       "      <td>31637780</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46588</th>\n",
       "      <td>protein_coding</td>\n",
       "      <td>PRRC2A</td>\n",
       "      <td>CHS.48910</td>\n",
       "      <td>chr6:31618007-31637780:+,chr6_GL000251v2_alt:3...</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>7916</td>\n",
       "      <td>proline rich coiled-coil 2A</td>\n",
       "      <td>CHS.48910</td>\n",
       "      <td>2925714</td>\n",
       "      <td>2942818</td>\n",
       "      <td>both</td>\n",
       "      <td>31618007</td>\n",
       "      <td>31637780</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46589</th>\n",
       "      <td>protein_coding</td>\n",
       "      <td>PRRC2A</td>\n",
       "      <td>CHS.49092</td>\n",
       "      <td>chr6:31618007-31637780:+,chr6_GL000251v2_alt:3...</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>7916</td>\n",
       "      <td>proline rich coiled-coil 2A</td>\n",
       "      <td>CHS.49092</td>\n",
       "      <td>2962555</td>\n",
       "      <td>2979657</td>\n",
       "      <td>both</td>\n",
       "      <td>31618007</td>\n",
       "      <td>31637780</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46590</th>\n",
       "      <td>protein_coding</td>\n",
       "      <td>PRRC2A</td>\n",
       "      <td>CHS.49312</td>\n",
       "      <td>chr6:31618007-31637780:+,chr6_GL000251v2_alt:3...</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>7916</td>\n",
       "      <td>proline rich coiled-coil 2A</td>\n",
       "      <td>CHS.49312</td>\n",
       "      <td>2876495</td>\n",
       "      <td>2893595</td>\n",
       "      <td>both</td>\n",
       "      <td>31618007</td>\n",
       "      <td>31637780</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46591</th>\n",
       "      <td>protein_coding</td>\n",
       "      <td>PRRC2A</td>\n",
       "      <td>CHS.49520</td>\n",
       "      <td>chr6:31618007-31637780:+,chr6_GL000251v2_alt:3...</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>7916</td>\n",
       "      <td>proline rich coiled-coil 2A</td>\n",
       "      <td>CHS.49520</td>\n",
       "      <td>2919955</td>\n",
       "      <td>2937058</td>\n",
       "      <td>both</td>\n",
       "      <td>31618007</td>\n",
       "      <td>31637780</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46614</th>\n",
       "      <td>protein_coding</td>\n",
       "      <td>ZNF100</td>\n",
       "      <td>CHS.27835</td>\n",
       "      <td>chr19:21722772-21767628:-,chr19_GL383575v2_alt...</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>163227</td>\n",
       "      <td>zinc finger protein 100</td>\n",
       "      <td>CHS.27835</td>\n",
       "      <td>85231</td>\n",
       "      <td>128818</td>\n",
       "      <td>both</td>\n",
       "      <td>21722772</td>\n",
       "      <td>21767628</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46629</th>\n",
       "      <td>lncRNA</td>\n",
       "      <td>LOC105375795</td>\n",
       "      <td>CHS.54880</td>\n",
       "      <td>chr8:143043083-143045256:-,chr8_KI270817v1_alt...</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>105375795</td>\n",
       "      <td>uncharacterized LOC105375795</td>\n",
       "      <td>CHS.54880</td>\n",
       "      <td>33371</td>\n",
       "      <td>35544</td>\n",
       "      <td>both</td>\n",
       "      <td>143043083</td>\n",
       "      <td>143045256</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46632</th>\n",
       "      <td>misc_RNA</td>\n",
       "      <td>HLA-L</td>\n",
       "      <td>CHS.48407</td>\n",
       "      <td>chr6:30259562-30273921:+,chr6_GL000251v2_alt:1...</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>3139</td>\n",
       "      <td>major histocompatibility complex%2C class I%2C...</td>\n",
       "      <td>CHS.48407</td>\n",
       "      <td>1739329</td>\n",
       "      <td>1746646</td>\n",
       "      <td>both</td>\n",
       "      <td>30259562</td>\n",
       "      <td>30273921</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46633</th>\n",
       "      <td>misc_RNA</td>\n",
       "      <td>HLA-L</td>\n",
       "      <td>CHS.48632</td>\n",
       "      <td>chr6:30259562-30273921:+,chr6_GL000251v2_alt:1...</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>3139</td>\n",
       "      <td>major histocompatibility complex%2C class I%2C...</td>\n",
       "      <td>CHS.48632</td>\n",
       "      <td>1515431</td>\n",
       "      <td>1522820</td>\n",
       "      <td>both</td>\n",
       "      <td>30259562</td>\n",
       "      <td>30273921</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46634</th>\n",
       "      <td>misc_RNA</td>\n",
       "      <td>HLA-L</td>\n",
       "      <td>CHS.48849</td>\n",
       "      <td>chr6:30259562-30273921:+,chr6_GL000251v2_alt:1...</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>3139</td>\n",
       "      <td>major histocompatibility complex%2C class I%2C...</td>\n",
       "      <td>CHS.48849</td>\n",
       "      <td>1569799</td>\n",
       "      <td>1577184</td>\n",
       "      <td>both</td>\n",
       "      <td>30259562</td>\n",
       "      <td>30273921</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46635</th>\n",
       "      <td>misc_RNA</td>\n",
       "      <td>HLA-L</td>\n",
       "      <td>CHS.49242</td>\n",
       "      <td>chr6:30259562-30273921:+,chr6_GL000251v2_alt:1...</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>3139</td>\n",
       "      <td>major histocompatibility complex%2C class I%2C...</td>\n",
       "      <td>CHS.49242</td>\n",
       "      <td>1514659</td>\n",
       "      <td>1522046</td>\n",
       "      <td>both</td>\n",
       "      <td>30259562</td>\n",
       "      <td>30273921</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46636</th>\n",
       "      <td>misc_RNA</td>\n",
       "      <td>HLA-L</td>\n",
       "      <td>CHS.49465</td>\n",
       "      <td>chr6:30259562-30273921:+,chr6_GL000251v2_alt:1...</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>3139</td>\n",
       "      <td>major histocompatibility complex%2C class I%2C...</td>\n",
       "      <td>CHS.49465</td>\n",
       "      <td>1558042</td>\n",
       "      <td>1565431</td>\n",
       "      <td>both</td>\n",
       "      <td>30259562</td>\n",
       "      <td>30273921</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46638</th>\n",
       "      <td>protein_coding</td>\n",
       "      <td>ATAT1</td>\n",
       "      <td>CHS.48420</td>\n",
       "      <td>chr6:30626842-30646823:+,chr6_GL000251v2_alt:2...</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>79969</td>\n",
       "      <td>alpha tubulin acetyltransferase 1</td>\n",
       "      <td>CHS.48420</td>\n",
       "      <td>2106566</td>\n",
       "      <td>2126545</td>\n",
       "      <td>both</td>\n",
       "      <td>30626842</td>\n",
       "      <td>30646823</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46639</th>\n",
       "      <td>protein_coding</td>\n",
       "      <td>ATAT1</td>\n",
       "      <td>CHS.48645</td>\n",
       "      <td>chr6:30626842-30646823:+,chr6_GL000251v2_alt:2...</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>79969</td>\n",
       "      <td>alpha tubulin acetyltransferase 1</td>\n",
       "      <td>CHS.48645</td>\n",
       "      <td>1882672</td>\n",
       "      <td>1902652</td>\n",
       "      <td>both</td>\n",
       "      <td>30626842</td>\n",
       "      <td>30646823</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46640</th>\n",
       "      <td>protein_coding</td>\n",
       "      <td>ATAT1</td>\n",
       "      <td>CHS.48862</td>\n",
       "      <td>chr6:30626842-30646823:+,chr6_GL000251v2_alt:2...</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>79969</td>\n",
       "      <td>alpha tubulin acetyltransferase 1</td>\n",
       "      <td>CHS.48862</td>\n",
       "      <td>1937071</td>\n",
       "      <td>1957052</td>\n",
       "      <td>both</td>\n",
       "      <td>30626842</td>\n",
       "      <td>30646823</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46641</th>\n",
       "      <td>protein_coding</td>\n",
       "      <td>ATAT1</td>\n",
       "      <td>CHS.49044</td>\n",
       "      <td>chr6:30626842-30646823:+,chr6_GL000251v2_alt:2...</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>79969</td>\n",
       "      <td>alpha tubulin acetyltransferase 1</td>\n",
       "      <td>CHS.49044</td>\n",
       "      <td>1970890</td>\n",
       "      <td>1990873</td>\n",
       "      <td>both</td>\n",
       "      <td>30626842</td>\n",
       "      <td>30646823</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46642</th>\n",
       "      <td>protein_coding</td>\n",
       "      <td>ATAT1</td>\n",
       "      <td>CHS.49255</td>\n",
       "      <td>chr6:30626842-30646823:+,chr6_GL000251v2_alt:2...</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>79969</td>\n",
       "      <td>alpha tubulin acetyltransferase 1</td>\n",
       "      <td>CHS.49255</td>\n",
       "      <td>1881930</td>\n",
       "      <td>1901911</td>\n",
       "      <td>both</td>\n",
       "      <td>30626842</td>\n",
       "      <td>30646823</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46643</th>\n",
       "      <td>protein_coding</td>\n",
       "      <td>ATAT1</td>\n",
       "      <td>CHS.49478</td>\n",
       "      <td>chr6:30626842-30646823:+,chr6_GL000251v2_alt:2...</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>79969</td>\n",
       "      <td>alpha tubulin acetyltransferase 1</td>\n",
       "      <td>CHS.49478</td>\n",
       "      <td>1927600</td>\n",
       "      <td>1947582</td>\n",
       "      <td>both</td>\n",
       "      <td>30626842</td>\n",
       "      <td>30646823</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46678</th>\n",
       "      <td>lncRNA</td>\n",
       "      <td>LOC101929494,AC124789.1</td>\n",
       "      <td>CHS.23852</td>\n",
       "      <td>chr17:38450626-38452413:-,chr17_KI270857v1_alt...</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>101929494</td>\n",
       "      <td>uncharacterized LOC101929494</td>\n",
       "      <td>CHS.23852</td>\n",
       "      <td>2485898</td>\n",
       "      <td>2488747</td>\n",
       "      <td>both</td>\n",
       "      <td>38450626</td>\n",
       "      <td>38452413</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46684</th>\n",
       "      <td>lncRNA</td>\n",
       "      <td>MIF-AS1</td>\n",
       "      <td>CHS.36357</td>\n",
       "      <td>chr22:23886114-23898930:-,chr22_KI270879v1_alt...</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>284889</td>\n",
       "      <td>MIF antisense RNA 1</td>\n",
       "      <td>CHS.36357</td>\n",
       "      <td>130072</td>\n",
       "      <td>135292</td>\n",
       "      <td>both</td>\n",
       "      <td>23886114</td>\n",
       "      <td>23898930</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46728</th>\n",
       "      <td>protein_coding</td>\n",
       "      <td>TBC1D3J</td>\n",
       "      <td>CHS.23809</td>\n",
       "      <td>chr17:36450649-36458000:-,chr17_KI270857v1_alt...</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>100510707</td>\n",
       "      <td>TBC1 domain family member 3J</td>\n",
       "      <td>CHS.23809</td>\n",
       "      <td>685553</td>\n",
       "      <td>692904</td>\n",
       "      <td>both</td>\n",
       "      <td>36450649</td>\n",
       "      <td>36458000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46755</th>\n",
       "      <td>protein_coding</td>\n",
       "      <td>LILRA3</td>\n",
       "      <td>CHS.27902</td>\n",
       "      <td>chr19_GL949746v1_alt:270735-275150:-,chr19_GL9...</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>11026</td>\n",
       "      <td>leukocyte immunoglobulin like receptor A3</td>\n",
       "      <td>CHS.27902</td>\n",
       "      <td>270616</td>\n",
       "      <td>275031</td>\n",
       "      <td>both</td>\n",
       "      <td>270735</td>\n",
       "      <td>275150</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46756</th>\n",
       "      <td>protein_coding</td>\n",
       "      <td>LILRA3</td>\n",
       "      <td>CHS.28088</td>\n",
       "      <td>chr19_GL949746v1_alt:270735-275150:-,chr19_GL9...</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>11026</td>\n",
       "      <td>leukocyte immunoglobulin like receptor A3</td>\n",
       "      <td>CHS.28088</td>\n",
       "      <td>270991</td>\n",
       "      <td>275406</td>\n",
       "      <td>both</td>\n",
       "      <td>270735</td>\n",
       "      <td>275150</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46757</th>\n",
       "      <td>protein_coding</td>\n",
       "      <td>LILRA3</td>\n",
       "      <td>CHS.28399</td>\n",
       "      <td>chr19_GL949746v1_alt:270735-275150:-,chr19_GL9...</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>11026</td>\n",
       "      <td>leukocyte immunoglobulin like receptor A3</td>\n",
       "      <td>CHS.28399</td>\n",
       "      <td>270964</td>\n",
       "      <td>275379</td>\n",
       "      <td>both</td>\n",
       "      <td>270735</td>\n",
       "      <td>275150</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46847</th>\n",
       "      <td>lncRNA</td>\n",
       "      <td>LOC150935</td>\n",
       "      <td>CHS.59478</td>\n",
       "      <td>chr2:239762860-239800664:+,KN538362.1:65957-10...</td>\n",
       "      <td>RefSeq,GENCODE</td>\n",
       "      <td>150935</td>\n",
       "      <td>uncharacterized LOC150935</td>\n",
       "      <td>CHS.59478</td>\n",
       "      <td>65957</td>\n",
       "      <td>103761</td>\n",
       "      <td>both</td>\n",
       "      <td>239762860</td>\n",
       "      <td>239800664</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46861</th>\n",
       "      <td>protein_coding</td>\n",
       "      <td>SALL3</td>\n",
       "      <td>CHS.25089</td>\n",
       "      <td>chr18:78980275-79002965:+,chr18_GL383572v1_alt...</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>27164</td>\n",
       "      <td>spalt-like transcription factor 3</td>\n",
       "      <td>CHS.25089</td>\n",
       "      <td>45390</td>\n",
       "      <td>67793</td>\n",
       "      <td>both</td>\n",
       "      <td>78980275</td>\n",
       "      <td>79002965</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46862</th>\n",
       "      <td>protein_coding</td>\n",
       "      <td>SALL3</td>\n",
       "      <td>CHS.25098</td>\n",
       "      <td>chr18:78980275-79002965:+,chr18_GL383572v1_alt...</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>27164</td>\n",
       "      <td>spalt-like transcription factor 3</td>\n",
       "      <td>CHS.25098</td>\n",
       "      <td>45390</td>\n",
       "      <td>64084</td>\n",
       "      <td>both</td>\n",
       "      <td>78980275</td>\n",
       "      <td>79002965</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46865</th>\n",
       "      <td>protein_coding</td>\n",
       "      <td>GATAD2B</td>\n",
       "      <td>CHS.28440</td>\n",
       "      <td>chr1:153793687-153923360:-,chr1_GL383518v1_alt...</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>57459</td>\n",
       "      <td>GATA zinc finger domain containing 2B</td>\n",
       "      <td>CHS.28440</td>\n",
       "      <td>121426</td>\n",
       "      <td>145051</td>\n",
       "      <td>both</td>\n",
       "      <td>153793687</td>\n",
       "      <td>153923360</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46884</th>\n",
       "      <td>lncRNA</td>\n",
       "      <td>LOC107986668</td>\n",
       "      <td>CHS.49643</td>\n",
       "      <td>chr6:166254686-166259194:-,chr6_KI270802v1_alt...</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>107986668</td>\n",
       "      <td>uncharacterized LOC107986668</td>\n",
       "      <td>CHS.49643</td>\n",
       "      <td>34107</td>\n",
       "      <td>38615</td>\n",
       "      <td>both</td>\n",
       "      <td>166254686</td>\n",
       "      <td>166259194</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3790 rows × 13 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            Gene_Type                Gene_Name     GFF_ID  \\\n",
       "48     protein_coding                   ZBTB12  CHS.48511   \n",
       "49     protein_coding                   ZBTB12  CHS.48732   \n",
       "50     protein_coding                   ZBTB12  CHS.49124   \n",
       "51     protein_coding                   ZBTB12  CHS.49345   \n",
       "75             lncRNA             LOC105375115  CHS.52474   \n",
       "76             lncRNA             LOC105375115  CHS.52499   \n",
       "85     protein_coding                 SERPINF1  CHS.23883   \n",
       "166    protein_coding                      ABR  CHS.23887   \n",
       "167    protein_coding                      ABR  CHS.23947   \n",
       "169    protein_coding                    OR9G1  CHS.10525   \n",
       "175            lncRNA             LOC107984555  CHS.14862   \n",
       "280          misc_RNA                    HLA-J  CHS.48258   \n",
       "281          misc_RNA                    HLA-J  CHS.48395   \n",
       "282          misc_RNA                    HLA-J  CHS.48620   \n",
       "283          misc_RNA                    HLA-J  CHS.48837   \n",
       "284          misc_RNA                    HLA-J  CHS.49023   \n",
       "285          misc_RNA                    HLA-J  CHS.49230   \n",
       "286          misc_RNA                    HLA-J  CHS.49451   \n",
       "333            lncRNA             LOC105378597  CHS.28495   \n",
       "342    protein_coding                   LILRA6  CHS.27857   \n",
       "343    protein_coding                   LILRA6  CHS.28035   \n",
       "344    protein_coding                   LILRA6  CHS.28085   \n",
       "345    protein_coding                   LILRA6  CHS.28395   \n",
       "403            lncRNA             LOC107984593  CHS.59635   \n",
       "418    protein_coding                   DNAJC9   CHS.6491   \n",
       "440    protein_coding                  TBC1D3B  CHS.23799   \n",
       "441    protein_coding                  TBC1D3B  CHS.23933   \n",
       "477            lncRNA                LOC574538  CHS.13397   \n",
       "481            lncRNA                 FLJ13224  CHS.13427   \n",
       "510    protein_coding                    OR8K3  CHS.10520   \n",
       "...               ...                      ...        ...   \n",
       "46586  protein_coding                   PRRC2A  CHS.48475   \n",
       "46587  protein_coding                   PRRC2A  CHS.48697   \n",
       "46588  protein_coding                   PRRC2A  CHS.48910   \n",
       "46589  protein_coding                   PRRC2A  CHS.49092   \n",
       "46590  protein_coding                   PRRC2A  CHS.49312   \n",
       "46591  protein_coding                   PRRC2A  CHS.49520   \n",
       "46614  protein_coding                   ZNF100  CHS.27835   \n",
       "46629          lncRNA             LOC105375795  CHS.54880   \n",
       "46632        misc_RNA                    HLA-L  CHS.48407   \n",
       "46633        misc_RNA                    HLA-L  CHS.48632   \n",
       "46634        misc_RNA                    HLA-L  CHS.48849   \n",
       "46635        misc_RNA                    HLA-L  CHS.49242   \n",
       "46636        misc_RNA                    HLA-L  CHS.49465   \n",
       "46638  protein_coding                    ATAT1  CHS.48420   \n",
       "46639  protein_coding                    ATAT1  CHS.48645   \n",
       "46640  protein_coding                    ATAT1  CHS.48862   \n",
       "46641  protein_coding                    ATAT1  CHS.49044   \n",
       "46642  protein_coding                    ATAT1  CHS.49255   \n",
       "46643  protein_coding                    ATAT1  CHS.49478   \n",
       "46678          lncRNA  LOC101929494,AC124789.1  CHS.23852   \n",
       "46684          lncRNA                  MIF-AS1  CHS.36357   \n",
       "46728  protein_coding                  TBC1D3J  CHS.23809   \n",
       "46755  protein_coding                   LILRA3  CHS.27902   \n",
       "46756  protein_coding                   LILRA3  CHS.28088   \n",
       "46757  protein_coding                   LILRA3  CHS.28399   \n",
       "46847          lncRNA                LOC150935  CHS.59478   \n",
       "46861  protein_coding                    SALL3  CHS.25089   \n",
       "46862  protein_coding                    SALL3  CHS.25098   \n",
       "46865  protein_coding                  GATAD2B  CHS.28440   \n",
       "46884          lncRNA             LOC107986668  CHS.49643   \n",
       "\n",
       "                                                Location        Database  \\\n",
       "48     chr6:31899194-31902331:-,chr6_GL000251v2_alt:3...          RefSeq   \n",
       "49     chr6:31899194-31902331:-,chr6_GL000251v2_alt:3...          RefSeq   \n",
       "50     chr6:31899194-31902331:-,chr6_GL000251v2_alt:3...          RefSeq   \n",
       "51     chr6:31899194-31902331:-,chr6_GL000251v2_alt:3...          RefSeq   \n",
       "75     chr7:174920-176013:+,chr7_KI270804v1_alt:13432...          RefSeq   \n",
       "76     chr7:174920-176013:+,chr7_KI270804v1_alt:13432...          RefSeq   \n",
       "85     chr17:1761965-1779392:+,chr17_KI270861v1_alt:1...          RefSeq   \n",
       "166    chr17:999953-1229713:-,chr17_KI270862v1_alt:42...          RefSeq   \n",
       "167    chr17:999953-1229713:-,chr17_KI270862v1_alt:42...          RefSeq   \n",
       "169    chr11:56700388-56701305:+,chr11_JH159137v1_alt...          RefSeq   \n",
       "175    chr13:112273139-112276195:+,chr13_KI270838v1_a...          RefSeq   \n",
       "280    chr6:30005971-30009956:+,chr6_GL000250v2_alt:1...          RefSeq   \n",
       "281    chr6:30005971-30009956:+,chr6_GL000250v2_alt:1...          RefSeq   \n",
       "282    chr6:30005971-30009956:+,chr6_GL000250v2_alt:1...          RefSeq   \n",
       "283    chr6:30005971-30009956:+,chr6_GL000250v2_alt:1...          RefSeq   \n",
       "284    chr6:30005971-30009956:+,chr6_GL000250v2_alt:1...          RefSeq   \n",
       "285    chr6:30005971-30009956:+,chr6_GL000250v2_alt:1...          RefSeq   \n",
       "286    chr6:30005971-30009956:+,chr6_GL000250v2_alt:1...          RefSeq   \n",
       "333    chr1:2633111-2641043:+,chr1_KI270762v1_alt:184...          RefSeq   \n",
       "342    chr19:54236592-54242768:-,chr19_GL949746v1_alt...          RefSeq   \n",
       "343    chr19:54236592-54242768:-,chr19_GL949746v1_alt...          RefSeq   \n",
       "344    chr19:54236592-54242768:-,chr19_GL949746v1_alt...          RefSeq   \n",
       "345    chr19:54236592-54242768:-,chr19_GL949746v1_alt...          RefSeq   \n",
       "403    chr13:114126120-114127591:+,KN538371.1:185211-...          RefSeq   \n",
       "418    chr10:73241954-73247331:-,chr10:73247342-73248...  RefSeq,GENCODE   \n",
       "440    chr17:36165681-36176624:-,chr17_KI270857v1_alt...          RefSeq   \n",
       "441    chr17:36165681-36176624:-,chr17_KI270857v1_alt...          RefSeq   \n",
       "477    chr12:137411-149169:-,chr12_GL877875v1_alt:131...          RefSeq   \n",
       "481    chr12:31324316-31325945:+,chr12_KI270835v1_alt...          RefSeq   \n",
       "510    chr11:56318307-56319245:+,chr11_JH159136v1_alt...          RefSeq   \n",
       "...                                                  ...             ...   \n",
       "46586  chr6:31618007-31637780:+,chr6_GL000251v2_alt:3...          RefSeq   \n",
       "46587  chr6:31618007-31637780:+,chr6_GL000251v2_alt:3...          RefSeq   \n",
       "46588  chr6:31618007-31637780:+,chr6_GL000251v2_alt:3...          RefSeq   \n",
       "46589  chr6:31618007-31637780:+,chr6_GL000251v2_alt:3...          RefSeq   \n",
       "46590  chr6:31618007-31637780:+,chr6_GL000251v2_alt:3...          RefSeq   \n",
       "46591  chr6:31618007-31637780:+,chr6_GL000251v2_alt:3...          RefSeq   \n",
       "46614  chr19:21722772-21767628:-,chr19_GL383575v2_alt...          RefSeq   \n",
       "46629  chr8:143043083-143045256:-,chr8_KI270817v1_alt...          RefSeq   \n",
       "46632  chr6:30259562-30273921:+,chr6_GL000251v2_alt:1...          RefSeq   \n",
       "46633  chr6:30259562-30273921:+,chr6_GL000251v2_alt:1...          RefSeq   \n",
       "46634  chr6:30259562-30273921:+,chr6_GL000251v2_alt:1...          RefSeq   \n",
       "46635  chr6:30259562-30273921:+,chr6_GL000251v2_alt:1...          RefSeq   \n",
       "46636  chr6:30259562-30273921:+,chr6_GL000251v2_alt:1...          RefSeq   \n",
       "46638  chr6:30626842-30646823:+,chr6_GL000251v2_alt:2...          RefSeq   \n",
       "46639  chr6:30626842-30646823:+,chr6_GL000251v2_alt:2...          RefSeq   \n",
       "46640  chr6:30626842-30646823:+,chr6_GL000251v2_alt:2...          RefSeq   \n",
       "46641  chr6:30626842-30646823:+,chr6_GL000251v2_alt:2...          RefSeq   \n",
       "46642  chr6:30626842-30646823:+,chr6_GL000251v2_alt:2...          RefSeq   \n",
       "46643  chr6:30626842-30646823:+,chr6_GL000251v2_alt:2...          RefSeq   \n",
       "46678  chr17:38450626-38452413:-,chr17_KI270857v1_alt...          RefSeq   \n",
       "46684  chr22:23886114-23898930:-,chr22_KI270879v1_alt...          RefSeq   \n",
       "46728  chr17:36450649-36458000:-,chr17_KI270857v1_alt...          RefSeq   \n",
       "46755  chr19_GL949746v1_alt:270735-275150:-,chr19_GL9...          RefSeq   \n",
       "46756  chr19_GL949746v1_alt:270735-275150:-,chr19_GL9...          RefSeq   \n",
       "46757  chr19_GL949746v1_alt:270735-275150:-,chr19_GL9...          RefSeq   \n",
       "46847  chr2:239762860-239800664:+,KN538362.1:65957-10...  RefSeq,GENCODE   \n",
       "46861  chr18:78980275-79002965:+,chr18_GL383572v1_alt...          RefSeq   \n",
       "46862  chr18:78980275-79002965:+,chr18_GL383572v1_alt...          RefSeq   \n",
       "46865  chr1:153793687-153923360:-,chr1_GL383518v1_alt...          RefSeq   \n",
       "46884  chr6:166254686-166259194:-,chr6_KI270802v1_alt...          RefSeq   \n",
       "\n",
       "      RefSeq_GeneID                                        Description  \\\n",
       "48           221527           zinc finger and BTB domain containing 12   \n",
       "49           221527           zinc finger and BTB domain containing 12   \n",
       "50           221527           zinc finger and BTB domain containing 12   \n",
       "51           221527           zinc finger and BTB domain containing 12   \n",
       "75        105375115                       uncharacterized LOC105375115   \n",
       "76        105375115                       uncharacterized LOC105375115   \n",
       "85             5176                           serpin family F member 1   \n",
       "166              29                                 active BCR-related   \n",
       "167              29                                 active BCR-related   \n",
       "169          390174   olfactory receptor family 9 subfamily G member 1   \n",
       "175       107984555                       uncharacterized LOC107984555   \n",
       "280            3137  major histocompatibility complex%2C class I%2C...   \n",
       "281            3137  major histocompatibility complex%2C class I%2C...   \n",
       "282            3137  major histocompatibility complex%2C class I%2C...   \n",
       "283            3137  major histocompatibility complex%2C class I%2C...   \n",
       "284            3137  major histocompatibility complex%2C class I%2C...   \n",
       "285            3137  major histocompatibility complex%2C class I%2C...   \n",
       "286            3137  major histocompatibility complex%2C class I%2C...   \n",
       "333       105378597                       uncharacterized LOC105378597   \n",
       "342           79168          leukocyte immunoglobulin like receptor A6   \n",
       "343           79168          leukocyte immunoglobulin like receptor A6   \n",
       "344           79168          leukocyte immunoglobulin like receptor A6   \n",
       "345           79168          leukocyte immunoglobulin like receptor A6   \n",
       "403       107984593                       uncharacterized LOC107984593   \n",
       "418           23234   DnaJ heat shock protein family (Hsp40) member C9   \n",
       "440          414059                       TBC1 domain family member 3B   \n",
       "441          414059                       TBC1 domain family member 3B   \n",
       "477          574538                          uncharacterized LOC574538   \n",
       "481           79857                           uncharacterized LOC79857   \n",
       "510          219473  olfactory receptor family 8 subfamily K member...   \n",
       "...             ...                                                ...   \n",
       "46586          7916                        proline rich coiled-coil 2A   \n",
       "46587          7916                        proline rich coiled-coil 2A   \n",
       "46588          7916                        proline rich coiled-coil 2A   \n",
       "46589          7916                        proline rich coiled-coil 2A   \n",
       "46590          7916                        proline rich coiled-coil 2A   \n",
       "46591          7916                        proline rich coiled-coil 2A   \n",
       "46614        163227                            zinc finger protein 100   \n",
       "46629     105375795                       uncharacterized LOC105375795   \n",
       "46632          3139  major histocompatibility complex%2C class I%2C...   \n",
       "46633          3139  major histocompatibility complex%2C class I%2C...   \n",
       "46634          3139  major histocompatibility complex%2C class I%2C...   \n",
       "46635          3139  major histocompatibility complex%2C class I%2C...   \n",
       "46636          3139  major histocompatibility complex%2C class I%2C...   \n",
       "46638         79969                  alpha tubulin acetyltransferase 1   \n",
       "46639         79969                  alpha tubulin acetyltransferase 1   \n",
       "46640         79969                  alpha tubulin acetyltransferase 1   \n",
       "46641         79969                  alpha tubulin acetyltransferase 1   \n",
       "46642         79969                  alpha tubulin acetyltransferase 1   \n",
       "46643         79969                  alpha tubulin acetyltransferase 1   \n",
       "46678     101929494                       uncharacterized LOC101929494   \n",
       "46684        284889                                MIF antisense RNA 1   \n",
       "46728     100510707                       TBC1 domain family member 3J   \n",
       "46755         11026          leukocyte immunoglobulin like receptor A3   \n",
       "46756         11026          leukocyte immunoglobulin like receptor A3   \n",
       "46757         11026          leukocyte immunoglobulin like receptor A3   \n",
       "46847        150935                          uncharacterized LOC150935   \n",
       "46861         27164                  spalt-like transcription factor 3   \n",
       "46862         27164                  spalt-like transcription factor 3   \n",
       "46865         57459              GATA zinc finger domain containing 2B   \n",
       "46884     107986668                       uncharacterized LOC107986668   \n",
       "\n",
       "              id     start       end _merge  start_old    end_old  \n",
       "48     CHS.48511   3377042   3379417   both   31899194   31902331  \n",
       "49     CHS.48732   3147391   3149766   both   31899194   31902331  \n",
       "50     CHS.49124   3241692   3244452   both   31899194   31902331  \n",
       "51     CHS.49345   3155590   3157967   both   31899194   31902331  \n",
       "75     CHS.52474    134324    135417   both     174920     176013  \n",
       "76     CHS.52499    167241    168334   both     174920     176013  \n",
       "85     CHS.23883    190927    195947   both    1761965    1779392  \n",
       "166    CHS.23887     42765    231794   both     999953    1229713  \n",
       "167    CHS.23947     42765    153205   both     999953    1229713  \n",
       "169    CHS.10525      7154      8071   both   56700388   56701305  \n",
       "175    CHS.14862     90257     92661   both  112273139  112276195  \n",
       "280    CHS.48258   1263680   1267665   both   30005971   30009956  \n",
       "281    CHS.48395   1485356   1489341   both   30005971   30009956  \n",
       "282    CHS.48620   1261506   1265491   both   30005971   30009956  \n",
       "283    CHS.48837   1266702   1270687   both   30005971   30009956  \n",
       "284    CHS.49023   1350027   1354012   both   30005971   30009956  \n",
       "285    CHS.49230   1261068   1265040   both   30005971   30009956  \n",
       "286    CHS.49451   1303778   1307763   both   30005971   30009956  \n",
       "333    CHS.28495    184276    187405   both    2633111    2641043  \n",
       "342    CHS.27857    212163    217575   both   54236592   54242768  \n",
       "343    CHS.28035    211581    217730   both   54236592   54242768  \n",
       "344    CHS.28085    213801    217756   both   54236592   54242768  \n",
       "345    CHS.28395    201519    218008   both   54236592   54242768  \n",
       "403    CHS.59635    185211    186682   both  114126120  114127591  \n",
       "418     CHS.6491  73247342  73248268   both   73241954   73247331  \n",
       "440    CHS.23799    400589    411512   both   36165681   36176624  \n",
       "441    CHS.23933    119784    130727   both   36165681   36176624  \n",
       "477    CHS.13397    131426    143184   both     137411     149169  \n",
       "481    CHS.13427    215110    216739   both   31324316   31325945  \n",
       "510    CHS.10520    129056    129994   both   56318307   56319245  \n",
       "...          ...       ...       ...    ...        ...        ...  \n",
       "46586  CHS.48475   3097961   3115064   both   31618007   31637780  \n",
       "46587  CHS.48697   2868437   2885541   both   31618007   31637780  \n",
       "46588  CHS.48910   2925714   2942818   both   31618007   31637780  \n",
       "46589  CHS.49092   2962555   2979657   both   31618007   31637780  \n",
       "46590  CHS.49312   2876495   2893595   both   31618007   31637780  \n",
       "46591  CHS.49520   2919955   2937058   both   31618007   31637780  \n",
       "46614  CHS.27835     85231    128818   both   21722772   21767628  \n",
       "46629  CHS.54880     33371     35544   both  143043083  143045256  \n",
       "46632  CHS.48407   1739329   1746646   both   30259562   30273921  \n",
       "46633  CHS.48632   1515431   1522820   both   30259562   30273921  \n",
       "46634  CHS.48849   1569799   1577184   both   30259562   30273921  \n",
       "46635  CHS.49242   1514659   1522046   both   30259562   30273921  \n",
       "46636  CHS.49465   1558042   1565431   both   30259562   30273921  \n",
       "46638  CHS.48420   2106566   2126545   both   30626842   30646823  \n",
       "46639  CHS.48645   1882672   1902652   both   30626842   30646823  \n",
       "46640  CHS.48862   1937071   1957052   both   30626842   30646823  \n",
       "46641  CHS.49044   1970890   1990873   both   30626842   30646823  \n",
       "46642  CHS.49255   1881930   1901911   both   30626842   30646823  \n",
       "46643  CHS.49478   1927600   1947582   both   30626842   30646823  \n",
       "46678  CHS.23852   2485898   2488747   both   38450626   38452413  \n",
       "46684  CHS.36357    130072    135292   both   23886114   23898930  \n",
       "46728  CHS.23809    685553    692904   both   36450649   36458000  \n",
       "46755  CHS.27902    270616    275031   both     270735     275150  \n",
       "46756  CHS.28088    270991    275406   both     270735     275150  \n",
       "46757  CHS.28399    270964    275379   both     270735     275150  \n",
       "46847  CHS.59478     65957    103761   both  239762860  239800664  \n",
       "46861  CHS.25089     45390     67793   both   78980275   79002965  \n",
       "46862  CHS.25098     45390     64084   both   78980275   79002965  \n",
       "46865  CHS.28440    121426    145051   both  153793687  153923360  \n",
       "46884  CHS.49643     34107     38615   both  166254686  166259194  \n",
       "\n",
       "[3790 rows x 13 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "ename": "AssertionError",
     "evalue": "something's wrong with the gene start and end coordinates",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mAssertionError\u001b[0m                            Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-19-e9eb8716d42c>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m()\u001b[0m\n\u001b[1;32m      7\u001b[0m \u001b[0mmdf\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'end_old'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mmdf\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'end_old'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mastype\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mint\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      8\u001b[0m \u001b[0mdisplay\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mmdf\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m~\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mmdf\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'start'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m==\u001b[0m\u001b[0mmdf\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'start_old'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m&\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mmdf\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'end'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m==\u001b[0m\u001b[0mmdf\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'end_old'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 9\u001b[0;31m \u001b[0;32massert\u001b[0m \u001b[0mlen\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mmdf\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mmdf\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'start'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m==\u001b[0m\u001b[0mmdf\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'start_old'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m&\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mmdf\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'end'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m==\u001b[0m\u001b[0mmdf\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'end_old'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m==\u001b[0m\u001b[0mlen\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mmdf\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\"something's wrong with the gene start and end coordinates\"\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;31mAssertionError\u001b[0m: something's wrong with the gene start and end coordinates"
     ]
    }
   ],
   "source": [
    "# genes file should be identical to the old one, since transcript coordinates should have never extended the old genes\n",
    "# so let's verify this assumption is also correct\n",
    "mdf['start']=mdf['start'].astype(int)\n",
    "mdf['end']=mdf['end'].astype(int)\n",
    "mdf[['start_old','end_old']]=mdf['Location'].str.split(':',expand=True)[1].str.split(\"-\",expand=True)\n",
    "mdf['start_old']=mdf['start_old'].astype(int)\n",
    "mdf['end_old']=mdf['end_old'].astype(int)\n",
    "display(mdf[~((mdf['start']==mdf['start_old'])&(mdf['end']==mdf['end_old']))])\n",
    "assert len(mdf[(mdf['start']==mdf['start_old'])&(mdf['end']==mdf['end_old'])])==len(mdf),\"something's wrong with the gene start and end coordinates\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [],
   "source": [
    "outFP=open(\"./chess2.04.genes\",\"w+\")\n",
    "\n",
    "genesEvaluated=[]\n",
    "wrongCoords=[]\n",
    "header=True\n",
    "with open(\"./chess2.03.genes\") as inFP:\n",
    "    for line in inFP.readlines():\n",
    "        if header: # skip header\n",
    "            outFP.write(line)\n",
    "            header=False\n",
    "            continue\n",
    "        lineCols=line.split(\"\\t\")\n",
    "        gids=lineCols[2].split(\",\")\n",
    "        locs=lineCols[3].split(\",\")\n",
    "        assert len(gids)==len(locs),\"something is wrong with the following locus: \"+lineCols[2]+\"\\t\"+lineCols[3]\n",
    "        newLocs=[]\n",
    "        for gl in zip(gids,locs):\n",
    "            genesEvaluated.append(gl[0]) # append geneID for verification later\n",
    "            location=gl[1].split(\":\")\n",
    "            coords=[int(x) for x in location[1].split(\"-\")]\n",
    "            ann_row=ann[ann['id']==gl[0]].reset_index(drop=True)\n",
    "            assert len(ann_row)==1,\"multiple rows?\"\n",
    "            if not int(ann_row['start'])==coords[0] and int(ann_row['end'])==coords[1]:\n",
    "                wrongCoords.append(gl[0])\n",
    "                \n",
    "            newLocs.append(ann_row['seqid'].iloc[0]+\":\"+str(int(ann_row['start'].iloc[0]))+\"-\"+str(int(ann_row['end'].iloc[0]))+\":\"+location[-1])\n",
    "        lineCols[3]=\",\".join(newLocs)\n",
    "        outFP.write('\\t'.join(lineCols))\n",
    "inFP.close()\n",
    "outFP.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['CHS.4228']\n",
      "46972\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "323817"
      ]
     },
     "execution_count": 41,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print(wrongCoords)\n",
    "print(len(genesEvaluated))\n",
    "len(ann[~(ann[\"id\"].isin(genesEvaluated))])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqid</th>\n",
       "      <th>source</th>\n",
       "      <th>type</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>phase</th>\n",
       "      <th>attributes</th>\n",
       "      <th>id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>transcript</td>\n",
       "      <td>11874.0</td>\n",
       "      <td>14409.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=CHS.1.1;Parent=CHS.1;gene_name=DDX11L1;Dbxr...</td>\n",
       "      <td>CHS.1.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>transcript</td>\n",
       "      <td>14362.0</td>\n",
       "      <td>29370.0</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=CHS.2.1;Parent=CHS.2;gene_name=WASH7P;Dbxre...</td>\n",
       "      <td>CHS.2.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>StringTie</td>\n",
       "      <td>transcript</td>\n",
       "      <td>14410.0</td>\n",
       "      <td>29369.0</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=CHS.2.2;Parent=CHS.2;STATUS=novel;ASSEMBLED...</td>\n",
       "      <td>CHS.2.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>StringTie</td>\n",
       "      <td>transcript</td>\n",
       "      <td>14410.0</td>\n",
       "      <td>29369.0</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=CHS.2.3;Parent=CHS.2;STATUS=novel;ASSEMBLED...</td>\n",
       "      <td>CHS.2.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>StringTie</td>\n",
       "      <td>transcript</td>\n",
       "      <td>14410.0</td>\n",
       "      <td>29369.0</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=CHS.2.4;Parent=CHS.2;STATUS=novel;ASSEMBLED...</td>\n",
       "      <td>CHS.2.4</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  seqid      source        type    start      end score strand phase  \\\n",
       "0  chr1  BestRefSeq  transcript  11874.0  14409.0     .      +     .   \n",
       "1  chr1  BestRefSeq  transcript  14362.0  29370.0     .      -     .   \n",
       "2  chr1   StringTie  transcript  14410.0  29369.0     .      -     .   \n",
       "3  chr1   StringTie  transcript  14410.0  29369.0     .      -     .   \n",
       "4  chr1   StringTie  transcript  14410.0  29369.0     .      -     .   \n",
       "\n",
       "                                          attributes       id  \n",
       "0  ID=CHS.1.1;Parent=CHS.1;gene_name=DDX11L1;Dbxr...  CHS.1.1  \n",
       "1  ID=CHS.2.1;Parent=CHS.2;gene_name=WASH7P;Dbxre...  CHS.2.1  \n",
       "2  ID=CHS.2.2;Parent=CHS.2;STATUS=novel;ASSEMBLED...  CHS.2.2  \n",
       "3  ID=CHS.2.3;Parent=CHS.2;STATUS=novel;ASSEMBLED...  CHS.2.3  \n",
       "4  ID=CHS.2.4;Parent=CHS.2;STATUS=novel;ASSEMBLED...  CHS.2.4  "
      ]
     },
     "execution_count": 39,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# now we need to replace old transcript coordinates with updated ones\n",
    "ann=pd.read_csv(\"./chess2.04.gff\",sep=\"\\t\",names=gff3Cols)\n",
    "ann.dropna(inplace=True,axis=0)\n",
    "ann=ann[ann[\"type\"]==\"transcript\"].reset_index(drop=True)\n",
    "ann[\"id\"]=ann.attributes.str.split(\"ID=\",expand=True)[1].str.split(\";\",expand=True)[0]\n",
    "ann[\"start\"]==ann['start'].astype(int)\n",
    "ann[\"end\"]==ann['end'].astype(int)\n",
    "ann.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [],
   "source": [
    "ann['coords']=ann[\"start\"].astype(int).astype(str)+\"-\"+ann[\"end\"].astype(int).astype(str)\n",
    "annDict=pd.Series(ann.coords.values,index=ann.id).to_dict()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {},
   "outputs": [],
   "source": [
    "outFP=open(\"./chess2.04.transcripts\",\"w+\")\n",
    "\n",
    "transcriptsEvaluated=[]\n",
    "wrongCoords=[]\n",
    "header=True\n",
    "with open(\"./chess2.03.transcripts\") as inFP:\n",
    "    for line in inFP.readlines():\n",
    "        if header: # skip header\n",
    "            outFP.write(line)\n",
    "            header=False\n",
    "            continue\n",
    "        lineCols=line.split(\"\\t\")\n",
    "        tid=lineCols[0]\n",
    "        location=lineCols[7].split(\":\")\n",
    "        transcriptsEvaluated.append(tid) # append geneID for verification later\n",
    "        coords=[int(x) for x in location[1].split(\"-\")]\n",
    "        annCoords=[int(x) for x in annDict[tid].split(\"-\")]\n",
    "        if not int(annCoords[0])==coords[0] and int(annCoords[1])==coords[1]:\n",
    "            wrongCoords.append(tl[0])\n",
    "        newLoc=location[0]+\":\"+str(annCoords[0])+\"-\"+str(annCoords[1])+\":\"+location[-1]\n",
    "        lineCols[7]=newLoc\n",
    "        outFP.write('\\t'.join(lineCols))\n",
    "inFP.close()\n",
    "outFP.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6', 'CHS.36462.6']\n",
      "323817\n",
      "0\n"
     ]
    }
   ],
   "source": [
    "print(wrongCoords)\n",
    "print(len(transcriptsEvaluated))\n",
    "print(len(ann[~(ann[\"id\"].isin(transcriptsEvaluated))]))"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
