{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import re\n",
    "import sys\n",
    "import csv\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "\n",
    "gff3Cols=[\"seqid\",\"source\",\"type\",\"start\",\"end\",\"score\",\"strand\",\"phase\",\"attributes\"]\n",
    "version=\"2.1\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'geneID', 'comment', 'gene_biotype', 'transcript_id', 'model_evidence', 'Parent', 'exception', 'ID', 'SIMILAR_TO', 'description', 'gbkey', 'TYPE', 'Dbxref', 'end_range', 'STATUS', 'transcript_support_level', 'ncrna_class', 'product', 'gene_status', 'gene_synonym', 'transcript_status', 'protein_id', 'Name', 'level', 'Note', 'gene_type', 'ASSEMBLED', 'havana_gene', 'partial', 'gene_name', 'pseudo', 'ccdsid', 'tag', 'havana_transcript', 'transcript_type', 'start_range'}\n",
      "{'FANTOM_coding_status', 'comment', 'FANTOM_phylocsf_call', 'NR_coding_status', 'Parent', 'ID', 'SIMILAR_TO', 'FANTOM_functional_evidence', 'description', 'FANTOM_ribo_call', 'TYPE', 'SP_coding_status', 'FANTOM_rnacode_call', 'STATUS', 'ASSEMBLED', 'RIBO_coding_status'}\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>TranscriptID</th>\n",
       "      <th>CHESS_gene</th>\n",
       "      <th>Gene_name</th>\n",
       "      <th>Gene_type</th>\n",
       "      <th>Gene_Database</th>\n",
       "      <th>GeneID</th>\n",
       "      <th>Transcript_Database</th>\n",
       "      <th>Location</th>\n",
       "      <th>No_of_exons</th>\n",
       "      <th>Transcript_Length</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CHS.45778.5</td>\n",
       "      <td>CHS.45778</td>\n",
       "      <td>ZSCAN16</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>80345</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>chr6:28124556-28130086:+</td>\n",
       "      <td>4</td>\n",
       "      <td>1336</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CHS.45778.2</td>\n",
       "      <td>CHS.45778</td>\n",
       "      <td>ZSCAN16</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>80345</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>chr6:28124556-28127572:+</td>\n",
       "      <td>3</td>\n",
       "      <td>1330</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CHS.45778.4</td>\n",
       "      <td>CHS.45778</td>\n",
       "      <td>ZSCAN16</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>80345</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>chr6:28124556-28130086:+</td>\n",
       "      <td>4</td>\n",
       "      <td>1394</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CHS.45778.3</td>\n",
       "      <td>CHS.45778</td>\n",
       "      <td>ZSCAN16</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>80345</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>chr6:28124556-28130086:+</td>\n",
       "      <td>4</td>\n",
       "      <td>1386</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CHS.45778.8</td>\n",
       "      <td>CHS.45778</td>\n",
       "      <td>-</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>-</td>\n",
       "      <td>FANTOM</td>\n",
       "      <td>chr6:28124622-28129805:+</td>\n",
       "      <td>3</td>\n",
       "      <td>1724</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  TranscriptID CHESS_gene Gene_name       Gene_type Gene_Database GeneID  \\\n",
       "0  CHS.45778.5  CHS.45778   ZSCAN16  protein_coding        RefSeq  80345   \n",
       "1  CHS.45778.2  CHS.45778   ZSCAN16  protein_coding        RefSeq  80345   \n",
       "2  CHS.45778.4  CHS.45778   ZSCAN16  protein_coding        RefSeq  80345   \n",
       "3  CHS.45778.3  CHS.45778   ZSCAN16  protein_coding        RefSeq  80345   \n",
       "4  CHS.45778.8  CHS.45778         -  protein_coding        RefSeq      -   \n",
       "\n",
       "  Transcript_Database                  Location  No_of_exons  \\\n",
       "0              RefSeq  chr6:28124556-28130086:+            4   \n",
       "1              RefSeq  chr6:28124556-28127572:+            3   \n",
       "2              RefSeq  chr6:28124556-28130086:+            4   \n",
       "3              RefSeq  chr6:28124556-28130086:+            4   \n",
       "4              FANTOM  chr6:28124622-28129805:+            3   \n",
       "\n",
       "   Transcript_Length  \n",
       "0               1336  \n",
       "1               1330  \n",
       "2               1394  \n",
       "3               1386  \n",
       "4               1724  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'lncRNA', 'antisense_RNA', 'misc_RNA', 'protein_coding'}\n",
      "['CHS.12888.1', 'CHS.6037.4', 'CHS.10214.1', 'CHS.12835.6', 'CHS.43009.3', 'CHS.19604.1', 'CHS.36508.8', 'CHS.313.7', 'CHS.57417.4', 'CHS.3341.3']\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqid</th>\n",
       "      <th>source</th>\n",
       "      <th>type</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>phase</th>\n",
       "      <th>attributes</th>\n",
       "      <th>id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>49303</th>\n",
       "      <td>KN196486.1</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>transcript</td>\n",
       "      <td>8758</td>\n",
       "      <td>10915</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=CHS.59946.1;ASSEMBLED=no;Parent=CHS.59946;D...</td>\n",
       "      <td>CHS.59946.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>49304</th>\n",
       "      <td>KN196486.1</td>\n",
       "      <td>Gnomon</td>\n",
       "      <td>transcript</td>\n",
       "      <td>15437</td>\n",
       "      <td>17657</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=CHS.59948.1;ASSEMBLED=no;Parent=CHS.59948;D...</td>\n",
       "      <td>CHS.59948.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>49305</th>\n",
       "      <td>KQ759761.1</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>transcript</td>\n",
       "      <td>934</td>\n",
       "      <td>3090</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=CHS.59954.1;ASSEMBLED=no;Parent=CHS.59954;D...</td>\n",
       "      <td>CHS.59954.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>49306</th>\n",
       "      <td>KQ759762.1</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>transcript</td>\n",
       "      <td>71945</td>\n",
       "      <td>74286</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=CHS.59960.1;ASSEMBLED=no;Parent=CHS.59960;D...</td>\n",
       "      <td>CHS.59960.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>49307</th>\n",
       "      <td>KN196487.1</td>\n",
       "      <td>Gnomon</td>\n",
       "      <td>transcript</td>\n",
       "      <td>31953</td>\n",
       "      <td>33940</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=CHS.59962.1;ASSEMBLED=no;Parent=CHS.59962;D...</td>\n",
       "      <td>CHS.59962.1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            seqid      source        type  start    end score strand phase  \\\n",
       "49303  KN196486.1  BestRefSeq  transcript   8758  10915     .      +     .   \n",
       "49304  KN196486.1      Gnomon  transcript  15437  17657     .      +     .   \n",
       "49305  KQ759761.1  BestRefSeq  transcript    934   3090     .      +     .   \n",
       "49306  KQ759762.1  BestRefSeq  transcript  71945  74286     .      -     .   \n",
       "49307  KN196487.1      Gnomon  transcript  31953  33940     .      -     .   \n",
       "\n",
       "                                              attributes           id  \n",
       "49303  ID=CHS.59946.1;ASSEMBLED=no;Parent=CHS.59946;D...  CHS.59946.1  \n",
       "49304  ID=CHS.59948.1;ASSEMBLED=no;Parent=CHS.59948;D...  CHS.59948.1  \n",
       "49305  ID=CHS.59954.1;ASSEMBLED=no;Parent=CHS.59954;D...  CHS.59954.1  \n",
       "49306  ID=CHS.59960.1;ASSEMBLED=no;Parent=CHS.59960;D...  CHS.59960.1  \n",
       "49307  ID=CHS.59962.1;ASSEMBLED=no;Parent=CHS.59962;D...  CHS.59962.1  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqid</th>\n",
       "      <th>source</th>\n",
       "      <th>type</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>phase</th>\n",
       "      <th>attributes</th>\n",
       "      <th>id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>267463</th>\n",
       "      <td>KQ759761.1</td>\n",
       "      <td>Gnomon</td>\n",
       "      <td>transcript</td>\n",
       "      <td>25184</td>\n",
       "      <td>30339</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=CHS.59956.26;ASSEMBLED=no;Parent=CHS.59956;...</td>\n",
       "      <td>CHS.59956.26</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>267464</th>\n",
       "      <td>KQ759761.1</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>transcript</td>\n",
       "      <td>51431</td>\n",
       "      <td>106747</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=CHS.59958.1;ASSEMBLED=no;Parent=CHS.59958;D...</td>\n",
       "      <td>CHS.59958.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>267465</th>\n",
       "      <td>KQ759761.1</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>transcript</td>\n",
       "      <td>51431</td>\n",
       "      <td>106745</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=CHS.59958.2;ASSEMBLED=no;Parent=CHS.59958;D...</td>\n",
       "      <td>CHS.59958.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>267466</th>\n",
       "      <td>KQ759762.1</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>transcript</td>\n",
       "      <td>10731</td>\n",
       "      <td>69328</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=CHS.59959.1;ASSEMBLED=no;Parent=CHS.59959;D...</td>\n",
       "      <td>CHS.59959.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>267467</th>\n",
       "      <td>KQ759762.1</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>transcript</td>\n",
       "      <td>74341</td>\n",
       "      <td>81410</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=CHS.59961.1;ASSEMBLED=no;Parent=CHS.59961;D...</td>\n",
       "      <td>CHS.59961.1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             seqid      source        type  start     end score strand phase  \\\n",
       "267463  KQ759761.1      Gnomon  transcript  25184   30339     .      +     .   \n",
       "267464  KQ759761.1  BestRefSeq  transcript  51431  106747     .      -     .   \n",
       "267465  KQ759761.1  BestRefSeq  transcript  51431  106745     .      -     .   \n",
       "267466  KQ759762.1  BestRefSeq  transcript  10731   69328     .      +     .   \n",
       "267467  KQ759762.1  BestRefSeq  transcript  74341   81410     .      +     .   \n",
       "\n",
       "                                               attributes            id  \n",
       "267463  ID=CHS.59956.26;ASSEMBLED=no;Parent=CHS.59956;...  CHS.59956.26  \n",
       "267464  ID=CHS.59958.1;ASSEMBLED=no;Parent=CHS.59958;D...   CHS.59958.1  \n",
       "267465  ID=CHS.59958.2;ASSEMBLED=no;Parent=CHS.59958;D...   CHS.59958.2  \n",
       "267466  ID=CHS.59959.1;ASSEMBLED=no;Parent=CHS.59959;D...   CHS.59959.1  \n",
       "267467  ID=CHS.59961.1;ASSEMBLED=no;Parent=CHS.59961;D...   CHS.59961.1  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'FANTOM_coding_status', 'comment', 'geneID', 'gene_biotype', 'NR_coding_status', 'transcript_id', 'model_evidence', 'Parent', 'exception', 'ID', 'SIMILAR_TO', 'description', 'gbkey', 'TYPE', 'Dbxref', 'end_range', 'SP_coding_status', 'STATUS', 'transcript_support_level', 'ncrna_class', 'RIBO_coding_status', 'product', 'FANTOM_phylocsf_call', 'gene_status', 'gene_synonym', 'transcript_status', 'protein_id', 'Name', 'level', 'FANTOM_functional_evidence', 'gene_type', 'ASSEMBLED', 'Note', 'FANTOM_ribo_call', 'havana_gene', 'partial', 'gene_name', 'pseudo', 'FANTOM_rnacode_call', 'ccdsid', 'tag', 'havana_transcript', 'transcript_type', 'start_range'}\n",
      "{'FANTOM_coding_status', 'comment', 'geneID', 'gene_biotype', 'NR_coding_status', 'transcript_id', 'model_evidence', 'Parent', 'exception', 'ID', 'SIMILAR_TO', 'description', 'gbkey', 'TYPE', 'Dbxref', 'end_range', 'SP_coding_status', 'STATUS', 'transcript_support_level', 'ncrna_class', 'RIBO_coding_status', 'product', 'FANTOM_phylocsf_call', 'gene_status', 'gene_synonym', 'transcript_status', 'protein_id', 'Name', 'level', 'FANTOM_functional_evidence', 'gene_type', 'ASSEMBLED', 'Note', 'FANTOM_ribo_call', 'havana_gene', 'partial', 'gene_name', 'pseudo', 'FANTOM_rnacode_call', 'ccdsid', 'tag', 'havana_transcript', 'transcript_type', 'start_range'}\n",
      "set()\n",
      "set()\n",
      "{'gbkey=', 'gene_synonym='}\n",
      "{'gene_type=', 'gene_biotype=', 'TYPE=', 'transcript_type='}\n"
     ]
    }
   ],
   "source": [
    "df=pd.read_csv(\"./chess2.05.gff\",sep=\"\\t\",names=gff3Cols)\n",
    "df=df[df[\"type\"]==\"transcript\"].reset_index(drop=True)\n",
    "df.dropna(inplace=True,axis=0)\n",
    "df.reset_index(inplace=True,drop=True)\n",
    "df[\"start\"]=df[\"start\"].astype(int)\n",
    "df[\"end\"]=df[\"end\"].astype(int)\n",
    "\n",
    "setAttributes_known=set()\n",
    "setAttributes_novel=set()\n",
    "\n",
    "def addToSet_known(row):\n",
    "    global setAttributes_known\n",
    "    setAttributes_known=setAttributes_known.union(set([x.split(\"=\")[0] for x in row.attributes.split(\";\")]))\n",
    "    \n",
    "def addToSet_novel(row):\n",
    "    global setAttributes_novel\n",
    "    setAttributes_novel=setAttributes_novel.union(set([x.split(\"=\")[0] for x in row.attributes.split(\";\")]))\n",
    "\n",
    "df_known=df[~(df[\"source\"].isin([\"StringTie\",\"FANTOM\"]))].reset_index(drop=True)\n",
    "df_novel=df[df[\"source\"].isin([\"StringTie\",\"FANTOM\"])].reset_index(drop=True)\n",
    "df_known.apply(lambda row: addToSet_known(row),axis=1)\n",
    "print(setAttributes_known)\n",
    "df_novel.apply(lambda row: addToSet_novel(row),axis=1)\n",
    "print(setAttributes_novel)\n",
    "\n",
    "# now to investigate the difference between lncRNA and protein-coding transcripts in terms of the attribute composition\n",
    "\n",
    "# first load the transcripts file\n",
    "transcripts=pd.read_csv(\"./chess2.0.transcripts\",sep=\"\\t\")\n",
    "display(transcripts.head())\n",
    "\n",
    "print(set(transcripts[\"Gene_type\"]))\n",
    "setLnc=set()\n",
    "setProt=set()\n",
    "def addTransLnc(row):\n",
    "    global setLnc\n",
    "    for t in row.TranscriptID.split(\",\"):\n",
    "        setLnc.add(t)\n",
    "\n",
    "def addTransProt(row):\n",
    "    global setProt\n",
    "    for t in row.TranscriptID.split(\",\"):\n",
    "        setProt.add(t)\n",
    "\n",
    "transcripts[transcripts[\"Gene_type\"]==\"lncRNA\"].apply(lambda row: addTransLnc(row),axis=1)\n",
    "transcripts[transcripts[\"Gene_type\"]==\"protein_coding\"].apply(lambda row: addTransProt(row),axis=1)\n",
    "print(list(setLnc)[:10])\n",
    "\n",
    "df[\"id\"]=df.attributes.str.split(\"ID=\",expand=True)[1].str.split(\";\",expand=True)[0]\n",
    "df_lnc=df[df['id'].isin(setLnc)].reset_index(drop=True)\n",
    "df_prot=df[df['id'].isin(setProt)].reset_index(drop=True)\n",
    "display(df_lnc.tail())\n",
    "display(df_prot.tail())\n",
    "\n",
    "setAttributes_lnc=set()\n",
    "setAttributes_prot=set()\n",
    "\n",
    "def addToSet_lnc(row):\n",
    "    global setAttributes_lnc\n",
    "    setAttributes_lnc=setAttributes_lnc.union(set([x.split(\"=\")[0] for x in row.attributes.split(\";\")]))\n",
    "    \n",
    "def addToSet_prot(row):\n",
    "    global setAttributes_prot\n",
    "    setAttributes_prot=setAttributes_prot.union(set([x.split(\"=\")[0] for x in row.attributes.split(\";\")]))\n",
    "\n",
    "df_lnc.apply(lambda row: addToSet_lnc(row),axis=1)\n",
    "print(setAttributes_lnc)\n",
    "df_prot.apply(lambda row: addToSet_prot(row),axis=1)\n",
    "print(setAttributes_prot)\n",
    "\n",
    "print(setAttributes_lnc-setAttributes_prot)\n",
    "print(setAttributes_prot-setAttributes_lnc)\n",
    "\n",
    "# investigate where lncRNA appears in attributes\n",
    "setLncRNA_tags=set()\n",
    "def getLncRNATags(row):\n",
    "    global setLncRNA_tags\n",
    "    atts=row.attributes.split(\"ncRNA\")\n",
    "    for a in atts[:-1]:\n",
    "        found=re.findall(r\"\\w+=$\",a)\n",
    "        for f in found:\n",
    "            setLncRNA_tags.add(f)\n",
    "df_lncRNA=df[df[\"attributes\"].str.contains(\"lncRNA\")].reset_index(drop=True)\n",
    "df_lncRNA.apply(lambda row: getLncRNATags(row),axis=1)\n",
    "# df_lncRNA[\"lncRNA_tag\"]=df_lncRNA.attributes.str.split(\"lncRNA\",expand=True)[0].str.extract(\";\\w+=$\",expand=True)[1]\n",
    "print(setLncRNA_tags)\n",
    "\n",
    "# do the same for protein_coding\n",
    "\n",
    "setProt_tags=set()\n",
    "def getProtTags(row):\n",
    "    global setProt_tags\n",
    "    atts=row.attributes.split(\"protein_coding\")\n",
    "    for a in atts[:-1]:\n",
    "        found=re.findall(r\"\\w+=$\",a)\n",
    "        for f in found:\n",
    "            setProt_tags.add(f)\n",
    "df_Prot=df[df[\"attributes\"].str.contains(\"protein_coding\")].reset_index(drop=True)\n",
    "df_Prot.apply(lambda row: getProtTags(row),axis=1)\n",
    "print(setProt_tags)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene_Type</th>\n",
       "      <th>Gene_Name</th>\n",
       "      <th>GFF_ID</th>\n",
       "      <th>Location</th>\n",
       "      <th>Database</th>\n",
       "      <th>RefSeq_GeneID</th>\n",
       "      <th>Description</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>protein_coding</td>\n",
       "      <td>LOC105371921</td>\n",
       "      <td>CHS.23496</td>\n",
       "      <td>chr17:80885776-80892131:-</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>105371921</td>\n",
       "      <td>uncharacterized LOC105371921</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>lncRNA</td>\n",
       "      <td>LOC107986127</td>\n",
       "      <td>CHS.38280</td>\n",
       "      <td>chr3:127744219-127751505:-</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>107986127</td>\n",
       "      <td>uncharacterized LOC107986127</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>protein_coding</td>\n",
       "      <td>RPS11</td>\n",
       "      <td>CHS.27254</td>\n",
       "      <td>chr19:49496365-49499712:+</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>6205</td>\n",
       "      <td>ribosomal protein S11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>protein_coding</td>\n",
       "      <td>CREB3L1</td>\n",
       "      <td>CHS.8579</td>\n",
       "      <td>chr11:46276634-46321430:+</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>90993</td>\n",
       "      <td>cAMP responsive element binding protein 3-like 1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>lncRNA</td>\n",
       "      <td>LOC101930053</td>\n",
       "      <td>CHS.54970</td>\n",
       "      <td>chr9:2385141-2494223:-</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>101930053</td>\n",
       "      <td>uncharacterized LOC101930053</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        Gene_Type     Gene_Name     GFF_ID                    Location  \\\n",
       "0  protein_coding  LOC105371921  CHS.23496   chr17:80885776-80892131:-   \n",
       "1          lncRNA  LOC107986127  CHS.38280  chr3:127744219-127751505:-   \n",
       "2  protein_coding         RPS11  CHS.27254   chr19:49496365-49499712:+   \n",
       "3  protein_coding       CREB3L1   CHS.8579   chr11:46276634-46321430:+   \n",
       "4          lncRNA  LOC101930053  CHS.54970      chr9:2385141-2494223:-   \n",
       "\n",
       "  Database RefSeq_GeneID                                       Description  \n",
       "0   RefSeq     105371921                      uncharacterized LOC105371921  \n",
       "1   RefSeq     107986127                      uncharacterized LOC107986127  \n",
       "2   RefSeq          6205                             ribosomal protein S11  \n",
       "3   RefSeq         90993  cAMP responsive element binding protein 3-like 1  \n",
       "4   RefSeq     101930053                      uncharacterized LOC101930053  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'lncRNA', 'antisense_RNA', 'misc_RNA', 'protein_coding'}\n",
      "['CHS.58729', 'CHS.19424', 'CHS.31101', 'CHS.27079', 'CHS.49695', 'CHS.30986', 'CHS.13150', 'CHS.50753', 'CHS.16710', 'CHS.10799']\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqid</th>\n",
       "      <th>source</th>\n",
       "      <th>type</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>phase</th>\n",
       "      <th>attributes</th>\n",
       "      <th>id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>19499</th>\n",
       "      <td>KN196486.1</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>gene</td>\n",
       "      <td>8758</td>\n",
       "      <td>10915</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=CHS.59946;GENE_TYPE=lncRNA;STATUS=known_ref...</td>\n",
       "      <td>CHS.59946</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19500</th>\n",
       "      <td>KN196486.1</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>gene</td>\n",
       "      <td>15437</td>\n",
       "      <td>17657</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=CHS.59948;GENE_TYPE=lncRNA;STATUS=known_ref...</td>\n",
       "      <td>CHS.59948</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19501</th>\n",
       "      <td>KQ759761.1</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>gene</td>\n",
       "      <td>934</td>\n",
       "      <td>3090</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=CHS.59954;GENE_TYPE=lncRNA;STATUS=known_ref...</td>\n",
       "      <td>CHS.59954</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19502</th>\n",
       "      <td>KQ759762.1</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>gene</td>\n",
       "      <td>71945</td>\n",
       "      <td>74286</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=CHS.59960;GENE_TYPE=lncRNA;STATUS=known_ref...</td>\n",
       "      <td>CHS.59960</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19503</th>\n",
       "      <td>KN196487.1</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>gene</td>\n",
       "      <td>31953</td>\n",
       "      <td>33940</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=CHS.59962;GENE_TYPE=lncRNA;STATUS=known_ref...</td>\n",
       "      <td>CHS.59962</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            seqid  source  type  start    end score strand phase  \\\n",
       "19499  KN196486.1  RefSeq  gene   8758  10915     .      +     .   \n",
       "19500  KN196486.1  RefSeq  gene  15437  17657     .      +     .   \n",
       "19501  KQ759761.1  RefSeq  gene    934   3090     .      +     .   \n",
       "19502  KQ759762.1  RefSeq  gene  71945  74286     .      -     .   \n",
       "19503  KN196487.1  RefSeq  gene  31953  33940     .      -     .   \n",
       "\n",
       "                                              attributes         id  \n",
       "19499  ID=CHS.59946;GENE_TYPE=lncRNA;STATUS=known_ref...  CHS.59946  \n",
       "19500  ID=CHS.59948;GENE_TYPE=lncRNA;STATUS=known_ref...  CHS.59948  \n",
       "19501  ID=CHS.59954;GENE_TYPE=lncRNA;STATUS=known_ref...  CHS.59954  \n",
       "19502  ID=CHS.59960;GENE_TYPE=lncRNA;STATUS=known_ref...  CHS.59960  \n",
       "19503  ID=CHS.59962;GENE_TYPE=lncRNA;STATUS=known_ref...  CHS.59962  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqid</th>\n",
       "      <th>source</th>\n",
       "      <th>type</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>phase</th>\n",
       "      <th>attributes</th>\n",
       "      <th>id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>23829</th>\n",
       "      <td>KQ759761.1</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>gene</td>\n",
       "      <td>4240</td>\n",
       "      <td>8622</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=CHS.59955;GENE_TYPE=protein_coding;STATUS=k...</td>\n",
       "      <td>CHS.59955</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23830</th>\n",
       "      <td>KQ759761.1</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>gene</td>\n",
       "      <td>6604</td>\n",
       "      <td>30680</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=CHS.59956;GENE_TYPE=protein_coding;STATUS=k...</td>\n",
       "      <td>CHS.59956</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23831</th>\n",
       "      <td>KQ759761.1</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>gene</td>\n",
       "      <td>51431</td>\n",
       "      <td>106747</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=CHS.59958;GENE_TYPE=protein_coding;STATUS=k...</td>\n",
       "      <td>CHS.59958</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23832</th>\n",
       "      <td>KQ759762.1</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>gene</td>\n",
       "      <td>10731</td>\n",
       "      <td>69328</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=CHS.59959;GENE_TYPE=protein_coding;STATUS=k...</td>\n",
       "      <td>CHS.59959</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23833</th>\n",
       "      <td>KQ759762.1</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>gene</td>\n",
       "      <td>74341</td>\n",
       "      <td>81410</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=CHS.59961;GENE_TYPE=protein_coding;STATUS=k...</td>\n",
       "      <td>CHS.59961</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            seqid  source  type  start     end score strand phase  \\\n",
       "23829  KQ759761.1  RefSeq  gene   4240    8622     .      -     .   \n",
       "23830  KQ759761.1  RefSeq  gene   6604   30680     .      +     .   \n",
       "23831  KQ759761.1  RefSeq  gene  51431  106747     .      -     .   \n",
       "23832  KQ759762.1  RefSeq  gene  10731   69328     .      +     .   \n",
       "23833  KQ759762.1  RefSeq  gene  74341   81410     .      +     .   \n",
       "\n",
       "                                              attributes         id  \n",
       "23829  ID=CHS.59955;GENE_TYPE=protein_coding;STATUS=k...  CHS.59955  \n",
       "23830  ID=CHS.59956;GENE_TYPE=protein_coding;STATUS=k...  CHS.59956  \n",
       "23831  ID=CHS.59958;GENE_TYPE=protein_coding;STATUS=k...  CHS.59958  \n",
       "23832  ID=CHS.59959;GENE_TYPE=protein_coding;STATUS=k...  CHS.59959  \n",
       "23833  ID=CHS.59961;GENE_TYPE=protein_coding;STATUS=k...  CHS.59961  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'GENE_NAME', 'STATUS', 'ID', 'GENE_TYPE', 'description', 'product', 'GENCODE_GENE_NAME'}\n",
      "{'GENE_NAME', 'STATUS', 'ID', 'GENE_TYPE', 'description', 'GENCODE_GENE_NAME'}\n",
      "{'product'}\n",
      "set()\n",
      "{'description=', 'GENE_TYPE='}\n",
      "{'GENE_TYPE='}\n"
     ]
    }
   ],
   "source": [
    "# now to investigate attributes of protein coding genes vs lncRNA\n",
    "df=pd.read_csv(\"./chess2.05.gff\",sep=\"\\t\",names=gff3Cols)\n",
    "df=df[df[\"type\"]==\"gene\"].reset_index(drop=True)\n",
    "df.dropna(inplace=True,axis=0)\n",
    "df.reset_index(inplace=True,drop=True)\n",
    "df[\"start\"]=df[\"start\"].astype(int)\n",
    "df[\"end\"]=df[\"end\"].astype(int)\n",
    "\n",
    "genes=pd.read_csv(\"./chess2.0.genes\",sep=\"\\t\")\n",
    "display(genes.head())\n",
    "\n",
    "print(set(genes[\"Gene_Type\"]))\n",
    "setLnc=set()\n",
    "setProt=set()\n",
    "def addGenesLnc(row):\n",
    "    global setLnc\n",
    "    for t in row.GFF_ID.split(\",\"):\n",
    "        setLnc.add(t)\n",
    "\n",
    "def addGenesProt(row):\n",
    "    global setProt\n",
    "    for t in row.GFF_ID.split(\",\"):\n",
    "        setProt.add(t)\n",
    "\n",
    "genes[genes[\"Gene_Type\"]==\"lncRNA\"].apply(lambda row: addGenesLnc(row),axis=1)\n",
    "genes[genes[\"Gene_Type\"]==\"protein_coding\"].apply(lambda row: addGenesProt(row),axis=1)\n",
    "print(list(setLnc)[:10])\n",
    "\n",
    "df[\"id\"]=df.attributes.str.split(\"ID=\",expand=True)[1].str.split(\";\",expand=True)[0]\n",
    "df_lnc=df[df['id'].isin(setLnc)].reset_index(drop=True)\n",
    "df_prot=df[df['id'].isin(setProt)].reset_index(drop=True)\n",
    "display(df_lnc.tail())\n",
    "display(df_prot.tail())\n",
    "\n",
    "setAttributes_lnc=set()\n",
    "setAttributes_prot=set()\n",
    "\n",
    "def addToSet_lnc(row):\n",
    "    global setAttributes_lnc\n",
    "    setAttributes_lnc=setAttributes_lnc.union(set([x.split(\"=\")[0] for x in row.attributes.split(\";\")]))\n",
    "    \n",
    "def addToSet_prot(row):\n",
    "    global setAttributes_prot\n",
    "    setAttributes_prot=setAttributes_prot.union(set([x.split(\"=\")[0] for x in row.attributes.split(\";\")]))\n",
    "\n",
    "df_lnc.apply(lambda row: addToSet_lnc(row),axis=1)\n",
    "print(setAttributes_lnc)\n",
    "df_prot.apply(lambda row: addToSet_prot(row),axis=1)\n",
    "print(setAttributes_prot)\n",
    "\n",
    "print(setAttributes_lnc-setAttributes_prot)\n",
    "print(setAttributes_prot-setAttributes_lnc)\n",
    "\n",
    "# investigate where lncRNA appears in attributes\n",
    "setLncRNA_tags=set()\n",
    "def getLncRNATags(row):\n",
    "    global setLncRNA_tags\n",
    "    atts=row.attributes.split(\"lncRNA\")\n",
    "    for a in atts[:-1]:\n",
    "        found=re.findall(r\"\\w+=$\",a)\n",
    "        for f in found:\n",
    "            setLncRNA_tags.add(f)\n",
    "df_lncRNA=df[df[\"attributes\"].str.contains(\"lncRNA\")].reset_index(drop=True)\n",
    "df_lncRNA.apply(lambda row: getLncRNATags(row),axis=1)\n",
    "# df_lncRNA[\"lncRNA_tag\"]=df_lncRNA.attributes.str.split(\"lncRNA\",expand=True)[0].str.extract(\";\\w+=$\",expand=True)[1]\n",
    "print(setLncRNA_tags)\n",
    "\n",
    "# do the same for protein_coding\n",
    "\n",
    "setProt_tags=set()\n",
    "def getProtTags(row):\n",
    "    global setProt_tags\n",
    "    atts=row.attributes.split(\"protein_coding\")\n",
    "    for a in atts[:-1]:\n",
    "        found=re.findall(r\"\\w+=$\",a)\n",
    "        for f in found:\n",
    "            setProt_tags.add(f)\n",
    "df_Prot=df[df[\"attributes\"].str.contains(\"protein_coding\")].reset_index(drop=True)\n",
    "df_Prot.apply(lambda row: getProtTags(row),axis=1)\n",
    "print(setProt_tags)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Status</th>\n",
       "      <th>Gene_ID</th>\n",
       "      <th>Location</th>\n",
       "      <th>NR_blast_Eval</th>\n",
       "      <th>NR_hit</th>\n",
       "      <th>SP_blast_Eval</th>\n",
       "      <th>SP_hit</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>discard</td>\n",
       "      <td>CHS.10012</td>\n",
       "      <td>chr11:113875694-113899029:+</td>\n",
       "      <td>1e-17</td>\n",
       "      <td>CAE91339.1, unnamed protein product [Homo sapi...</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>non-coding</td>\n",
       "      <td>CHS.10056</td>\n",
       "      <td>chr11:116307531-116308521:+</td>\n",
       "      <td>3e-22</td>\n",
       "      <td>CAE89781.1, unnamed protein product [Homo sapi...</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>non-coding</td>\n",
       "      <td>CHS.1008</td>\n",
       "      <td>chr1:29124945-29139372:+</td>\n",
       "      <td>4e-19</td>\n",
       "      <td>EAW78405.1, hCG2021310 [Homo sapiens]</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>non-coding</td>\n",
       "      <td>CHS.10158</td>\n",
       "      <td>chr11:119057437-119059239:+</td>\n",
       "      <td>4e-81</td>\n",
       "      <td>BAC85254.1, unnamed protein product [Homo sapi...</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>discard</td>\n",
       "      <td>CHS.10242</td>\n",
       "      <td>chr11:121958716-121961024:+</td>\n",
       "      <td>2e-16</td>\n",
       "      <td>AAL23680.1, non-small cell lung cancer RimL3b ...</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       Status    Gene_ID                     Location NR_blast_Eval  \\\n",
       "0     discard  CHS.10012  chr11:113875694-113899029:+         1e-17   \n",
       "1  non-coding  CHS.10056  chr11:116307531-116308521:+         3e-22   \n",
       "2  non-coding   CHS.1008     chr1:29124945-29139372:+         4e-19   \n",
       "3  non-coding  CHS.10158  chr11:119057437-119059239:+         4e-81   \n",
       "4     discard  CHS.10242  chr11:121958716-121961024:+         2e-16   \n",
       "\n",
       "                                              NR_hit SP_blast_Eval SP_hit  \n",
       "0  CAE91339.1, unnamed protein product [Homo sapi...             -      -  \n",
       "1  CAE89781.1, unnamed protein product [Homo sapi...             -      -  \n",
       "2              EAW78405.1, hCG2021310 [Homo sapiens]             -      -  \n",
       "3  BAC85254.1, unnamed protein product [Homo sapi...             -      -  \n",
       "4  AAL23680.1, non-small cell lung cancer RimL3b ...             -      -  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Status</th>\n",
       "      <th>Transcript_ID</th>\n",
       "      <th>Gene_ID</th>\n",
       "      <th>Location</th>\n",
       "      <th>Transcript_Length</th>\n",
       "      <th>No_of_exons</th>\n",
       "      <th>No_of_samples</th>\n",
       "      <th>Average_TPM</th>\n",
       "      <th>Maximum_TPM</th>\n",
       "      <th>NR_Eval</th>\n",
       "      <th>...</th>\n",
       "      <th>SP_Eval</th>\n",
       "      <th>SP_hit</th>\n",
       "      <th>CDD_Eval</th>\n",
       "      <th>Pfam_Eval</th>\n",
       "      <th>CDSlen</th>\n",
       "      <th>CDSreps</th>\n",
       "      <th>CDSrepeatcov</th>\n",
       "      <th>ALU_cdcov</th>\n",
       "      <th>LINE_cdcov</th>\n",
       "      <th>GencodeCall</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>coding</td>\n",
       "      <td>CHS.55578.5</td>\n",
       "      <td>CHS.55578</td>\n",
       "      <td>chr9:61672749-61674874:-</td>\n",
       "      <td>1705</td>\n",
       "      <td>2</td>\n",
       "      <td>4</td>\n",
       "      <td>1.195119</td>\n",
       "      <td>1.687937</td>\n",
       "      <td>6e-63</td>\n",
       "      <td>...</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>0.00837726</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>non-coding</td>\n",
       "      <td>CHS.20892.1</td>\n",
       "      <td>CHS.20892</td>\n",
       "      <td>chr17:722531-732316:+</td>\n",
       "      <td>9786</td>\n",
       "      <td>1</td>\n",
       "      <td>1354</td>\n",
       "      <td>3.256216</td>\n",
       "      <td>22.569984</td>\n",
       "      <td>9e-18</td>\n",
       "      <td>...</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>333</td>\n",
       "      <td>Simple_repeat:49,SINE|Alu:248</td>\n",
       "      <td>89.2</td>\n",
       "      <td>74.5</td>\n",
       "      <td>0</td>\n",
       "      <td>SINE/Alu;Simple_repeat</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>non-coding</td>\n",
       "      <td>CHS.24098.4</td>\n",
       "      <td>CHS.24098</td>\n",
       "      <td>chr18:8974487-8998422:+</td>\n",
       "      <td>2451</td>\n",
       "      <td>3</td>\n",
       "      <td>187</td>\n",
       "      <td>5.632762</td>\n",
       "      <td>28.178564</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>SINE/Alu</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>non-coding</td>\n",
       "      <td>CHS.24098.3</td>\n",
       "      <td>CHS.24098</td>\n",
       "      <td>chr18:8974487-8980729:+</td>\n",
       "      <td>436</td>\n",
       "      <td>2</td>\n",
       "      <td>59</td>\n",
       "      <td>3.200027</td>\n",
       "      <td>9.097818</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>SINE/Alu</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>non-coding</td>\n",
       "      <td>CHS.24098.2</td>\n",
       "      <td>CHS.24098</td>\n",
       "      <td>chr18:8974487-8979364:+</td>\n",
       "      <td>454</td>\n",
       "      <td>2</td>\n",
       "      <td>17</td>\n",
       "      <td>2.209827</td>\n",
       "      <td>3.991817</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>SINE/Alu</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 21 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       Status Transcript_ID    Gene_ID                  Location  \\\n",
       "0      coding   CHS.55578.5  CHS.55578  chr9:61672749-61674874:-   \n",
       "1  non-coding   CHS.20892.1  CHS.20892     chr17:722531-732316:+   \n",
       "2  non-coding   CHS.24098.4  CHS.24098   chr18:8974487-8998422:+   \n",
       "3  non-coding   CHS.24098.3  CHS.24098   chr18:8974487-8980729:+   \n",
       "4  non-coding   CHS.24098.2  CHS.24098   chr18:8974487-8979364:+   \n",
       "\n",
       "   Transcript_Length  No_of_exons  No_of_samples  Average_TPM  Maximum_TPM  \\\n",
       "0               1705            2              4     1.195119     1.687937   \n",
       "1               9786            1           1354     3.256216    22.569984   \n",
       "2               2451            3            187     5.632762    28.178564   \n",
       "3                436            2             59     3.200027     9.097818   \n",
       "4                454            2             17     2.209827     3.991817   \n",
       "\n",
       "  NR_Eval           ...           SP_Eval SP_hit    CDD_Eval Pfam_Eval CDSlen  \\\n",
       "0   6e-63           ...                 -      -  0.00837726         -      -   \n",
       "1   9e-18           ...                 -      -           -         -    333   \n",
       "2       -           ...                 -      -           -         -      -   \n",
       "3       -           ...                 -      -           -         -      -   \n",
       "4       -           ...                 -      -           -         -      -   \n",
       "\n",
       "                         CDSreps CDSrepeatcov ALU_cdcov LINE_cdcov  \\\n",
       "0                              -            -         -          -   \n",
       "1  Simple_repeat:49,SINE|Alu:248         89.2      74.5          0   \n",
       "2                              -            -         -          -   \n",
       "3                              -            -         -          -   \n",
       "4                              -            -         -          -   \n",
       "\n",
       "              GencodeCall  \n",
       "0                       -  \n",
       "1  SINE/Alu;Simple_repeat  \n",
       "2                SINE/Alu  \n",
       "3                SINE/Alu  \n",
       "4                SINE/Alu  \n",
       "\n",
       "[5 rows x 21 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "genesDF=pd.read_csv(\"./chess2.0.novel_protein.genes.conservative.txt\",sep=\"\\t\")\n",
    "transDF=pd.read_csv(\"./chess2.0.novel_protein.transcripts.conservative.txt\",sep=\"\\t\")\n",
    "display(genesDF.head())\n",
    "display(transDF.head())\n",
    "\n",
    "genesDiscard=set(genesDF[genesDF['Status']=='discard']['Gene_ID']) # set of all geneIDs that need to be removed from the annotation\n",
    "genesC=set(genesDF[genesDF['Status']=='coding']['Gene_ID']) # set of all coding genes - used to verify that they all are annotated as coding and that they all have corresponding CDS\n",
    "genesNC=set(genesDF[genesDF['Status']=='non-coding']['Gene_ID']) # set of all geneIDs that need to be labeled as lncRNA\n",
    "\n",
    "transDiscard=set(transDF[transDF['Status']=='discard']['Transcript_ID'])\n",
    "transC=set(transDF[transDF['Status']=='coding']['Transcript_ID'])\n",
    "transNC=set(transDF[transDF['Status']=='non-coding']['Transcript_ID'])\n",
    "\n",
    "# need to make sure the geneIDS extracted from transDF of discard, NC and C match the corresponding ones from genesDF\n",
    "assert(len(genesDiscard.symmetric_difference(set([\"CHS.\"+t.split(\".\")[1] for t in transDiscard])))==0),\"wrong discard\"\n",
    "assert(len(genesC.symmetric_difference(set([\"CHS.\"+t.split(\".\")[1] for t in transC])))==0),\"wrong coding\"\n",
    "assert(len(genesNC.symmetric_difference(set([\"CHS.\"+t.split(\".\")[1] for t in transNC])))==0),\"wrong non-coding\"\n",
    "\n",
    "outFP=open(\"./chess2.1.gff\",\"w+\")\n",
    "\n",
    "with open(\"./chess2.05.gff\",\"r\") as inFP:\n",
    "    for line in inFP.readlines():\n",
    "        if line[0]==\"#\":\n",
    "            if line[:7]==\"##CHESS\":\n",
    "                outFP.write(\"##CHESS\"+version+\"\\n\")\n",
    "            else:\n",
    "                outFP.write(line)\n",
    "            continue\n",
    "        else:\n",
    "            lineCols=line.split(\"\\t\")\n",
    "            if lineCols[2]==\"gene\":\n",
    "                attTmp=lineCols[8].rstrip(\"\\n\").split(\";\")\n",
    "                attributes=dict([tuple(x.split(\"=\")) for x in attTmp])\n",
    "                attOrder=[x.split(\"=\")[0] for x in attTmp]\n",
    "                gid=attributes[\"ID\"]\n",
    "                if gid in genesDiscard: # do not write since gene is the targeted for discarding\n",
    "                    continue\n",
    "                elif gid in genesNC: # need to alter the attributes in a way that makes sense\n",
    "                    attributes[\"GENE_TYPE\"]=\"lncRNA\"\n",
    "                    newLine=\"\"\n",
    "                    for att in attOrder:\n",
    "                        newLine+=att+\"=\"+attributes[att]+\";\"\n",
    "                    newLine=newLine[:-1] # remove trailing ';'\n",
    "                    outFP.write(\"\\t\".join(lineCols[:-1]+[newLine+\"\\n\"]))\n",
    "                    continue\n",
    "                else:\n",
    "                    if attributes[\"GENE_TYPE\"]==\"protein_coding\" and attributes[\"STATUS\"]==\"novel\": # coding - make sure it is in the genesC set\n",
    "                        assert(gid in genesC),\"coding protein found; not in genesC\"\n",
    "                    outFP.write(line)\n",
    "                    continue\n",
    "            elif lineCols[2]==\"transcript\":\n",
    "                attTmp=lineCols[8].rstrip(\"\\n\").split(\";\")\n",
    "                attributes=dict([tuple(x.split(\"=\")) for x in attTmp])\n",
    "                attOrder=[x.split(\"=\")[0] for x in attTmp]\n",
    "                tid=attributes[\"ID\"]\n",
    "                if tid in transDiscard: # do not write since gene is the targeted for discarding\n",
    "                    assert(\"CHS.\"+tid.split(\".\")[1] in genesDiscard),\"wrong transcript marked to be discarded\"\n",
    "                    continue\n",
    "                elif tid in transNC: # need to alter the attributes in a way that makes sense\n",
    "                    assert(\"CHS.\"+tid.split(\".\")[1] in genesNC),\"wrong transcript marked to be changed to protein-coding\"\n",
    "                    if \"GENE_TYPE\" in attributes:\n",
    "                        attributes.pop(\"GENE_TYPE\",None)\n",
    "                    if \"gbkey\" in attributes:\n",
    "                        attributes[\"gbkey\"]=\"lncRNA\"\n",
    "                    if \"gene_biotype\" in attributes:\n",
    "                        attributes[\"gene_biotype\"]=\"lncRNA\"\n",
    "                    if \"transcript_type\" in attributes:\n",
    "                        attributes.pop('transcript_type',None)\n",
    "                    if \"gene_type\" in attributes:\n",
    "                        attributes.pop('gene_type',None)\n",
    "                    attributes[\"TYPE\"]=\"non_coding\"\n",
    "                    newLine=\"\"\n",
    "                    for att in attOrder:\n",
    "                        newLine+=att+\"=\"+attributes[att]+\";\"\n",
    "                    newLine=newLine[:-1] # remove trailing ';'\n",
    "                    outFP.write(\"\\t\".join(lineCols[:-1]+[newLine+\"\\n\"]))\n",
    "                    continue\n",
    "                else:\n",
    "                    outFP.write(line)\n",
    "                    continue\n",
    "            elif lineCols[2]==\"exon\":\n",
    "                attTmp=lineCols[8].rstrip(\"\\n\").split(\";\")\n",
    "                attributes=dict([tuple(x.split(\"=\")) for x in attTmp])\n",
    "                attOrder=[x.split(\"=\")[0] for x in attTmp]\n",
    "                tid=attributes[\"Parent\"]\n",
    "                if tid in transDiscard: # do not write since gene is the targeted for discarding\n",
    "                    continue\n",
    "                else:\n",
    "                    outFP.write(line)\n",
    "                    continue\n",
    "            elif lineCols[2]==\"CDS\":\n",
    "                attTmp=lineCols[8].rstrip(\"\\n\").split(\";\")\n",
    "                attributes=dict([tuple(x.split(\"=\")) for x in attTmp])\n",
    "                attOrder=[x.split(\"=\")[0] for x in attTmp]\n",
    "                tid=attributes[\"Parent\"]\n",
    "                if tid in transDiscard or tid in transNC: # do not write since gene is the targeted for discarding\n",
    "                    continue\n",
    "                else:\n",
    "                    outFP.write(line)\n",
    "                    continue\n",
    "            else:\n",
    "                print('something is wrong with the type of a record: '+line)\n",
    "                \n",
    "\n",
    "inFP.close()\n",
    "outFP.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Status</th>\n",
       "      <th>Gene_ID</th>\n",
       "      <th>Location</th>\n",
       "      <th>NR_blast_Eval</th>\n",
       "      <th>NR_hit</th>\n",
       "      <th>SP_blast_Eval</th>\n",
       "      <th>SP_hit</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>discard</td>\n",
       "      <td>CHS.10012</td>\n",
       "      <td>chr11:113875694-113899029:+</td>\n",
       "      <td>1e-17</td>\n",
       "      <td>CAE91339.1, unnamed protein product [Homo sapi...</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>non-coding</td>\n",
       "      <td>CHS.10056</td>\n",
       "      <td>chr11:116307531-116308521:+</td>\n",
       "      <td>3e-22</td>\n",
       "      <td>CAE89781.1, unnamed protein product [Homo sapi...</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>non-coding</td>\n",
       "      <td>CHS.1008</td>\n",
       "      <td>chr1:29124945-29139372:+</td>\n",
       "      <td>4e-19</td>\n",
       "      <td>EAW78405.1, hCG2021310 [Homo sapiens]</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>non-coding</td>\n",
       "      <td>CHS.10158</td>\n",
       "      <td>chr11:119057437-119059239:+</td>\n",
       "      <td>4e-81</td>\n",
       "      <td>BAC85254.1, unnamed protein product [Homo sapi...</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>discard</td>\n",
       "      <td>CHS.10242</td>\n",
       "      <td>chr11:121958716-121961024:+</td>\n",
       "      <td>2e-16</td>\n",
       "      <td>AAL23680.1, non-small cell lung cancer RimL3b ...</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       Status    Gene_ID                     Location NR_blast_Eval  \\\n",
       "0     discard  CHS.10012  chr11:113875694-113899029:+         1e-17   \n",
       "1  non-coding  CHS.10056  chr11:116307531-116308521:+         3e-22   \n",
       "2  non-coding   CHS.1008     chr1:29124945-29139372:+         4e-19   \n",
       "3  non-coding  CHS.10158  chr11:119057437-119059239:+         4e-81   \n",
       "4     discard  CHS.10242  chr11:121958716-121961024:+         2e-16   \n",
       "\n",
       "                                              NR_hit SP_blast_Eval SP_hit  \n",
       "0  CAE91339.1, unnamed protein product [Homo sapi...             -      -  \n",
       "1  CAE89781.1, unnamed protein product [Homo sapi...             -      -  \n",
       "2              EAW78405.1, hCG2021310 [Homo sapiens]             -      -  \n",
       "3  BAC85254.1, unnamed protein product [Homo sapi...             -      -  \n",
       "4  AAL23680.1, non-small cell lung cancer RimL3b ...             -      -  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Status</th>\n",
       "      <th>Transcript_ID</th>\n",
       "      <th>Gene_ID</th>\n",
       "      <th>Location</th>\n",
       "      <th>Transcript_Length</th>\n",
       "      <th>No_of_exons</th>\n",
       "      <th>No_of_samples</th>\n",
       "      <th>Average_TPM</th>\n",
       "      <th>Maximum_TPM</th>\n",
       "      <th>NR_Eval</th>\n",
       "      <th>...</th>\n",
       "      <th>SP_Eval</th>\n",
       "      <th>SP_hit</th>\n",
       "      <th>CDD_Eval</th>\n",
       "      <th>Pfam_Eval</th>\n",
       "      <th>CDSlen</th>\n",
       "      <th>CDSreps</th>\n",
       "      <th>CDSrepeatcov</th>\n",
       "      <th>ALU_cdcov</th>\n",
       "      <th>LINE_cdcov</th>\n",
       "      <th>GencodeCall</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>coding</td>\n",
       "      <td>CHS.55578.5</td>\n",
       "      <td>CHS.55578</td>\n",
       "      <td>chr9:61672749-61674874:-</td>\n",
       "      <td>1705</td>\n",
       "      <td>2</td>\n",
       "      <td>4</td>\n",
       "      <td>1.195119</td>\n",
       "      <td>1.687937</td>\n",
       "      <td>6e-63</td>\n",
       "      <td>...</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>0.00837726</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>non-coding</td>\n",
       "      <td>CHS.20892.1</td>\n",
       "      <td>CHS.20892</td>\n",
       "      <td>chr17:722531-732316:+</td>\n",
       "      <td>9786</td>\n",
       "      <td>1</td>\n",
       "      <td>1354</td>\n",
       "      <td>3.256216</td>\n",
       "      <td>22.569984</td>\n",
       "      <td>9e-18</td>\n",
       "      <td>...</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>333</td>\n",
       "      <td>Simple_repeat:49,SINE|Alu:248</td>\n",
       "      <td>89.2</td>\n",
       "      <td>74.5</td>\n",
       "      <td>0</td>\n",
       "      <td>SINE/Alu;Simple_repeat</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>non-coding</td>\n",
       "      <td>CHS.24098.4</td>\n",
       "      <td>CHS.24098</td>\n",
       "      <td>chr18:8974487-8998422:+</td>\n",
       "      <td>2451</td>\n",
       "      <td>3</td>\n",
       "      <td>187</td>\n",
       "      <td>5.632762</td>\n",
       "      <td>28.178564</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>SINE/Alu</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>non-coding</td>\n",
       "      <td>CHS.24098.3</td>\n",
       "      <td>CHS.24098</td>\n",
       "      <td>chr18:8974487-8980729:+</td>\n",
       "      <td>436</td>\n",
       "      <td>2</td>\n",
       "      <td>59</td>\n",
       "      <td>3.200027</td>\n",
       "      <td>9.097818</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>SINE/Alu</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>non-coding</td>\n",
       "      <td>CHS.24098.2</td>\n",
       "      <td>CHS.24098</td>\n",
       "      <td>chr18:8974487-8979364:+</td>\n",
       "      <td>454</td>\n",
       "      <td>2</td>\n",
       "      <td>17</td>\n",
       "      <td>2.209827</td>\n",
       "      <td>3.991817</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>SINE/Alu</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 21 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       Status Transcript_ID    Gene_ID                  Location  \\\n",
       "0      coding   CHS.55578.5  CHS.55578  chr9:61672749-61674874:-   \n",
       "1  non-coding   CHS.20892.1  CHS.20892     chr17:722531-732316:+   \n",
       "2  non-coding   CHS.24098.4  CHS.24098   chr18:8974487-8998422:+   \n",
       "3  non-coding   CHS.24098.3  CHS.24098   chr18:8974487-8980729:+   \n",
       "4  non-coding   CHS.24098.2  CHS.24098   chr18:8974487-8979364:+   \n",
       "\n",
       "   Transcript_Length  No_of_exons  No_of_samples  Average_TPM  Maximum_TPM  \\\n",
       "0               1705            2              4     1.195119     1.687937   \n",
       "1               9786            1           1354     3.256216    22.569984   \n",
       "2               2451            3            187     5.632762    28.178564   \n",
       "3                436            2             59     3.200027     9.097818   \n",
       "4                454            2             17     2.209827     3.991817   \n",
       "\n",
       "  NR_Eval           ...           SP_Eval SP_hit    CDD_Eval Pfam_Eval CDSlen  \\\n",
       "0   6e-63           ...                 -      -  0.00837726         -      -   \n",
       "1   9e-18           ...                 -      -           -         -    333   \n",
       "2       -           ...                 -      -           -         -      -   \n",
       "3       -           ...                 -      -           -         -      -   \n",
       "4       -           ...                 -      -           -         -      -   \n",
       "\n",
       "                         CDSreps CDSrepeatcov ALU_cdcov LINE_cdcov  \\\n",
       "0                              -            -         -          -   \n",
       "1  Simple_repeat:49,SINE|Alu:248         89.2      74.5          0   \n",
       "2                              -            -         -          -   \n",
       "3                              -            -         -          -   \n",
       "4                              -            -         -          -   \n",
       "\n",
       "              GencodeCall  \n",
       "0                       -  \n",
       "1  SINE/Alu;Simple_repeat  \n",
       "2                SINE/Alu  \n",
       "3                SINE/Alu  \n",
       "4                SINE/Alu  \n",
       "\n",
       "[5 rows x 21 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "genesDF=pd.read_csv(\"./chess2.0.novel_protein.genes.conservative.txt\",sep=\"\\t\")\n",
    "transDF=pd.read_csv(\"./chess2.0.novel_protein.transcripts.conservative.txt\",sep=\"\\t\")\n",
    "display(genesDF.head())\n",
    "display(transDF.head())\n",
    "\n",
    "genesDiscard=set(genesDF[genesDF['Status']=='discard']['Gene_ID']) # set of all geneIDs that need to be removed from the annotation\n",
    "genesC=set(genesDF[genesDF['Status']=='coding']['Gene_ID']) # set of all coding genes - used to verify that they all are annotated as coding and that they all have corresponding CDS\n",
    "genesNC=set(genesDF[genesDF['Status']=='non-coding']['Gene_ID']) # set of all geneIDs that need to be labeled as lncRNA\n",
    "\n",
    "transDiscard=set(transDF[transDF['Status']=='discard']['Transcript_ID'])\n",
    "transC=set(transDF[transDF['Status']=='coding']['Transcript_ID'])\n",
    "transNC=set(transDF[transDF['Status']=='non-coding']['Transcript_ID'])\n",
    "\n",
    "# need to make sure the geneIDS extracted from transDF of discard, NC and C match the corresponding ones from genesDF\n",
    "assert(len(genesDiscard.symmetric_difference(set([\"CHS.\"+t.split(\".\")[1] for t in transDiscard])))==0),\"wrong discard\"\n",
    "assert(len(genesC.symmetric_difference(set([\"CHS.\"+t.split(\".\")[1] for t in transC])))==0),\"wrong coding\"\n",
    "assert(len(genesNC.symmetric_difference(set([\"CHS.\"+t.split(\".\")[1] for t in transNC])))==0),\"wrong non-coding\"\n",
    "\n",
    "outFP=open(\"./chess2.1_and_refseq.gff\",\"w+\")\n",
    "\n",
    "version=\"2.1\"\n",
    "\n",
    "with open(\"./chess2.05_and_refseq.gff\",\"r\") as inFP:\n",
    "    for line in inFP.readlines():\n",
    "        if line[0]==\"#\":\n",
    "            if line[:7]==\"##CHESS\":\n",
    "                outFP.write(\"##CHESS\"+version+\"\\n\")\n",
    "            else:\n",
    "                outFP.write(line)\n",
    "            continue\n",
    "        else:\n",
    "            lineCols=line.split(\"\\t\")\n",
    "            if lineCols[2]==\"gene\":\n",
    "                attTmp=lineCols[8].rstrip(\"\\n\").split(\";\")\n",
    "                attributes=dict([tuple(x.split(\"=\")) for x in attTmp])\n",
    "                attOrder=[x.split(\"=\")[0] for x in attTmp]\n",
    "                gid=attributes[\"ID\"]\n",
    "                if gid in genesDiscard: # do not write since gene is the targeted for discarding\n",
    "                    continue\n",
    "                elif gid in genesNC: # need to alter the attributes in a way that makes sense\n",
    "                    attributes[\"GENE_TYPE\"]=\"lncRNA\"\n",
    "                    newLine=\"\"\n",
    "                    for att in attOrder:\n",
    "                        newLine+=att+\"=\"+attributes[att]+\";\"\n",
    "                    newLine=newLine[:-1] # remove trailing ';'\n",
    "                    outFP.write(\"\\t\".join(lineCols[:-1]+[newLine+\"\\n\"]))\n",
    "                    continue\n",
    "                else:\n",
    "                    if attributes[\"GENE_TYPE\"]==\"protein_coding\" and attributes[\"STATUS\"]==\"novel\": # coding - make sure it is in the genesC set\n",
    "                        assert(gid in genesC),\"coding protein found; not in genesC\"\n",
    "                    outFP.write(line)\n",
    "                    continue\n",
    "            elif lineCols[2]==\"transcript\":\n",
    "                attTmp=lineCols[8].rstrip(\"\\n\").split(\";\")\n",
    "                attributes=dict([tuple(x.split(\"=\")) for x in attTmp])\n",
    "                attOrder=[x.split(\"=\")[0] for x in attTmp]\n",
    "                tid=attributes[\"ID\"]\n",
    "                if tid in transDiscard: # do not write since gene is the targeted for discarding\n",
    "                    assert(\"CHS.\"+tid.split(\".\")[1] in genesDiscard),\"wrong transcript marked to be discarded\"\n",
    "                    continue\n",
    "                elif tid in transNC: # need to alter the attributes in a way that makes sense\n",
    "                    assert(\"CHS.\"+tid.split(\".\")[1] in genesNC),\"wrong transcript marked to be changed to protein-coding\"\n",
    "                    if \"GENE_TYPE\" in attributes:\n",
    "                        attributes.pop(\"GENE_TYPE\",None)\n",
    "                    if \"gbkey\" in attributes:\n",
    "                        attributes[\"gbkey\"]=\"lncRNA\"\n",
    "                    if \"gene_biotype\" in attributes:\n",
    "                        attributes[\"gene_biotype\"]=\"lncRNA\"\n",
    "                    if \"transcript_type\" in attributes:\n",
    "                        attributes.pop('transcript_type',None)\n",
    "                    if \"gene_type\" in attributes:\n",
    "                        attributes.pop('gene_type',None)\n",
    "                    attributes[\"TYPE\"]=\"non_coding\"\n",
    "                    newLine=\"\"\n",
    "                    for att in attOrder:\n",
    "                        newLine+=att+\"=\"+attributes[att]+\";\"\n",
    "                    newLine=newLine[:-1] # remove trailing ';'\n",
    "                    outFP.write(\"\\t\".join(lineCols[:-1]+[newLine+\"\\n\"]))\n",
    "                    continue\n",
    "                else:\n",
    "                    outFP.write(line)\n",
    "                    continue\n",
    "            elif lineCols[2]==\"exon\":\n",
    "                attTmp=lineCols[8].rstrip(\"\\n\").split(\";\")\n",
    "                attributes=dict([tuple(x.split(\"=\")) for x in attTmp])\n",
    "                attOrder=[x.split(\"=\")[0] for x in attTmp]\n",
    "                tid=attributes[\"Parent\"]\n",
    "                if tid in transDiscard: # do not write since gene is the targeted for discarding\n",
    "                    continue\n",
    "                else:\n",
    "                    outFP.write(line)\n",
    "                    continue\n",
    "            elif lineCols[2]==\"CDS\":\n",
    "                attTmp=lineCols[8].rstrip(\"\\n\").split(\";\")\n",
    "                attributes=dict([tuple(x.split(\"=\")) for x in attTmp])\n",
    "                attOrder=[x.split(\"=\")[0] for x in attTmp]\n",
    "                tid=attributes[\"Parent\"]\n",
    "                if tid in transDiscard or tid in transNC: # do not write since gene is the targeted for discarding\n",
    "                    continue\n",
    "                else:\n",
    "                    outFP.write(line)\n",
    "                    continue\n",
    "            else:\n",
    "                outFP.write(line)\n",
    "                \n",
    "\n",
    "inFP.close()\n",
    "outFP.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Status</th>\n",
       "      <th>Gene_ID</th>\n",
       "      <th>Location</th>\n",
       "      <th>NR_blast_Eval</th>\n",
       "      <th>NR_hit</th>\n",
       "      <th>SP_blast_Eval</th>\n",
       "      <th>SP_hit</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>discard</td>\n",
       "      <td>CHS.10012</td>\n",
       "      <td>chr11:113875694-113899029:+</td>\n",
       "      <td>1e-17</td>\n",
       "      <td>CAE91339.1, unnamed protein product [Homo sapi...</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>non-coding</td>\n",
       "      <td>CHS.10056</td>\n",
       "      <td>chr11:116307531-116308521:+</td>\n",
       "      <td>3e-22</td>\n",
       "      <td>CAE89781.1, unnamed protein product [Homo sapi...</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>non-coding</td>\n",
       "      <td>CHS.1008</td>\n",
       "      <td>chr1:29124945-29139372:+</td>\n",
       "      <td>4e-19</td>\n",
       "      <td>EAW78405.1, hCG2021310 [Homo sapiens]</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>non-coding</td>\n",
       "      <td>CHS.10158</td>\n",
       "      <td>chr11:119057437-119059239:+</td>\n",
       "      <td>4e-81</td>\n",
       "      <td>BAC85254.1, unnamed protein product [Homo sapi...</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>discard</td>\n",
       "      <td>CHS.10242</td>\n",
       "      <td>chr11:121958716-121961024:+</td>\n",
       "      <td>2e-16</td>\n",
       "      <td>AAL23680.1, non-small cell lung cancer RimL3b ...</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       Status    Gene_ID                     Location NR_blast_Eval  \\\n",
       "0     discard  CHS.10012  chr11:113875694-113899029:+         1e-17   \n",
       "1  non-coding  CHS.10056  chr11:116307531-116308521:+         3e-22   \n",
       "2  non-coding   CHS.1008     chr1:29124945-29139372:+         4e-19   \n",
       "3  non-coding  CHS.10158  chr11:119057437-119059239:+         4e-81   \n",
       "4     discard  CHS.10242  chr11:121958716-121961024:+         2e-16   \n",
       "\n",
       "                                              NR_hit SP_blast_Eval SP_hit  \n",
       "0  CAE91339.1, unnamed protein product [Homo sapi...             -      -  \n",
       "1  CAE89781.1, unnamed protein product [Homo sapi...             -      -  \n",
       "2              EAW78405.1, hCG2021310 [Homo sapiens]             -      -  \n",
       "3  BAC85254.1, unnamed protein product [Homo sapi...             -      -  \n",
       "4  AAL23680.1, non-small cell lung cancer RimL3b ...             -      -  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Status</th>\n",
       "      <th>Transcript_ID</th>\n",
       "      <th>Gene_ID</th>\n",
       "      <th>Location</th>\n",
       "      <th>Transcript_Length</th>\n",
       "      <th>No_of_exons</th>\n",
       "      <th>No_of_samples</th>\n",
       "      <th>Average_TPM</th>\n",
       "      <th>Maximum_TPM</th>\n",
       "      <th>NR_Eval</th>\n",
       "      <th>...</th>\n",
       "      <th>SP_Eval</th>\n",
       "      <th>SP_hit</th>\n",
       "      <th>CDD_Eval</th>\n",
       "      <th>Pfam_Eval</th>\n",
       "      <th>CDSlen</th>\n",
       "      <th>CDSreps</th>\n",
       "      <th>CDSrepeatcov</th>\n",
       "      <th>ALU_cdcov</th>\n",
       "      <th>LINE_cdcov</th>\n",
       "      <th>GencodeCall</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>coding</td>\n",
       "      <td>CHS.55578.5</td>\n",
       "      <td>CHS.55578</td>\n",
       "      <td>chr9:61672749-61674874:-</td>\n",
       "      <td>1705</td>\n",
       "      <td>2</td>\n",
       "      <td>4</td>\n",
       "      <td>1.195119</td>\n",
       "      <td>1.687937</td>\n",
       "      <td>6e-63</td>\n",
       "      <td>...</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>0.00837726</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>non-coding</td>\n",
       "      <td>CHS.20892.1</td>\n",
       "      <td>CHS.20892</td>\n",
       "      <td>chr17:722531-732316:+</td>\n",
       "      <td>9786</td>\n",
       "      <td>1</td>\n",
       "      <td>1354</td>\n",
       "      <td>3.256216</td>\n",
       "      <td>22.569984</td>\n",
       "      <td>9e-18</td>\n",
       "      <td>...</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>333</td>\n",
       "      <td>Simple_repeat:49,SINE|Alu:248</td>\n",
       "      <td>89.2</td>\n",
       "      <td>74.5</td>\n",
       "      <td>0</td>\n",
       "      <td>SINE/Alu;Simple_repeat</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>non-coding</td>\n",
       "      <td>CHS.24098.4</td>\n",
       "      <td>CHS.24098</td>\n",
       "      <td>chr18:8974487-8998422:+</td>\n",
       "      <td>2451</td>\n",
       "      <td>3</td>\n",
       "      <td>187</td>\n",
       "      <td>5.632762</td>\n",
       "      <td>28.178564</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>SINE/Alu</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>non-coding</td>\n",
       "      <td>CHS.24098.3</td>\n",
       "      <td>CHS.24098</td>\n",
       "      <td>chr18:8974487-8980729:+</td>\n",
       "      <td>436</td>\n",
       "      <td>2</td>\n",
       "      <td>59</td>\n",
       "      <td>3.200027</td>\n",
       "      <td>9.097818</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>SINE/Alu</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>non-coding</td>\n",
       "      <td>CHS.24098.2</td>\n",
       "      <td>CHS.24098</td>\n",
       "      <td>chr18:8974487-8979364:+</td>\n",
       "      <td>454</td>\n",
       "      <td>2</td>\n",
       "      <td>17</td>\n",
       "      <td>2.209827</td>\n",
       "      <td>3.991817</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>SINE/Alu</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 21 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       Status Transcript_ID    Gene_ID                  Location  \\\n",
       "0      coding   CHS.55578.5  CHS.55578  chr9:61672749-61674874:-   \n",
       "1  non-coding   CHS.20892.1  CHS.20892     chr17:722531-732316:+   \n",
       "2  non-coding   CHS.24098.4  CHS.24098   chr18:8974487-8998422:+   \n",
       "3  non-coding   CHS.24098.3  CHS.24098   chr18:8974487-8980729:+   \n",
       "4  non-coding   CHS.24098.2  CHS.24098   chr18:8974487-8979364:+   \n",
       "\n",
       "   Transcript_Length  No_of_exons  No_of_samples  Average_TPM  Maximum_TPM  \\\n",
       "0               1705            2              4     1.195119     1.687937   \n",
       "1               9786            1           1354     3.256216    22.569984   \n",
       "2               2451            3            187     5.632762    28.178564   \n",
       "3                436            2             59     3.200027     9.097818   \n",
       "4                454            2             17     2.209827     3.991817   \n",
       "\n",
       "  NR_Eval           ...           SP_Eval SP_hit    CDD_Eval Pfam_Eval CDSlen  \\\n",
       "0   6e-63           ...                 -      -  0.00837726         -      -   \n",
       "1   9e-18           ...                 -      -           -         -    333   \n",
       "2       -           ...                 -      -           -         -      -   \n",
       "3       -           ...                 -      -           -         -      -   \n",
       "4       -           ...                 -      -           -         -      -   \n",
       "\n",
       "                         CDSreps CDSrepeatcov ALU_cdcov LINE_cdcov  \\\n",
       "0                              -            -         -          -   \n",
       "1  Simple_repeat:49,SINE|Alu:248         89.2      74.5          0   \n",
       "2                              -            -         -          -   \n",
       "3                              -            -         -          -   \n",
       "4                              -            -         -          -   \n",
       "\n",
       "              GencodeCall  \n",
       "0                       -  \n",
       "1  SINE/Alu;Simple_repeat  \n",
       "2                SINE/Alu  \n",
       "3                SINE/Alu  \n",
       "4                SINE/Alu  \n",
       "\n",
       "[5 rows x 21 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "genesDF=pd.read_csv(\"./chess2.0.novel_protein.genes.conservative.txt\",sep=\"\\t\")\n",
    "transDF=pd.read_csv(\"./chess2.0.novel_protein.transcripts.conservative.txt\",sep=\"\\t\")\n",
    "display(genesDF.head())\n",
    "display(transDF.head())\n",
    "\n",
    "genesDiscard=set(genesDF[genesDF['Status']=='discard']['Gene_ID']) # set of all geneIDs that need to be removed from the annotation\n",
    "genesC=set(genesDF[genesDF['Status']=='coding']['Gene_ID']) # set of all coding genes - used to verify that they all are annotated as coding and that they all have corresponding CDS\n",
    "genesNC=set(genesDF[genesDF['Status']=='non-coding']['Gene_ID']) # set of all geneIDs that need to be labeled as lncRNA\n",
    "\n",
    "transDiscard=set(transDF[transDF['Status']=='discard']['Transcript_ID'])\n",
    "transC=set(transDF[transDF['Status']=='coding']['Transcript_ID'])\n",
    "transNC=set(transDF[transDF['Status']=='non-coding']['Transcript_ID'])\n",
    "\n",
    "# need to make sure the geneIDS extracted from transDF of discard, NC and C match the corresponding ones from genesDF\n",
    "assert(len(genesDiscard.symmetric_difference(set([\"CHS.\"+t.split(\".\")[1] for t in transDiscard])))==0),\"wrong discard\"\n",
    "assert(len(genesC.symmetric_difference(set([\"CHS.\"+t.split(\".\")[1] for t in transC])))==0),\"wrong coding\"\n",
    "assert(len(genesNC.symmetric_difference(set([\"CHS.\"+t.split(\".\")[1] for t in transNC])))==0),\"wrong non-coding\"\n",
    "\n",
    "outFP=open(\"./chess2.1_assembly.gff\",\"w+\")\n",
    "\n",
    "version=\"2.1\"\n",
    "\n",
    "with open(\"./chess2.05_assembly.gff\",\"r\") as inFP:\n",
    "    for line in inFP.readlines():\n",
    "        if line[0]==\"#\":\n",
    "            if line[:7]==\"##CHESS\":\n",
    "                outFP.write(\"##CHESS\"+version+\"\\n\")\n",
    "            else:\n",
    "                outFP.write(line)\n",
    "            continue\n",
    "        else:\n",
    "            lineCols=line.split(\"\\t\")\n",
    "            if lineCols[2]==\"gene\":\n",
    "                attTmp=lineCols[8].rstrip(\"\\n\").split(\";\")\n",
    "                attributes=dict([tuple(x.split(\"=\")) for x in attTmp])\n",
    "                attOrder=[x.split(\"=\")[0] for x in attTmp]\n",
    "                gid=attributes[\"ID\"]\n",
    "                if gid in genesDiscard: # do not write since gene is the targeted for discarding\n",
    "                    continue\n",
    "                elif gid in genesNC: # need to alter the attributes in a way that makes sense\n",
    "                    attributes[\"GENE_TYPE\"]=\"lncRNA\"\n",
    "                    newLine=\"\"\n",
    "                    for att in attOrder:\n",
    "                        newLine+=att+\"=\"+attributes[att]+\";\"\n",
    "                    newLine=newLine[:-1] # remove trailing ';'\n",
    "                    outFP.write(\"\\t\".join(lineCols[:-1]+[newLine+\"\\n\"]))\n",
    "                    continue\n",
    "                else:\n",
    "                    if attributes[\"GENE_TYPE\"]==\"protein_coding\" and attributes[\"STATUS\"]==\"novel\": # coding - make sure it is in the genesC set\n",
    "                        assert(gid in genesC),\"coding protein found; not in genesC\"\n",
    "                    outFP.write(line)\n",
    "                    continue\n",
    "            elif lineCols[2]==\"transcript\":\n",
    "                attTmp=lineCols[8].rstrip(\"\\n\").split(\";\")\n",
    "                attributes=dict([tuple(x.split(\"=\")) for x in attTmp])\n",
    "                attOrder=[x.split(\"=\")[0] for x in attTmp]\n",
    "                tid=attributes[\"ID\"]\n",
    "                if tid in transDiscard: # do not write since gene is the targeted for discarding\n",
    "                    assert(\"CHS.\"+tid.split(\".\")[1] in genesDiscard),\"wrong transcript marked to be discarded\"\n",
    "                    continue\n",
    "                elif tid in transNC: # need to alter the attributes in a way that makes sense\n",
    "                    assert(\"CHS.\"+tid.split(\".\")[1] in genesNC),\"wrong transcript marked to be changed to protein-coding\"\n",
    "                    if \"GENE_TYPE\" in attributes:\n",
    "                        attributes.pop(\"GENE_TYPE\",None)\n",
    "                    if \"gbkey\" in attributes:\n",
    "                        attributes[\"gbkey\"]=\"lncRNA\"\n",
    "                    if \"gene_biotype\" in attributes:\n",
    "                        attributes[\"gene_biotype\"]=\"lncRNA\"\n",
    "                    if \"transcript_type\" in attributes:\n",
    "                        attributes.pop('transcript_type',None)\n",
    "                    if \"gene_type\" in attributes:\n",
    "                        attributes.pop('gene_type',None)\n",
    "                    attributes[\"TYPE\"]=\"non_coding\"\n",
    "                    newLine=\"\"\n",
    "                    for att in attOrder:\n",
    "                        newLine+=att+\"=\"+attributes[att]+\";\"\n",
    "                    newLine=newLine[:-1] # remove trailing ';'\n",
    "                    outFP.write(\"\\t\".join(lineCols[:-1]+[newLine+\"\\n\"]))\n",
    "                    continue\n",
    "                else:\n",
    "                    outFP.write(line)\n",
    "                    continue\n",
    "            elif lineCols[2]==\"exon\":\n",
    "                attTmp=lineCols[8].rstrip(\"\\n\").split(\";\")\n",
    "                attributes=dict([tuple(x.split(\"=\")) for x in attTmp])\n",
    "                attOrder=[x.split(\"=\")[0] for x in attTmp]\n",
    "                tid=attributes[\"Parent\"]\n",
    "                if tid in transDiscard: # do not write since gene is the targeted for discarding\n",
    "                    continue\n",
    "                else:\n",
    "                    outFP.write(line)\n",
    "                    continue\n",
    "            elif lineCols[2]==\"CDS\":\n",
    "                attTmp=lineCols[8].rstrip(\"\\n\").split(\";\")\n",
    "                attributes=dict([tuple(x.split(\"=\")) for x in attTmp])\n",
    "                attOrder=[x.split(\"=\")[0] for x in attTmp]\n",
    "                tid=attributes[\"Parent\"]\n",
    "                if tid in transDiscard or tid in transNC: # do not write since gene is the targeted for discarding\n",
    "                    continue\n",
    "                else:\n",
    "                    outFP.write(line)\n",
    "                    continue\n",
    "            else:\n",
    "                outFP.write(line)\n",
    "                \n",
    "\n",
    "inFP.close()\n",
    "outFP.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "# lastly need to modify all three supplementary files: mapfile, genes and transcripts\n",
    "outFP=open('./mapfile2.1.txt',\"w+\")\n",
    "with open('./mapfile2.05.txt',\"r\") as inFP:\n",
    "    for line in inFP.readlines():\n",
    "        lineCols=line.split(\"\\t\")\n",
    "        if lineCols[-1] in transDiscard or lineCols[-1] in genesDiscard:\n",
    "            continue\n",
    "        else:\n",
    "            outFP.write(line)\n",
    "            \n",
    "inFP.close()\n",
    "outFP.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "outFP=open('./chess2.1.genes',\"w+\")\n",
    "genesEvaluated=[]\n",
    "wrongCoords=[]\n",
    "header=True\n",
    "with open('./chess2.05.genes','r') as inFP:\n",
    "    for line in inFP.readlines():\n",
    "        if header: # skip header\n",
    "            outFP.write(line)\n",
    "            header=False\n",
    "            continue\n",
    "        lineCols=line.split(\"\\t\")\n",
    "        gids=lineCols[2].split(\",\")\n",
    "        locs=lineCols[3].split(\",\")\n",
    "        assert len(gids)==len(locs),\"something is wrong with the following locus: \"+lineCols[2]+\"\\t\"+lineCols[3]\n",
    "        newLocs=[]\n",
    "        newGIDs=[]\n",
    "        genesNC_flag=[]\n",
    "        skip=False\n",
    "        changeStatus=False\n",
    "        for gl in zip(gids,locs):\n",
    "            genesEvaluated.append(gl[0]) # append geneID for verification later\n",
    "            if gl[0] in genesDiscard:\n",
    "                skip=True\n",
    "                continue\n",
    "            elif gl[0] in genesNC: # if change of status - need to verify that all in that grouping have a change of status - otherwise a problem\n",
    "                changeStatus=True\n",
    "                assert lineCols[0]==\"protein_coding\",\"wrong status: \"+lineCols[0]\n",
    "                lineCols[0]=\"lncRNA\"\n",
    "                genesNC_flag.append(1)\n",
    "            newGIDs.append(gl[0])\n",
    "            newLocs.append(gl[1])\n",
    "        if not skip:\n",
    "            if changeStatus:\n",
    "                assert len(genesNC_flag)==len(gids),\"wrong change of status on a locus, maybe separate then?: \"+\",\".join([str(x) for x in genesNC_flag])\n",
    "            lineCols[3]=\",\".join(newLocs)\n",
    "            lineCols[2]=\",\".join(newGIDs)\n",
    "            outFP.write('\\t'.join(lineCols))\n",
    "        \n",
    "inFP.close()\n",
    "outFP.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "outFP=open(\"./chess2.1.transcripts\",\"w+\")\n",
    "\n",
    "transcriptsEvaluated=[]\n",
    "wrongCoords=[]\n",
    "header=True\n",
    "with open(\"./chess2.05.transcripts\") as inFP:\n",
    "    for line in inFP.readlines():\n",
    "        if header: # skip header\n",
    "            outFP.write(line)\n",
    "            header=False\n",
    "            continue\n",
    "        lineCols=line.split(\"\\t\")\n",
    "        tid=lineCols[0]\n",
    "        \n",
    "        if tid in transDiscard:\n",
    "            skip=True\n",
    "        elif tid in transNC:\n",
    "            assert lineCols[3]==\"protein_coding\",\"wrong status: \"+tid\n",
    "            lineCols[3]='lncRNA'\n",
    "        outFP.write('\\t'.join(lineCols))\n",
    "\n",
    "inFP.close()\n",
    "outFP.close()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
